---
title: "Figures in KURAMOCHI manuscript"
author: "Shuyu Zheng and Jing Tang"
date: "`r Sys.Date()`"
output: html_notebook
---

# Load pakcages

```{r}
library(Seurat)
library(plyr)
library(tidyverse)
library(readxl)
library(ggpubr)
library(ggalt)
library(ggrepel)
library(ggforce)
library(wordspace)
require(scales)
library(viridis)
library(ggsci)
library(synergyfinder)
library(ggsankey)
library(survminer)
library(survival)
library(scales)
library(sparseMatrixStats) # sparse matrix calculation
library(distances)
library(vegan) # distance function
library(irlba) # for fast pca
library(gridExtra) # Merge multiple ggplots with 'grid.arrange' function
library(reshape2)
library(car) # scatterplot combined with barplot
output_dir <- "./data/output/"
if (!dir.exists(output_dir)){
  dir.create(output_dir, recursive = TRUE)
}
```

# Color code

```{r}
blue = "#3C5488FF"
red =  "#DC0000FF"
green = "#00A087FF"
brown = "#B09C85FF"

blue2 <- "#4DBBD5FF"
red2 <- "#E64B35FF"
green2 <- "#91D1C2FF"
brown2 <- "#7E6148FF"

blue3 <- "#8491B4FF"
red3 <- "#F39B7FFF"
```

Some abbreviations used in the codes

* preR: pre-resistant
* preS: pre-sensitive
* BT: pre-treatment
* AT: post-treatment
* Carbo: Carboplatin
* Control/ctrl: non-treatment control

# Figure 1: ReSisTrace leverages sister cell similarity to discover cellular states primed for resistance

## B2: Sister distance in pre-treatment sample (violin plot)

Load auxiliary codes

```{r}
source("./auxiliary_codes.R") # Definition for some auxiliary codes
sub_output_dir <- paste0(output_dir, "Figure1/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir, recursive = TRUE)
}
```

Load data

```{r}
merge <- readRDS("./data/input/all_bt_sample_merge_filtered_gene_sisters.RDS")

# All of the non-sister cells are labeled with the "NA" at the end of values in "sisters_sample" column
all_sister_pairs <- merge$sisters_sample[!endsWith(merge$sisters_sample, "NA")]
```

Number of lineages containing sister cells

```{r}
t <- table(table(all_sister_pairs))
t
```

Percentage of lineages/cells containing twins (%)

```{r}
2424/(sum(!is.na(merge$drugSens)) - sum(!is.na(merge$sisters)) + sum(t)) * 100 # lineages
2424 * 2/sum(!is.na(merge$drugSens)) * 100 # cells
```

Percentage of lineages containing sisters (%)

```{r}
sum(t)/(sum(!is.na(merge$drugSens)) - sum(!is.na(merge$sisters)) + sum(t)) * 100 # lineages
sum(!is.na(merge$sisters))/sum(!is.na(merge$drugSens)) * 100 # cells
```

Percentage of cells preR/(preR + preS) (%)

```{r}
sum(merge$drugSens == "pre-resistant", na.rm = TRUE) / 
  sum(merge$drugSens %in% c("pre-resistant", "pre-sensitive"), na.rm = TRUE) * 
  100
```

Percentage of lineages preR/(preR + preS) (%)

All pre-treatment sample

```{r}
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens) %>% 
  filter(!is.na(drugSens) & drugSens == "pre-resistant")
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens) %>% 
  filter(!is.na(drugSens) & drugSens %in% c("pre-resistant", "pre-sensitive"))
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

Carboplatin

```{r}
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens == "pre-resistant",
    sample %in% c("Carbo1", "Carbo2")
  )
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens %in% c("pre-resistant", "pre-sensitive"),
    sample %in% c("Carbo1", "Carbo2")
    )
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

Olaparib

```{r}
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens == "pre-resistant",
    sample %in% c("Olaparib1", "Olaparib2")
  )
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens %in% c("pre-resistant", "pre-sensitive"),
    sample %in% c("Olaparib1", "Olaparib2")
    )
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

NK

```{r}
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens == "pre-resistant",
    sample %in% c("NK1", "NK2")
  )
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens %in% c("pre-resistant", "pre-sensitive"),
    sample %in% c("NK1", "NK2")
    )
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

Non-treatment control

```{r}
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens == "pre-resistant",
    sample %in% c("ctrl1", "ctrl2")
  )
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens %in% c("pre-resistant", "pre-sensitive"),
    sample %in% c("ctrl1", "ctrl2")
    )
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

1. Prepare expression matrix

```{r}
expression_mat <- as.matrix(GetAssayData(merge, slot = "data"))
```

2. Prepare sister table (including all sisters)

```{r}
sister_table <- data.frame(cell_barcode = colnames(merge), 
                     sisters = merge$sisters_sample,
                     stringsAsFactors = FALSE) %>% 
  dplyr::filter(sisters %in% all_sister_pairs) # remove singletons

sisters <- NULL
for (i in unique(sister_table$sisters)) {
  tmp_set <- sister_table$cell_barcode[which(sister_table$sisters  == i)]
  tmp_set <- combn(tmp_set, 2)
  tmp_df <- data.frame(group = rep(i, ncol(tmp_set)),
                       sister1 = tmp_set[1, ],
                       sister2 = tmp_set[2, ],
                       stringsAsFactors = FALSE)
  sisters <- rbind.data.frame(sisters, tmp_df)
  tmp_df <- NULL
  tmp_set <- NULL
}
```

```{r}
rm(merge) # remove large seurat object to release memory
rm(pca, tsne, umap)
```

3. Statistic analysis 

We can use function `CalculateSisterSimilarity` (defined in 'script/Auxiliary_codes.R" file) to generate analysis sister similarities by setting the parameters:

* **expression_mat**: the gene expression matrix in log2(normalized_count + 1) scale containing gene names as rows and cell barcodes as columns.
* **n_gene**: number of top variable genes used to calculate the similarity.
* **sister_table**: a data frame containing the group number of twins and corresponding sisters' cell barcodes.

Following is an example using Euclidean distance and top 1000 variable genes.

1. Calculate difference matrix

**Note** This function will take a long time to calculate the difference matrix. Some times it takes hours when there are a lot genes and cells.

```{r}
# time_stamp <- Sys.time()
plot_data <- CalculateSisterSimilarity(
  expression_mat = expression_mat,
  n_gene = 1000,
  sister_table = sisters,
  similarity = "euclidean"
  )
# Sys.time() - time_stamp
```

```{r}
sister_d <- data.frame(
    sister_id = names(plot_data$sister_distance),
    distance = plot_data$sister_distance,
    stringsAsFactors = FALSE
    )
```

```{r}
rm(expression_mat)
```

4. (Batch job) Generate the violin plot

This part takes a lot of time to finish. The codes were submited as a batch job on the CSC Puhti cluster. The codes are in: "./codes_for_batch_job/plot_sister_similarity.R" and "./codes_for_batch_job/plot_sister_similarity.sh"

```{r eval=FALSE}
sister_d <- readRDS("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_distance.RDS")
# p1 <- PlotSisterSimilarity(distance = sister_d$distance,
#                           similarity = sister_d$similarity,
#                                  n_gene = sister_d$n_gene,
#                                  cutoff = NULL)
m <- switch (sister_d$similarity,
             "euclidean" = "Euclidean Distance",
             "pearson" = "Pearson Coefficient",
             "spearman" = "Spearman Coefficient"
)

cutoff <- 14.94
test <- t.test(sister_d$distance$similarity[sister_d$distance$group == "sisters"], sister_d$distance$similarity[sister_d$distance$group == "non-sisters"])
print(test)

p1 <- ggplot(data = sister_d$distance, aes(x = group, y = similarity, fill = group)) +
  geom_violin(trim = FALSE) + 
  labs(title=paste0(m, " (genes: ", sister_d$n_gene, ")"),x="", y = m) +
  geom_boxplot(width=0.1, fill="white", outlier.shape = NA) + 
  # geom_segment(aes(y = cutoff, x = 1.75, xend = 2.25, yend = 14.93)) + 
  scale_fill_manual(values=c("#3C5488FF", "#DC0000FF")) +
  theme_classic() + 
  theme(
    legend.position = "none",
    plot.title = element_text(hjust = 0.5),
    axis.title.y = element_text(size = 30),
    axis.text.x = element_text(size = 30)
  ) +
  annotate(
    geom = "text",
    x = 1.5,
    y = 40,
    label = ifelse(
      test$p.value == 0,
      "P < 2.2E-16",
      paste0("P = ", test$p.value)
    ))
ggsave(
  paste0(
    sub_output_dir, "Figure1B2_",
    sister_d$sister_d, "_", sister_d$n_gene, "_genes.png"
  ), 
  plot = p1, device = "png", width = 6, height = 6, dpi = 300
)
```

## B3: UMAP for sister pairs

Clean environment

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

Load data

```{r}
merge <- readRDS("./data/input/all_bt_sample_merge_filtered_gene_sisters.RDS")
```

Prepare plot table

```{r}
all_sister_pairs <- merge$sisters_sample[!endsWith(merge$sisters_sample, "NA")]
all_sister_pairs <- unique(all_sister_pairs)
set.seed(123)
n <- 10
groups <- sample(all_sister_pairs, n)
highlight_group <- vector("list", length(groups))
names(highlight_group) <- groups
for (i in groups){
  highlight_group[[i]] <- colnames(merge)[which(merge$sisters_sample == i)]
}

myCol <- ggsci::pal_npg("nrc")(n)
names(myCol) <- groups
plot_table <- as.data.frame(merge@reductions$umap@cell.embeddings)
plot_table$sisters <- merge$sisters_sample
plot_table$sisters[!plot_table$sisters %in% groups] <- NA
```

UMAP plot

```{r}
umap <- ggplot(data = plot_table, mapping = aes(x = UMAP_1, y = UMAP_2)) +
  geom_point(
    data = plot_table,
    mapping = aes(x = UMAP_1, y = UMAP_2),
    color = "grey90") +
  geom_point(
    data = plot_table %>% filter(!is.na(sisters)),
    size = 6,
    mapping = aes(x = UMAP_1, y = UMAP_2, color=sisters)
  ) +
  scale_color_manual(values = myCol, labels = seq(1, length(myCol))) + 
  geom_line(
    data = plot_table %>% filter(!is.na(sisters)),
    mapping = aes(color = sisters),
    size = 2
  ) +
  labs(
    x = "UMAP 1",
    y = "UMAP 2",
    color = "Lineage label"
  ) +
  theme_classic() +
  theme(
    axis.text = element_text(size = 20, colour = "black"),
    axis.title = element_text(size = 20),
    legend.text = element_text(size = 20),
    legend.title = element_text(size = 25)
  )
umap
ggsave(paste0(sub_output_dir, "/Figure1B3_highlight_sister_group_",paste0(groups, collapse = "."), "_filtered_sisters_umap.pdf"), plot = umap, device = "pdf", width = 7.5, height = 4.5)
```

## C: Scatter plot for sister-concordant genes

The calculation of sister-concordant genes were performed on the CSC Puhti server with a batch job. The codes are in: "./codes_for_batch_job/calculate_conserve_gene.R" and "./codes_for_batch_job/calculate_conserve_gene.sh".

The output files from this batch job are saved in "./data/input/sister_concordant_gene/"

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

Load data

```{r}
conserve_gene <- read.csv(
  "./data/input/sister_concordant_gene/sister_conserve_gene_all_bt_sample_twin.csv",
  stringsAsFactors = FALSE)
```

Prepare data table for plot

```{r}
highlight_gene_symbols <- conserve_gene %>% 
  arrange(desc(log2FC_dif)) %>% 
  head(10) %>% 
  select(symbol) %>% 
  unlist()
data_plot <- conserve_gene %>% 
  mutate(log10_p_value = -log10(adj_p_value)) %>% 
  mutate(
    highlight = ifelse(
      log10_p_value > -log10(0.05),
      "sister-concordant gene",
      "sister-disconcordant gene"
    ),
    label = (
      symbol %in% highlight_gene_symbols
    )
  )
data_plot$highlight <- factor(data_plot$highlight, levels = c("sister-concordant gene", "sister-disconcordant gene"))
cor <- cor.test(data_plot$mean_log2FC_sister, data_plot$mean_log2FC_non_sister)
```

Visualization

```{r}
p <- 
  ggplot(
    data_plot,
    aes(
      x = mean_log2FC_sister,
      y = mean_log2FC_non_sister,
      color = highlight
    )
  ) +
  geom_point() +
  lims(x = c(0, 1.5), y = c(0, 1.5)) +
  scale_color_manual(values = c("#E64B3555", "#3C5488")) + 
  geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
  annotate(
    geom = "text",
    x = 1.0,
    y = 0.2,
    label = paste0("R = ", signif(cor$estimate, 2), "\n", ifelse(cor$p.value == 0, "P < 2.2E-16", paste0("P =", format(cor$p.value, scientific = T)))),
    size = 6.5
  ) +
  geom_point(
    data=dplyr::filter(data_plot, label==TRUE),
    size = 6,
    color = "#DC0000") +
  geom_point(
    data=dplyr::filter(data_plot, label==TRUE),
    shape = 1,
    size = 6,
    show.legend = FALSE,
    color = "black"
  ) +
  ggrepel::geom_text_repel(
    data=dplyr::filter(
      data_plot, label==TRUE),
    aes(label = symbol),
    box.padding = 0.5,
    colour="black", point.padding = 0.25, max.overlaps = 30,
    size = 6.5,
    fontface="bold",
    show.legend=FALSE) +
  labs(
       x = expression("Sister cell pair log"[2]~"(FC)"),
       y = expression("Random cell pair log"[2]~"(FC)")
       ) +
  theme_classic() +
  theme(legend.position="right",
        axis.ticks.x = element_blank(),
        axis.text = element_text(size = 20, color = "black"),
        axis.title = element_text(size = 20, color = "black")
      ) +
  guides(colour = "none")
p
ggsave(paste0(sub_output_dir, "Figure1C_conserve_gene_adjust_p.pdf"), p,
       width = 6, height = 6)
```

## D: RNA dynamics from RNA velocity analysis

The alignment is done with velocyto. The codes are in: "./codes_for_batch_job/runVelo.sh"

The scVelo is used to do the RNA velocity analysis. The codes are in "./codes_for_batch_job/runscvelo.py" and "./codes_for_batch_job/runscvelo.sh"

The outputs are saved in "./data/input/scvelo/"

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

Load data

```{r}
dynamic <- read.csv("./data/input/scvelo/RNA_velocity_var_all_BT.csv")
# conserve_gene <- read.csv("./data/input/sister_concordant_gene/sister_conserve_gene_all_bt_sample_twin.csv")
selected_genes <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_second_bin.csv")
```

Prepare table for plot

```{r}
plot_table <- selected_genes %>%
  mutate(conserve_gene = ifelse(Sister.conserved.gene, "Sister-concordant", "Sister-discordant")) %>% 
  right_join(dynamic, by = c("ensembl" = "X")) %>% 
  select(
    ensembl, 
    transcription = fit_alpha, 
    splicing = fit_beta, 
    degradation = fit_gamma, 
    conserve_gene) %>% 
  na.omit()
plot_table$conserve_gene <- factor(plot_table$conserve_gene, levels = c("Sister-concordant", "Sister-discordant"))
t <- plot_table %>% 
  select(ensembl, transcription, value = splicing, conserve_gene) %>%
  mutate(dynamic = "splicing") %>% 
  rbind.data.frame(
    plot_table %>% 
      select(ensembl, transcription, value = degradation, conserve_gene) %>% 
      mutate(dynamic = "degradation"))
t$dynamic <- str_to_title(t$dynamic)
t$dynamic <- factor(t$dynamic, levels = c("Splicing", "Degradation"))
```

Visualization

```{r}
min_value <- min(c(t$transcription, t$value))
max_value <- max(c(t$transcription, t$value))
p <- ggplot(t, aes(x = transcription, y = value)) +
  stat_density_2d(aes(fill = after_stat(ndensity)), geom = "raster", contour = FALSE,
                  ) +
  scale_fill_viridis() +
  labs(x = "Transcription") +
  geom_abline(
    intercept = 0,
    slope = 1,
    color = "#FFFFFFFF",
    linetype = "dashed",
    size = 1) + 
  scale_y_continuous(
    expand = c(0, 0),
    limits = c(min_value, max_value),
    trans = log2_trans(),
    breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x))) +
  scale_x_continuous(
    expand = c(0, 0),
    limits = c(min_value, max_value),
    trans = log2_trans(),
    breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x))) +
  coord_equal(xlim=c(min_value, max_value),ylim=c(min_value, max_value)) +
  facet_grid(vars(dynamic), vars(conserve_gene)) +
  theme(
    legend.position = "none",
    strip.text.x = element_text(
      size = 20
    ),
    strip.text.y = element_text(
      size = 20
    ),
    axis.text = element_text(size = 20, colour = "black"),
    axis.title.y = element_blank(),
    axis.title.x = element_text(
      size = 20
    )
  )
p
ggsave(
  paste0(
    sub_output_dir,
    "Figure1D_sister_conserve_gene_dynamic_log_scale_second_bin.pdf"
  ),
  p, width = 6, height = 6
)
```

## E: Gene set analysis for sister-discordant genes

1. We calculated the summation of all of the gene's expression across all of the cells in all BT samples. (results in './data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv') 
2. We generated a histogram for the summation of expression for all of the genes (see below Figure S2A). Genes are grouped by sister-concordant genes (adj_p_value < 0.05 and log2FC_diff > 0) or sister-discordant genes (adj_p_value > 0.05 or log2FC_diff < 0).
3. The genes in the second bin (second bar for each group) in the histogram are extracted and the gene set enrichment analysis is done on each group (sister-concordant genes n = 2059 [enrichr result](https://maayanlab.cloud/Enrichr/enrich?dataset=fe932ded91e8f8298f674699e8b007eb), and non-sister-concordant genes n = 1332 [enrichr result](https://maayanlab.cloud/Enrichr/enrich?dataset=7b522be89fef4b4a4f08355a0ef3703d))
4. The tables for Pathways "KEGG 2021 Human" are downloaded and saved in "./data/input/enrichr/KEGG_2021_Human_table_sister_conserved_gene.txt" and "./data/input/enrichr/KEGG_2021_Human_table_non_conserved_gene.txt"

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure1/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir, recursive = TRUE)
}
```

```{r}
cons <- data.table::fread("./data/input/enrichr/KEGG_2021_Human_table_non_conserved_genes.txt")
plot_table <- cons %>% 
    select(Term, x = `Odds Ratio`, p = `Adjusted P-value`) %>% 
    mutate(y = -log10(p),
           sig = p < 0.05)
# plot_table$Term <- paste0("bold('", plot_table$Term, "')")
highlight <- plot_table %>% 
    # filter(Term %in% c("Inositol phosphate metabolism", "Phosphatidylinositol signaling system"))
    filter(sig)

p <- ggplot() +
   theme_classic() +
   theme(
       legend.key = element_blank(), 
       axis.text = element_text(size = 20, colour = "black"),
       axis.title = element_text(size = 20),
       plot.title = element_text(size = 20, hjust = 0.5)
       ) +
   labs(x = expression("Odds Ratio"),
        y = expression("-log"[10]~"(P value)"),
        title="") +
   # geom_hline(yintercept=-log10(0.05), linetype="dashed") +
   # geom_vline(xintercept = 0, linetype="dashed") +
 geom_point(
     aes(x = x, y = y, colour=factor(sig)),
     colour = "grey70",
     size = 6,
     data = plot_table,
     show.legend = FALSE) +
  geom_point(
    aes(x = x, y = y),
    size = 8,
    data = highlight,
    color = "#3C5488FF",
    show.legend = FALSE) +
  geom_point(
    aes(x = x, y = y),
    shape = 1,
    size = 8,
    data = highlight,
    show.legend = FALSE) +
  geom_text_repel(
    aes(x = x, y = y, label = Term),
    size = 6.5,
    data = highlight,
    # parse = TRUE,
    box.padding = 1
  )
p
ggsave(paste0(sub_output_dir, "Figure1E_non_conserved_KEGG_volcano.pdf"), p,
       width = 6, height = 6)
```

# Figure S1: Cell doubling, lineage labelling efficiency, and quality assessment of ReSisTrace

## B1: Number of divisions after 1x doubling

The data table comes from Jun. He manually counted the number of cells which divided zero, once, twice or more times after "1 x doubling time". Totally 6 views are counted.

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

Load data

```{r}
data <- read_xlsx("./data/input/ReSisTrace_figS1 - Copy.xlsx", range = "B1:G4")
data <- t(data)
colnames(data) <- c("0", "1", "> 1")
data <- data * 100
data <- data %>% 
  as.data.frame() %>% 
  mutate(view = 1:6) %>% 
  pivot_longer(!view, names_to = "Division", values_to = "Proportion")
data$Division <- factor(data$Division, levels = c("0", "1", "> 1"))
```

Prepare table for plot

```{r}
plot_table <- data %>%
  dplyr::group_by(Division) %>% 
  dplyr::summarise(
    mean = mean(Proportion),
    sd = sd(Proportion)
  )
write.csv(plot_table, paste0(sub_output_dir, "FigureS1A_cell_division_times.csv"), row.names = FALSE)
col = c("#d1d7e3", "#8f9dbb", "#3c5488")
```


```{r}
p <- ggplot(plot_table, aes(x = Division, y = mean, fill = Division)) +
  geom_bar(
    stat="identity", color="black",
    position=position_dodge()
  ) +
  geom_errorbar(
    aes(ymin = mean - sd, ymax = mean + sd, width = 0.2), color="black"
  ) +
  scale_fill_manual(values = col) +
  labs(
    x = "Division(s)",
    y = "Proportion of clones (%)"
  ) +
  theme_classic() +
  theme(
    legend.position = "None",
    text = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 20, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "FigureS1B1_cell_division_times.pdf"),
  width = 5, height = 5
)
```

## B2: Counts of lineages in the BT sample

The information about lineage labels for each samples are saved in "./data/input/preprocess/[sample_name]/lineage_label_tables.RData". After loading the ".RData" files, a data frame "l_b_fre_merge_filtered" will be in the working environment. This table contains the lineage label information for 
```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

Prepare table for plot

```{r}
plot_table <- NULL

samples <- c("Control_1", "Control_2", "Carboplatin_1", "Carboplatin_2", "Olaparib_1", "Olaparib_2", "NK_1", "NK_2")

for (s in samples){
  load(paste0("./data/input/preprocess/", s, "/lineage_label_tables.RData"))
  tmp <- l_b_fre_merge_filtered %>% 
    select(lineage_id) %>% 
    table() %>% 
    table() %>% 
    prop.table()
  tmp2 <- l_b_fre_merge_filtered %>% 
    select(lineage_id) %>% 
    table() %>% 
    table()
  cat(s, "\n", tmp2, "\n")
  tmp <- data.frame(
    Size_of_lineage = tmp,
    Sample = s,
    label_at = length(unique(l_a_fre_merge_filtered$lineage_id)),
    label_bt = length(unique(l_b_fre_merge_filtered$lineage_id)),
    label_btat = length(unique(intersect(l_b_fre_merge_filtered$lineage_id, l_a_fre_merge_filtered$lineage_id)))
  )
  plot_table <- tmp
}

colnames(plot_table) <- c("Size_of_lineage", "Frequency", "Sample", "labelAT", "labelBT", "labelBTAT")
write.csv(
  plot_table,
  paste0(sub_output_dir, "FigureS1B2_frequence_of_lineage_lable.csv"),
  row.names = FALSE
)
```

Visualization

```{r}
plot_table$Frequency <- plot_table$Frequency * 100
plot_table <- plot_table %>%
  dplyr::group_by(Size_of_lineage) %>% 
  dplyr::summarise(
    mean = mean(Frequency),
    sd = sd(Frequency)
  )
col = c("#d1d7e3", "#8897b7", "#5e729d", "#3c5488")
p <- ggplot(plot_table, aes(x = Size_of_lineage, y = mean, fill = Size_of_lineage)) +
  geom_bar(
    stat="identity", color="black",
    position=position_dodge()
  ) +
  geom_errorbar(
    aes(ymin = mean - sd, ymax = mean + sd, width = 0.2), color="black"
  ) +
  scale_fill_manual(values = col) +
  labs(
    x = "Clone size",
    y = "Proportion of lineages (%)"
  ) +
  theme_classic() +
  theme(
    legend.position = "None",
    text = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 20, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "FigureS1B_frequence_of_lineage_lable.pdf"),
  width = 5, height = 5
)
```

## C: DEGs correlation between replicates

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir, recursive = TRUE)
}
```

## Carboplatin

```{r}
treatments <- c("Carboplatin", "Olaparib", "NK", "Control")
for (t in treatments){
  t_short <- switch (t,
    "Carboplatin" = "Carbo",
    "Control" = "scLT_ctrl",
    "NK" = "NK",
    "Olaparib" = "Olaparib"
  )
  deg <- read.csv(
    paste0(
      "./data/input/DEG/deg_prer_pres_in_singleton_",
      t,
      "_intersection.csv"
    )
  )

  cat("Number of genes ", nrow(deg), "\n")
  cor <- cor.test(
    deg[[paste0(t_short, "_1_avg_log2FC")]],
    deg[[paste0(t_short, "_2_avg_log2FC")]]
  )
  range <- range(
    c(
      deg[[paste0(t_short, "_1_avg_log2FC")]],
      deg[[paste0(t_short, "_2_avg_log2FC")]]
    )
  )
  p <- ggplot(
    data = deg,
    aes_string(
      x = paste0(t_short, "_1_avg_log2FC"),
      y = paste0(t_short, "_2_avg_log2FC")
    )
    ) +
    geom_point(
      color = "#3C548855",
      size = 6,
      show.legend = FALSE
    ) +
    ylim(range) +
    xlim(range) +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
    annotate(
      geom = "text",
      x = 0.15,
      y = - 0.3,
      label = paste0(
        "R = ", signif(cor$estimate, 2), "\n",
        ifelse(
          cor$p.value == 0, "P < 2.2E-16",
          paste0("P =", format(cor$p.value, scientific = T))
        )
      ),
      size = 6.5
    ) + 
    labs(
      x = expression("Replicate 1 log"[2]~"(FC)"),
      y = expression("Replicate 2 log"[2]~"(FC)"),
      title = t
    ) +
    theme_classic() +
    theme(
      legend.position="none",
      axis.ticks.x = element_blank(),
      axis.text = element_text(size = 20, colour = "black"),
      axis.title = element_text(size = 20, colour = "black"),
      plot.title = element_text(size = 20, hjust = 0.5, colour = "black", face = "bold")
    )
  p
  ggsave(paste0(sub_output_dir, "FigureS1C_correlation_between_replicates_", t, ".pdf"), p,
         width = 4.5, height = 4.5)
}

```

# Figure S2: Characteristics of sister-concordant and sister-discordant genes

## A: Histogram for gene expressions of sister-conserved/non-conserved genes

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS2/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir, recursive = TRUE)
}
```

```{r}
plot_table <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv")

t_test <- t.test(plot_table$mean_expression[plot_table$conserve_gene == TRUE], plot_table$mean_expression[plot_table$conserve_gene == FALSE])
print(t_test)

plot_table <-  plot_table %>%  
  mutate(`Sister-conserved gene` = ifelse(conserve_gene, "Sister-concordant gene", "Sister-discordant gene")) %>% 
  select(ensembl, `Sister-conserved gene`, sum_expression, mean_expression)

mylog10_trans <- function (base = 10)
{
  trans <- function(x) log(x + 1, base)
  inv <- function(x) base^x
  trans_new(paste0("log-", format(base)), trans, inv, log_breaks(base = base),
            domain = c(1e-100, Inf))
}

p <- ggplot(
  plot_table,
  aes(
    x = sum_expression,
    fill = `Sister-conserved gene`,
    color = `Sister-conserved gene`
    )
  ) +
  geom_histogram(
    alpha=0.2,
    bins = 50,
    position="identity"
  ) +
  scale_fill_manual(values = c("#DC000055", "#3C548855")) +
  scale_color_manual(values = c("#DC0000", "#3C5488")) +
  # ?(
  #   breaks = trans_breaks("log10", function(x) 10^x),
  #   labels = trans_format("log10", math_format(10^.x))
  # ) +
  geom_segment(
    data = NULL, 
    aes(x = 10000, y = 5 * 10^3, xend = 5000, yend = 2.5* 10^3),
    arrow = arrow(length = unit(0.25, "cm"), type = "closed"),
    show.legend = FALSE) +
  scale_y_continuous(
    trans = "mylog10",
    breaks = trans_breaks("log10", function(x) 10^(x - 1)),
    labels = trans_format("log10", function(x) math_format()(round(x)))
    ) +
  annotate(
    "text",
    x = 100000,
    y = 1000,
    label = paste0(
      "T-test, ",
      ifelse(
        t_test$p.value == 0,
        "P < 2.2E10-16",
        paste0("P = ", t_test$p.value)
      )
    )
  )
  labs(
    x = "Expression in pre-treatment sample",
    y = "Number of genes",
    fill = "",
    color = "") +
  theme_classic() +
  theme(
    legend.position = "top",
    axis.text = element_text(size = 20, colour = "black"),
    axis.title = element_text(size = 20, colour = "black"),
    legend.title = element_text(size = 20, colour = "black"),
    legend.text = element_text(size = 20, colour = "black"),
    axis.line = element_line(colour = "black"),
    axis.ticks = element_line(colour = "black")
  )
p
ggsave(
  paste0(sub_output_dir, "FigureS2A_sister_conserve_gene_exp_distribution.pdf"), p,
  width = 9, height = 5
)
```

## B: Unspliced RNA propotion for sister-conserved/non-conserved genes

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS2/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
selected_genes <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_second_bin.csv")
```

```{r eval = FALSE}
unsplice <- read.csv("./data/input/scvelo/RNA_velocity_unspliced_all_BT.csv", row.names = 1)
unsplice <- as.matrix(unsplice)
unsplice  <- unsplice[, colnames(unsplice) %in% selected_genes$ensembl]
splice <- read.csv("./data/input/scvelo/RNA_velocity_spliced_all_BT.csv", row.names = 1)
splice <- as.matrix(splice)
splice <- splice[, colnames(splice) %in% selected_genes$ensembl]
u_s_propotion <- unsplice / (unsplice + splice)
write.csv(u_s_propotion, paste0(sub_output_dir, "FigureS2B_unspliced_RNA_propotion_second_bin.csv"))
```

```{r}
u_s_propotion <- read.csv(
  paste0(sub_output_dir, "FigureS2B_unspliced_RNA_propotion_second_bin.csv"),
  row.names = 1
)
```

```{r}
u_s_propotion <- as.matrix(u_s_propotion)
selected_genes$conserve_gene <- ifelse(selected_genes$Sister.conserved.gene, "Sister-concordant\ngene", "Sister-discordant\ngene")
selected_genes$conserve_gene <- factor(selected_genes$conserve_gene, levels = c("Sister-concordant\ngene", "Sister-discordant\ngene"))
u_s_propotion_mean <- colMeans(u_s_propotion, na.rm = T)
u_s_propotion_mean <- data.frame(
  gene = names(u_s_propotion_mean), 
  unsplice_propotion = u_s_propotion_mean
) %>% 
  filter(gene %in% selected_genes$ensembl) %>% 
  left_join(select(selected_genes, gene = ensembl, conserve_gene), by = "gene") %>% 
  filter(!is.na(conserve_gene))
  
means <- aggregate(unsplice_propotion ~ conserve_gene, u_s_propotion_mean, mean)
p <- ggplot(
  data = u_s_propotion_mean,
  aes(x = conserve_gene, y = unsplice_propotion, fill = conserve_gene)
) + 
  geom_violin(trim = FALSE) + 
  labs(x="", y = "Unspliced RNA propotion") +
  geom_boxplot(width=0.1, fill="white", outlier.shape = NA) + 
  scale_fill_manual(values=c(red, blue)) +
  theme_classic() + 
  theme(
    legend.position = "none",
    plot.title = element_text(hjust = 0.5),
    axis.title.y = element_text(size = 20, color = "black"),
    axis.text.x = element_text(size = 20, color = "black"),
    axis.text.y = element_text(size = 20, color = "black")
  ) + 
  ggpubr::stat_compare_means(
    method = "t.test",
    label.y = 1.3,
    label.x.npc = "center",
    size = 6.5,
    hjust = 0.5) +
  geom_text(
    data = means,
    aes(label = signif(unsplice_propotion, 3),y = unsplice_propotion), 
    nudge_x = 0.18, color = "white", size = 6.5, fontface = "bold")

ggsave(
  paste0(sub_output_dir, "FigureS2B_unspliced_RNA_propotion_second_bin.pdf"), 
  plot = p, width = 8, height = 5
)
```

## C: Degradation/Transcription violin plot for sister-conserved/non-conserved genes

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS2/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
selected_genes <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_second_bin.csv")
```

```{r}
dynamic <- read.csv("./data/input/scvelo/RNA_velocity_var_all_BT.csv")
selected_genes$conserve_gene <- ifelse(selected_genes$Sister.conserved.gene, "Sister-concordant\ngene", "Sister-discordant\ngene")
selected_genes$conserve_gene <- factor(selected_genes$conserve_gene, levels = c("Sister-concordant\ngene", "Sister-discordant\ngene"))
plot_table <- selected_genes %>%
  select(ensembl, conserve_gene) %>% 
  right_join(dynamic, by = c("ensembl" = "X")) %>% 
  mutate(
    Degradation_Transcription = fit_gamma/fit_alpha
  ) %>% 
  select(
    ensembl, 
    Degradation_Transcription, 
    conserve_gene) %>% 
  na.omit()
means <- aggregate(Degradation_Transcription ~ conserve_gene, plot_table, mean)
```

```{r}
p <- ggplot(
  data = plot_table,
  aes(x = conserve_gene, y = Degradation_Transcription, fill = conserve_gene)
) + 
  geom_violin(trim = FALSE) + 
  labs(x="", y = "Degradation rate / Transcription rate") +
  geom_boxplot(width=0.1, fill="white", outlier.shape = NA) + 
  scale_fill_manual(values=c(red, blue)) +
  theme_classic() + 
  theme(
    legend.position = "none",
    axis.title.y = element_text(size = 20, color = "black"),
    axis.text.x = element_text(size = 20, color = "black"),
    axis.text.y = element_text(size = 20, color = "black")
  ) + 
  ggpubr::stat_compare_means(
    method = "t.test",
    label.x.npc = "center",
    size = 6.5,
    hjust = 0.5
  ) +
  geom_text(
    data = means,
    aes(label = signif(Degradation_Transcription, 3), y = Degradation_Transcription), 
    nudge_x = 0.15, color = "white", size = 6.5, fontface = "bold")

ggsave(
  paste0(sub_output_dir, "FigureS2C_Degradation_Transcription_second_bin.pdf"), 
  plot = p, width = 8, height = 5.5
)
```

## D: Gene set analysis for sister-concordant genes

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS2/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

```{r}
cons <- data.table::fread("./data/input/enrichr/KEGG_2021_Human_table_sister_conserved_gene.txt")
plot_table <- cons %>% 
    select(Term, x = `Odds Ratio`, p = `Adjusted P-value`) %>% 
    mutate(y = -log10(p),
           sig = p < 0.05)
# plot_table$Term <- paste0("bold('", plot_table$Term, "')")
highlight <- plot_table %>% 
    # filter(Term %in% c("TGF-beta signaling pathway", "Fanconi anemia pathway",
    #            "TNF signaling pathway"))
    filter(sig)

p <- ggplot() +
   theme_classic() +
   theme(
       legend.key = element_blank(), 
       axis.text = element_text(size = 20, colour = "black"),
       axis.title = element_text(size = 20),
       plot.title = element_text(size = 20, hjust = 0.5)
       ) +
   labs(x = expression("Odds Ratio"),
        y = expression("-log"[10]~"(P value)"),
        title="") +
   # geom_hline(yintercept=-log10(0.05), linetype="dashed") +
   # geom_vline(xintercept = 0, linetype="dashed") +
 geom_point(
     aes(x = x, y = y, colour=factor(sig)),
     colour = "grey70",
     size = 6,
     data = plot_table,
     show.legend = FALSE) +
  geom_point(
    aes(x = x, y = y),
    size = 8,
    data = highlight,
    color = "#3C5488FF",
    show.legend = FALSE) +
  geom_point(
    aes(x = x, y = y),
    shape = 1,
    size = 8,
    data = highlight,
    show.legend = FALSE) +
  geom_text_repel(
    aes(x = x, y = y, label = Term),
    size = 6.5,
    data = highlight,
    # parse = TRUE,
    box.padding = 1
  )
p
ggsave(paste0(sub_output_dir, "FigureS2D_sister_conserved_KEGG_volcano.pdf"), p,
       width = 8, height = 5.5)
```

# Figure 2: ReSisTrace uncovers the interplay between genetic and transcriptomic features in primed resistance

## A: UMAP drug sensitivity groups

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure2/")

if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

We merged the scRNA-seq data from all of the 8 samples into one Seurat object. FindVariableFeatures and ScaleData are performed on this object. It is saved in "./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS". PCA and UMAP projections are calculated for this object and the results are saved in "./data/input/merged_KURAMOCHI_samples1_SCT_umap_2000_n50_scaleT.RDS". The codes for processing this Seurat object can be found in file "./UMAP_visualization.R"

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
data2 = data@assays$RNA@scale.data  # scale data is for dimension reduction or clustering
dim(data2) # 2000 77724

### UMAP visualization ----------------------------------------------------

data$drugSens[which(data$drugSens == "resistant")] = "post-treatment"
data2.umap = readRDS("./data/input/merged_KURAMOCHI_samples1_SCT_umap_2000_n50_scaleT.RDS")

index = which(data$drugSens %in% c("pre-sensitive", "pre-resistant","post-treatment"))

pdf(paste0(sub_output_dir, "Figure2A_umap_all.pdf"), width = 5, height = 5, pointsize = 12)
col = c("snow2", "snow4", "slategray4")
size = c(1,3,3)
tmp = factor(data$drugSens[index], levels = c("pre-sensitive", "pre-resistant", "post-treatment"))
plot(data2.umap$layout[index,], col = col[tmp], pch = '.', cex = size[tmp], xlab = "UMAP 1", ylab = "UMAP 2")
legend("bottomright", legend = levels(tmp), pch = 16, col = col[1:length(tmp)], cex = 1, pt.cex = 1)
dev.off()
table(data$drugSens[index])
```

## B: UMAP CNV subclone

```{r}
index = which(data$drugSens %in% c("pre-sensitive", "pre-resistant","post-treatment"))

pdf(paste0(sub_output_dir, "Figure2B_umap_all_cnv.pdf"), width = 5, height = 5, pointsize = 12)
col <- c(
  "A" = "#DC0000FF", "B" = "#B09C85FF", "C" = "#00A087FF",
  "D" = "#91D1C2FF",  "E" = "#3C5488FF"
  )
tmp = factor(data$CNV_subclone[index], levels = c("A", "B", "C", "D", "E"))
plot(data2.umap$layout[index,], col = col[tmp], pch = '.', xlab = "UMAP 1", ylab = "UMAP 2")
legend("topright", legend = levels(tmp), pch = 16, col = col[1:length(tmp)], cex = 1, pt.cex = 1)
dev.off()
table(data$CNV_subclone[index])
```

## C: Proportion of CNV subclones

```{r}
# proportion of sub-clones in pre-S
subclone = sort(unique(data$CNV_subclone))
# treatment = unique(data$treatment)
treatment = c("Carboplatin", "Olaparib", "NK", "Control") 
res2 = data.frame(mat.or.vec(length(treatment), length(subclone)))
count2 = data.frame(mat.or.vec(length(treatment), length(subclone)))
for(i in 1:length(treatment)){
  tmp = data$CNV_subclone[which(data$drugSens == "pre-sensitive" & data$treatment == treatment[i])]
  count2[i,] = table(tmp)
  x = table(tmp)/length(tmp)
  cat(treatment[i], "\n", table(tmp), "\n")
  res2[i,] = x
  names(res2) = names(x)
  names(count2) <- names(x)
}

rownames(res2) = treatment
res2 = melt(as.matrix(res2))
colnames(res2) <- c("treatment", "subclone", "proportion")
res2$drugSens <- "pre-sensitive"

rownames(count2) = treatment
count2 = melt(as.matrix(count2))
colnames(count2) <- c("treatment", "subclone", "count")
count2$drugSens <- "pre-sensitive"

values1 = c(red, brown, green, green2, blue)

p <- ggplot(res2, aes(x = treatment, y = proportion, fill = subclone)) +
  geom_col(position = "fill") +
  labs(
    y = "Proportion",
    fill = ""
  ) + 
  ggtitle("pre-sensitive") +
  scale_fill_manual(values = values1) +
  theme_bw() +
  theme(
    text = element_text(size = 24),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 21,
      color = "black"
    ),
    axis.text.y = element_text(
      size = 21,
      color = "black"
    ),
    axis.title.x = element_blank(),
    legend.text = element_text(size = 21),
    plot.title = element_text(hjust = 0.5, size = 28),
    panel.grid = element_blank()
  )
ggsave(paste0(sub_output_dir, "Figure2C_BRCA_CNV_preS.pdf"), width = 5, height = 5)

# pre-R
res3 = data.frame(mat.or.vec(length(treatment), length(subclone)))
count3 = data.frame(mat.or.vec(length(treatment), length(subclone)))
for(i in 1:length(treatment)){
  tmp = data$CNV_subclone[which(data$drugSens == "pre-resistant" & data$treatment == treatment[i])]
  count3[i, ] <- table(tmp)
  x = table(tmp)/length(tmp)
  res3[i,] = x
  names(res3) = names(x)
  names(count3) <- names(x)
}
rownames(res3) = treatment
res3
res3 = melt(as.matrix(res3))
colnames(res3) <- c("treatment", "subclone", "proportion")
res3$drugSens <- "pre-resistant"

rownames(count3) = treatment
count3 = melt(as.matrix(count3))
colnames(count3) <- c("treatment", "subclone", "count")
count3$drugSens <- "pre-resistant"

p <- ggplot(res3, aes(x = treatment, y = proportion, fill = subclone)) +
  geom_col(position = "fill") +
  labs(
    y = "Proportion",
    fill = "") + 
  ggtitle("pre-resistant") +
  scale_fill_manual(values = values1) +
  theme_bw() +
  theme(
    text = element_text(size = 24),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 21,
      color = "black"
    ),
    axis.text.y = element_text(
      size = 21,
      color = "black"
    ),
    axis.title.x = element_blank(),
    plot.title = element_text(hjust = 0.5, size = 28),
    panel.grid = element_blank()
  )
ggsave(paste0(sub_output_dir, "Figure2C_BRCA_CNV_preR.pdf"), width = 5, height = 5)

# Post-treatment
res4 = data.frame(mat.or.vec(length(treatment), length(subclone)))
count4 = data.frame(mat.or.vec(length(treatment), length(subclone)))
for(i in 1:length(treatment)){
  tmp = data$CNV_subclone[which(data$drugSens == "post-treatment" & data$treatment == treatment[i])]
  count4[i, ] <- table(tmp)
  x = table(tmp)/length(tmp)
  res4[i,] = x
  names(res4) = names(x)
  names(count4) <- names(x)
}
rownames(res4) = treatment
res4
res4 = melt(as.matrix(res4))
colnames(res4) <- c("treatment", "subclone", "proportion")
res4$drugSens <- "post-treatment"

rownames(count4) = treatment
count4 = melt(as.matrix(count4))
colnames(count4) <- c("treatment", "subclone", "count")
count4$drugSens <- "post-treatment"

p <- ggplot(res4, aes(x = treatment, y = proportion, fill = subclone)) +
  geom_col(position = "fill") +
  labs(
    y = "Proportion",
    fill = "") + 
  ggtitle("post-treatment") +
  scale_fill_manual(values = values1) +
  theme_bw() +
  theme(
    text = element_text(size = 24),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 21,
      color = "black"
    ),
    axis.text.y = element_text(
      size = 21,
      color = "black"
    ),
    axis.title.x = element_blank(),
    plot.title = element_text(hjust = 0.5, size = 28),
    panel.grid = element_blank()
  )
ggsave(paste0(sub_output_dir, "Figure2C_BRCA_CNV_postT.pdf"), width = 5, height = 5)
```

```{r}
# paired t tes
res <- res2 %>% 
  rbind.data.frame(res3) %>% 
  rbind.data.frame(res4)
write.csv(res, paste0(sub_output_dir, "Figure2C_cell_proportion_table.csv"), row.names = FALSE)

preR_A <- res$proportion[res$subclone == "A" & res$drugSens == "pre-resistant"]
preS_A <- res$proportion[res$subclone == "A" & res$drugSens == "pre-sensitive"]
t_test <- t.test(preR_A, preS_A, paired = TRUE)
t_test
```

```{r}
# proportion test
count <- count2 %>% 
  rbind.data.frame(count3) %>% 
  rbind.data.frame(count4)
write.csv(count, paste0(sub_output_dir, "Figure2C_cell_count_table.csv"), row.names = FALSE)

preR_A_control <- count$count[count$subclone == "A" & count$drugSens == "pre-resistant" & treatment == "Control"]
preR_control <- sum(count$count[count$drugSens == "pre-resistant" & treatment == "Control"])
preS_A_control <- count$count[count$subclone == "A" & count$drugSens == "pre-sensitive" & treatment == "Control"]
preS_control <- sum(count$count[count$drugSens == "pre-sensitive" & treatment == "Control"])
cat("preR vs preS control", "\n")

for (t in c("Carboplatin", "Olaparib", "NK")){
  cat(t, "\n")
  cat("pre-resistant", "\n")
  preR_A <- count$count[count$subclone == "A" & count$drugSens == "pre-resistant" & treatment == t]
  preR_treatment <- sum(count$count[count$drugSens == "pre-resistant" & treatment == t])
  p <- prop.test(x = c(preR_A, preR_A_control), n = c(preR_treatment, preR_control))
  print(p)
}

```

## D: Correlation between CNV subclone DEG and preR vs preS DEG

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure2/")

if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
conserve <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  select(ensembl, conserve_gene)
```

The DEG list from different replicates are merged for each treatment. The values for non-treated control samples are appended to each DEG list for each anti-cancer treatment. The final lists are saved in "./data/input/DEG_filter_by_bootstrap_permutation_in_singleton/"

```{r}
treatment <- c("Carbo", "Olaparib", "NK")
deg <- NULL
for (t in treatment){
  tmp <- read.csv(
  paste0(
    "./data/input/DEG_filter_by_bootstrap_permutation_in_singleton/prer_pres_",
    t,
    "_control_bootstrap_permutation_summary.csv")
  )
  tmp[[paste0(t, "_log2FC_preR_preS_adj")]] <- tmp$mean_log2FC_treat - tmp$mean_log2FC_ctrl
  tmp <- tmp[, c("ensembl", "symbol", paste0(t, "_log2FC_preR_preS_adj"))]
  if (is.null(deg)){
    deg <- tmp
  } else {
    deg <- left_join(deg, tmp, by = c("ensembl", "symbol"))
  }
}
```

DEG lists for each CNV groups vs. other cells are calculated on the CSC Puhti server with the batch job. The codes are available in the files: "./codes_for_batch_job/CNV_subclone_DEG.R", "./codes_for_batch_job/CNV_subclone_DEG.sh", and "./codes_for_batch_job/sub_CNV_subclone_DEG.sh"

```{r}
for (s in LETTERS[1:5]){
  CNV_deg <- read.csv(
    paste0(
      "./data/input/CNV_DEG/",
      s,
      "_vs_others_DEG.csv")
    ) %>% 
    select(X, avg_log2FC)
  colnames(CNV_deg) <- c("ensembl", paste0("log2FC_", s))
  deg <- deg %>% 
    left_join(CNV_deg, by = "ensembl")
}

deg <- deg %>% 
  left_join(conserve, by = "ensembl")
write.csv(deg, paste0(sub_output_dir, "Figure2D_log2FC_preR_preS_CNV_subclone.csv"), row.names = FALSE)

deg <- deg %>% 
  filter(conserve_gene)
```

```{r}
treatments <- c("NK", "Carbo", "Olaparib")
subclones <- c("A", "B", "C", "D", "E")
plot_table <- NULL
for (t in treatments){
  for (s in subclones){
    tmp <- cor.test(
      deg[[paste0(t, "_log2FC_preR_preS_adj")]],
      deg[[paste0("log2FC_", s)]]
    )
    df <- data.frame(
      treatment = case_when(
        t == "Carbo" ~ "Carboplatin", 
        TRUE ~ t
      ),
      subclone = s,
      correlation = tmp$estimate,
      CI_lower = tmp$conf.int[1],
      CI_upper = tmp$conf.int[2],
      p_value = tmp$p.value,
      sig = case_when(
        tmp$p.value < .05 & tmp$p.value >= .01 ~ "*", 
        tmp$p.value < .01 & tmp$p.value >= .001 ~ "**", 
        tmp$p.value < .001 ~ "***", 
        TRUE ~ ""
      ),
      n = sum(
        !is.na(deg[[paste0("log2FC_", s)]]) & 
        !is.na(deg[[paste0(t, "_log2FC_preR_preS_adj")]])
      )
    )
    if (is.null(plot_table)){
      plot_table <- df
    } else {
      plot_table <- rbind.data.frame(plot_table, df)
    }
  }
}
plot_table$sig_y <- plot_table$CI_upper + 0.05
plot_table$sig_y[plot_table$correlation < 0] <- plot_table$CI_lower[plot_table$correlation < 0] - 0.05
write.csv(plot_table, paste0(sub_output_dir, "Figure2D_Pearson_corelation_coefficient.csv"),
          row.names = FALSE)
plot_table$treatment <- factor(plot_table$treatment, levels = c("Carboplatin", "Olaparib", "NK"))
```

```{r}
color <- c("A" = "#DC0000FF", "B" = "#B09C85FF", "C" = "#00A087FF", "D" = "#91D1C2FF", "E" = "#3C5488FF")

ggplot(
  plot_table,
  aes(y = correlation, x = treatment, fill = subclone)
  ) +
  scale_fill_manual(
    values = color
  ) + 
  geom_bar(stat="identity", position="dodge") + 
  geom_errorbar(
    aes(ymin=CI_lower, ymax=CI_upper),
    position=position_dodge(width=.9),
    width=.15
  ) + 
  geom_text(
    aes(y=sig_y, label=sig),
    position=position_dodge(width=.9),
    size = 6
  ) + 
  labs(x="", y="Correlation coefficient", fill = "") +
  theme_classic() +
  theme(
     legend.key = element_blank(), 
     axis.text = element_text(size = 21, colour = "black"),
     axis.title = element_text(size = 21),
     legend.text = element_text(size = 21)
     )
ggsave(
  paste0(sub_output_dir, "Figure2D_Pearson_correlation_coefficients_color_by_subclones.pdf"),
  width = 10, height = 5)
```

## E: Representative gene ontologies enriched in the pre-resistance signatures as well as the signatures of subclones A and E

This figure is generated by Matias.

# Figure S3

## A: UMAP projection of pre-sensitive, pre-resistant, and post-treatment cells for each treatment

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS3/")

if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
data2 = data@assays$RNA@scale.data  # scale data is for dimension reduction or clustering
dim(data2) # 2000 77724

### UMAP visualization ----------------------------------------------------

data$drugSens[which(data$drugSens == "resistant")] = "post-treatment"
data2.umap = readRDS("./data/input/merged_KURAMOCHI_samples1_SCT_umap_2000_n50_scaleT.RDS")
index = which(data$drugSens %in% c("pre-sensitive", "pre-resistant","post-treatment"))
```

```{r}
index = which(data$drugSens %in% c("pre-sensitive", "pre-resistant","post-treatment"))
```

```{r}
# individual treatment conditions
for(i in unique(data$treatment)){
  for(j in c("pre-sensitive", "pre-resistant", "post-treatment")){
    pdf(
      paste(sub_output_dir, "FigureS3A_", paste(i, j, sep="_"),".pdf",sep=""),
      width = 3.7, height = 3.5, pointsize = 10)
    
    index1 = which(data$drugSens == j & data$treatment == i)
    # if(j == "pre-resistant") color = red
    # if(j == "pre-sensitive") color = blue
    # if(j == "post-treatment") color = brown
    if(i == "Olaparib") color = red
    if(i == "Carboplatin") color = blue
    if(i == "NK") color = brown
    if(i == "Control") color = green
    scatterplot(
      data2.umap$layout[index1, 1],
      data2.umap$layout[index1, 2], cex = 0.5, cex.main = 1.4, cex.axis = 1.2,
      col = color, main = paste(i,j),
      boxplots = FALSE, las = 1, smooth = FALSE, grid = TRUE,
      regLine = FALSE, ellipse = TRUE, pch = 20, 
      xlim = c(-10,10), ylim = c(-10,10),
      legend = list(title = "treatment", coords = "topleft"),
      # xlab = "UMAP 1", ylab = "UMAP 2"
      xlab = "", ylab = ""
      )
    dev.off()
    cat(i, ", ", j, ", ", sum(index1), "\n")
  }
  print(i)
  print(t.test(data2.umap$layout[which(data$drugSens == "pre-resistant" & data$treatment == i), 1], data2.umap$layout[which(data$drugSens == "pre-sensitive" & data$treatment == i),1]))
  print(t.test(data2.umap$layout[which(data$drugSens == "pre-resistant" & data$treatment == i), 2], data2.umap$layout[which(data$drugSens == "pre-sensitive" & data$treatment == i),2]))
}
```

## B: Cumulative distribution

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS3/")

if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS") 

# options(scipen = 999)
# memory.limit(size = 54000)
data2 = data@assays$RNA@data

sink(paste0(sub_output_dir, "FigureS3B_shannon.txt"))
res = list()
k = 1
for (i in unique(data$sample)){
  for (j in unique(data$time_point)){
    index = intersect(grep(i, data@meta.data$prer_r_group), which(data$time_point == j))
    tmp = data.frame(table(data@meta.data$prer_r_group[index]))
    shannon = diversity(tmp$Freq)
    print(c(gsub('[0-9]+', '', i), i,j,shannon))
    tmp$sample = i
    tmp$time_point = j
    tmp$treatment = gsub('[0-9]+', '', i)
    res[[k]] = tmp
    k = k + 1
  }
}
sink()
```

Reference: https://rgraphs.com/how-to-create-a-cumulative-frequency-graph-in-r/

```{r}
tmp = do.call(rbind, res)
tmp$treatment[tmp$treatment == "Carbo"] <- "Carboplatin"
tmp$treatment[tmp$treatment == "ctrl"] <- "Control"
tmp$treatment = factor(tmp$treatment, levels = c("Carboplatin", "Olaparib", "NK", "Control"))
tmp$time_point[tmp$time_point == "BeforeTreatment"] <- "pre-treatment"
tmp$time_point[tmp$time_point == "AfterTreatment"] <- "post-treatment"
tmp$time_point = factor(tmp$time_point, levels = c("pre-treatment", "post-treatment"))
values1 = c(blue, red, brown, green)

fig1 = ggplot(tmp, aes(Freq, y = 1 - ..y.., color = time_point)) +
  stat_ecdf(geom = "step", size = 1) + 
  labs(
    x = "Clone size",
    y = "Cumulative proportion",
    color = "") +
  scale_y_continuous(trans = 'log10') + scale_x_log10() +
  theme_classic() +
  theme(
    text = element_text(size = 20),
    axis.title = element_text(size = 20),
    axis.text = element_text(color = "black", size = 20),
    legend.text = element_text(size = 20),
    legend.title = element_blank(),
    # legend.position=c(.7,.9)
  ) +
  scale_color_manual(values = values1)
ggsave(
  paste0(sub_output_dir, "FigureS3C_cumulate_all.pdf"),
  fig1, width = 9, height = 5
)
```

## C: Shannon diversity

```{r}
shannon = read.table(paste0(sub_output_dir, "FigureS3B_shannon.txt"))
names(shannon) = c("V1", "Treatment","sample","time_point","value")
shannon$Treatment[shannon$Treatment == "Carbo"] <- "Carboplatin"
shannon$Treatment[shannon$Treatment == "ctrl"] <- "Control"
shannon$Treatment = factor(shannon$Treatment, levels = c("Carboplatin", "Olaparib", "NK", "Control"))
shannon$time_point[shannon$time_point == "BeforeTreatment"] <- "pre-treatment"
shannon$time_point[shannon$time_point == "AfterTreatment"] <- "post-treatment"
shannon$time_point = factor(shannon$time_point, levels = c("pre-treatment", "post-treatment"))
values1 = c(blue, red, brown, green)

p <- ggplot(shannon, aes(x = time_point, y = value)) +
  geom_boxplot() + 
  geom_point(aes(x = time_point, color = Treatment), size = 4) +
  geom_line(aes(group = sample, color = Treatment)) +
  xlab("") + 
  ylab("Shannon diversity index") +
  theme_classic() +
  theme(
    text = element_text(size = 20),
    axis.title = element_text(size = 20),
    axis.text = element_text(color = "black", size = 20),
    legend.text = element_text(size = 20),
    legend.title = element_blank(),
    # legend.position = c(0.9, 0.9),
    legend.background = element_blank()
  ) +
  scale_color_manual(values = values1)
ggsave(
  paste0(sub_output_dir, "FiguerS3B_shannon.pdf"),
  p,
  width = 9, height = 5
  )

# Paired t test

pre <- shannon$value[shannon$time_point == "pre-treatment"]
post <- shannon$value[shannon$time_point == "post-treatment"]
t_test <- t.test(pre, post, paired = TRUE)
t_test
```

# Figure 3: Functional validations/interpretations

## B: TCVA BRCAness/HRD survival plot

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure3/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

Load lineage tracing data

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS") # Shuyu update on 1.4.2022 after FindVariableFeatures and Scale, check test.R in the server
sister_conserve_gene <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  filter(conserve_gene) %>% 
  select(ensembl) %>% 
  unlist()
data2 = data@assays$RNA@data[rownames(data) %in% sister_conserve_gene, ]
dim(data2)
unique(data$drugSens)
unique(data$sample)
unique(data$treatment)
unique(data$time_point)
```

load TCGA data

The clinical data for patients from TCGA is saved in "./data/input/BRCAness/data_clinical_patient.txt". The clinical data for sequencing samples from TCGA is saved in "./data/input/BRCAness/data_clinical_sample.txt". The gene expression matrix after deconvolution with single-cell RNA sequencing data are saved in "./data/input/BRCAness/EOC_deconv_TCGA_expr_mat.RDS"

```{r}
data_clinical_patient = read.delim("./data/input/BRCAness/data_clinical_patient.txt", skip = 4)
data_clinical_sample = read.delim("./data/input/BRCAness/data_clinical_sample.txt", skip = 4)
data_mrna = readRDS("./data/input/BRCAness/EOC_deconv_TCGA_expr_mat.RDS")
colnames(data_mrna) = gsub("\\.","\\-", colnames(data_mrna))
colnames(data_mrna) = substr(colnames(data_mrna), 1, 15)
data_mrna <- data_mrna[apply(data_mrna, 1, function(x){
  sum(is.na(x))
}) < 300, ]

identical(data_clinical_patient$PATIENT_ID, data_clinical_sample$PATIENT_ID) # data aligned already
table(data_clinical_sample$GRADE)
high_grade = grepl("G2|G3|G4", data_clinical_sample$GRADE)

data_clinical_sample = data_clinical_sample[high_grade, ]
data_clinical_patient = data_clinical_patient[high_grade, ]

patients = names(data_mrna) # 307 patients
time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]
table(status, exclude = NULL)
status[which(status == "")] = NA

length(which(is.na(time)==F)) # 270 has survival
length(which(is.na(status)==F)) # 271 has status
```

The DEG list for HRD vs HRP groups from deconvoluted TCGA data is saved in "./data/input/BRCAness/TCGA_deconvoluted_HRD_DEGs.tsv".

```{r}
tcga = read.csv("./data/input/BRCAness/TCGA_deconvoluted_HRD_DEGs.tsv", sep = "\t")
tcga[which(rownames(tcga) == "BRCA1"), ] # should be negative logFC
tcga$ensembl = data@misc$gene_meta$ensembl[match(rownames(tcga), data@misc$gene_meta$symbol)]
tcga$symbol = rownames(tcga)
length(which(is.na(tcga$ensembl)==T)) # 383 genes have no emsembl
```

### Test for good p value

```{r}
p_threshold = c(1*10^-(2:4), 5*10^-(2:4))
fc_threshold = 0
conserve_gene_only <- FALSE

res <- mat.or.vec(length(p_threshold), 3)
colnames(res) <- c("p_threshold", "n_gene", "survival_p")

for (i in 1:length(p_threshold)){
  data_mrna2 <- data_mrna
  if (conserve_gene_only){
    tcga_test <- filter(tcga, ensembl %in% sister_conserve_gene)
  } else {
    tcga_test <- tcga
  }
  
  res[i, 1] <- p_threshold[i]
  index1 = which(tcga_test$logFC <= -fc_threshold & tcga_test$adj.P.Val <= p_threshold[i]) # down-regulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(tcga_test$logFC > fc_threshold & tcga_test$adj.P.Val <= p_threshold[i]) # up-regulated genes when BRCA1 is knocked out
  index2_symbol = tcga_test$symbol[index2]
  
  index = c(index1, index2)
  length(index)
  tcga_test = tcga_test[index, ]
  # tcga_test$symbol
  
  tcga_test = tcga_test[match(rownames(data_mrna2), tcga_test$symbol), ]
  
  res[i, 2] <- nrow(na.omit(tcga_test))
  
  # reverse index2 gene expressions
  data_mrna2[which(rownames(rownames(data_mrna2)) %in% index2_symbol), ] = -data_mrna2[which(rownames(rownames(data_mrna2)) %in% index2_symbol), ]
  
  # gene selection
  index = which(is.na(tcga_test$symbol) == F)
  res1 = c()
  for (j in 1:length(patients)){
    cat('\r', j)
    res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
    res1[j] = weighted.mean(res.tmp[index], w = abs(tcga_test$logFC[index]), na.rm = T) # rank of the BRCAness genes
    # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362$avg_log2FC[index])*(-log10(cov362$p_val_adj[index]))) # rank of the BRCAness genes
    
  }
  
  group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
  # fiveyear = which(time < 60)
  plot_data = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
  fit <- survfit(Surv(time, status) ~ group, data = plot_data)
  fit.coxph = coxph(Surv(time, status) ~ group, data = plot_data)
  s <- summary(fit.coxph)
  res[i, 3] <- s$logtest[3]
}

write.csv(
  res,
  paste0(sub_output_dir, "Figure3B_BRCA_survival_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

### Survival plot with selected parameters

```{r}
conserve_gene_only <- FALSE

if (conserve_gene_only) {
  selected_p_threshold = 0.005
  selected_fc_threshold = 0
} else {
  selected_p_threshold = 0.005
  selected_fc_threshold = 0
}

if (conserve_gene_only){
  tcga_test <- filter(tcga, ensembl %in% sister_conserve_gene)
} else {
  tcga_test <- tcga
}

data_mrna2 <- data_mrna
index1 = which(tcga_test$logFC <= -selected_fc_threshold & tcga_test$adj.P.Val <= selected_p_threshold) # down-regulated genes when BRCA1 is knocked out (BRCA-)
index2 = which(tcga_test$logFC > selected_fc_threshold & tcga_test$adj.P.Val <= selected_p_threshold) # up-regulated genes when BRCA1 is knocked out
index2_symbol = tcga_test$symbol[index2]

index = c(index1, index2)
length(index)
tcga_test = tcga_test[index, ]
# tcga_test$symbol

tcga_test = tcga_test[match(rownames(data_mrna2), tcga_test$symbol), ]

# reverse index2 gene expressions
data_mrna2[which(rownames(rownames(data_mrna2)) %in% index2_symbol), ] = -data_mrna2[which(rownames(rownames(data_mrna2)) %in% index2_symbol), ]

# gene selection
index = which(is.na(tcga_test$symbol) == F)
res1 = c()
for (j in 1:length(patients)){
  cat('\r', j)
  res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
  res1[j] = weighted.mean(res.tmp[index], w = abs(tcga_test$logFC[index]), na.rm = T) # rank of the BRCAness genes
  # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362$avg_log2FC[index])*(-log10(cov362$p_val_adj[index]))) # rank of the BRCAness genes
  
}

group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
# fiveyear = which(time < 60)
plot_data = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
fit <- survfit(Surv(time, status) ~ group, data = plot_data)
fit.coxph = coxph(Surv(time, status) ~ group, data = plot_data)

s <- summary(fit.coxph)
  
dim(tcga_test)
concordance(fit.coxph)$concordance

pdf(paste0(sub_output_dir, "Figure3B_threshold_", selected_p_threshold, "_", ifelse(conserve_gene_only, "BRCA_TCGA_OS_conserve_gene.pdf", "BRCA_TCGA_OS.pdf")), width = 4, height = 5, pointsize = 20)
ggsurvplot(
  fit,
  data = plot_data,
  palette = c(red, blue),
  risk.table = T,
  legend.title = "BRCAness (TCGA)",# - undeconvoluted",
  legend.labs = c("high","low"),
  xlab = "Time (months)", ylab = "Overall survival",
  ggtheme = theme_survminer(
    base_size = 12,
    font.main = c(12, "plain", "black"),
    font.x = c(12, "plain", "black"),
    font.y = c(12, "plain", "black"),
    font.caption = c(12, "plain", "black"),
    font.legend = c(12, "plain", "black"),
    font.tickslab = c(12, "plain", "black"),
    font.submain = c(12, "plain", "black")
  ),
  pval = signif(s$logtest[3], 2),
  pval.coord = c(100, 0.75)
)+
  # annotate("text", x = 150, y =0.75, label = paste0("P = ", round(s$logtest[3], 2))) +
  guides(colour = guide_legend(nrow = 2))
dev.off()
```

## D: Box plot for correlation coefficients of BRCAness signature vs. pre-sensitivity signatures (seurat logFC)

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure3/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}

lseq <- function(from=1, to=100000, length.out=6) {
  # logarithmic spaced sequence
  # blatantly stolen from library("emdbook"), because need only this
  exp(seq(log(from), log(to), length.out = length.out))
}
treatments <- c("Olaparib", "Carboplatin", "NK")
```

### load data

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
log2FC_table <- read.csv("./data/output/Figure2/Figure2D_log2FC_preR_preS_CNV_subclone.csv")
```

At vs BT logFC adjust growth control

```{r}
# for (t in treatments){
#   if (t == "Carboplatin"){
#     t <- "Carbo"
#   }
#   tmp <- read.csv(paste0("./data/at_bt_", t, "_control_wilcox_summary.csv"))
#   tmp[[paste0(t, "_log2FC_at_bt_adj")]] <- tmp$mean_log2FC_treat - tmp$mean_log2FC_ctrl
#   tmp <- select(tmp, ensembl, symbol, paste0(t, "_log2FC_at_bt_adj"))
#   log2FC_table <- log2FC_table %>% 
#     left_join(tmp, by = c("ensembl", "symbol"))
# }
```

TCGA BRCAness signature

```{r}
tcga = read.csv("./data/input/BRCAness/TCGA_deconvoluted_HRD_DEGs.tsv", sep = "\t")
tcga[which(rownames(tcga) == "BRCA1"), ] # should be negative logFC
tcga$ensembl = data@misc$gene_meta$ensembl[match(rownames(tcga), data@misc$gene_meta$symbol)]
length(which(is.na(tcga$ensembl)==T)) # 383 genes have no emsembl

summary(tcga$adj.P.Val)
```

```{r}
log2FC_table <- left_join(
  log2FC_table,
  tcga %>% 
    rownames_to_column("symbol") %>% 
    select(ensembl, symbol, BRCAness_tcga_logFC = logFC, BRCAness_tcga_p_adj = adj.P.Val),
  by = c("ensembl", "symbol")
)
```

COV362 BRCAness signature

```{r}
cov362 <- read.csv("./data/input/BRCAness/deg_COV362.csv")
cov362[which(cov362$symbol == "BRCA1"),] # should be negative logFC
```

```{r}
log2FC_table <- left_join(
  log2FC_table,
  cov362 %>% 
    select(ensembl, symbol, BRCAness_cov_logFC = avg_log2FC, BRCAness_cov_p_adj = p_val_adj),
  by = c("ensembl", "symbol")
)
```

```{r}
write.csv(log2FC_table, paste0(sub_output_dir, "Figure3D_logFC_table_for_plots.csv"), row.names = FALSE)
```

TCGA BRCAness signature vs. CNV subclone signature

```{r}
data <- read.csv(paste0(sub_output_dir, "Figure3D_logFC_table_for_plots.csv")) %>%
  filter(BRCAness_tcga_p_adj <= 0.01) %>% 
  select("log2FC_A", "log2FC_B", "log2FC_C", "log2FC_D",
         "log2FC_E", "BRCAness_tcga_logFC", "BRCAness_cov_logFC")

cor(data, method = "spearman", use = "complete")
```

COV362 signature vs. CNV subclone signature

```{r}
data <- read.csv(paste0(sub_output_dir, "Figure3D_logFC_table_for_plots.csv")) %>%
  filter(BRCAness_cov_p_adj <= 1 * 10^-8) %>% 
  select("log2FC_A", "log2FC_B", "log2FC_C", "log2FC_D",
         "log2FC_E", "BRCAness_tcga_logFC", "BRCAness_cov_logFC")

cor(data, method = "spearman", use = "complete")
```

```{r}
conserve_gene_only <- TRUE

if (conserve_gene_only){
  plot_table <- filter(log2FC_table, conserve_gene)
} else {
  plot_table <- log2FC_table
}
```

### Test for good P threshold

#### TCGA signature

```{r}
min_p = 0.005 # around 200 sister conserved BRCAness genes 
max_p = 0.05

p_threshold = c(lseq(min_p, max_p, 20), 1*10^-(2:4), 5*10^-(2:4))

res = mat.or.vec(length(p_threshold), 5)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK")

for(i in 1:length(p_threshold)){
  table_tests <- plot_table %>% 
    filter(BRCAness_tcga_p_adj <= p_threshold[i]) %>% 
    select(starts_with("BRCAness_tcga"), ends_with("log2FC_preR_preS_adj")) %>% 
    na.omit()
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  res[i, 2] <- nrow(table_tests)
  res[i, 3] <- cor(table_tests$BRCAness_tcga_logFC, -table_tests$Olaparib_log2FC_preR_preS_adj, method = "spearman", use = "complete")
  res[i, 4] <- cor(table_tests$BRCAness_tcga_logFC, -table_tests$Carbo_log2FC_preR_preS_adj, method = "spearman", use = "complete")
  res[i, 5] <- cor(table_tests$BRCAness_tcga_logFC, -table_tests$NK_log2FC_preR_preS_adj, method = "spearman", use = "complete")
}
write.csv(
  res,
  paste0(sub_output_dir,"Figure3D_BRCA_preRpreS_cor_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

#### COV362 signature

```{r}
min_p = 5 * 10^-16 # around 200 sister concordant BRCAness genes
max_p = 0.05
p_threshold <- c(lseq(min_p, max_p, 20), 1 * 10 ^ -c(2:14), 5 * 10 ^ -c(2:14))

res = mat.or.vec(length(p_threshold), 5)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK")

for(i in 1:length(p_threshold)){
  table_tests <- plot_table %>% 
    filter(BRCAness_cov_p_adj <= p_threshold[i]) %>% 
    select(starts_with("BRCAness_cov"), ends_with("log2FC_preR_preS_adj")) %>% 
    na.omit()
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  res[i, 2] <- nrow(table_tests)
  res[i, 3] <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Olaparib_log2FC_preR_preS_adj, method = "spearman", use = "complete")
  res[i, 4] <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Carbo_log2FC_preR_preS_adj, method = "spearman", use = "complete")
  res[i, 5] <- cor(table_tests$BRCAness_cov_logFC, -table_tests$NK_log2FC_preR_preS_adj, method = "spearman", use = "complete")
}
write.csv(
  res,
  paste0(sub_output_dir,"Figure3D_BRCA_preRpreS_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

### Box plot preS/preR

```{r}
res1 <- read.csv(
  paste0(sub_output_dir,"Figure3D_BRCA_preRpreS_cor_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "p_threshold")
res1 <- res1[1:20, ]
res1.df = reshape2::melt(res1, id.vars = "p_threshold", variable.name = "treatment")
res1.df$group = "TCGA"

res2 <- read.csv(
  paste0(sub_output_dir,"Figure3D_BRCA_preRpreS_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "p_threshold")
res2 <- res2[1:20, ]
res2.df = reshape2::melt(res2, id.vars = "p_threshold", variable.name = "treatment")
res2.df$group = "COV362"

res.df = rbind.data.frame(res1.df, res2.df)
res.df$group <- factor(res.df$group, levels = c("TCGA", "COV362"))
res.df$treatment <- factor(res.df$treatment, levels = c("Carboplatin", "Olaparib", "NK"))

ggplot(
  res.df[res.df$treatment !="Control",],
  aes(x = group, y = value, fill = treatment)
) +
  geom_boxplot() +
  labs(x = "", y = "Correlation coefficient", fill = "")+
  theme_bw() + 
  theme(
    text = element_text(size = 20),
    axis.title = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 20, color = "black"),
    legend.text = element_text(size = 20, color = "black")
  ) +
  scale_fill_manual(values = c(blue, red, brown))
ggsave(
  paste0(
    sub_output_dir,
    "Figure3D_BRCA_preRpreS_cor_cov_tcga",
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 9, height = 5
)
```

## E: NK killing rate on COV362 and KURAMOCHI

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure3/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

The tables for cell signals after treated by NK cells are from the experiments performed by Jun Dai. The background signal has been removed from the data. They are saved in "

### COV362

```{r}
table <- read_xlsx(
  "./data/input/NK_cell_killing_experiment/COV362-NK-20220417.xlsx", sheet = 2,
  range = "B1:Z7")
table[1, ] <- t(rep(c("0", "0.5", "1", "2", "3"), each = 5))
table <- t(table)
colnames(table) <- c("rate", "sample", "rep1", "rep2", "rep3", "rep4")
table <- table %>% 
  as.data.frame() %>% 
  filter(sample %in% c("COV362-GFP-M", "COV362-B1-GFP-M")) %>% 
  pivot_longer(cols = starts_with("rep"), values_to = "read", names_to = "replicates") %>% 
  na.omit()
table$read <- as.numeric(table$read)
NC_COV362_GFP_M <- mean(table$read[table$sample == "COV362-GFP-M" & table$rate == "0"])
NC_COV362_B1_GFP_M <- mean(table$read[table$sample == "COV362-B1-GFP-M" & table$rate == "0"])
table <- table %>% 
  mutate(
    cell_viability = ifelse(sample == "COV362-GFP-M", read/NC_COV362_GFP_M, read/NC_COV362_B1_GFP_M) * 100
  )
write.csv(table, paste0(sub_output_dir, "Figure3E_cell_viablility_COV362.csv"))
```

```{r}
plot_table <- table %>%
  dplyr::group_by(rate, sample) %>% 
  dplyr::summarise(
    mean = mean(cell_viability),
    sd = sd(cell_viability)
  )
plot_table$sample[plot_table$sample == "COV362-GFP-M"] <- "COV362-control"
plot_table$sample[plot_table$sample == "COV362-B1-GFP-M"] <- "COV362-BRCA1 restored"
plot_table$sample <- factor(plot_table$sample, levels = c("COV362-control", "COV362-BRCA1 restored"))
col = c("#B36981", "#659698")
p <- ggplot(plot_table, aes(x = rate, y = mean, fill = sample)) +
  geom_bar(
    stat="identity",
    color = "white",
    position=position_dodge()
  ) +
  geom_errorbar(
    aes(ymin=mean-sd, ymax=mean+sd),
    width=.2,
    position=position_dodge(.9)
  ) +
  scale_fill_manual(values = col) +
  labs(
    x = "NK : target cell",
    y = "Cell viability (%)"
  ) +
  theme_classic() +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 15),
    text = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 15, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "Figure3E_COV362_NK_killing.pdf"),
  width = 4.5, height = 4
)
```

### KURAMOCHI

```{r}
table <- read_xlsx(
    "./data/input/NK_cell_killing_experiment/Kura-NK-20220417.xlsx", sheet = 2,
  range = "B3:Z9")
table[1, ] <- t(rep(c("0", "0.5", "1", "2", "3"), each = 5))
table <- t(table)
colnames(table) <- c("rate", "sample", "rep1", "rep2", "rep3", "rep4")
table <- table %>% 
  as.data.frame() %>% 
  filter(sample %in% c("Kura-GFP-H", "Kura-B2-GFP-H")) %>% 
  pivot_longer(cols = starts_with("rep"), values_to = "read", names_to = "replicates") %>% 
  na.omit()
table$read <- as.numeric(table$read)
NC_Kura_GFP_M <- mean(table$read[table$sample == "Kura-GFP-H" & table$rate == "0"])
NC_Kura_B1_GFP_M <- mean(table$read[table$sample == "Kura-B2-GFP-H" & table$rate == "0"])
table <- table %>% 
  mutate(
    cell_viability = ifelse(sample == "Kura-GFP-H", read/NC_Kura_GFP_M, read/NC_Kura_B1_GFP_M) * 100
  )
write.csv(table, paste0(sub_output_dir, "Figure3E_cell_viablility_Kura.csv"))
```

```{r}
plot_table <- table %>%
  dplyr::group_by(rate, sample) %>% 
  dplyr::summarise(
    mean = mean(cell_viability),
    sd = sd(cell_viability)
  )
plot_table$sample[plot_table$sample == "Kura-GFP-H"] <- "KURAMOCHI-control"
plot_table$sample[plot_table$sample == "Kura-B2-GFP-H"] <- "KURAMOCHI-BRCA2 restored"
plot_table$sample <- factor(plot_table$sample, levels = c("KURAMOCHI-control", "KURAMOCHI-BRCA2 restored"))
col = c("#B36981", "#659698")
p <- ggplot(plot_table, aes(x = rate, y = mean, fill = sample)) +
  geom_bar(
    stat="identity",
    color = "white",
    position=position_dodge()
  ) +
  geom_errorbar(
    aes(ymin=mean-sd, ymax=mean+sd),
    width=.2,
    position=position_dodge(.9)
  ) +
  scale_fill_manual(values = col) +
  labs(
    x = "NK : target cell",
    y = "Cell viability (%)"
  ) +
  theme_classic() +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 15),
    text = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 15, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "Figure3E_KURAMOCHI_NK_killing.pdf"),
  width = 4.5, height = 4
)
```

## F: Hazard ratios of overall survival for the interaction between NK transcriptional signature, and patients’ HRD/HRP status

This figure is generated by Matias.

# Figure S4

## A: Heatmap showing the subclonal CNV profiles of the carboplatin samples inferred from their scRNA-seq data.

This figure is generated by Matias.

## B: Sankey plot for CNV subgroup transition

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS4/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
data <- data@meta.data
```

```{r}
# Merge replicates
samples <- c("Carboplatin", "Control", "NK", "Olaparib")
for (s in samples) {
  cat(s, "\n")
  data_tmp <- data %>% 
    filter(treatment == s)

  # Sankey diagram
  CNV_subclone <- data_tmp %>% 
    select(
      drugSens, sisters, prer_r_group, CNV_subclone, time_point, treatment) %>% 
    filter(!is.na(prer_r_group))
  CNV_subclone$CNV_subclone <- as.character(CNV_subclone$CNV_subclone)
  
  prer_r_group <- unique(CNV_subclone$prer_r_group)
  
  ## sankeyNetwork ggplot
  
  links <- data.frame(
    prer_r_group = prer_r_group,
    source = NA,
    target = NA
  )
  for (i in prer_r_group){
    # subclone in BT
    subclone_bt <- CNV_subclone %>% 
      filter(time_point == "BeforeTreatment" & prer_r_group == i) %>% 
      select(CNV_subclone) %>% 
      unique() %>% 
      unlist()
    if (length(subclone_bt) == 0 ){
      links$source[links$prer_r_group == i] <- NA
    } else if (length(subclone_bt) > 1) {
      links$source[links$prer_r_group == i] <- paste0(sort(subclone_bt), collapse = "")
    } else {
      links$source[links$prer_r_group == i] <- paste0(rep(subclone_bt, sum(links$prer_r_group == i)))
    }
    # subclone in AT
    subclone_at <- CNV_subclone %>% 
      filter(time_point == "AfterTreatment" & prer_r_group == i) %>% 
      select(CNV_subclone) %>% 
      unique() %>% 
      unlist()
    if (length(subclone_at) == 0 ){
      links$target[links$prer_r_group == i] <- NA
    } else if (length(subclone_at) > 1) {
      links$target[links$prer_r_group == i] <- paste0(sort(subclone_at), collapse = "")
    } else {
      links$target[links$prer_r_group == i] <- paste0(rep(subclone_at, sum(links$prer_r_group == i)))
    }
  }
  cols <- c(
    "A" = "#DC0000FF", "B" = "#B09C85FF", "C" = "#00A087FF", "BC" = "#B09C85FF",
    "AB" = "#F39B7FFF", "D" = "#91D1C2FF",  "E" = "#3C5488FF", "AD" = "#4DBBD5FF",
    "AC" = "#8491B4FF", "ACD" = "#E64B35FF")
  # remove NA subclone
  links <- na.omit(links)
  df <- links %>%
    select(`pre-treatment` = "source", `post-treatment` = "target") %>% 
    make_long(`pre-treatment`, `post-treatment`)
  ggplot(
    df,
    aes(x = x, 
        next_x = next_x, 
        node = node, 
        next_node = next_node,
        fill = factor(node),
        label = node)) +
  geom_sankey(flow.alpha = 0.5) +
  scale_fill_manual(
    values = cols) +
  geom_sankey_text(
    mapping = aes(x = ifelse(as.numeric(x) == 1, 0.9, 2.25)),
    size = 5,
    color = 1,
    show.legend = FALSE,
    width = 0.01,
    hjust = 1) +
  theme_sankey(base_size = 20) +
  ggtitle(s) +
  theme(
    legend.position = "none",
    plot.title = element_text(hjust = 0.5, size = 20, face = "bold"),
    text = element_text(size = 20),
    axis.text = element_text(size = 20, color = "black"),
    axis.title.x = element_blank()
  )
  ggsave(paste0(sub_output_dir, "FigureS4B_CNV_subclone_transfer_", s, ".pdf"), width = 6, height = 6)
  print(nrow(links))
}
```

# Figure S5 cell-cell distances between drug-sensitive groups within CNV subclones

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS5/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
data2 = data@assays$RNA@scale.data  # scale data is for dimension reduction or clustering
dim(data2) # 2000 77724
treatments <- c("Olaparib", "Carboplatin", "NK", "Control")
subclone <- LETTERS[1:5]
```

Calculate cell-cell distance

It costs a lot of time to calculate the cell-cell distances. Here I comment out the codes for calculation and directly read the previously calculated matrix. 

```{r eval = FALSE}
# data2.dist = distances(t(data2), normalize = "mahalanobize")
# saveRDS(data2.dist, file = paste0(sub_output_dir, "all_cell_distance.RDS"))
data2.dist <- readRDS(paste0(sub_output_dir, "all_cell_distance.RDS"))
```

```{r eval = FALSE}
res <- NULL
for (t in treatments){
  tmp_res <- vector(mode = "list", length = 5)
  names(tmp_res) <- subclone
  for(i in subclone){
    # cat("\r",i)
    preS = which(
      data$drugSens == "pre-sensitive" &
      data$CNV_subclone == i &
      data$treatment == t)
    preR = which(
      data$drugSens == "pre-resistant" &
      data$CNV_subclone == i &
      data$treatment == t
    )
    posT = which(
      data$drugSens == "resistant" &
      data$CNV_subclone == i &
      data$treatment == t
    )
    tmp = data2.dist[preS, preS]
    distS = tmp[upper.tri(tmp, diag = F)]
    tmp = data2.dist[preR, preR]
    distR = tmp[upper.tri(tmp, diag = F)]
    tmp = data2.dist[posT, posT]
    distT = tmp[upper.tri(tmp, diag = F)]
    tmp = cbind(dist = c(distS, distR, distT),
    drugSens = c(rep(1, length(distS)), rep(2, length(distR)), rep(3, length(distT))),
    subclone = i)
    if (nrow(tmp) > 100000){
      set.seed(123)
      tmp_small = tmp[sample(1:nrow(tmp), 100000),]
    } else {
      tmp_small <- tmp
    }
    tmp_small = data.frame(tmp_small)
    tmp_small$drugSens[which(tmp_small$drugSens == 1)] = "pre-sensitive"
    tmp_small$drugSens[which(tmp_small$drugSens == 2)] = "pre-resistant"
    tmp_small$drugSens[which(tmp_small$drugSens == 3)] = "post-treatment"
    tmp_res[[i]] = tmp_small
  }
  tmp_res = do.call("rbind", tmp_res)
  tmp_res$drugSens <- factor(tmp_res$drugSens, levels = c("pre-sensitive", "pre-resistant", "post-treatment"))
  tmp_res$subclone <- factor(tmp_res$subclone, levels = c("A","B","C","D","E"))
  tmp_res$dist <- as.numeric(tmp_res$dist)
  tmp_res$treatment <- t
  if (is.null(res)){
    res <- tmp_res
  } else {
    res <- rbind.data.frame(res, tmp_res)
  }
}
View(res)

write.csv(res, paste0(sub_output_dir, "cell_distance_table_by_treatment_for_plot.csv"))
```

```{r}
res <- read.csv(paste0(sub_output_dir, "cell_distance_table_by_treatment_for_plot.csv"))
color = c("pre-sensitive" = "snow2", "pre-resistant" = "snow4", "post-treatment" = "slategray4")
res$drugSens <- factor(res$drugSens, levels = c("pre-sensitive", "pre-resistant", "post-treatment"))
res$subclone <- factor(res$subclone, levels = c("A","B","C","D","E"))
res <- res %>% 
  filter(drugSens != "post-treatment")
for (t in treatments) {
  tmp_plot <- res[res$treatment == t,]
  test <- rep(NA, 5)
  names(test) <- subclone
  for (i in subclone){
    tmp_test <- t.test(
        tmp_plot$dist[tmp_plot$drugSens == "pre-resistant" & tmp_plot$subclone == i],
        tmp_plot$dist[tmp_plot$drugSens == "pre-sensitive" & tmp_plot$subclone == i]
      )
    test[i] <- tmp_test$p.value
  }
  test <- data.frame(
    subclone = names(test),
    p_value = test,
    sig = case_when(
    test < .1 & test >= .05 ~ "*", 
    test < .05 & test >= .01 ~ "**", 
    test < .01 ~ "***", 
    TRUE ~ ""
    ),
    x = 1:5 - 0.1,
    y = rep(90, 5)
  )
  fig <- ggplot(
    data = tmp_plot,
    aes(y = dist, x = subclone, group = interaction(subclone, drugSens))
    ) +
    geom_violin(bw = 2, width = 0.8, aes(fill = drugSens)) +
    geom_boxplot(
      position = position_dodge(width = 0.8),
      fill = "white", width = 0.1,
      outlier.shape = NA,# alpha = 1
    ) +
    scale_y_continuous(name = "Normalized distance", limits = c(30, 90)) +  # There are 4.71% cell pairs has the distance larger than 100
    scale_x_discrete(name = "Subclone") +
    scale_fill_manual(values = color) + 
    # annotate("text", x = test$x, y = test$y, label = test$sig) +
    ggtitle(t) +
    theme_bw() +
    theme(
      text = element_text(size = 20, colour = "black"),
      plot.title = element_text(hjust = 0.5, size = 20),
      axis.text = element_text(size = 18, color = "black"),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      legend.position = "none"
    )
  ggsave(
    paste0(sub_output_dir, t, "_dist_cnv.pdf"),
    fig, width = 9, height = 4
  )
  ggsave(
    paste0(sub_output_dir, t, "_dist_cnv.svg"),
    fig, width = 9, height = 4
  )
}
```

# Figure S6

## A: COV362 survival plot

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS6/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

```{r}
data_clinical_patient = read.delim("./data/input/BRCAness/data_clinical_patient.txt", skip = 4)
data_clinical_sample = read.delim("./data/input/BRCAness/data_clinical_sample.txt", skip = 4)
data_mrna = readRDS("./data/input/BRCAness/EOC_deconv_TCGA_expr_mat.RDS")
colnames(data_mrna) = gsub("\\.","\\-", colnames(data_mrna))
colnames(data_mrna) = substr(colnames(data_mrna), 1, 15)
data_mrna <- data_mrna[apply(data_mrna, 1, function(x){
  sum(is.na(x))
}) < 300, ]

identical(data_clinical_patient$PATIENT_ID, data_clinical_sample$PATIENT_ID) # data aligned already
table(data_clinical_sample$GRADE)
high_grade = grepl("G2|G3|G4", data_clinical_sample$GRADE)

data_clinical_sample = data_clinical_sample[high_grade, ]
data_clinical_patient = data_clinical_patient[high_grade, ]

patients = names(data_mrna) # 307 patients
time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]
table(status, exclude = NULL)
status[which(status == "")] = NA

length(which(is.na(time)==F)) # 270 has survival
length(which(is.na(status)==F)) # 271 has status
```

#### Test for good p value

```{r}
cov362 <- read.csv("./data/input/BRCAness/deg_COV362.csv")

p_threshold = c(1*10^-(2:14), 5*10^-(2:14))
fc_threshold = 0
conserve_gene_only <- FALSE

res <- mat.or.vec(length(p_threshold), 3)
colnames(res) <- c("p_threshold", "n_gene", "survival_p")

for (i in 1:length(p_threshold)){
  data_mrna2 <- data_mrna
  if (conserve_gene_only){
    cov362_test <- filter(cov362, ensembl %in% sister_conserve_gene)
  } else {
    cov362_test <- cov362
  }
  res[i, 1] <- p_threshold[i]
  
  index1 = which(cov362_test$avg_log2FC <= -fc_threshold & cov362_test$p_val_adj <= p_threshold[i]) # downregulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(cov362_test$avg_log2FC > fc_threshold & cov362_test$p_val_adj <= p_threshold[i]) # upregulated genes when BRCA1 is knocked out
  index2_symbol = cov362_test$symbol[index2]
  
  index = c(index1, index2)
  length(index)
  cov362_test = cov362_test[index, ]
  # cov362_test$symbol
  
  cov362_test = cov362_test[match(rownames(data_mrna2), cov362_test$symbol), ]
  
  res[i, 2] <- nrow(na.omit(cov362_test))
  
  # reverse index2 gene expressions
  data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ] = -data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ]
  
  # gene selection
  index = which(is.na(cov362_test$symbol) == F)
  
  res1 = c()
  for (j in 1:length(patients)){
    cat('\r', j)
    res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
    res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])) # rank of the BRCAness genes
    # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])*(-log10(cov362_test$p_val_adj[index]))) # rank of the BRCAness genes
    
  }
  time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
  status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]
  
  group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
  data_plot = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
  set.seed(123)
  fit <- survfit(Surv(time, status) ~ group, data = data_plot)
  fit.coxph = coxph(Surv(time, status) ~ group, data = data_plot)
  s <- summary(fit.coxph)
  res[i, 3] <- s$logtest[3]
}

write.csv(
  res,
  paste0(sub_output_dir, "FigureS6A_BRCA_survival_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

#### Survival plot with selected parameters

```{r}
conserve_gene_only <- FALSE

if (conserve_gene_only) {
  p_threshold = 1*10^(-8)
  fc_threshold = 0
} else {
  p_threshold = 1*10^(-8)
  fc_threshold = 0
}
data_mrna2 <- data_mrna
  
if (conserve_gene_only){
  cov362_test <- filter(cov362, ensembl %in% sister_conserve_gene)
} else {
  cov362_test <- cov362
}

index1 = which(cov362_test$avg_log2FC <= -fc_threshold & cov362_test$p_val_adj <= p_threshold) # downregulated genes when BRCA1 is knocked out (BRCA-)
index2 = which(cov362_test$avg_log2FC > fc_threshold & cov362_test$p_val_adj <= p_threshold) # upregulated genes when BRCA1 is knocked out
index2_symbol = cov362_test$symbol[index2]

index = c(index1, index2)
length(index)
cov362_test = cov362_test[index, ]
# cov362_test$symbol

cov362_test = cov362_test[match(rownames(data_mrna2), cov362_test$symbol), ]

# reverse index2 gene expressions
data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ] = -data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ]

# gene selection
index = which(is.na(cov362_test$symbol) == F)

res1 = c()
for (j in 1:length(patients)){
  cat('\r', j)
  res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
  res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])) # rank of the BRCAness genes
  # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])*(-log10(cov362_test$p_val_adj[index]))) # rank of the BRCAness genes
  
}
time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]

group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
data_plot = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
set.seed(123)
fit <- survfit(Surv(time, status) ~ group, data = data_plot)
fit.coxph = coxph(Surv(time, status) ~ group, data = data_plot)
s <- summary(fit.coxph)
# res[i, 3] <- s$logtest[3]
  
pdf(paste0(sub_output_dir,"FigureS6A_threshold_", p_threshold, ifelse(conserve_gene_only, "BRCA_COV_OS_conserve_gene.pdf", "BRCA_COV_OS.pdf")), width = 4, height = 5, pointsize = 20)
fig <- ggsurvplot(
  fit,
  data = data_plot,
  palette = c(red, blue),
  risk.table = T,
  legend.title = "BRCAness (COV362)", # - undeconvoluted",
  legend.labs = c("high", "low"),
  xlab = "Time (months)", ylab = "Overall survival",
  ggtheme = theme_survminer(
    base_size = 12,
    font.main = c(12, "plain", "black"),
    font.x = c(12, "plain", "black"),
    font.y = c(12, "plain", "black"),
    font.caption = c(12, "plain", "black"),
    font.legend = c(12, "plain", "black"),
    font.tickslab = c(12, "plain", "black"),
    font.submain = c(12, "plain", "black")
  ),
  pval = signif(s$logtest[3], 2),
  pval.coord = c(100, 0.75)
  ) +
  guides(colour = guide_legend(nrow = 2))
fig
dev.off()
```

## B: Bar plot for TCGA BRCAness signatute's log2FC

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS6/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}

lseq <- function(from=1, to=100000, length.out=6) {
  # logarithmic spaced sequence
  # blatantly stolen from library("emdbook"), because need only this
  exp(seq(log(from), log(to), length.out = length.out))
}

conserve_gene_only <- TRUE
```

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
log2FC_table <- read.csv("./data/output/Figure3/Figure3D_logFC_table_for_plots.csv")
```

```{r}
conserve_gene_only <- TRUE

if (conserve_gene_only){
  plot_table <- filter(log2FC_table, conserve_gene)
} else {
  plot_table <- log2FC_table
}
```

```{r}
selected_p_threshold <- 0.005

table_tests <- plot_table %>% 
  filter(BRCAness_tcga_p_adj <= selected_p_threshold) %>% 
  select(ensembl, starts_with("BRCAness_tcga"), ends_with("log2FC_preR_preS_adj")) %>% 
  na.omit() %>% 
  arrange(BRCAness_tcga_logFC)

# BRCAness
pdf(
  paste0(
    sub_output_dir, "FigureS6B_BRCA_signature_tcga_p_", signif(selected_p_threshold, 2),
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(table_tests$BRCAness_tcga_logFC > 0, red, blue)
fig = barplot(
  table_tests$BRCAness_tcga_logFC,
  names.arg = table_tests$ensembl,
  las = 2,
  main = "BRCAness signature (TCGA)",
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# Olaparib
r <- cor(
  table_tests$BRCAness_tcga_logFC,
  -table_tests$Olaparib_log2FC_preR_preS_adj,
  method = "spearman", use = "complete"
  )
pdf(paste0(sub_output_dir,"FigureS6B_Olaparib_signature_tcga_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$Olaparib_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$Olaparib_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("Olaparib to TCGA, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# Carboplatin
r <- cor(
  table_tests$BRCAness_tcga_logFC,
  -table_tests$Carbo_log2FC_preR_preS_adj,
  method = "spearman", use = "complete"
)
pdf(paste0(sub_output_dir,"FigureS6B_Carboplatin_signature_tcga_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$Carbo_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$Carbo_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("Carboplatin to TCGA, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# NK
r <- cor(
  table_tests$BRCAness_tcga_logFC,
  -table_tests$NK_log2FC_preR_preS_adj,
  method = "spearman", use = "complete"
  )
pdf(paste0(sub_output_dir,"FigureS6B_NK_signature_tcga_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$NK_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$NK_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("NK to TCGA, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()
```

## C: Bar plot for COV362 BRCAness signatute's log2FC

```{r}
selected_p_threshold <- 1 * 10^-8

table_tests <- plot_table %>% 
  filter(BRCAness_cov_p_adj <= selected_p_threshold) %>% 
  select(ensembl, starts_with("BRCAness_cov"), ends_with("log2FC_preR_preS_adj")) %>% 
  na.omit() %>% 
  arrange(BRCAness_cov_logFC)

# BRCAness
pdf(
  paste0(
    sub_output_dir, "FigureS6C_BRCA_signature_cov_p_", signif(selected_p_threshold, 2),
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(table_tests$BRCAness_cov_logFC > 0, red, blue)
fig = barplot(
  table_tests$BRCAness_cov_logFC,
  names.arg = table_tests$ensembl,
  las = 2,
  main = "BRCAness signature (COV362)",
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# Olaparib
r <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Olaparib_log2FC_preR_preS_adj)
pdf(paste0(sub_output_dir,"FigureS6C_Olaparib_signature_cov_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$Olaparib_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$Olaparib_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("Olaparib to COV362, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# Carboplatin
r <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Carbo_log2FC_preR_preS_adj)
pdf(paste0(sub_output_dir,"FigureS6C_Carboplatin_signature_cov_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$Carbo_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$Carbo_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("Carboplatin to COV362, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# NK
r <- cor(table_tests$BRCAness_cov_logFC, -table_tests$NK_log2FC_preR_preS_adj)
pdf(paste0(sub_output_dir,"FigureS6C_NK_signature_cov_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$NK_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$NK_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("NK to COV362, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()
```

# Session information

```{r}
sessionInfo()
```


