---
title: "Figures in KURAMOCHI manuscript"
author: "Shuyu Zheng and Jing Tang"
date: "`r Sys.Date()`"
output: html_notebook
---

# Load pakcages

```{r}
library(Seurat)
library(plyr)
library(tidyverse)
library(readxl)
library(ggpubr)
library(rstatix)
library(ggalt)
library(ggrepel)
library(ggforce)
library(wordspace)
require(scales)
library(viridis)
library(ggsci)
library(synergyfinder)
library(ggsankey)
library(survminer)
library(survival)
library(scales)
library(sparseMatrixStats) # sparse matrix calculation
library(distances)
library(vegan) # distance function
library(irlba) # for fast pca
library(gridExtra) # Merge multiple ggplots with 'grid.arrange' function
library(reshape2)
library(car) # scatterplot combined with barplot
library(PharmacoGx)
library(synergyfinder)
output_dir <- "./data/output/"
if (!dir.exists(output_dir)){
  dir.create(output_dir, recursive = TRUE)
}
```

# Color code

```{r}
blue = "#3C5488FF"
red =  "#DC0000FF"
green = "#00A087FF"
brown = "#B09C85FF"
purple = "#8491B4FF"

blue2 <- "#4DBBD5FF"
red2 <- "#E64B35FF"
green2 <- "#91D1C2FF"
brown2 <- "#7E6148FF"

blue3 <- "#8491B4FF"
red3 <- "#F39B7FFF"
# green and velvet
# hrd_color <- "#C6809E"
# hrd_vector_color <- "#B36981"
# hrp_color <- "#659698"
# orange and blue
# hrd_color <- "#FFAC64FF"
# hrd_vector_color <- "#FA7702FF"
# hrp_color <- "#1BBAFFFF"
# red and blue
hrd_color <- "#E64B35FF"
hrd_vector_color <- "#EC8F84FF"
hrp_color <- "#3C5488FF"
```

Some abbreviations used in the codes

* preR: pre-resistant
* preS: pre-sensitive
* BT: pre-treatment
* AT: post-treatment
* Carbo: Carboplatin
* Control/ctrl: non-treatment control

# Figure 1: ReSisTrace leverages sister cell similarity to discover cellular states primed for resistance

## B2: Sister distance in pre-treatment sample (violin plot)

Load auxiliary codes

```{r}
source("./auxiliary_codes.R") # Definition for some auxiliary codes
sub_output_dir <- paste0(output_dir, "Figure1/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir, recursive = TRUE)
}
```

Load data

```{r eval=FALSE}
merge <- readRDS("./data/input/all_bt_sample_merge_filtered_gene_sisters.RDS")

# All of the non-sister cells are labeled with the "NA" at the end of values in "sisters_sample" column
all_sister_pairs <- merge$sisters_sample[!endsWith(merge$sisters_sample, "NA")]
```

Number of lineages containing sister cells

```{r eval=FALSE}
t <- table(table(all_sister_pairs))
t
```

Percentage of lineages/cells containing twins (%)

```{r eval=FALSE}
2424/(sum(!is.na(merge$drugSens)) - sum(!is.na(merge$sisters)) + sum(t)) * 100 # lineages
2424 * 2/sum(!is.na(merge$drugSens)) * 100 # cells
```

Percentage of lineages containing sisters (%)

```{r eval=FALSE}
sum(t)/(sum(!is.na(merge$drugSens)) - sum(!is.na(merge$sisters)) + sum(t)) * 100 # lineages
sum(!is.na(merge$sisters))/sum(!is.na(merge$drugSens)) * 100 # cells
```

Percentage of cells preR/(preR + preS) (%)

```{r eval=FALSE}
sum(merge$drugSens == "pre-resistant", na.rm = TRUE) / 
  sum(merge$drugSens %in% c("pre-resistant", "pre-sensitive"), na.rm = TRUE) * 
  100
```

Percentage of lineages preR/(preR + preS) (%)

All pre-treatment sample

```{r}
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens) %>% 
  filter(!is.na(drugSens) & drugSens == "pre-resistant")
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens) %>% 
  filter(!is.na(drugSens) & drugSens %in% c("pre-resistant", "pre-sensitive"))
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

Carboplatin

```{r}
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens == "pre-resistant",
    sample %in% c("Carbo1", "Carbo2")
  )
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens %in% c("pre-resistant", "pre-sensitive"),
    sample %in% c("Carbo1", "Carbo2")
    )
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

Olaparib

```{r}
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens == "pre-resistant",
    sample %in% c("Olaparib1", "Olaparib2")
  )
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens %in% c("pre-resistant", "pre-sensitive"),
    sample %in% c("Olaparib1", "Olaparib2")
    )
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

NK

```{r}
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens == "pre-resistant",
    sample %in% c("NK1", "NK2")
  )
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens %in% c("pre-resistant", "pre-sensitive"),
    sample %in% c("NK1", "NK2")
    )
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

Non-treatment control

```{r}
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens == "pre-resistant",
    sample %in% c("ctrl1", "ctrl2")
  )
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens %in% c("pre-resistant", "pre-sensitive"),
    sample %in% c("ctrl1", "ctrl2")
    )
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

1. Prepare expression matrix

```{r}
expression_mat <- as.matrix(GetAssayData(merge, slot = "data"))
```

2. Prepare sister table (including all sisters)

```{r}
sister_table <- data.frame(cell_barcode = colnames(merge), 
                     sisters = merge$sisters_sample,
                     stringsAsFactors = FALSE) %>% 
  dplyr::filter(sisters %in% all_sister_pairs) # remove singletons

sisters <- NULL
for (i in unique(sister_table$sisters)) {
  tmp_set <- sister_table$cell_barcode[which(sister_table$sisters  == i)]
  tmp_set <- combn(tmp_set, 2)
  tmp_df <- data.frame(group = rep(i, ncol(tmp_set)),
                       sister1 = tmp_set[1, ],
                       sister2 = tmp_set[2, ],
                       stringsAsFactors = FALSE)
  sisters <- rbind.data.frame(sisters, tmp_df)
  tmp_df <- NULL
  tmp_set <- NULL
}
```

```{r}
rm(merge) # remove large seurat object to release memory
rm(pca, tsne, umap)
```

3. Statistic analysis 

We can use function `CalculateSisterSimilarity` (defined in 'script/Auxiliary_codes.R" file) to generate analysis sister similarities by setting the parameters:

* **expression_mat**: the gene expression matrix in log2(normalized_count + 1) scale containing gene names as rows and cell barcodes as columns.
* **n_gene**: number of top variable genes used to calculate the similarity.
* **sister_table**: a data frame containing the group number of twins and corresponding sisters' cell barcodes.

Following is an example using Euclidean distance and top 1000 variable genes.

1. Calculate difference matrix

**Note** This function will take a long time to calculate the difference matrix. Some times it takes hours when there are a lot genes and cells.

```{r}
# time_stamp <- Sys.time()
plot_data <- CalculateSisterSimilarity(
  expression_mat = expression_mat,
  n_gene = 1000,
  sister_table = sisters,
  similarity = "euclidean"
  )
# Sys.time() - time_stamp
```

```{r}
sister_d <- data.frame(
    sister_id = names(plot_data$sister_distance),
    distance = plot_data$sister_distance,
    stringsAsFactors = FALSE
    )
```

```{r}
rm(expression_mat)
```

4. (Batch job) Generate the violin plot

This part takes a lot of time to finish. The codes were submited as a batch job on the CSC Puhti cluster. The codes are in: "./codes_for_batch_job/plot_sister_similarity.R" and "./codes_for_batch_job/plot_sister_similarity.sh"

```{r eval=FALSE}
sister_d <- readRDS("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_distance.RDS")
# p1 <- PlotSisterSimilarity(distance = sister_d$distance,
#                           similarity = sister_d$similarity,
#                                  n_gene = sister_d$n_gene,
#                                  cutoff = NULL)
m <- switch (sister_d$similarity,
             "euclidean" = "Euclidean Distance",
             "pearson" = "Pearson Coefficient",
             "spearman" = "Spearman Coefficient"
)

cutoff <- 14.94
test <- t.test(sister_d$distance$similarity[sister_d$distance$group == "sisters"], sister_d$distance$similarity[sister_d$distance$group == "non-sisters"])
print(test)

p1 <- ggplot(data = sister_d$distance, aes(x = group, y = similarity, fill = group)) +
  geom_violin(trim = FALSE) + 
  labs(title=paste0(m, " (genes: ", sister_d$n_gene, ")"),x="", y = m) +
  geom_boxplot(width=0.1, fill="white", outlier.shape = NA) + 
  # geom_segment(aes(y = cutoff, x = 1.75, xend = 2.25, yend = 14.93)) + 
  scale_fill_manual(values=c("#3C5488FF", "#DC0000FF")) +
  theme_classic() + 
  theme(
    legend.position = "none",
    plot.title = element_text(hjust = 0.5),
    axis.title.y = element_text(size = 30),
    axis.text.x = element_text(size = 30)
  ) +
  annotate(
    geom = "text",
    x = 1.5,
    y = 40,
    label = ifelse(
      test$p.value == 0,
      "P < 2.2E-16",
      paste0("P = ", test$p.value)
    ))
ggsave(
  paste0(
    sub_output_dir, "Figure1B2_",
    sister_d$sister_d, "_", sister_d$n_gene, "_genes.png"
  ), 
  plot = p1, device = "png", width = 6, height = 6, dpi = 300
)
```

## B3: UMAP for sister pairs

Clean environment

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "Figure1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

Load data

```{r}
merge <- readRDS("./data/input/all_bt_sample_merge_filtered_gene_sisters.RDS")
```

Prepare plot table

```{r}
all_sister_pairs <- merge$sisters_sample[!endsWith(merge$sisters_sample, "NA")]
all_sister_pairs <- unique(all_sister_pairs)
set.seed(123)
n <- 10
groups <- sample(all_sister_pairs, n)
highlight_group <- vector("list", length(groups))
names(highlight_group) <- groups
for (i in groups){
  highlight_group[[i]] <- colnames(merge)[which(merge$sisters_sample == i)]
}

myCol <- ggsci::pal_npg("nrc")(n)
names(myCol) <- groups
plot_table <- as.data.frame(merge@reductions$umap@cell.embeddings)
plot_table$sisters <- merge$sisters_sample
plot_table$sisters[!plot_table$sisters %in% groups] <- NA
```

UMAP plot

```{r}
umap <- ggplot(data = plot_table, mapping = aes(x = UMAP_1, y = UMAP_2)) +
  geom_point(
    data = plot_table,
    mapping = aes(x = UMAP_1, y = UMAP_2),
    color = "grey90") +
  geom_point(
    data = plot_table %>% filter(!is.na(sisters)),
    size = 6,
    mapping = aes(x = UMAP_1, y = UMAP_2, color=sisters)
  ) +
  scale_color_manual(values = myCol, labels = seq(1, length(myCol))) + 
  geom_line(
    data = plot_table %>% filter(!is.na(sisters)),
    mapping = aes(color = sisters),
    size = 2
  ) +
  labs(
    x = "UMAP 1",
    y = "UMAP 2",
    color = "Lineage label"
  ) +
  theme_classic() +
  theme(
    axis.text = element_text(size = 21, colour = "black"),
    axis.title = element_text(size = 21),
    legend.text = element_text(size = 15),
    legend.title = element_text(size = 18)
  )
umap
ggsave(paste0(sub_output_dir, "/Figure1B3_highlight_sister_group_",paste0(groups, collapse = "."), "_filtered_sisters_umap.pdf"), plot = umap, device = "pdf", width = 6, height = 4)
```

## C: Scatter plot for sister-concordant genes

The calculation of sister-concordant genes were performed on the CSC Puhti server with a batch job. The codes are in: "./codes_for_batch_job/calculate_conserve_gene.R" and "./codes_for_batch_job/calculate_conserve_gene.sh".

The output files from this batch job are saved in "./data/input/sister_concordant_gene/"

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "Figure1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

Load data

```{r}
conserve_gene <- read.csv(
  "./data/input/sister_concordant_gene/sister_conserve_gene_all_bt_sample_twin.csv",
  stringsAsFactors = FALSE)
```

Prepare data table for plot

```{r}
highlight_gene_symbols <- conserve_gene %>% 
  arrange(desc(log2FC_dif)) %>% 
  head(10) %>% 
  select(symbol) %>% 
  unlist()
data_plot <- conserve_gene %>% 
  mutate(log10_p_value = -log10(adj_p_value)) %>% 
  mutate(
    highlight = ifelse(
      log10_p_value > -log10(0.05),
      "sister-concordant gene",
      "sister-disconcordant gene"
    ),
    label = (
      symbol %in% highlight_gene_symbols
    )
  )
data_plot$highlight <- factor(data_plot$highlight, levels = c("sister-concordant gene", "sister-disconcordant gene"))
cor <- cor.test(data_plot$mean_log2FC_sister, data_plot$mean_log2FC_non_sister)
```

Visualization

```{r}
p <- 
  ggplot(
    data_plot,
    aes(
      x = mean_log2FC_sister,
      y = mean_log2FC_non_sister,
      color = highlight
    )
  ) +
  geom_point() +
  lims(x = c(0, 1.5), y = c(0, 1.5)) +
  scale_color_manual(values = c("#E64B3555", "#3C5488")) + 
  geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
  annotate(
    geom = "text",
    x = 1.0,
    y = 0.2,
    label = paste0("R = ", signif(cor$estimate, 2), "\n", ifelse(cor$p.value == 0, "P < 2.2E-16", paste0("P =", format(cor$p.value, scientific = T)))),
    size = 15/.pt
  ) +
  geom_point(
    data=dplyr::filter(data_plot, label==TRUE),
    size = 3,
    color = "#DC0000") +
  geom_point(
    data=dplyr::filter(data_plot, label==TRUE),
    shape = 1,
    size = 3,
    show.legend = FALSE,
    color = "black"
  ) +
  ggrepel::geom_text_repel(
    data=dplyr::filter(
      data_plot, label==TRUE),
    aes(label = symbol),
    box.padding = 0.5,
    colour="black", point.padding = 0.25, max.overlaps = 30,
    size = 15/.pt,
    fontface="bold",
    show.legend=FALSE) +
  labs(
       x = expression("Sister cell pair log"[2]~"(FC)"),
       y = expression("Random cell pair log"[2]~"(FC)")
       ) +
  theme_classic() +
  theme(legend.position="right",
        axis.ticks.x = element_blank(),
        axis.text = element_text(size = 21, color = "black"),
        text = element_text(size = 22.5, color = "black"),
        axis.title = element_text(size = 21, color = "black")
      ) +
  guides(colour = "none")
p
ggsave(paste0(sub_output_dir, "Figure1C_conserve_gene_adjust_p.pdf"), p,
       width = 4, height = 4)
```

## D: RNA dynamics from RNA velocity analysis

The alignment is done with velocyto. The codes are in: "./codes_for_batch_job/runVelo.sh"

The scVelo is used to do the RNA velocity analysis. The codes are in "./codes_for_batch_job/runscvelo.py" and "./codes_for_batch_job/runscvelo.sh"

The outputs are saved in "./data/input/scvelo/"

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "Figure1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

Load data

```{r}
dynamic <- read.csv("./data/input/scvelo/RNA_velocity_var_all_BT.csv")
# conserve_gene <- read.csv("./data/input/sister_concordant_gene/sister_conserve_gene_all_bt_sample_twin.csv")
selected_genes <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_second_bin.csv")
```

Prepare table for plot

```{r}
plot_table <- selected_genes %>%
  mutate(conserve_gene = ifelse(Sister.conserved.gene, "Sister-concordant", "Sister-discordant")) %>% 
  right_join(dynamic, by = c("ensembl" = "X")) %>%
  select(
    ensembl,
    transcription = fit_alpha,
    splicing = fit_beta,
    degradation = fit_gamma,
    conserve_gene) %>% 
  na.omit()
plot_table$conserve_gene <- factor(plot_table$conserve_gene, levels = c("Sister-concordant", "Sister-discordant"))
t <- plot_table %>% 
  select(ensembl, transcription, value = splicing, conserve_gene) %>%
  mutate(dynamic = "splicing") %>% 
  rbind.data.frame(
    plot_table %>% 
      select(ensembl, transcription, value = degradation, conserve_gene) %>% 
      mutate(dynamic = "degradation"))
t$dynamic <- str_to_title(t$dynamic)
t$dynamic <- factor(t$dynamic, levels = c("Splicing", "Degradation"))
table(plot_table$conserve_gene)
```

Visualization

```{r}
min_value <- min(c(t$transcription, t$value))
max_value <- max(c(t$transcription, t$value))
p <- ggplot(t, aes(x = transcription, y = value)) +
  stat_density_2d(aes(fill = after_stat(ndensity)), geom = "raster", contour = FALSE,
                  ) +
  scale_fill_viridis() +
  labs(x = "Transcription") +
  geom_abline(
    intercept = 0,
    slope = 1,
    color = "#FFFFFFFF",
    linetype = "dashed",
    size = 1) + 
  scale_y_continuous(
    expand = c(0, 0),
    limits = c(min_value, max_value),
    trans = log2_trans(),
    breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x))) +
  scale_x_continuous(
    expand = c(0, 0),
    limits = c(min_value, max_value),
    trans = log2_trans(),
    breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x))) +
  coord_equal(xlim=c(min_value, max_value),ylim=c(min_value, max_value)) +
  facet_grid(vars(dynamic), vars(conserve_gene)) +
  theme(
    legend.position = "none",
    strip.text.x = element_text(
      size = 15
    ),
    strip.text.y = element_text(
      size = 15
    ),
    axis.text = element_text(size = 15, colour = "black"),
    axis.title.y = element_blank(),
    axis.title.x = element_text(
      size = 21, face = "plain"
    )
  )
p
ggsave(
  paste0(
    sub_output_dir,
    "Figure1D_sister_conserve_gene_dynamic_log_scale_second_bin.pdf"
  ),
  p, width = 4.5, height = 4.5
)
```

## E: Gene set analysis for sister-discordant genes

1. We calculated the summation of all of the gene's expression across all of the cells in all BT samples. (results in './data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv') 
2. We generated a histogram for the summation of expression for all of the genes (see below Figure S2A). Genes are grouped by sister-concordant genes (adj_p_value < 0.05 and log2FC_diff > 0) or sister-discordant genes (adj_p_value > 0.05 or log2FC_diff < 0).
3. The genes in the second bin (second bar for each group) in the histogram are extracted and the gene set enrichment analysis is done on each group (sister-concordant genes n = 2059 [enrichr result](https://maayanlab.cloud/Enrichr/enrich?dataset=fe932ded91e8f8298f674699e8b007eb), and non-sister-concordant genes n = 1332 [enrichr result](https://maayanlab.cloud/Enrichr/enrich?dataset=7b522be89fef4b4a4f08355a0ef3703d))
4. The tables for Pathways "KEGG 2021 Human" are downloaded and saved in "./data/input/enrichr/KEGG_2021_Human_table_sister_conserved_gene.txt" and "./data/input/enrichr/KEGG_2021_Human_table_non_conserved_gene.txt"

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "Figure1/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir, recursive = TRUE)
}
```

```{r}
cons <- data.table::fread("./data/input/enrichr/KEGG_2021_Human_table_non_conserved_genes.txt")
plot_table <- cons %>% 
    select(Term, x = `Odds Ratio`, p = `Adjusted P-value`) %>% 
    mutate(y = -log10(p),
           sig = p < 0.05)
# plot_table$Term <- paste0("bold('", plot_table$Term, "')")
highlight <- plot_table %>% 
    # filter(Term %in% c("Inositol phosphate metabolism", "Phosphatidylinositol signaling system"))
    filter(sig)

p <- ggplot() +
   theme_classic() +
   theme(
       legend.key = element_blank(), 
       axis.text = element_text(size = 21, colour = "black"),
       axis.title = element_text(size = 21),
       plot.title = element_text(size = 21, hjust = 0.5)
       ) +
   labs(x = expression("Odds Ratio"),
        y = expression("-log"[10]~"(P value)"),
        title="") +
   # geom_hline(yintercept=-log10(0.05), linetype="dashed") +
   # geom_vline(xintercept = 0, linetype="dashed") +
 geom_point(
     aes(x = x, y = y, colour=factor(sig)),
     colour = "grey70",
     size = 2,
     data = plot_table,
     show.legend = FALSE) +
  geom_point(
    aes(x = x, y = y),
    size = 3,
    data = highlight,
    color = "#3C5488FF",
    show.legend = FALSE) +
  geom_point(
    aes(x = x, y = y),
    shape = 1,
    size = 3,
    data = highlight,
    show.legend = FALSE) +
  geom_text_repel(
    aes(x = x, y = y, label = Term),
    size = 15/.pt,
    data = highlight,
    # parse = TRUE,
    box.padding = 1
  )
p
ggsave(paste0(sub_output_dir, "Figure1E_non_conserved_KEGG_volcano.pdf"), p,
       width = 4, height = 4)
```

# Figure S1: Cell doubling, lineage labelling efficiency, and quality assessment of ReSisTrace

## Number of unique barcode per sample

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

Prepare data table

```{r}
samples <- c("Control_1", "Control_2", "Carboplatin_1", "Carboplatin_2", "Olaparib_1", "Olaparib_2", "NK_1", "NK_2")

label_sequences <- vector(mode = "list", length = length(samples))
names(label_sequences) <- samples

for (s in samples){
  load(paste0("./data/input/preprocess/", s, "/lineage_label_tables.RData"))
  label_id_bt <- sapply(l_b_fre_merge_filtered[["lineage_id"]], function(x)(strsplit(x, ":")))
  label_id_bt <- unique(purrr::reduce(label_id_bt, `c`))
  label_index <- label_index %>% 
    filter(seq_id %in% label_id_bt)
  label_index$sample <- s
  label_index$total <- nrow(label_index)
  label_sequences[[s]] <- label_index
}

label_sharing <- bind_rows(label_sequences)
frequency <- as.data.frame(table(label_sharing$seq))
names(frequency) <- c("seq", "freq")
label_sharing <- left_join(label_sharing, frequency, by = "seq")
max(label_sharing$freq)
table(label_sharing$freq)
table(label_sharing$freq)/nrow(label_sharing) * 100
```

```{r}
# proportion of different frequency in each sample
plot_table <- label_sharing %>%
  mutate(unique = freq == 1) %>% 
  group_by(sample, unique) %>% 
  summarise(
    n = n(),
    total = mean(total),
    .groups = "keep") %>% 
  mutate(
    proportion = n/total * 100
  )
# plot_table$freq <- factor(plot_table$freq, levels = seq(6, 1))
plot_table$sample <- gsub("_", " ", plot_table$sample)
plot_table$sample <- factor(
  plot_table$sample,
  levels = c("Carboplatin 1", "Carboplatin 2", "Olaparib 1", "Olaparib 2",
              "NK 1", "NK 2", "Control 1", "Control 2")
  )
write.csv(plot_table, paste0(sub_output_dir, "unique_label_per_bt_sample.csv"), row.names = FALSE)

# colors = c("#3C5488","#5A6E9A","#7888AC","#95A3BF","#B3BDD1","#D1D7E3")
colors = c("#8f9dbb","#d1d7e3")

p <- ggplot(plot_table, aes(x = sample, y = proportion, fill = unique)) +
  geom_bar(position = "stack", stat = "identity") +
  labs(
    y = "Proportion of lineage labels (%)",
    fill = ""
  ) + 
  # ylim(c(0, 100)) +
  scale_fill_manual(values = colors, labels = c("Shared by samples", "Unique per sample")) +
  theme_classic() +
  theme(
    text = element_text(size = 20),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 20,
      color = "black"
    ),
    axis.text.y = element_text(
      size = 20,
      color = "black"
    ),
    axis.title.x = element_blank(),
    legend.text = element_text(size = 20),
    # legend.title = element_text(size = 15),
    # legend.position = "top",
    panel.grid = element_blank()
  )
ggsave(paste0(sub_output_dir, "unique_label_per_bt_sample.pdf"), width = 7.5, height = 5)

plot_table %>%
  ungroup() %>% 
  filter(unique == TRUE) %>% 
  summarise(mean = mean(proportion))
```

## Percentage of cells have barcodes

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

Prepare data table

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
```

```{r}
table <- data@meta.data %>%
  mutate(label_detect = !is.na(drugSens)) %>% 
  select(sample, time_point, label_detect) %>% 
  group_by(sample, time_point, label_detect) %>% 
  summarize(n = n(), .groups = "keep")

total <- table %>% 
  group_by(sample, time_point) %>% 
  summarise(total = sum(n), .groups = "keep")
```

Visualization

```{r}
plot_table <- table %>% 
  left_join(total, by = c("sample", "time_point")) %>% 
  mutate(proportion = n/total * 100)
plot_table <- plot_table %>% 
  mutate(
    sample = case_when(
      sample == "Carbo1" ~ "Carboplatin 1", 
      sample == "Carbo2" ~ "Carboplatin 2", 
      sample == "ctrl1" ~ "Control 1", 
      sample == "ctrl2" ~ "Control 2", 
      sample == "Olaparib1" ~ "Olaparib 1", 
      sample == "Olaparib2" ~ "Olaparib 2", 
      sample == "NK1" ~ "NK 1", 
      sample == "NK2" ~ "NK 2"
    ),
    time_point = case_when(
      time_point == "AfterTreatment" ~ "Post-treatment",
      time_point == "BeforeTreatment" ~ "Pre-treatment"
    ))
plot_table$sample <- factor(
  plot_table$sample,
  levels = c("Carboplatin 1", "Carboplatin 2", "Olaparib 1", "Olaparib 2",
             "NK 1", "NK 2", "Control 1", "Control 2")
  )
write.csv(plot_table, paste0(sub_output_dir, "proportion_of_cells_having_label.csv"), row.names = FALSE)


colors = c("#8f9dbb","#d1d7e3")

p <- ggplot(
  plot_table %>% filter(time_point == "Pre-treatment"),
  aes(x = sample, y = proportion, fill = label_detect)
  ) +
  geom_bar(position = "stack", stat = "identity") +
  labs(
    y = "Proportion of cells (%)",
    fill = "Lineage labels detected",
    title = "Pre-treatment"
  ) + 
  ylim(c(0, 100)) +
  scale_fill_manual(values = colors) +
  theme_classic() +
  theme(
    text = element_text(size = 20),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 20,
      color = "black"
    ),
    axis.text.y = element_text(
      size = 20,
      color = "black"
    ),
    axis.title.x = element_blank(),
    legend.text = element_text(size = 15),
    legend.title = element_text(size = 20),
    plot.title = element_text(size = 20, hjust = 0.5),
    panel.grid = element_blank()
  )
ggsave(paste0(sub_output_dir, "proportion_of_cells_having_label_preTreatment.pdf"), width = 8, height = 5.3)

p <- ggplot(
  plot_table %>% filter(time_point == "Post-treatment"),
  aes(x = sample, y = proportion, fill = label_detect)
  ) +
  geom_bar(position = "stack", stat = "identity") +
  labs(
    y = "Proportion of cells (%)",
    fill = "Lineage labels detected",
    title = "Post-treatment"
  ) + 
  ylim(c(0, 100)) +
  scale_fill_manual(values = colors) +
  theme_classic() +
  theme(
    text = element_text(size = 20),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 20,
      color = "black"
    ),
    axis.text.y = element_text(
      size = 20,
      color = "black"
    ),
    axis.title.x = element_blank(),
    legend.text = element_text(size = 15),
    legend.title = element_text(size = 20),
    plot.title = element_text(size = 20, hjust = 0.5),
    panel.grid = element_blank()
  )
ggsave(paste0(sub_output_dir, "proportion_of_cells_having_label_postTreatment.pdf"), width = 8, height = 5.3)
```

## B1: Number of divisions after 1x doubling

The data table comes from Jun. He manually counted the number of cells which divided zero, once, twice or more times after "1 x doubling time". Totally 6 views are counted.

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

Load data

```{r}
data <- read_xlsx("./data/input/ReSisTrace_figS1 - Copy.xlsx", range = "B1:G4")
data <- t(data)
colnames(data) <- c("0", "1", "> 1")
data <- data * 100
data <- data %>% 
  as.data.frame() %>% 
  mutate(view = 1:6) %>% 
  pivot_longer(!view, names_to = "Division", values_to = "Proportion")
data$Division <- factor(data$Division, levels = c("0", "1", "> 1"))
```

Prepare table for plot

```{r}
plot_table <- data %>%
  dplyr::group_by(Division) %>% 
  dplyr::summarise(
    mean = mean(Proportion),
    sd = sd(Proportion)
  )
write.csv(plot_table, paste0(sub_output_dir, "FigureS1A_cell_division_times.csv"), row.names = FALSE)
col = c("#d1d7e3", "#8f9dbb", "#3c5488")
```


```{r}
p <- ggplot(plot_table, aes(x = Division, y = mean, fill = Division)) +
  geom_bar(
    stat="identity", color="black",
    position=position_dodge()
  ) +
  geom_errorbar(
    aes(ymin = mean - sd, ymax = mean + sd, width = 0.2), color="black"
  ) +
  geom_point(
    data = data,
    aes(x = Division, y = Proportion),
    position = position_jitterdodge(jitter.width = 2)
  ) +
  scale_fill_manual(values = col) +
  labs(
    x = "Division(s)",
    y = "Proportion of clones (%)"
  ) +
  theme_classic() +
  theme(
    legend.position = "None",
    text = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 20, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "FigureS1B1_cell_division_times.pdf"),
  width = 5, height = 4.8
)
```

## B2: Counts of lineages in the BT sample

The information about lineage labels for each samples are saved in "./data/input/preprocess/[sample_name]/lineage_label_tables.RData". After loading the ".RData" files, a data frame "l_b_fre_merge_filtered" will be in the working environment. This table contains the lineage label information for pre-treatment samples from different treatments.

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

Prepare table for plot

```{r}
plot_table <- NULL

samples <- c("Control_1", "Control_2", "Carboplatin_1", "Carboplatin_2", "Olaparib_1", "Olaparib_2", "NK_1", "NK_2")

for (s in samples){
  load(paste0("./data/input/preprocess/", s, "/lineage_label_tables.RData"))
  tmp <- l_b_fre_merge_filtered %>% 
    select(lineage_id) %>% 
    table() %>% 
    table() %>% 
    prop.table()
  tmp2 <- l_b_fre_merge_filtered %>% 
    select(lineage_id) %>% 
    table() %>% 
    table()
  cat(s, "\n", tmp2, "\n")
  tmp <- data.frame(
    Size_of_lineage = tmp,
    Sample = s,
    label_at = length(unique(l_a_fre_merge_filtered$lineage_id)),
    label_bt = length(unique(l_b_fre_merge_filtered$lineage_id)),
    label_btat = length(unique(intersect(l_b_fre_merge_filtered$lineage_id, l_a_fre_merge_filtered$lineage_id)))
  )
  plot_table <- rbind.data.frame(plot_table, tmp)
}

colnames(plot_table) <- c("Size_of_lineage", "Frequency", "Sample", "labelAT", "labelBT", "labelBTAT")
write.csv(
  plot_table,
  paste0(sub_output_dir, "FigureS1B2_frequence_of_lineage_label.csv"),
  row.names = FALSE
)
```

Visualization

```{r}
plot_table$Frequency <- plot_table$Frequency * 100
data <- plot_table
plot_table <- plot_table %>%
  dplyr::group_by(Size_of_lineage) %>% 
  dplyr::summarise(
    mean = mean(Frequency),
    sd = sd(Frequency)
  )
col = c("#d1d7e3", "#8897b7", "#5e729d", "#3c5488")
p <- ggplot(plot_table, aes(x = Size_of_lineage, y = mean, fill = Size_of_lineage)) +
  geom_bar(
    stat="identity", color="black",
    position=position_dodge()
  ) +
  geom_errorbar(
    aes(ymin = mean - sd, ymax = mean + sd, width = 0.2), color="black"
  ) +
  geom_point(
    data = data,
    aes(x = Size_of_lineage, y = Frequency),
    position = position_jitterdodge(jitter.width = 2)
  )+
  scale_fill_manual(values = col) +
  labs(
    x = "Clone size",
    y = "Proportion of lineages (%)"
  ) +
  theme_classic() +
  theme(
    legend.position = "None",
    text = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 20, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "FigureS1B2_frequence_of_lineage_label.pdf"),
  width = 5, height = 4.8
)
```

## C: DEGs correlation between replicates

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir, recursive = TRUE)
}
```

## Carboplatin

```{r}
treatments <- c("Carboplatin", "Olaparib", "NK", "Control")
for (t in treatments){
  t_short <- switch (t,
    "Carboplatin" = "Carbo",
    "Control" = "scLT_ctrl",
    "NK" = "NK",
    "Olaparib" = "Olaparib"
  )
  deg <- read.csv(
    paste0(
      "./data/input/DEG/deg_prer_pres_in_singleton_",
      t,
      "_intersection.csv"
    )
  )

  cat("Number of genes ", nrow(deg), "\n")
  cor <- cor.test(
    deg[[paste0(t_short, "_1_avg_log2FC")]],
    deg[[paste0(t_short, "_2_avg_log2FC")]]
  )
  range <- range(
    c(
      deg[[paste0(t_short, "_1_avg_log2FC")]],
      deg[[paste0(t_short, "_2_avg_log2FC")]]
    )
  )
  p <- ggplot(
    data = deg,
    aes_string(
      x = paste0(t_short, "_1_avg_log2FC"),
      y = paste0(t_short, "_2_avg_log2FC")
    )
    ) +
    geom_point(
      color = "#3C548855",
      size = 6,
      show.legend = FALSE
    ) +
    ylim(range) +
    xlim(range) +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
    annotate(
      geom = "text",
      x = 0.15,
      y = - 0.3,
      label = paste0(
        "R = ", signif(cor$estimate, 2), "\n",
        ifelse(
          cor$p.value == 0, "P < 2.2E-16",
          paste0("P =", format(cor$p.value, scientific = T))
        )
      ),
      size = 6.5
    ) + 
    labs(
      x = expression("Replicate 1 log"[2]~"(FC)"),
      y = expression("Replicate 2 log"[2]~"(FC)"),
      title = t
    ) +
    theme_classic() +
    theme(
      legend.position="none",
      axis.ticks.x = element_blank(),
      axis.text = element_text(size = 21, colour = "black"),
      axis.title = element_text(size = 21, colour = "black"),
      plot.title = element_text(size = 21, hjust = 0.5, colour = "black", face = "bold")
    )
  p
  ggsave(paste0(sub_output_dir, "FigureS1C_correlation_between_replicates_", t, ".pdf"), p,
         width = 5, height = 5)
}

```

# Figure S2: Characteristics of sister-concordant and sister-discordant genes

## A: Histogram for gene expressions of sister-conserved/non-conserved genes

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS2/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir, recursive = TRUE)
}
```

```{r}
plot_table <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv")

t_test <- t.test(plot_table$mean_expression[plot_table$conserve_gene == TRUE], plot_table$mean_expression[plot_table$conserve_gene == FALSE])
print(t_test)

plot_table <-  plot_table %>%  
  mutate(`Sister-conserved gene` = ifelse(conserve_gene, "Sister-concordant gene", "Sister-discordant gene")) %>% 
  select(ensembl, `Sister-conserved gene`, sum_expression, mean_expression)

mylog10_trans <- function (base = 10)
{
  trans <- function(x) log(x + 1, base)
  inv <- function(x) base^x
  trans_new(paste0("log-", format(base)), trans, inv, log_breaks(base = base),
            domain = c(1e-100, Inf))
}

p <- ggplot(
  plot_table,
  aes(
    x = sum_expression,
    fill = `Sister-conserved gene`,
    color = `Sister-conserved gene`
    )
  ) +
  geom_histogram(
    alpha=0.2,
    bins = 50,
    position="identity"
  ) +
  scale_fill_manual(values = c("#DC000055", "#3C548855")) +
  scale_color_manual(values = c("#DC0000", "#3C5488")) +
  # ?(
  #   breaks = trans_breaks("log10", function(x) 10^x),
  #   labels = trans_format("log10", math_format(10^.x))
  # ) +
  geom_segment(
    data = NULL, 
    aes(x = 10000, y = 5 * 10^3, xend = 5000, yend = 2.5* 10^3),
    arrow = arrow(length = unit(0.25, "cm"), type = "closed"),
    show.legend = FALSE) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(
    trans = "mylog10",
    breaks = trans_breaks("log10", function(x) 10^(x - 1)),
    labels = trans_format("log10", function(x) math_format()(round(x))),
    expand = c(0, 0)
    ) +
  annotate(
    "text",
    x = 100000,
    y = 1000,
    label = paste0(
      "T-test, ",
      ifelse(
        t_test$p.value == 0,
        "P < 2.2E10-16",
        paste0("P = ", t_test$p.value)
      )
    )
  ) +
  labs(
    x = "Expression in pre-treatment sample",
    y = "Number of genes",
    fill = "",
    color = "") +
  theme_classic() +
  theme(
    legend.position = "top",
    axis.text = element_text(size = 20, colour = "black"),
    axis.title = element_text(size = 20, colour = "black"),
    legend.title = element_text(size = 20, colour = "black"),
    legend.text = element_text(size = 20, colour = "black"),
    axis.line = element_line(colour = "black"),
    axis.ticks = element_line(colour = "black")
  )
p
ggsave(
  paste0(sub_output_dir, "FigureS2A_sister_conserve_gene_exp_distribution.pdf"), p,
  width = 9, height = 5
)
```

## old B: Unspliced RNA propotion for sister-conserved/non-conserved genes

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS2/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
selected_genes <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_second_bin.csv")
```

```{r eval = FALSE}
unsplice <- read.csv("./data/input/scvelo/RNA_velocity_unspliced_all_BT.csv", row.names = 1)
unsplice <- as.matrix(unsplice)
unsplice  <- unsplice[, colnames(unsplice) %in% selected_genes$ensembl]
splice <- read.csv("./data/input/scvelo/RNA_velocity_spliced_all_BT.csv", row.names = 1)
splice <- as.matrix(splice)
splice <- splice[, colnames(splice) %in% selected_genes$ensembl]
u_s_propotion <- unsplice / (unsplice + splice)
write.csv(u_s_propotion, paste0(sub_output_dir, "FigureS2B_unspliced_RNA_propotion_second_bin.csv"))
```

```{r}
u_s_propotion <- read.csv(
  paste0(sub_output_dir, "FigureS2B_unspliced_RNA_propotion_second_bin.csv"),
  row.names = 1
)
```

```{r}
u_s_propotion <- as.matrix(u_s_propotion)

dynamic <- read.csv("./data/input/scvelo/RNA_velocity_var_all_BT.csv")
selected_genes$conserve_gene <- ifelse(selected_genes$Sister.conserved.gene, "Sister-concordant\ngene", "Sister-discordant\ngene")
selected_genes$conserve_gene <- factor(selected_genes$conserve_gene, levels = c("Sister-concordant\ngene", "Sister-discordant\ngene"))
selected_velocity_genes <- selected_genes %>%
  select(ensembl, conserve_gene) %>% 
  right_join(dynamic, by = c("ensembl" = "X")) %>% 
  mutate(
    Degradation_Transcription = fit_gamma/fit_alpha
  ) %>% 
  select(
    ensembl, 
    fit_gamma, fit_alpha, fit_beta,
    conserve_gene) %>% 
  na.omit()

u_s_propotion_mean <- colMeans(u_s_propotion, na.rm = T)
u_s_propotion_mean <- data.frame(
  gene = names(u_s_propotion_mean), 
  unsplice_propotion = u_s_propotion_mean
) %>% 
  filter(gene %in% selected_velocity_genes$ensembl) %>% 
  left_join(select(selected_genes, gene = ensembl, conserve_gene), by = "gene") %>% 
  filter(!is.na(conserve_gene))
  
means <- aggregate(unsplice_propotion ~ conserve_gene, u_s_propotion_mean, mean)
p <- ggplot(
  data = u_s_propotion_mean,
  aes(x = conserve_gene, y = unsplice_propotion, fill = conserve_gene)
) + 
  geom_violin(trim = FALSE) + 
  labs(x="", y = "Unspliced RNA propotion") +
  geom_boxplot(width=0.1, fill="white", outlier.shape = NA) + 
  scale_fill_manual(values=c(red, blue)) +
  theme_classic() + 
  theme(
    legend.position = "none",
    plot.title = element_text(hjust = 0.5),
    axis.title.y = element_text(size = 20, color = "black"),
    axis.text.x = element_text(size = 20, color = "black"),
    axis.text.y = element_text(size = 20, color = "black")
  ) + 
  ggpubr::stat_compare_means(
    method = "t.test",
    label.y = 1.3,
    label.x.npc = "center",
    size = 6.5,
    hjust = 0.5) +
  geom_text(
    data = means,
    aes(label = signif(unsplice_propotion, 3),y = unsplice_propotion), 
    nudge_x = 0.18, color = "white", size = 6.5, fontface = "bold")

ggsave(
  paste0(sub_output_dir, "FigureS2B_unspliced_RNA_propotion_second_bin.pdf"), 
  plot = p, width = 8, height = 5
)
```

## B: Degradation/Transcription violin plot for sister-conserved or non-conserved genes

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS2/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
selected_genes <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_second_bin.csv")
```

```{r}
dynamic <- read.csv("./data/input/scvelo/RNA_velocity_var_all_BT.csv")
selected_genes$conserve_gene <- ifelse(selected_genes$Sister.conserved.gene, "Sister-concordant\ngene", "Sister-discordant\ngene")
selected_genes$conserve_gene <- factor(selected_genes$conserve_gene, levels = c("Sister-concordant\ngene", "Sister-discordant\ngene"))
plot_table <- selected_genes %>%
  select(ensembl, conserve_gene) %>% 
  right_join(dynamic, by = c("ensembl" = "X")) %>% 
  mutate(
    Degradation_Transcription = fit_gamma/fit_alpha
  ) %>% 
  select(
    ensembl, 
    Degradation_Transcription, 
    conserve_gene) %>% 
  na.omit()
means <- aggregate(Degradation_Transcription ~ conserve_gene, plot_table, mean)
```

```{r}
p <- ggplot(
  data = plot_table,
  aes(x = conserve_gene, y = Degradation_Transcription, fill = conserve_gene)
) + 
  geom_violin(trim = FALSE) + 
  labs(x="", y = "Degradation rate / Transcription rate") +
  geom_boxplot(width=0.1, fill="white", outlier.shape = NA) + 
  scale_fill_manual(values=c(red, blue)) +
  theme_classic() + 
  theme(
    legend.position = "none",
    axis.title.y = element_text(size = 20, color = "black"),
    axis.text.x = element_text(size = 20, color = "black"),
    axis.text.y = element_text(size = 20, color = "black")
  ) + 
  ggpubr::stat_compare_means(
    method = "t.test",
    label.x.npc = "center",
    size = 6.5,
    hjust = 0.5
  ) +
  geom_text(
    data = means,
    aes(label = signif(Degradation_Transcription, 3), y = Degradation_Transcription), 
    nudge_x = 0.15, color = "white", size = 6.5, fontface = "bold")

ggsave(
  paste0(sub_output_dir, "FigureS2C_Degradation_Transcription_second_bin.pdf"), 
  plot = p, width = 8, height = 5.5
)
```

## C: Splicing rate/Transcription rate violin plot for sister-conserved or non-conserved genes 

```{r}
plot_table <- selected_genes %>%
  select(ensembl, conserve_gene) %>% 
  right_join(dynamic, by = c("ensembl" = "X")) %>% 
  mutate(
    Splicing_Transcription = fit_beta/fit_alpha
  ) %>% 
  select(
    ensembl, 
    Splicing_Transcription, 
    conserve_gene) %>% 
  na.omit()
means <- aggregate(Splicing_Transcription ~ conserve_gene, plot_table, mean)
```

```{r}
p <- ggplot(
  data = plot_table,
  aes(x = conserve_gene, y = Splicing_Transcription, fill = conserve_gene)
) + 
  geom_violin(trim = FALSE) + 
  labs(x="", y = "Splicing rate / Transcription rate") +
  geom_boxplot(width=0.1, fill="white", outlier.shape = NA) + 
  scale_fill_manual(values=c(red, blue)) +
  theme_classic() + 
  theme(
    legend.position = "none",
    axis.title.y = element_text(size = 20, color = "black"),
    axis.text.x = element_text(size = 20, color = "black"),
    axis.text.y = element_text(size = 20, color = "black")
  ) + 
  ggpubr::stat_compare_means(
    method = "t.test",
    label.x.npc = "center",
    size = 6.5,
    hjust = 0.5
  ) +
  geom_text(
    data = means,
    aes(label = signif(Splicing_Transcription, 3), y = Splicing_Transcription), 
    nudge_x = 0.15, color = "black", size = 6.5, fontface = "bold")

ggsave(
  paste0(sub_output_dir, "FigureS2_Splicing_Transcription_second_bin.pdf"), 
  plot = p, width = 8, height = 5.5
)
```

## D: Gene set analysis for sister-concordant genes

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS2/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

```{r}
cons <- data.table::fread("./data/input/enrichr/KEGG_2021_Human_table_sister_conserved_gene.txt")
plot_table <- cons %>% 
    select(Term, x = `Odds Ratio`, p = `Adjusted P-value`) %>% 
    mutate(y = -log10(p),
           sig = p < 0.05)
# plot_table$Term <- paste0("bold('", plot_table$Term, "')")
highlight <- plot_table %>% 
    # filter(Term %in% c("TGF-beta signaling pathway", "Fanconi anemia pathway",
    #            "TNF signaling pathway"))
    filter(sig)

p <- ggplot() +
   theme_classic() +
   theme(
       legend.key = element_blank(), 
       axis.text = element_text(size = 20, colour = "black"),
       axis.title = element_text(size = 20),
       plot.title = element_text(size = 20, hjust = 0.5)
       ) +
   labs(x = expression("Odds Ratio"),
        y = expression("-log"[10]~"(P value)"),
        title="") +
   # geom_hline(yintercept=-log10(0.05), linetype="dashed") +
   # geom_vline(xintercept = 0, linetype="dashed") +
 geom_point(
     aes(x = x, y = y, colour=factor(sig)),
     colour = "grey70",
     size = 6,
     data = plot_table,
     show.legend = FALSE) +
  geom_point(
    aes(x = x, y = y),
    size = 8,
    data = highlight,
    color = "#3C5488FF",
    show.legend = FALSE) +
  geom_point(
    aes(x = x, y = y),
    shape = 1,
    size = 8,
    data = highlight,
    show.legend = FALSE) +
  geom_text_repel(
    aes(x = x, y = y, label = Term),
    size = 6.5,
    data = highlight,
    # parse = TRUE,
    box.padding = 1
  )
p
ggsave(paste0(sub_output_dir, "FigureS2D_sister_conserved_KEGG_volcano.pdf"), p,
       width = 8, height = 5.5)
```

# Figure 2: ReSisTrace uncovers the interplay between genetic and transcriptomic features in primed resistance

## A: UMAP drug sensitivity groups

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "Figure2/")

if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

We merged the scRNA-seq data from all of the 8 samples into one Seurat object. FindVariableFeatures and ScaleData are performed on this object. It is saved in "./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS". PCA and UMAP projections are calculated for this object and the results are saved in "./data/input/merged_KURAMOCHI_samples1_SCT_umap_2000_n50_scaleT.RDS". The codes for processing this Seurat object can be found in file "./UMAP_visualization.R"

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
data2 = data@assays$RNA@scale.data  # scale data is for dimension reduction or clustering
dim(data2) # 2000 77724

### UMAP visualization ----------------------------------------------------

data$drugSens[which(data$drugSens == "resistant")] = "post-treatment"
data2.umap = readRDS("./data/input/merged_KURAMOCHI_samples1_SCT_umap_2000_n50_scaleT.RDS")

index = which(data$drugSens %in% c("pre-sensitive", "pre-resistant","post-treatment"))

pdf(paste0(sub_output_dir, "Figure2A_umap_all.pdf"), width = 5, height = 5, pointsize = 12)
col = c("snow2", "snow4", "slategray4")
size = c(1,3,3)
tmp = factor(data$drugSens[index], levels = c("pre-sensitive", "pre-resistant", "post-treatment"))
plot(data2.umap$layout[index,], col = col[tmp], pch = '.', cex = size[tmp], xlab = "UMAP 1", ylab = "UMAP 2")
legend("bottomright", legend = levels(tmp), pch = 16, col = col[1:length(tmp)], cex = 1, pt.cex = 1)
dev.off()
table(data$drugSens[index])
```

## B: UMAP CNV subclone

```{r}
index = which(data$drugSens %in% c("pre-sensitive", "pre-resistant","post-treatment"))

pdf(paste0(sub_output_dir, "Figure2B_umap_all_cnv.pdf"), width = 5, height = 5, pointsize = 12)
col <- c(
  "A" = "#DC0000FF", "B" = "#B09C85FF", "C" = "#00A087FF",
  "D" = "#91D1C2FF",  "E" = "#3C5488FF"
  )
tmp = factor(data$CNV_subclone[index], levels = c("A", "B", "C", "D", "E"))
plot(data2.umap$layout[index,], col = col[tmp], pch = '.', xlab = "UMAP 1", ylab = "UMAP 2")
legend("topright", legend = levels(tmp), pch = 16, col = col[1:length(tmp)], cex = 1, pt.cex = 1)
dev.off()
table(data$CNV_subclone[index])
```

## C: Proportion of CNV subclones

```{r}
# proportion of sub-clones in pre-S
subclone = sort(unique(data$CNV_subclone))
# treatment = unique(data$treatment)
treatment = c("Carboplatin", "Olaparib", "NK", "Control") 
res2 = data.frame(mat.or.vec(length(treatment), length(subclone)))
count2 = data.frame(mat.or.vec(length(treatment), length(subclone)))
for(i in 1:length(treatment)){
  tmp = data$CNV_subclone[which(data$drugSens == "pre-sensitive" & data$treatment == treatment[i])]
  count2[i,] = table(tmp)
  x = table(tmp)/length(tmp)
  cat(treatment[i], "\n", table(tmp), "\n")
  res2[i,] = x
  names(res2) = names(x)
  names(count2) <- names(x)
}

rownames(res2) = treatment
res2 = melt(as.matrix(res2))
colnames(res2) <- c("treatment", "subclone", "proportion")
res2$drugSens <- "pre-sensitive"

rownames(count2) = treatment
count2 = melt(as.matrix(count2))
colnames(count2) <- c("treatment", "subclone", "count")
count2$drugSens <- "pre-sensitive"

values1 = c(red, brown, green, green2, blue)

p <- ggplot(res2, aes(x = treatment, y = proportion, fill = subclone)) +
  geom_col(position = "fill") +
  labs(
    y = "Proportion",
    fill = ""
  ) + 
  ggtitle("pre-sensitive") +
  scale_fill_manual(values = values1) +
  theme_bw() +
  theme(
    text = element_text(size = 21),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 18,
      color = "black"
    ),
    axis.text.y = element_text(
      size = 18,
      color = "black"
    ),
    axis.title.x = element_blank(),
    legend.text = element_text(size = 21),
    plot.title = element_text(hjust = 0.5, size = 21),
    panel.grid = element_blank()
  )
ggsave(paste0(sub_output_dir, "Figure2C_BRCA_CNV_preS.pdf"), width = 4, height = 4)

# pre-R
res3 = data.frame(mat.or.vec(length(treatment), length(subclone)))
count3 = data.frame(mat.or.vec(length(treatment), length(subclone)))
for(i in 1:length(treatment)){
  tmp = data$CNV_subclone[which(data$drugSens == "pre-resistant" & data$treatment == treatment[i])]
  count3[i, ] <- table(tmp)
  x = table(tmp)/length(tmp)
  res3[i,] = x
  names(res3) = names(x)
  names(count3) <- names(x)
}
rownames(res3) = treatment
res3
res3 = melt(as.matrix(res3))
colnames(res3) <- c("treatment", "subclone", "proportion")
res3$drugSens <- "pre-resistant"

rownames(count3) = treatment
count3 = melt(as.matrix(count3))
colnames(count3) <- c("treatment", "subclone", "count")
count3$drugSens <- "pre-resistant"

p <- ggplot(res3, aes(x = treatment, y = proportion, fill = subclone)) +
  geom_col(position = "fill") +
  labs(
    y = "Proportion",
    fill = "") + 
  ggtitle("pre-resistant") +
  scale_fill_manual(values = values1) +
  theme_bw() +
  theme(
    text = element_text(size = 21),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 18,
      color = "black"
    ),
    axis.text.y = element_text(
      size = 18,
      color = "black"
    ),
    axis.title.x = element_blank(),
    plot.title = element_text(hjust = 0.5, size = 21),
    panel.grid = element_blank()
  )
ggsave(paste0(sub_output_dir, "Figure2C_BRCA_CNV_preR.pdf"), width = 4, height = 4)

# Post-treatment
res4 = data.frame(mat.or.vec(length(treatment), length(subclone)))
count4 = data.frame(mat.or.vec(length(treatment), length(subclone)))
for(i in 1:length(treatment)){
  tmp = data$CNV_subclone[which(data$drugSens == "post-treatment" & data$treatment == treatment[i])]
  count4[i, ] <- table(tmp)
  x = table(tmp)/length(tmp)
  res4[i,] = x
  names(res4) = names(x)
  names(count4) <- names(x)
}
rownames(res4) = treatment
res4
res4 = melt(as.matrix(res4))
colnames(res4) <- c("treatment", "subclone", "proportion")
res4$drugSens <- "post-treatment"

rownames(count4) = treatment
count4 = melt(as.matrix(count4))
colnames(count4) <- c("treatment", "subclone", "count")
count4$drugSens <- "post-treatment"

p <- ggplot(res4, aes(x = treatment, y = proportion, fill = subclone)) +
  geom_col(position = "fill") +
  labs(
    y = "Proportion",
    fill = "") + 
  ggtitle("post-treatment") +
  scale_fill_manual(values = values1) +
  theme_bw() +
  theme(
    text = element_text(size = 21),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 18,
      color = "black"
    ),
    axis.text.y = element_text(
      size = 18,
      color = "black"
    ),
    axis.title.x = element_blank(),
    plot.title = element_text(hjust = 0.5, size = 21),
    panel.grid = element_blank()
  )
ggsave(paste0(sub_output_dir, "Figure2C_BRCA_CNV_postT.pdf"), width = 4, height = 4)
```

```{r}
# paired t tes
res <- res2 %>% 
  rbind.data.frame(res3) %>% 
  rbind.data.frame(res4)
write.csv(res, paste0(sub_output_dir, "Figure2C_cell_proportion_table.csv"), row.names = FALSE)

preR_A <- res$proportion[res$subclone == "A" & res$drugSens == "pre-resistant"]
preS_A <- res$proportion[res$subclone == "A" & res$drugSens == "pre-sensitive"]
t_test <- t.test(preR_A, preS_A, paired = TRUE)
t_test
```

```{r}
# proportion test
count <- count2 %>% 
  rbind.data.frame(count3) %>% 
  rbind.data.frame(count4)
write.csv(count, paste0(sub_output_dir, "Figure2C_cell_count_table.csv"), row.names = FALSE)

preR_control <- sum(count$count[count$drugSens == "pre-resistant" & treatment == "Control"])
preS_control <- sum(count$count[count$drugSens == "pre-sensitive" & treatment == "Control"])
cat("preR vs preS control", "\n")

p_value <- NULL
for (i in LETTERS[1:5]){
  preR_sub_control <- count$count[count$subclone == i & count$drugSens == "pre-resistant" & treatment == "Control"]
  preS_sub_control <- count$count[count$subclone == i & count$drugSens == "pre-sensitive" & treatment == "Control"]
  for (t in c("Carboplatin", "Olaparib", "NK")){
    preR_sub <- count$count[count$subclone == i & count$drugSens == "pre-resistant" & treatment == t]
    preR_treatment <- sum(count$count[count$drugSens == "pre-resistant" & treatment == t])
    p_r <- prop.test(x = c(preR_sub, preR_sub_control), n = c(preR_treatment, preR_control))
    
    tmp <- data.frame(
      subclone = i,
      treatment = t,
      drugSens = "pre-resistant",
      p_value = p_r$p.value,
      prop_treat_or_preR = p_r$estimate[1],
      prop_control_or_preS = p_r$estimate[2]
    )
    p_value <- rbind.data.frame(p_value, tmp)
        
    preS_sub <- count$count[count$subclone == i & count$drugSens == "pre-sensitive" & treatment == t]
    preS_treatment <- sum(count$count[count$drugSens == "pre-sensitive" & treatment == t])
    p_s <- prop.test(x = c(preS_sub, preS_sub_control), n = c(preS_treatment, preS_control))
    
    tmp <- data.frame(
      subclone = i,
      treatment = t,
      drugSens = "pre-sensitive",
      p_value = p_s$p.value,
      prop_treat_or_preR = p_s$estimate[1],
      prop_control_or_preS = p_s$estimate[2]
    )
    p_value <- rbind.data.frame(p_value, tmp)
    
    p_rs <- prop.test(x = c(preR_sub, preS_sub), n = c(preR_treatment, preS_treatment))
    
    tmp <- data.frame(
      subclone = i,
      treatment = t,
      drugSens = "preR vs preS",
      p_value = p_rs$p.value,
      prop_treat_or_preR = p_rs$estimate[1],
      prop_control_or_preS = p_rs$estimate[2]
    )
    p_value <- rbind.data.frame(p_value, tmp)
  }
}

write.csv(p_value, paste0(sub_output_dir, "Figure2C_propoetion_test_table.csv"), row.names = FALSE)
```

## D: Correlation between CNV subclone DEG and preR vs preS DEG

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "Figure2/")

if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
conserve <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  select(ensembl, conserve_gene)
```

The DEG list from different replicates are merged for each treatment. The values for non-treated control samples are appended to each DEG list for each anti-cancer treatment. The final lists are saved in "./data/input/DEG_filter_by_bootstrap_permutation_in_singleton/"

```{r}
treatment <- c("Carbo", "Olaparib", "NK")
deg <- NULL
for (t in treatment){
  tmp <- read.csv(
  paste0(
    "./data/input/DEG_filter_by_bootstrap_permutation_in_singleton/prer_pres_",
    t,
    "_control_bootstrap_permutation_summary.csv")
  )
  tmp[[paste0(t, "_log2FC_preR_preS_adj")]] <- tmp$mean_log2FC_treat - tmp$mean_log2FC_ctrl
  tmp[[paste0(t, "_boot_p")]] <- tmp[[paste0("boot_", t, "_ctrl_p_value")]]
  tmp[[paste0(t, "_perm_p")]] <- tmp[[paste0(t, "_perm_p")]]
  tmp <- tmp[, c("ensembl", "symbol", paste0(t, "_log2FC_preR_preS_adj"), paste0(t, "_boot_p"), paste0(t, "_perm_p"))]
  if (is.null(deg)){
    deg <- tmp
  } else {
    deg <- left_join(deg, tmp, by = c("ensembl", "symbol"))
  }
}
```

DEG lists for each CNV groups vs. other cells are calculated on the CSC Puhti server with the batch job. The codes are available in the files: "./codes_for_batch_job/CNV_subclone_DEG.R", "./codes_for_batch_job/CNV_subclone_DEG.sh", and "./codes_for_batch_job/sub_CNV_subclone_DEG.sh"

```{r}
for (s in LETTERS[1:5]){
  CNV_deg <- read.csv(
    paste0(
      "./data/input/CNV_DEG/",
      s,
      "_vs_others_DEG.csv")
    ) %>% 
    select(X, avg_log2FC, p_val_adj)
  colnames(CNV_deg) <- c("ensembl", paste0("log2FC_", s), paste0("p_adj_", s))
  deg <- deg %>% 
    left_join(CNV_deg, by = "ensembl")
}

deg <- deg %>% 
  left_join(conserve, by = "ensembl")
write.csv(deg, paste0(sub_output_dir, "Figure2D_log2FC_preR_preS_CNV_subclone.csv"), row.names = FALSE)

deg <- deg %>% 
  filter(conserve_gene)
```

#### Correlation coefficient

```{r}
treatments <- c("NK", "Carbo", "Olaparib")
subclones <- c("A", "B", "C", "D", "E")
plot_table <- NULL
for (t in treatments){
  for (s in subclones){
    tmp <- cor.test(
      deg[[paste0(t, "_log2FC_preR_preS_adj")]],
      deg[[paste0("log2FC_", s)]]
    )
    df <- data.frame(
      treatment = case_when(
        t == "Carbo" ~ "Carboplatin", 
        TRUE ~ t
      ),
      subclone = s,
      correlation = tmp$estimate,
      CI_lower = tmp$conf.int[1],
      CI_upper = tmp$conf.int[2],
      p_value = tmp$p.value,
      sig = case_when(
        tmp$p.value < .05 & tmp$p.value >= .01 ~ "*", 
        tmp$p.value < .01 & tmp$p.value >= .001 ~ "**", 
        tmp$p.value < .001 ~ "***", 
        TRUE ~ ""
      ),
      n = sum(
        !is.na(deg[[paste0("log2FC_", s)]]) & 
        !is.na(deg[[paste0(t, "_log2FC_preR_preS_adj")]])
      )
    )
    if (is.null(plot_table)){
      plot_table <- df
    } else {
      plot_table <- rbind.data.frame(plot_table, df)
    }
  }
}
plot_table$sig_y <- plot_table$CI_upper + 0.05
plot_table$sig_y[plot_table$correlation < 0] <- plot_table$CI_lower[plot_table$correlation < 0] - 0.05
write.csv(plot_table, paste0(sub_output_dir, "Figure2D_Pearson_corelation_coefficient.csv"),
          row.names = FALSE)
plot_table$treatment <- factor(plot_table$treatment, levels = c("Carboplatin", "Olaparib", "NK"))
```

```{r}
color <- c("A" = "#DC0000FF", "B" = "#B09C85FF", "C" = "#00A087FF", "D" = "#91D1C2FF", "E" = "#3C5488FF")

ggplot(
  plot_table,
  aes(y = correlation, x = treatment, fill = subclone)
  ) +
  scale_fill_manual(
    values = color
  ) + 
  geom_bar(stat="identity", position="dodge") + 
  geom_errorbar(
    aes(ymin=CI_lower, ymax=CI_upper),
    position=position_dodge(width=.9),
    width=.15
  ) + 
  geom_text(
    aes(y=sig_y, label=sig),
    position=position_dodge(width=.9),
    size = 6
  ) + 
  labs(x="", y="Correlation coefficient", fill = "") +
  theme_classic() +
  theme(
     legend.key = element_blank(), 
     axis.text = element_text(size = 21, colour = "black"),
     axis.title = element_text(size = 21),
     legend.text = element_text(size = 21)
     )
ggsave(
  paste0(sub_output_dir, "Figure2D_Pearson_correlation_coefficients_color_by_subclones.pdf"),
  width = 10, height = 5)
```

#### Connectivity score

```{r include=FALSE}
treatments <- c("NK", "Carbo", "Olaparib")
subclones <- c("A", "B", "C", "D", "E")
plot_table <- NULL
for (t in treatments){
  for (s in subclones){
    xx <- data.frame(
      value = deg[[paste0(t, "_log2FC_preR_preS_adj")]],
      sig = deg[[paste0(t, "_boot_p")]]
    )
    rownames(xx) <- deg$ensembl
    xx <- na.omit(xx)
    yy <- data.frame(
      value = deg[[paste0("log2FC_", s)]],
      sig = deg[[paste0("p_adj_", s)]]
    )
    rownames(yy) <- deg$ensembl
    yy <- na.omit(yy)
    connect <- PharmacoGx::connectivityScore(
      x = xx,
      y = yy,
      method='gwc',
      gwc.method='spearman',
      nperm = 1000
      )
    
    df <- data.frame(
      treatment = case_when(
        t == "Carbo" ~ "Carboplatin", 
        TRUE ~ t
      ),
      subclone = s,
      connectivity_score = connect[1],
      p_value = connect[2],
      sig = case_when(
        connect[2] < .05 & connect[2] >= .01 ~ "*", 
        connect[2] < .01 & connect[2] >= .001 ~ "**", 
        connect[2] < .001 ~ "***", 
        TRUE ~ ""
      ),
      n = sum(
        !is.na(deg[[paste0("log2FC_", s)]]) & 
        !is.na(deg[[paste0(t, "_log2FC_preR_preS_adj")]])
      )
    )
    if (is.null(plot_table)){
      plot_table <- df
    } else {
      plot_table <- rbind.data.frame(plot_table, df)
    }
  }
}
plot_table$sig_y <- plot_table$connectivity_score + 0.05
plot_table$sig_y[plot_table$connectivity_score < 0] <- plot_table$connectivity_score[plot_table$connectivity_score < 0] - 0.05
write.csv(plot_table, paste0(sub_output_dir, "Figure2D_connectivity_score.csv"),
          row.names = FALSE)
```

```{r}
plot_table <- read.csv(paste0(sub_output_dir, "Figure2D_connectivity_score.csv"))
plot_table$treatment <- factor(plot_table$treatment, levels = c("Carboplatin", "Olaparib", "NK"))
color <- c("A" = "#DC0000FF", "B" = "#B09C85FF", "C" = "#00A087FF", "D" = "#91D1C2FF", "E" = "#3C5488FF")

ggplot(
  plot_table,
  aes(y = connectivity_score, x = treatment, fill = subclone)
  ) +
  scale_fill_manual(
    values = color
  ) + 
  geom_bar(stat="identity", position="dodge") + 
  # geom_errorbar(
  #   aes(ymin=CI_lower, ymax=CI_upper),
  #   position=position_dodge(width=.9),
  #   width=.15
  # ) + 
  geom_text(
    aes(y=sig_y, label=sig),
    position=position_dodge(width=.9),
    size = 21/.pt
  ) + 
  labs(x="", y="Connectivity score", fill = "") +
  theme_classic() +
  theme(
     legend.key = element_blank(), 
     axis.text = element_text(size = 18, colour = "black"),
     axis.title = element_text(size = 21),
     legend.text = element_text(size = 15)
     )
# ggsave(
#   paste0(sub_output_dir, "Figure2D_connectivity_score_color_by_subclones.pdf"),
#   width = 7, height = 4) # Science format
ggsave(
  paste0(sub_output_dir, "Figure2D_connectivity_score_color_by_subclones.pdf"),
  width = 9.3, height = 5.5)
```

## E: Representative gene ontologies enriched in the pre-resistance signatures as well as the signatures of subclones A and E

This figure is generated by Matias.

# Figure S3

## A: UMAP projection of pre-sensitive, pre-resistant, and post-treatment cells for each treatment

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS3/")

if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
data2 = data@assays$RNA@scale.data  # scale data is for dimension reduction or clustering
dim(data2) # 2000 77724

### UMAP visualization ----------------------------------------------------

data$drugSens[which(data$drugSens == "resistant")] = "post-treatment"
data2.umap = readRDS("./data/input/merged_KURAMOCHI_samples1_SCT_umap_2000_n50_scaleT.RDS")
index = which(data$drugSens %in% c("pre-sensitive", "pre-resistant","post-treatment"))
```

```{r}
index = which(data$drugSens %in% c("pre-sensitive", "pre-resistant","post-treatment"))
```

```{r}
# individual treatment conditions
for(i in unique(data$treatment)){
  for(j in c("pre-sensitive", "pre-resistant", "post-treatment")){
    pdf(
      paste(sub_output_dir, "FigureS3A_", paste(i, j, sep="_"),".pdf",sep=""),
      width = 3.7, height = 3.5, pointsize = 10)
    
    index1 = which(data$drugSens == j & data$treatment == i)
    # if(j == "pre-resistant") color = red
    # if(j == "pre-sensitive") color = blue
    # if(j == "post-treatment") color = brown
    if(i == "Olaparib") color = red
    if(i == "Carboplatin") color = blue
    if(i == "NK") color = brown
    if(i == "Control") color = green
    scatterplot(
      data2.umap$layout[index1, 1],
      data2.umap$layout[index1, 2], cex = 0.5, cex.main = 1.4, cex.axis = 1.2,
      col = color, main = paste(i,j),
      boxplots = FALSE, las = 1, smooth = FALSE, grid = TRUE,
      regLine = FALSE, ellipse = TRUE, pch = 20, 
      xlim = c(-10,10), ylim = c(-10,10),
      legend = list(title = "treatment", coords = "topleft"),
      # xlab = "UMAP 1", ylab = "UMAP 2"
      xlab = "", ylab = ""
      )
    dev.off()
    cat(i, ", ", j, ", ", sum(index1), "\n")
  }
  print(i)
  print(t.test(data2.umap$layout[which(data$drugSens == "pre-resistant" & data$treatment == i), 1], data2.umap$layout[which(data$drugSens == "pre-sensitive" & data$treatment == i),1]))
  print(t.test(data2.umap$layout[which(data$drugSens == "pre-resistant" & data$treatment == i), 2], data2.umap$layout[which(data$drugSens == "pre-sensitive" & data$treatment == i),2]))
}
```

## B: Cumulative distribution

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS3/")

if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

```{r message=FALSE}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS") 

# options(scipen = 999)
# memory.limit(size = 54000)
data2 = data@assays$RNA@data

# sink(paste0(sub_output_dir, "FigureS3B_shannon.txt"))
res = list()
k = 1
for (i in unique(data$sample)){
  for (j in unique(data$time_point)){
    index = intersect(grep(i, data@meta.data$prer_r_group), which(data$time_point == j))
    tmp = data.frame(table(data@meta.data$prer_r_group[index]))
    shannon = diversity(tmp$Freq)
    print(c(gsub('[0-9]+', '', i), i,j,shannon))
    tmp$sample = i
    tmp$time_point = j
    tmp$treatment = gsub('[0-9]+', '', i)
    res[[k]] = tmp
    k = k + 1
  }
}
tmp = do.call(rbind, res)
tmp$treatment[tmp$treatment == "Carbo"] <- "Carboplatin"
tmp$treatment[tmp$treatment == "ctrl"] <- "Control"
tmp$treatment = factor(tmp$treatment, levels = c("Carboplatin", "Olaparib", "NK", "Control"))
tmp$time_point[tmp$time_point == "BeforeTreatment"] <- "pre-treatment"
tmp$time_point[tmp$time_point == "AfterTreatment"] <- "post-treatment"
tmp$time_point = factor(tmp$time_point, levels = c("pre-treatment", "post-treatment"))
write.csv(paste0(sub_output_dir, "FigureS3B_shannon.txt"), row.names = FALSE)
# sink()
```

```{r}
ks.test(tmp$Freq[tmp$time_point == "pre-treatment"], tmp$Freq[tmp$time_point == "post-treatment"], exact = TRUE)
```

Reference: https://rgraphs.com/how-to-create-a-cumulative-frequency-graph-in-r/

```{r message=FALSE}
res <- read.table(paste0(sub_output_dir, "FigureS3B_shannon.txt"))

values1 = c(blue, red, brown, green)

fig1 = ggplot(tmp, aes(Freq, y = 1 - ..y.., color = time_point)) +
  stat_ecdf(geom = "step", size = 1) + 
  labs(
    x = "Clone size",
    y = "Cumulative proportion",
    color = "") +
  scale_y_continuous(trans = 'log10') + scale_x_log10() +
  theme_classic() +
  theme(
    text = element_text(size = 20),
    axis.title = element_text(size = 20),
    axis.text = element_text(color = "black", size = 20),
    legend.text = element_text(size = 20),
    legend.title = element_blank(),
    # legend.position=c(.7,.9)
  ) +
  scale_color_manual(values = values1)
ggsave(
  paste0(sub_output_dir, "FigureS3C_cumulate_all.pdf"),
  fig1, width = 9, height = 5
)


```

## C: Shannon diversity

```{r}
shannon = read.table(paste0(sub_output_dir, "FigureS3B_shannon.txt"))
names(shannon) = c("V1", "Treatment","sample","time_point","value")
shannon$Treatment[shannon$Treatment == "Carbo"] <- "Carboplatin"
shannon$Treatment[shannon$Treatment == "ctrl"] <- "Control"
shannon$Treatment = factor(shannon$Treatment, levels = c("Carboplatin", "Olaparib", "NK", "Control"))
shannon$time_point[shannon$time_point == "BeforeTreatment"] <- "pre-treatment"
shannon$time_point[shannon$time_point == "AfterTreatment"] <- "post-treatment"
shannon$time_point = factor(shannon$time_point, levels = c("pre-treatment", "post-treatment"))
values1 = c(blue, red, brown, green)

p <- ggplot(shannon, aes(x = time_point, y = value)) +
  geom_boxplot() + 
  geom_point(aes(x = time_point, color = Treatment), size = 4) +
  geom_line(aes(group = sample, color = Treatment)) +
  xlab("") + 
  ylab("Shannon diversity index") +
  theme_classic() +
  theme(
    text = element_text(size = 20),
    axis.title = element_text(size = 20),
    axis.text = element_text(color = "black", size = 20),
    legend.text = element_text(size = 20),
    legend.title = element_blank(),
    # legend.position = c(0.9, 0.9),
    legend.background = element_blank()
  ) +
  scale_color_manual(values = values1)
ggsave(
  paste0(sub_output_dir, "FiguerS3B_shannon.pdf"),
  p,
  width = 9, height = 5
  )

# Paired t test

pre <- shannon$value[shannon$time_point == "pre-treatment"]
post <- shannon$value[shannon$time_point == "post-treatment"]
t_test <- t.test(pre, post, paired = TRUE)
t_test
```

# Figure 3: Functional validations/interpretations

## C: TCVA BRCAness/HRD survival plot

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "Figure3/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

Load lineage tracing data

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS") # Shuyu update on 1.4.2022 after FindVariableFeatures and Scale, check test.R in the server
sister_conserve_gene <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  filter(conserve_gene) %>% 
  select(ensembl) %>% 
  unlist()
data2 = data@assays$RNA@data[rownames(data) %in% sister_conserve_gene, ]
dim(data2)
unique(data$drugSens)
unique(data$sample)
unique(data$treatment)
unique(data$time_point)
```

load TCGA data

The clinical data for patients from TCGA is saved in "./data/input/BRCAness/data_clinical_patient.txt". The clinical data for sequencing samples from TCGA is saved in "./data/input/BRCAness/data_clinical_sample.txt". The gene expression matrix after deconvolution with single-cell RNA sequencing data are saved in "./data/input/BRCAness/EOC_deconv_TCGA_expr_mat.RDS"

```{r}
data_clinical_patient = read.delim("./data/input/BRCAness/data_clinical_patient.txt", skip = 4)
data_clinical_sample = read.delim("./data/input/BRCAness/data_clinical_sample.txt", skip = 4)
data_mrna = readRDS("./data/input/BRCAness/EOC_deconv_TCGA_expr_mat.RDS")
colnames(data_mrna) = gsub("\\.","\\-", colnames(data_mrna))
colnames(data_mrna) = substr(colnames(data_mrna), 1, 15)
data_mrna <- data_mrna[apply(data_mrna, 1, function(x){
  sum(is.na(x))
}) < 300, ]

identical(data_clinical_patient$PATIENT_ID, data_clinical_sample$PATIENT_ID) # data aligned already
table(data_clinical_sample$GRADE)
high_grade = grepl("G2|G3|G4", data_clinical_sample$GRADE)

data_clinical_sample = data_clinical_sample[high_grade, ]
data_clinical_patient = data_clinical_patient[high_grade, ]

patients = names(data_mrna) # 307 patients
time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]
table(status, exclude = NULL)
status[which(status == "")] = NA

length(which(is.na(time)==F)) # 270 has survival
length(which(is.na(status)==F)) # 271 has status
```

The DEG list for HRD vs HRP groups from deconvoluted TCGA data is saved in "./data/input/BRCAness/TCGA_deconvoluted_HRD_DEGs.tsv".

```{r}
tcga = read.csv("./data/input/BRCAness/TCGA_deconvoluted_HRD_DEGs.tsv", sep = "\t")
tcga[which(rownames(tcga) == "BRCA1"), ] # should be negative logFC
tcga$ensembl = data@misc$gene_meta$ensembl[match(rownames(tcga), data@misc$gene_meta$symbol)]
tcga$symbol = rownames(tcga)
length(which(is.na(tcga$ensembl)==T)) # 383 genes have no emsembl
```

### Test for good p value

```{r}
p_threshold = c(1*10^-(2:4), 5*10^-(2:4))
fc_threshold = 0
conserve_gene_only <- FALSE

res <- mat.or.vec(length(p_threshold), 3)
colnames(res) <- c("p_threshold", "n_gene", "survival_p")

for (i in 1:length(p_threshold)){
  data_mrna2 <- data_mrna
  if (conserve_gene_only){
    tcga_test <- filter(tcga, ensembl %in% sister_conserve_gene)
  } else {
    tcga_test <- tcga
  }
  
  res[i, 1] <- p_threshold[i]
  index1 = which(tcga_test$logFC <= -fc_threshold & tcga_test$adj.P.Val <= p_threshold[i]) # down-regulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(tcga_test$logFC > fc_threshold & tcga_test$adj.P.Val <= p_threshold[i]) # up-regulated genes when BRCA1 is knocked out
  index2_symbol = tcga_test$symbol[index2]
  
  index = c(index1, index2)
  length(index)
  tcga_test = tcga_test[index, ]
  # tcga_test$symbol
  
  tcga_test = tcga_test[match(rownames(data_mrna2), tcga_test$symbol), ]
  
  res[i, 2] <- nrow(na.omit(tcga_test))
  
  # reverse index2 gene expressions
  data_mrna2[which(rownames(rownames(data_mrna2)) %in% index2_symbol), ] = -data_mrna2[which(rownames(rownames(data_mrna2)) %in% index2_symbol), ]
  
  # gene selection
  index = which(is.na(tcga_test$symbol) == F)
  res1 = c()
  for (j in 1:length(patients)){
    cat('\r', j)
    res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
    res1[j] = weighted.mean(res.tmp[index], w = abs(tcga_test$logFC[index]), na.rm = T) # rank of the BRCAness genes
    # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362$avg_log2FC[index])*(-log10(cov362$p_val_adj[index]))) # rank of the BRCAness genes
    
  }
  
  group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
  # fiveyear = which(time < 60)
  plot_data = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
  fit <- survfit(Surv(time, status) ~ group, data = plot_data)
  fit.coxph = coxph(Surv(time, status) ~ group, data = plot_data)
  s <- summary(fit.coxph)
  res[i, 3] <- s$logtest[3]
}

write.csv(
  res,
  paste0(sub_output_dir, "Figure3B_BRCA_survival_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

### Survival plot with selected parameters

```{r}
conserve_gene_only <- FALSE

if (conserve_gene_only) {
  selected_p_threshold = 0.005
  selected_fc_threshold = 0
} else {
  selected_p_threshold = 0.005
  selected_fc_threshold = 0
}

if (conserve_gene_only){
  tcga_test <- filter(tcga, ensembl %in% sister_conserve_gene)
} else {
  tcga_test <- tcga
}

data_mrna2 <- data_mrna
index1 = which(tcga_test$logFC <= -selected_fc_threshold & tcga_test$adj.P.Val <= selected_p_threshold) # down-regulated genes when BRCA1 is knocked out (BRCA-)
index2 = which(tcga_test$logFC > selected_fc_threshold & tcga_test$adj.P.Val <= selected_p_threshold) # up-regulated genes when BRCA1 is knocked out
index2_symbol = tcga_test$symbol[index2]

index = c(index1, index2)
length(index)
tcga_test = tcga_test[index, ]
# tcga_test$symbol

tcga_test = tcga_test[match(rownames(data_mrna2), tcga_test$symbol), ]

# reverse index2 gene expressions
data_mrna2[which(rownames(rownames(data_mrna2)) %in% index2_symbol), ] = -data_mrna2[which(rownames(rownames(data_mrna2)) %in% index2_symbol), ]

# gene selection
index = which(is.na(tcga_test$symbol) == F)
res1 = c()
for (j in 1:length(patients)){
  cat('\r', j)
  res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
  res1[j] = weighted.mean(res.tmp[index], w = abs(tcga_test$logFC[index]), na.rm = T) # rank of the BRCAness genes
  # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362$avg_log2FC[index])*(-log10(cov362$p_val_adj[index]))) # rank of the BRCAness genes
  
}

group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
# fiveyear = which(time < 60)
plot_data = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
fit <- survfit(Surv(time, status) ~ group, data = plot_data)
fit.coxph = coxph(Surv(time, status) ~ group, data = plot_data)

s <- summary(fit.coxph)
  
dim(tcga_test)
concordance(fit.coxph)$concordance

pdf(
  paste0(sub_output_dir, "Figure3B_threshold_", selected_p_threshold, "_", ifelse(conserve_gene_only, "BRCA_TCGA_OS_conserve_gene.pdf", "BRCA_TCGA_OS.pdf")),
  width = 4, height = 5.5, pointsize = 20
)
ggsurvplot(
  fit,
  data = plot_data,
  palette = c(hrd_color, hrp_color),
  risk.table = T,
  legend.title = "HRD (TCGA)",# - undeconvoluted",
  legend.labs = c("high","low"),
  xlab = "Time (months)", ylab = "Overall survival",
  ggtheme = theme_survminer(
    base_size = 12,
    font.main = c(15, "plain", "black"),
    font.x = c(14, "plain", "black"),
    font.y = c(14, "plain", "black"),
    font.caption = c(14, "plain", "black"),
    font.legend = c(14, "plain", "black"),
    font.tickslab = c(14, "plain", "black"),
    font.submain = c(14, "plain", "black")
  ),
  pval = signif(s$logtest[3], 2),
  pval.coord = c(100, 0.75)
)+
  # annotate("text", x = 150, y =0.75, label = paste0("P = ", round(s$logtest[3], 2))) +
  guides(colour = guide_legend(nrow = 2))
dev.off()
```

## B: Validate for cell line model

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "Figure3/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}

data_summary <- function(data, varname, groupnames){
  require(plyr)
  summary_func <- function(x, col){
    c(mean = mean(x[[col]], na.rm=TRUE),
      sd = sd(x[[col]], na.rm=TRUE))
  }
  data_sum<-ddply(data, groupnames, .fun=summary_func,
                  varname)
  data_sum <- rename(data_sum, c("mean" = varname))
 return(data_sum)
}
```

The experiment tables are from Jun

### COV362

```{r}
plot_table <- openxlsx::read.xlsx("./data/input/HRD_experiment/COV362-4-5-6-20220914-6DAY-BLACK.xlsx", rows = 59:73, cols = 2:6)
plot_table$`Olaparib.(uM)` <- rep(c("COV362", "COV362-control", "COV362-BRCA1"), each = 4)
plot_table <- plot_table %>% 
  dplyr::rename("cell" = "Olaparib.(uM)") %>% 
  pivot_longer(!cell, names_to = "Olaparib", values_to = "Viability")
plot_table$Viability <- plot_table$Viability * 100
plot_table$cell <- factor(plot_table$cell, levels = c("COV362", "COV362-control", "COV362-BRCA1"))
col = c(hrd_color, hrd_vector_color, hrp_color)

stat.test <- plot_table %>%
  group_by(Olaparib) %>%
  t_test(Viability ~ cell) %>%
  add_xy_position(
    fun = "mean_sd", x = "Olaparib", dodge = 0.8,
    step.increase = 0.05) %>%
  add_significance("p")

# Bar plot
p <- ggbarplot(
  plot_table,
  x = "Olaparib",
  y = "Viability",
  add = "mean_sd",
  fill = "cell",
  palette = col,
  position = position_dodge(0.8)) +
  stat_pvalue_manual(
    stat.test,  label = "p.signif", tip.length = 0.01,
    # bracket.nudge.y = - 2,
    hide.ns = TRUE,
    size = 14/.pt
  ) +
  labs(
    x = "Olaparib (uM)",
    y = "Cell viability (%)"
  ) +
  # geom_point(
  #   data = plot_table,
  #   aes(
  #     x = Olaparib, y = Viability,
  #     group = interaction(Olaparib,cell)
  #   ),
  #   position = position_dodge(0.8)#,
  #   # size = 1,
  #   # alpha = 0.5
  # ) +
  scale_y_continuous(expand = c(0, 0)) +
  theme_classic() +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 21),
    legend.title = element_blank(),
    text = element_text(size = 21, color = "black", face = "plain"),
    axis.text = element_text(size = 21, color = "black", face = "plain")
  )
ggsave(
  paste0(sub_output_dir, "Figure3C_COV362_Olaparib_bar_plot.pdf"),
  width = 4.5, height = 4
)
ggsave(
  paste0(sub_output_dir, "Figure3C_COV362_Olaparib_bar_plot.png"),
  width = 6, height = 4
)

# line plot
plot_table$Olaparib <- as.numeric(plot_table$Olaparib)
plot_table_line <- data_summary(plot_table, varname="Viability", 
                    groupnames=c("cell", "Olaparib"))
p <- ggplot(
  plot_table_line,
  aes(
    x = Olaparib,
    y = Viability,
    color = cell,
    group = cell
  )
) +
  geom_errorbar(
    aes(ymin = Viability - sd, ymax = Viability + sd),
    width = 0.2,
    size = 1) +
  geom_line(size = 1) +
  geom_point() +
  scale_color_manual(values = col) +
  scale_y_continuous(limits = c(0, 120), breaks=seq(0, 100, 25), expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0)) +
  labs(
    x = "Olaparib (uM)",
    y = "Cell viability (%)"
  ) +
  theme_classic() +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 21),
    legend.title = element_blank(),
    text = element_text(size = 21, color = "black"),
    axis.text = element_text(size = 21, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "Figure3C_COV362_Olaparib_line_plot.pdf"),
  width = 7.2, height = 3
)
ggsave(
  paste0(sub_output_dir, "Figure3C_COV362_Olaparib_line_plot.png"),
  width = 7, height = 3.5
)
```

### UWB1.289

```{r}
control <- openxlsx::read.xlsx(
  "./data/input/HRD_experiment/UWB1.289 UWB1.289+BRCA1 Olaparib 4days-20210517-new.xlsx",
  rows = 46:49,
  cols = 4:12
)
colnames(control) <- rep(c(0, 1, 5), each = 3)
control <- control %>% 
  pivot_longer(everything(), names_to = "Olaparib", values_to = "Viability") %>% 
  mutate(cell = "UWB1.289")

BRCA <- openxlsx::read.xlsx(
  "./data/input/HRD_experiment/UWB1.289 UWB1.289+BRCA1 Olaparib 4days-20210517-new.xlsx",
  rows = 51:53,
  cols = 4:12
)
colnames(BRCA) <- rep(c(0, 1, 5), each = 3)
BRCA <- BRCA %>% 
  pivot_longer(everything(), names_to = "Olaparib", values_to = "Viability") %>% 
  mutate(cell = "UWB1.289 + BRCA1")

plot_table <- control %>% 
  rbind.data.frame(BRCA)
plot_table$Viability <- plot_table$Viability * 100
plot_table$cell <- factor(plot_table$cell, levels = c("UWB1.289", "UWB1.289 + BRCA1"))

col = c(hrd_color, hrp_color)

stat.test <- plot_table %>%
  group_by(Olaparib) %>%
  t_test(Viability ~ cell) %>%
  add_xy_position(
    fun = "mean_sd", x = "Olaparib", dodge = 0.8,
    step.increase = 0.05) %>%
  add_significance("p")

# Bar plot
p <- ggbarplot(
  plot_table,
  x = "Olaparib",
  y = "Viability",
  add = "mean_sd",
  fill = "cell",
  palette = col,
  position = position_dodge(0.8)) +
  stat_pvalue_manual(
    stat.test,  label = "p.signif", tip.length = 0.01,
    # bracket.nudge.y = - 2,
    hide.ns = TRUE,
    size = 5
  ) +
  labs(
    x = "Olaparib (uM)",
    y = "Cell viability (%)"
  ) +
  # geom_point(
  #   data = plot_table,
  #   aes(
  #     x = Olaparib, y = Viability,
  #     group = interaction(Olaparib,cell)
  #   ),
  #   position = position_dodge(0.8)#,
  #   # size = 1,
  #   # alpha = 0.5
  # ) +
  theme_classic() +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 21),
    legend.title = element_blank(),
    text = element_text(size = 21, color = "black"),
    axis.text = element_text(size = 21, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "Figure3C_UWB1.289_Olaparib_bar_plot.pdf"),
  width = 4.5, height = 4
)
ggsave(
  paste0(sub_output_dir, "Figure3C_UWB1.289_Olaparib_bar_plot.png"),
  width = 6, height = 4
)

# line plot
plot_table$Olaparib <- as.integer(plot_table$Olaparib)
plot_table_line <- data_summary(plot_table, varname="Viability", 
                    groupnames=c("cell", "Olaparib"))
p <- ggplot(
  plot_table_line,
  aes(
    x = Olaparib,
    y = Viability,
    color = cell,
    group = cell
  )
) +
  geom_errorbar(
    aes(ymin = Viability - sd, ymax = Viability + sd),
    width = 0.2,
    size = 1) +
  geom_line(size = 1) +
  geom_point() +
  scale_color_manual(values = col) +
  scale_y_continuous(limits = c(0, 120), breaks=seq(0, 100, 25), expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0)) +
  labs(
    x = "Olaparib (uM)",
    y = "Cell viability (%)"
  ) +
  theme_classic() +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 21),
    legend.title = element_blank(),
    text = element_text(size = 21, color = "black"),
    axis.text = element_text(size = 21, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "Figure3C_UWB1.289_Olaparib_line_plot.pdf"),
    width = 7.2, height = 3
)
ggsave(
  paste0(sub_output_dir, "Figure3C_UWB1.289_Olaparib_line_plot.png"),
  width = 7.3, height = 3.5
)
```

### HRD experiment results

Data table comes from Jun from HRD experiment.

#### COV362

```{r}
cov <- openxlsx::read.xlsx("./data/input/HRD_experiment/HR_test_Sofia_20221011_to_Shuyu.xlsx", sheet = 1, rows = 1:56, cols = c(1, 10))
names(cov) <- c("cell", "RAD51")
# cov$cell <- factor(
#   rep(c("WT", "WT-control", "WT-BRCA1"), each = 4),
#   levels = c("WT", "WT-control", "WT-BRCA1")
# )
cov$cell <- factor(
  rep(c("COV362", "COV362 + ctrl vector", "COV362 + BRCA1"), each = 4),
  levels = c("COV362", "COV362 + ctrl vector", "COV362 + BRCA1")
)
cov$group
cov <- na.omit(cov)
cov$group <- "COV362"
```

```{r}
# plot_table <- Reduce(
#   rbind.data.frame,
#   # list(cov, uwb, kuramochi)
#   list(cov, uwb))
plot_table <- cov
# plot_table$group <- factor(
#   plot_table$group,
#   levels = c("COV362", "UWB1.289")#, "KURAMOCHI")
#   )
col = c(
  hrd_color, hrd_vector_color, hrp_color#, hrd_color, hrp_color#,
  #hrd_color, hrd_vector_color, hrp_color
)

stat.test <- plot_table %>%
  group_by(group) %>% 
  t_test(RAD51 ~ cell) %>%
  add_xy_position(
    fun = "mean_sd", x = "cell", dodge = 0.8,
    step.increase = 0.1) %>%
  add_significance("p") %>% 
  select(-p.adj, -p.adj.signif)

# Bar plot
p <- ggbarplot(
  plot_table,
  x = "cell",
  y = "RAD51",
  add = "mean_sd",
  fill = "cell",
  palette = col,
  position = position_dodge(0.8)) +
  stat_pvalue_manual(
    stat.test,
    label = "p.signif",
    tip.length = 0.01,
    hide.ns = TRUE,
    size = 5,
    label.size = 10
  ) +
  # geom_point(
  #   data = plot_table,
  #   aes(
  #     x = group, y = RAD51,
  #     group = interaction(group,cell)
  #   ),
  #   alpha = 0.5,
  #   position = position_dodge(0.8)
  # ) +
  scale_y_continuous(limits = c(0, 64), expand = c(0, 0)) + 
  labs(
    x = "Cell line",
    y = expression("% RAD51 + cells")
  ) +
  theme_classic() +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 21),
    legend.title = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(), 
    text = element_text(size = 21, color = "black"),
    axis.text = element_text(size = 21, color = "black")#,
    # axis.text.x = element_text(size = 18, color = "black", angle = 30, vjust = 1, hjust=1)
  )
ggsave(
  paste0(sub_output_dir, "Figure3C_COV362_HRD_bar.pdf"),
  p,
  width = 7.2, height = 2.5
)
```

#### UWB

```{r}
uwb <- openxlsx::read.xlsx("./data/input/HRD_experiment/HR_test_Sofia_20221011_to_Shuyu.xlsx", sheet = 1, rows = c(1, 62:139), cols = c(1, 10))
names(uwb) <- c("cell", "RAD51")
uwb$cell <- factor(
  rep(c("UWB1.289", "UWB1.289 + BRCA1"), each = 4),
  levels = c("UWB1.289", "UWB1.289 + BRCA1")
)
uwb <- na.omit(uwb)
```

```{r}
# plot_table <- Reduce(
#   rbind.data.frame,
#   # list(cov, uwb, kuramochi)
#   list(cov, uwb))
plot_table <- uwb
# plot_table$group <- factor(
#   plot_table$group,
#   levels = c("COV362", "UWB1.289")#, "KURAMOCHI")
#   )
col = c(
  hrd_color, hrp_color
)

stat.test <- plot_table %>%
  t_test(RAD51 ~ cell) %>%
  add_xy_position(
    fun = "mean_sd", x = "cell", dodge = 0.8,
    step.increase = 0.1) %>%
  add_significance("p")# %>% 
  #select(-p.adj, -p.adj.signif)

# Bar plot
p <- ggbarplot(
  plot_table,
  x = "cell",
  y = "RAD51",
  add = "mean_sd",
  fill = "cell",
  palette = col,
  position = position_dodge(0.8)) +
  stat_pvalue_manual(
    stat.test,
    label = "p.signif",
    tip.length = 0.01,
    hide.ns = TRUE,
    size = 5,
    label.size = 10
  ) +
  # geom_point(
  #   data = plot_table,
  #   aes(
  #     x = group, y = RAD51,
  #     group = interaction(group,cell)
  #   ),
  #   alpha = 0.5,
  #   position = position_dodge(0.8)
  # ) +
  scale_y_continuous(limits = c(0, 64), expand = c(0, 0)) + 
  labs(
    x = "Cell line",
    y = expression("% RAD51 + cells")
  ) +
  theme_classic() +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 21),
    legend.title = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(), 
    text = element_text(size = 21, color = "black"),
    axis.text = element_text(size = 21, color = "black")#,
    # axis.text.x = element_text(size = 18, color = "black", angle = 30, vjust = 1, hjust=1)
  )
ggsave(
  paste0(sub_output_dir, "Figure3C_UWB_HRD_bar.pdf"),
  p,
  width = 7, height = 2.5
)
```

#### KURAMOCHI

```{r}
kuramochi <- openxlsx::read.xlsx("./data/input/HRD_experiment/HR_test_Sofia_20221011_to_Shuyu.xlsx", sheet = 2, rows = 1:99, cols = c(2, 11))
names(kuramochi) <- c("cell", "RAD51")
kuramochi$cell <- factor(
  rep(c("Kuramochi", "Kuramochi + ctrl vector", "Kuramochi + BRCA2"), each = 4),
  levels = c("Kuramochi", "Kuramochi + ctrl vector", "Kuramochi + BRCA2")
)
kuramochi <- na.omit(kuramochi)
kuramochi$group <- "KURAMOCHI"
```

```{r}
# plot_table <- Reduce(
#   rbind.data.frame,
#   # list(cov, uwb, kuramochi)
#   list(cov, uwb))
plot_table <- kuramochi
# plot_table$group <- factor(
#   plot_table$group,
#   levels = c("COV362", "UWB1.289")#, "KURAMOCHI")
#   )
col = c(
  hrd_color, hrd_vector_color, hrp_color
)

stat.test <- plot_table %>%
  t_test(RAD51 ~ cell) %>%
  add_xy_position(
    fun = "mean_sd", x = "cell", dodge = 0.8,
    step.increase = 0.1) %>%
  add_significance("p")# %>% 
  #select(-p.adj, -p.adj.signif)

# Bar plot
p <- ggbarplot(
  plot_table,
  x = "cell",
  y = "RAD51",
  add = "mean_sd",
  fill = "cell",
  palette = col,
  position = position_dodge(0.8)) +
  stat_pvalue_manual(
    stat.test,
    label = "p.signif",
    tip.length = 0.01,
    hide.ns = TRUE,
    size = 5,
    label.size = 10
  ) +
  # geom_point(
  #   data = plot_table,
  #   aes(
  #     x = group, y = RAD51,
  #     group = interaction(group,cell)
  #   ),
  #   alpha = 0.5,
  #   position = position_dodge(0.8)
  # ) +
  scale_y_continuous(limits = c(0, 64), expand = c(0, 0)) + 
  labs(
    x = "Cell line",
    y = expression("% RAD51 + cells")
  ) +
  theme_classic() +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 21),
    legend.title = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(), 
    text = element_text(size = 21, color = "black"),
    axis.text = element_text(size = 21, color = "black")#,
    # axis.text.x = element_text(size = 18, color = "black", angle = 30, vjust = 1, hjust=1)
  )
ggsave(
  paste0(sub_output_dir, "Figure3C_Kura_HRD_bar.pdf"),
  p,
  width = 7.3, height = 2.5
)
```

#### Merge cell lines

```{r eval=FALSE}
plot_table <- Reduce(
  rbind.data.frame,
  # list(cov, uwb, kuramochi)
  list(uwb, kuramochi))
plot_table$group <- factor(
  plot_table$group,
  levels = c("UWB1.289", "KURAMOCHI")
  )
col = c(
  hrd_color, hrd_vector_color, hrp_color,
  hrd_color, hrd_vector_color, hrp_color
)

stat.test <- plot_table %>%
  group_by(group) %>% 
  t_test(RAD51 ~ cell) %>%
  add_xy_position(
    fun = "mean_sd", x = "group", dodge = 0.8,
    step.increase = 0.11) %>%
  add_significance("p") %>% 
  select(-p.adj, -p.adj.signif)

# Bar plot
p <- ggbarplot(
  plot_table,
  x = "group",
  y = "RAD51",
  add = "mean_sd",
  fill = "cell",
  palette = col,
  position = position_dodge(0.8)) +
  stat_pvalue_manual(
    stat.test,
    label = "p.signif",
    tip.length = 0.01,
    hide.ns = TRUE,
    size = 5,
    label.size = 10
  ) +
  # geom_point(
  #   data = plot_table,
  #   aes(
  #     x = group, y = RAD51,
  #     group = interaction(group,cell)
  #   ),
  #   alpha = 0.5,
  #   position = position_dodge(0.8)
  # ) +
  scale_y_continuous(limits = c(0, 64), expand = c(0, 0)) + 
  labs(
    x = "Cell line",
    y = expression("% RAD51 + cells")
  ) +
  theme_classic() +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 18),
    legend.title = element_blank(),
    axis.title.x = element_blank(),
    text = element_text(size = 22.5, color = "black"),
    axis.text = element_text(size = 18, color = "black")#,
    # axis.text.x = element_text(size = 18, color = "black", angle = 30, vjust = 1, hjust=1)
  )
ggsave(
  paste0(sub_output_dir, "Figure3C_HRD_bar_uwb_kura.pdf"),
  p,
  width = 6, height = 4
)
```

## D: Box plot for correlation coefficients of BRCAness signature vs. pre-sensitivity signatures (seurat logFC)

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "Figure3/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}

lseq <- function(from=1, to=100000, length.out=6) {
  # logarithmic spaced sequence
  # blatantly stolen from library("emdbook"), because need only this
  exp(seq(log(from), log(to), length.out = length.out))
}
treatments <- c("Olaparib", "Carboplatin", "NK")
conserve_gene_only <- TRUE
```

### load data

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
log2FC_table <- read.csv("./data/output/Figure2/Figure2D_log2FC_preR_preS_CNV_subclone.csv")
conserve <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  dplyr::select(ensembl, conserve_gene)
```

```{r}
# treatment <- c("Carbo", "Olaparib", "NK")
# for (t in treatment){
#   tmp <- read.csv(
#   paste0(
#     "./data/input/DEG_filter_by_bootstrap_permutation_in_singleton/prer_pres_",
#     t,
#     "_control_bootstrap_permutation_summary.csv")
#   )
#   tmp[[paste0(t, "_log2FC_preR_preS_adj")]] <- tmp$mean_log2FC_treat - tmp$mean_log2FC_ctrl
#   tmp <- tmp[, c("ensembl", "symbol", # paste0(t, "_log2FC_preR_preS_adj"),
#                 paste0("boot_", t, "_ctrl_p_value"),
#                 paste0(t, "_perm_p"))]
#   if (is.null(log2FC_table)){
#     log2FC_table <- tmp
#   } else {
#     log2FC_table <- left_join(log2FC_table, tmp, by = c("ensembl", "symbol"))
#   }
# }
# # log2FC_table <- left_join(log2FC_table, conserve, by = "ensembl")
```

At vs BT logFC adjust growth control

```{r}
# for (t in treatments){
#   if (t == "Carboplatin"){
#     t <- "Carbo"
#   }
#   tmp <- read.csv(paste0("./data/at_bt_", t, "_control_wilcox_summary.csv"))
#   tmp[[paste0(t, "_log2FC_at_bt_adj")]] <- tmp$mean_log2FC_treat - tmp$mean_log2FC_ctrl
#   tmp <- select(tmp, ensembl, symbol, paste0(t, "_log2FC_at_bt_adj"))
#   log2FC_table <- log2FC_table %>% 
#     left_join(tmp, by = c("ensembl", "symbol"))
# }
```

TCGA BRCAness signature

```{r}
tcga = read.csv("./data/input/BRCAness/TCGA_deconvoluted_HRD_DEGs.tsv", sep = "\t")
tcga[which(rownames(tcga) == "BRCA1"), ] # should be negative logFC
tcga$ensembl = data@misc$gene_meta$ensembl[match(rownames(tcga), data@misc$gene_meta$symbol)]
length(which(is.na(tcga$ensembl)==T)) # 383 genes have no emsembl

summary(tcga$adj.P.Val)
```

```{r}
log2FC_table <- left_join(
  log2FC_table,
  tcga %>% 
    rownames_to_column("symbol") %>% 
    select(
      ensembl, symbol, BRCAness_tcga_logFC = logFC,
      BRCAness_tcga_p_adj = adj.P.Val
    ),
  by = c("ensembl", "symbol")
)
```

COV362 BRCAness signature

```{r}
cov362 <- read.csv("./data/input/BRCAness/deg_COV362.csv")
cov362[which(cov362$symbol == "BRCA1"),] # should be negative logFC
```

```{r}
log2FC_table <- left_join(
  log2FC_table,
  cov362 %>% 
    select(
      ensembl, symbol, BRCAness_cov_logFC = avg_log2FC,
      BRCAness_cov_p_adj = p_val_adj
    ),
  by = c("ensembl", "symbol")
)
```

```{r}
write.csv(log2FC_table, paste0(sub_output_dir, "Figure3D_logFC_table_for_plots.csv"), row.names = FALSE)
```

TCGA BRCAness signature vs. CNV subclone signature

```{r}
data <- read.csv(paste0(sub_output_dir, "Figure3D_logFC_table_for_plots.csv")) %>%
  filter(BRCAness_tcga_p_adj <= 0.01) %>% 
  select("log2FC_A", "log2FC_B", "log2FC_C", "log2FC_D",
         "log2FC_E", "BRCAness_tcga_logFC", "BRCAness_cov_logFC")

cor(data, method = "spearman", use = "complete")
```

COV362 signature vs. CNV subclone signature

```{r}
data <- read.csv(paste0(sub_output_dir, "Figure3D_logFC_table_for_plots.csv")) %>%
  filter(BRCAness_cov_p_adj <= 1 * 10^-8) %>% 
  select("log2FC_A", "log2FC_B", "log2FC_C", "log2FC_D",
         "log2FC_E", "BRCAness_tcga_logFC", "BRCAness_cov_logFC")

cor(data, method = "spearman", use = "complete")
```

```{r}
if (conserve_gene_only){
  plot_table <- filter(log2FC_table, conserve_gene)
} else {
  plot_table <- log2FC_table
}
```

### Correlation coefficient
#### Test for good P threshold

##### TCGA signature

```{r}
min_p = 0.005 # around 200 sister conserved BRCAness genes 
max_p = 0.05

p_threshold = c(lseq(min_p, max_p, 20), 1*10^-(2:4), 5*10^-(2:4))

res = mat.or.vec(length(p_threshold), 5)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK")

for(i in 1:length(p_threshold)){
  table_tests <- plot_table %>% 
    filter(BRCAness_tcga_p_adj <= p_threshold[i]) %>% 
    select(starts_with("BRCAness_tcga"), ends_with("log2FC_preR_preS_adj")) %>% 
    na.omit()
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  res[i, 2] <- nrow(table_tests)
  res[i, 3] <- cor(table_tests$BRCAness_tcga_logFC, -table_tests$Olaparib_log2FC_preR_preS_adj, method = "spearman", use = "complete")
  res[i, 4] <- cor(table_tests$BRCAness_tcga_logFC, -table_tests$Carbo_log2FC_preR_preS_adj, method = "spearman", use = "complete")
  res[i, 5] <- cor(table_tests$BRCAness_tcga_logFC, -table_tests$NK_log2FC_preR_preS_adj, method = "spearman", use = "complete")
}
write.csv(
  res,
  paste0(sub_output_dir,"Figure3D_BRCA_preRpreS_cor_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

##### COV362 signature

```{r}
min_p = 5 * 10^-16 # around 200 sister concordant BRCAness genes
max_p = 0.05
p_threshold <- c(lseq(min_p, max_p, 20), 1 * 10 ^ -c(2:14), 5 * 10 ^ -c(2:14))

res = mat.or.vec(length(p_threshold), 5)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK")

for(i in 1:length(p_threshold)){
  table_tests <- plot_table %>% 
    filter(BRCAness_cov_p_adj <= p_threshold[i]) %>% 
    select(starts_with("BRCAness_cov"), ends_with("log2FC_preR_preS_adj")) %>% 
    na.omit()
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  res[i, 2] <- nrow(table_tests)
  res[i, 3] <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Olaparib_log2FC_preR_preS_adj, method = "spearman", use = "complete")
  res[i, 4] <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Carbo_log2FC_preR_preS_adj, method = "spearman", use = "complete")
  res[i, 5] <- cor(table_tests$BRCAness_cov_logFC, -table_tests$NK_log2FC_preR_preS_adj, method = "spearman", use = "complete")
}
write.csv(
  res,
  paste0(sub_output_dir,"Figure3D_BRCA_preRpreS_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

#### Box plot preS/preR

```{r}
res1 <- read.csv(
  paste0(sub_output_dir,"Figure3D_BRCA_preRpreS_cor_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "p_threshold")
res1 <- res1[1:20, ]
res1.df = reshape2::melt(res1, id.vars = "p_threshold", variable.name = "treatment")
res1.df$group = "TCGA"

res2 <- read.csv(
  paste0(sub_output_dir,"Figure3D_BRCA_preRpreS_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "p_threshold")
res2 <- res2[1:20, ]
res2.df = reshape2::melt(res2, id.vars = "p_threshold", variable.name = "treatment")
res2.df$group = "COV362"

res.df = rbind.data.frame(res1.df, res2.df)
res.df$group <- factor(res.df$group, levels = c("TCGA", "COV362"))
res.df$treatment <- factor(res.df$treatment, levels = c("Carboplatin", "Olaparib", "NK"))

stat.test <- res.df[res.df$treatment !="Control",] %>%
  group_by(group) %>%
  t_test(value ~ treatment, ref.group = "NK") %>%
  add_xy_position(
    x = "group", dodge = 0.8,
    step.increase = 0.05
  ) %>%
  add_significance("p")
stat.test

p <- ggboxplot(
  res.df[res.df$treatment !="Control",],
  x = "group", y = "value", fill = "treatment",
  outlier.shape = NA
) +
  stat_pvalue_manual(
    stat.test,
    label = "p.signif", tip.length = 0.01,
    bracket.nudge.y = -0.01,
    size = 5
  ) +
  geom_point(
    data = res.df[res.df$treatment !="Control",],
    mapping = aes(x = group, y = value, fill = treatment),
    position = position_jitterdodge(jitter.width = 0.5),
    size = 0.7
  ) +
  labs(x = "", y = "Correlation coefficient", fill = "")+
  theme_bw() + 
  theme(
    text = element_text(size = 20),
    axis.title = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 20, color = "black"),
    legend.text = element_text(size = 20, color = "black")
  ) +
  scale_fill_manual(values = c(blue, red, brown))
ggsave(
  paste0(
    sub_output_dir,
    "Figure3D_BRCA_preRpreS_cor_cov_tcga",
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 4
)
```

### GWC connectivity score

#### Test for good P threshold

##### TCGA signature

```{r}
min_p = 0.005 # around 200 sister conserved BRCAness genes 
max_p = 0.05

p_threshold = c(lseq(min_p, max_p, 20), 1*10^-(2:3), 5*10^-(2:4), 1)
```

```{r}
res = mat.or.vec(length(p_threshold), 8)
colnames(res) <- c(
  "p_threshold", "n_gene", "Olaparib_score", "Carboplatin_score", "NK_score",
  "Olaparib_p", "Carboplatin_p", "NK_p")

for(i in 1:length(p_threshold)){
  table_tests <- plot_table %>% 
    dplyr::filter(BRCAness_tcga_p_adj <= p_threshold[i]) %>% 
    dplyr::select(ensembl, starts_with("BRCAness_tcga"), ends_with("log2FC_preR_preS_adj"), ends_with("_boot_p")) %>% 
    tibble::column_to_rownames("ensembl") %>% 
    na.omit()
  cat('\r', i)
  BRCA_genes <- table_tests %>% 
    dplyr::mutate(value = -BRCAness_tcga_logFC, sig = BRCAness_tcga_p_adj) %>% 
    dplyr::select(value, sig)
  
  res[i, "p_threshold"] <- p_threshold[i]
  res[i, "n_gene"] <- nrow(table_tests)
  
  # Olaparib
  tmp <- table_tests %>% 
    dplyr::mutate(value = Olaparib_log2FC_preR_preS_adj, sig = Olaparib_boot_p) %>% 
    dplyr::select(value, sig)
  connect <- connectivityScore(x = tmp, y = BRCA_genes, method = "gwc", gwc.method='spearman', nperm=300)
  res[i, "Olaparib_score"] <- connect["score"]
  res[i, "Olaparib_p"] <- connect["p"]
  
  # Carboplatin
  tmp <- table_tests %>% 
    dplyr::mutate(value = Carbo_log2FC_preR_preS_adj, sig = Carbo_boot_p) %>% 
    dplyr::select(value, sig)
  connect <- connectivityScore(x = tmp, y = BRCA_genes, method = "gwc", gwc.method='spearman', nperm=300)
  res[i, "Carboplatin_score"] <- connect["score"]
  res[i, "Carboplatin_p"] <- connect["p"]
  
  # NK
  tmp <- table_tests %>% 
    dplyr::mutate(value = NK_log2FC_preR_preS_adj, sig = NK_boot_p) %>% 
    dplyr::select(value, sig)
  connect <- connectivityScore(x = tmp, y = BRCA_genes, method = "gwc", gwc.method='spearman', nperm=300)
  res[i, "NK_score"] <- connect["score"]
  res[i, "NK_p"] <- connect["p"]

}
write.csv(
  res,
  paste0(sub_output_dir,"Figure3D_BRCA_preRpreS_gwc_connect_tcga_weighted_spearman", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

##### COV362 signature

```{r}
min_p = 5 * 10^-16 # around 200 sister concordant BRCAness genes
max_p = 0.05
p_threshold <- c(lseq(min_p, max_p, 20), 1 * 10 ^ -c(2:14), 5 * 10 ^ -c(2:14))

res = mat.or.vec(length(p_threshold), 5)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK")
```

```{r}
res = mat.or.vec(length(p_threshold), 8)
colnames(res) <- c(
  "p_threshold", "n_gene", "Olaparib_score", "Carboplatin_score", "NK_score",
  "Olaparib_p", "Carboplatin_p", "NK_p")

for(i in 1:length(p_threshold)){
  table_tests <- plot_table %>% 
    dplyr::filter(BRCAness_cov_p_adj <= p_threshold[i]) %>% 
    dplyr::select(ensembl, starts_with("BRCAness_cov"), ends_with("log2FC_preR_preS_adj"), ends_with("_boot_p")) %>% 
    tibble::column_to_rownames("ensembl") %>% 
    na.omit()
  cat('\r', i)
  BRCA_genes <- table_tests %>% 
    dplyr::mutate(value = -BRCAness_cov_logFC, sig = BRCAness_cov_p_adj) %>% 
    dplyr::select(value, sig)
  
  res[i, "p_threshold"] <- p_threshold[i]
  res[i, "n_gene"] <- nrow(table_tests)
  
  # Olaparib
  tmp <- table_tests %>% 
    dplyr::mutate(value = Olaparib_log2FC_preR_preS_adj, sig = Olaparib_boot_p) %>% 
    dplyr::select(value, sig)
  connect <- connectivityScore(x = tmp, y = BRCA_genes, method = "gwc", gwc.method='spearman', nperm=300)
  res[i, "Olaparib_score"] <- connect["score"]
  res[i, "Olaparib_p"] <- connect["p"]
  
  # Carboplatin
  tmp <- table_tests %>% 
    dplyr::mutate(value = Carbo_log2FC_preR_preS_adj, sig = Carbo_boot_p) %>% 
    dplyr::select(value, sig)
  connect <- connectivityScore(x = tmp, y = BRCA_genes, method = "gwc", gwc.method='spearman', nperm=300)
  res[i, "Carboplatin_score"] <- connect["score"]
  res[i, "Carboplatin_p"] <- connect["p"]
  
  # NK
  tmp <- table_tests %>% 
    dplyr::mutate(value = NK_log2FC_preR_preS_adj, sig = NK_boot_p) %>% 
    dplyr::select(value, sig)
  connect <- connectivityScore(x = tmp, y = BRCA_genes, method = "gwc", gwc.method='spearman', nperm=300)
  res[i, "NK_score"] <- connect["score"]
  res[i, "NK_p"] <- connect["p"]

}
write.csv(
  res,
  paste0(sub_output_dir,"Figure3D_BRCA_preRpreS_gwc_connect_cov_weighted_spearman", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

#### Box plot preS/preR

```{r}
res1 <- read.csv(
  paste0(sub_output_dir,"Figure3D_BRCA_preRpreS_gwc_connect_tcga_weighted_spearman", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>%
  dplyr::select(
    Olaparib = "Olaparib_score", Carboplatin = "Carboplatin_score",
    NK = "NK_score", "p_threshold"
  )
res1 <- res1[1:20, ]
res1.df = reshape2::melt(res1, id.vars = "p_threshold", variable.name = "treatment")
res1.df$group = "TCGA"
res1.df %>%
  filter(treatment == "NK") %>% 
  summarise(mean = mean(value), sd = sd(value))

res2 <- read.csv(
  paste0(sub_output_dir,"Figure3D_BRCA_preRpreS_gwc_connect_cov_weighted_spearman", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  dplyr::select(
    Olaparib = "Olaparib_score", Carboplatin = "Carboplatin_score",
    NK = "NK_score", "p_threshold")
res2 <- res2[1:20, ]
res2.df = reshape2::melt(res2, id.vars = "p_threshold", variable.name = "treatment")
res2.df$group = "COV362"

# res.df = rbind.data.frame(res1.df, res2.df)
res.df = res2.df
res.df$group <- factor(res.df$group, levels = c("TCGA", "COV362"))
res.df$treatment <- factor(res.df$treatment, levels = c("Carboplatin", "Olaparib", "NK"))

stat.test <- res.df %>%
  # dplyr::group_by(group) %>%
  t_test(value ~ treatment, ref.group = "NK") %>%
  # add_xy_position(
  #   x = "group", dodge = 0.8,
  #   step.increase = 0.05
  # ) %>%
  add_xy_position(
  x = "treatment", dodge = 0.8,
  step.increase = 0.12) %>%
  add_significance("p")
stat.test

# TCGA + COV362 signatures
# p <- ggboxplot(
#   res.df,
#   x = "group", y = "value", fill = "treatment",
#   outlier.shape = NA
# ) +
#   stat_pvalue_manual(
#     stat.test,
#     label = "p.signif", tip.length = 0.01,
#     bracket.nudge.y = -0.01,
#     size = 4,
#     label.size = 10
#   ) +
#   scale_y_continuous(limits = c(0, 0.62), expand = c(0 , 0)) +
#   # geom_point(
#   #   data = res.df[res.df$treatment !="Control",],
#   #   mapping = aes(x = group, y = value, fill = treatment),
#   #   position = position_jitterdodge(jitter.width = 0.5),
#   #   size = 1,
#   #   alpha = 0.5
#   # ) +
#   labs(x = "", y = "Connectivity score", fill = "")+
#   theme_classic() + 
#   theme(
#     legend.position = "top",
#     text = element_text(size = 22.5),
#     axis.title = element_text(size = 22.5, color = "black"),
#     axis.text = element_text(size = 22.5, color = "black"),
#     legend.text = element_text(size = 21, color = "black")
#   ) +
#   scale_fill_manual(values = c(green, brown, purple))
# ggplot2::ggsave(
#   paste0(
#     sub_output_dir,
#     "Figure3D_BRCA_preRpreS_gwc_connect_cov_tcga_weighted_spearman",
#     ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
#   ),
#   p,
#   width = 7, height = 7
#   #width = 6, height = 5
# )

# COV362 signatures only
p <- ggboxplot(
  res.df,
  x = "treatment", y = "value", fill = "treatment",
  outlier.shape = NA
) +
  stat_pvalue_manual(
    stat.test,
    label = "p.signif", tip.length = 0.01,
    # bracket.nudge.y = -0.01,
    size = 4,
    label.size = 10
  ) +
  scale_y_continuous(limits = c(0, 0.62), expand = c(0 , 0)) +
  # geom_point(
  #   data = res.df[res.df$treatment !="Control",],
  #   mapping = aes(x = group, y = value, fill = treatment),
  #   position = position_jitterdodge(jitter.width = 0.5),
  #   size = 1,
  #   alpha = 0.5
  # ) +
  labs(x = "", y = "Connectivity score", fill = "")+
  theme_classic() + 
  theme(
    # legend.position = "top",
    legend.position = "none",
    text = element_text(size = 21),
    axis.title = element_text(size = 21, color = "black"),
    axis.text = element_text(
      size = 20,
      color = "black"#,
      #angle = 45,
      #hjust = 1
    ),
    legend.text = element_text(size = 21, color = "black")
  ) +
  scale_fill_manual(values = c(green, brown, purple))
ggplot2::ggsave(
  paste0(
    sub_output_dir,
    "Figure3D_BRCA_preRpreS_gwc_connect_cov_weighted_spearman",
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  p,
  width = 5, height = 4
  #width = 6, height = 5
)

```

## E: NK killing rate on COV362 and KURAMOCHI

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "Figure3/four_donor_NK/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}

data_summary <- function(data, varname, groupnames){
  require(plyr)
  summary_func <- function(x, col){
    c(mean = mean(x[[col]], na.rm=TRUE),
      sd = sd(x[[col]], na.rm=TRUE))
  }
  data_sum<-ddply(data, groupnames, .fun=summary_func,
                  varname)
  data_sum <- rename(data_sum, c("mean" = varname))
 return(data_sum)
}
```

The tables for cell signals after treated by NK cells are from the experiments performed by Jun Dai. The background signal has been removed from the data.

### Four donors

```{r}
data_area <- data.frame(
  control = c(
    rep("Kuramochi + ctrl vector", 4),
    rep("COV362 + ctrl vector", 4),
    rep("UWB1.289", 4)
  ),
  brca = c(
    rep("Kuramochi + BRCA2", 4),
    rep("COV362 + BRCA1", 4),
    rep("UWB1.289 + BRCA1", 4)
  ),
  donor = rep(1:4, 3),
  replicate = c(3, 3, 6, 6, 4, 3, 6, 6, 4, 3, 6, 6),
  region = c(
    "B4:H8", "B13:H17", "B22:N26", "B31:N35",
    "B45:J49", "B54:H58", "B63:N67", "B72:N76",
    "B86:J90", "B95:H99", "B104:N108", "B113:N117"
  )
)
for (i in 1:nrow(data_area)){
  table <- read_xlsx(
    "./data/input/NK_cell_killing_experiment/HRD_NK from 4 donors_JD.xlsx",
    col_names = c(
      "rate", rep(data_area$control[i], data_area$replicate[i]),
      rep(data_area$brca[i], data_area$replicate[i])
    ),
    range = data_area$region[i])
  table <- table %>% 
    pivot_longer(!rate, names_to = "sample", values_to = "cell_viability") %>% 
    na.omit()
  table$sample <- gsub("\\.\\..*", "", table$sample)
  table$cell_viability <- table$cell_viability * 100
  table$sample <- factor(
    table$sample,
    levels = c(data_area$control[i], data_area$brca[i])
  )
  
  col = c(hrd_vector_color, hrp_color)
  
  stat.test <- table %>%
    group_by(rate) %>%
    t_test(cell_viability ~ sample) %>%
    add_xy_position(
      fun = "mean_sd", x = "rate", dodge = 0.8,
      step.increase = 0.05) %>%
    add_significance("p")
  
  # bar plot
  p <- ggbarplot(
    table,
    x = "rate",
    y = "cell_viability",
    add = "mean_sd",
    fill = "sample",
    palette = col,
    position = position_dodge(0.8)) +
    ggtitle(paste0("Donor ", data_area$donor[i])) +
    stat_pvalue_manual(
      stat.test,
      label = "p.signif", tip.length = 0.01,
      bracket.nudge.y = - 2, hide.ns = TRUE,
      label.size = 10, bracket.size = 0.8
    ) +
    labs(
      x = "NK : target cell",
      y = "Cell viability (%)",
      fill = ""
    ) +
    scale_y_continuous(limits = c(0, 110), breaks=seq(0, 100, 25), expand = c(0, 0)) +
    theme_classic() +
    theme(
      legend.position = "right",
      legend.text = element_text(size = 21),
      text = element_text(size = 21, color = "black"),
      axis.text = element_text(size = 21, color = "black"),
      plot.title = element_text(size = 21, hjust = 0.5)
    )
  ggsave(
    paste0(sub_output_dir, gsub("\\ \\+.*", "", data_area$control[i]), "_donor_", data_area$donor[i], "_NK_bar_plot.pdf"),
    width = 7.5, height = 3.5
  )

  # line plot
  table$rate <- as.numeric(table$rate)
  plot_table_line <- data_summary(
    table,
    varname="cell_viability", 
    groupnames=c("sample", "rate")
  )
  sig_label <- stat.test %>% 
    select(rate, sample = group1, p.signif)
  sig_label$rate <- as.numeric(sig_label$rate)
  sig_label <- sig_label %>% 
    left_join(
      plot_table_line %>% select(sample, cell_viability, rate),
      by = c("rate", "sample")
    )
  sig_label$p.signif[sig_label$p.signif == "ns"] <- NA
  p <- ggplot(
    plot_table_line,
    aes(
      x = rate,
      y = cell_viability,
      color = sample,
      group = sample
    )
  ) +
    ggtitle(paste0("Donor ", data_area$donor[i])) +
    geom_errorbar(
      aes(ymin = cell_viability - sd, ymax = cell_viability + sd),
      width = 0.2,
      linewidth = 1) +
    geom_line(linewidth = 1) +
    geom_point() +
    scale_color_manual(values = col) +
    scale_y_continuous(limits = c(0, 110), breaks=seq(0, 100, 25), expand = c(0, 0)) +
    scale_x_continuous(expand = c(0, 0)) +
    labs(
      x = "NK : target cell",
      y = "Cell viability (%)"
    ) +
    annotate(
      "text",
      x = sig_label$rate,
      y = sig_label$cell_viability + 15,
      label = sig_label$p.signif,
      color = "black",
      size = 10
    ) +
    theme_classic() +
    theme(
      legend.position = "right",
      legend.text = element_text(size = 21),
      legend.title = element_blank(),
      text = element_text(size = 21, color = "black"),
      axis.text = element_text(size = 21, color = "black"),
      plot.title = element_text(size = 21, hjust = 0.5)
    )
  ggsave(
    paste0(sub_output_dir, gsub("\\ \\+.*", "", data_area$control[i]), "_donor_", data_area$donor[i], "_NK_line_plot.pdf"),
    width = 7.5, height = 3.5
  )
}

```

### COV362

```{r}
table <- read_xlsx(
  "./data/input/NK_cell_killing_experiment/COV362-NK-20220417.xlsx", sheet = 2,
  range = "B1:Z7")
table[1, ] <- t(rep(c("0", "0.5", "1", "2", "3"), each = 5))
table <- t(table)
colnames(table) <- c("rate", "sample", "rep1", "rep2", "rep3", "rep4")
table <- table %>% 
  as.data.frame() %>% 
  filter(sample %in% c("COV362-GFP-M", "COV362-B1-GFP-M")) %>% 
  pivot_longer(cols = starts_with("rep"), values_to = "read", names_to = "replicates") %>% 
  na.omit()
table$read <- as.numeric(table$read)
NC_COV362_GFP_M <- mean(table$read[table$sample == "COV362-GFP-M" & table$rate == "0"])
NC_COV362_B1_GFP_M <- mean(table$read[table$sample == "COV362-B1-GFP-M" & table$rate == "0"])
table <- table %>% 
  mutate(
    cell_viability = ifelse(sample == "COV362-GFP-M", read/NC_COV362_GFP_M, read/NC_COV362_B1_GFP_M) * 100
  )
table$sample[table$sample == "COV362-GFP-M"] <- "COV362 + ctrl vector"
table$sample[table$sample == "COV362-B1-GFP-M"] <- "COV362 + BRCA1"
table$sample <- factor(table$sample, levels = c("COV362 + ctrl vector", "COV362 + BRCA1"))
write.csv(table, paste0(sub_output_dir, "Figure3E_cell_viablility_COV362.csv"))
```

```{r}
col = c(hrd_vector_color, hrp_color)

stat.test <- table %>%
  group_by(rate) %>%
  t_test(cell_viability ~ sample) %>%
  add_xy_position(
    fun = "mean_sd", x = "rate", dodge = 0.8,
    step.increase = 0.05) %>%
  add_significance("p")

p <- ggbarplot(
  table,
  x = "rate",
  y = "cell_viability",
  add = "mean_sd",
  fill = "sample",
  palette = col,
  position = position_dodge(0.8)) +
  stat_pvalue_manual(
    stat.test,
    label = "p.signif", tip.length = 0.01,
    bracket.nudge.y = - 2, hide.ns = TRUE,
    label.size = 10, bracket.size = 0.8
  ) +
  labs(
    x = "NK : target cell",
    y = "Cell viability (%)",
    fill = ""
  ) +
  scale_y_continuous(limits = c(0, 110), breaks=seq(0, 100, 25), expand = c(0, 0)) +
  theme_classic() +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 18),
    text = element_text(size = 22.5, color = "black"),
    axis.text = element_text(size = 21, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "Figure3E_COV362_NK_killing.pdf"),
  width = 7, height = 3.5
)

# line plot
table$rate <- as.numeric(table$rate)
plot_table_line <- data_summary(table, varname="cell_viability", 
                    groupnames=c("sample", "rate"))
sig_label <- stat.test %>% 
  select(rate, sample = group1, p.signif)
sig_label$rate <- as.numeric(sig_label$rate)
sig_label <- sig_label %>% 
  left_join(
    plot_table_line %>% select(sample, cell_viability, rate),
    by = c("rate", "sample")
  )
sig_label$p.signif[sig_label$p.signif == "ns"] <- NA
p <- ggplot(
  plot_table_line,
  aes(
    x = rate,
    y = cell_viability,
    color = sample,
    group = sample
  )
) +
  geom_errorbar(
    aes(ymin = cell_viability - sd, ymax = cell_viability + sd),
    width = 0.2,
    size = 1) +
  geom_line(size = 1) +
  geom_point() +
  scale_color_manual(values = col) +
  scale_y_continuous(limits = c(0, 110), breaks=seq(0, 100, 25), expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0)) +
  labs(
    x = "NK : target cell",
    y = "Cell viability (%)"
  ) +
  annotate(
    "text",
    x = sig_label$rate,
    y = sig_label$cell_viability + 15,
    label = sig_label$p.signif,
    color = "black",
    size = 10
  ) +
  theme_classic() +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 21),
    legend.title = element_blank(),
    text = element_text(size = 21, color = "black"),
    axis.text = element_text(size = 21, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "Figure3E_COV362_NK_line_plot.pdf"),
  width = 8.4, height = 3.5
)
ggsave(
  paste0(sub_output_dir, "Figure3E_COV362_NK_line_plot.png"),
  width = 6.5, height = 3.5
)
```

### UWB1.289

```{r}
table <- read_xlsx(
    "./data/input/NK_cell_killing_experiment/NK-UWB-UWB+B1-RPE-B40-FM2587.xlsx", sheet = 2,
  range = "B14:G22")
table$`NK : cancer` <- rep(c("UWB1.289", "UWB1.289 + BRCA1"), each = 4)
colnames(table)[1] <- "sample"
table <- table %>% 
  pivot_longer(cols = -sample, values_to = "cell_viability", names_to = "rate")
table$rate <- as.numeric(table$rate)
table$cell_viability <- table$cell_viability * 100

table$sample <- factor(table$sample, levels = c("UWB1.289", "UWB1.289 + BRCA1"))
write.csv(table, paste0(sub_output_dir, "Figure3E_cell_viablility_NK_UWB.csv"))
```

```{r}
col = c(hrd_color, hrp_color)

stat.test <- table %>%
  group_by(rate) %>%
  t_test(cell_viability ~ sample) %>%
  add_xy_position(
    fun = "mean_sd", x = "rate", dodge = 0.8,
    step.increase = 0.05) %>%
  add_significance("p")

p <- ggbarplot(
  table,
  x = "rate",
  y = "cell_viability",
  add = "mean_sd",
  fill = "sample",
  palette = col,
  position = position_dodge(0.8)) +
  stat_pvalue_manual(
    stat.test,  label = "p.signif", tip.length = 0.01,
    bracket.nudge.y = - 2, hide.ns = TRUE,
    label.size = 12, bracket.size = 0.8
  ) +
  labs(
    x = "NK : target cell",
    y = "Cell viability (%)",
    fill = ""
  ) +
  scale_y_continuous(limits = c(0, 110), breaks=seq(0, 100, 25), expand = c(0, 0)) +
  theme_classic() +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 21),
    text = element_text(size = 21, color = "black"),
    axis.text = element_text(size = 21, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "Figure3E_UWB_NK_killing.pdf"),
  width = 7.8, height = 3.5
)

# line plot
table$rate <- as.numeric(table$rate)
plot_table_line <- data_summary(table, varname="cell_viability", 
                    groupnames=c("sample", "rate"))
sig_label <- stat.test %>% 
  select(rate, sample = group1, p.signif)
sig_label$rate <- as.numeric(sig_label$rate)
sig_label <- sig_label %>% 
  left_join(
    plot_table_line %>% select(sample, cell_viability, rate),
    by = c("rate", "sample")
  )
sig_label$p.signif[sig_label$p.signif == "ns"] <- NA
p <- ggplot(
  plot_table_line,
  aes(
    x = rate,
    y = cell_viability,
    color = sample,
    group = sample
  )
) +
  geom_errorbar(
    aes(ymin = cell_viability - sd, ymax = cell_viability + sd),
    width = 0.2,
    size = 1) +
  geom_line(size = 1) +
  geom_point() +
  scale_color_manual(values = col) +
  scale_y_continuous(limits = c(0, 110), breaks=seq(0, 100, 25), expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0)) +
  labs(
    x = "NK : target cell",
    y = "Cell viability (%)"
  ) +
  annotate(
    "text",
    x = sig_label$rate,
    y = sig_label$cell_viability + 15,
    label = sig_label$p.signif,
    color = "black",
    size = 10
  ) +
  theme_classic() +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 21),
    legend.title = element_blank(),
    text = element_text(size = 21, color = "black"),
    axis.text = element_text(size = 21, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "Figure3E_UWB_NK_line_plot.pdf"),
  width = 7.8, height = 3.5
)
ggsave(
  paste0(sub_output_dir, "Figure3E_UWB_NK_line_plot.png"),
  width = 6.8, height = 3.5
)
```

### KURAMOCHI

```{r}
table <- read_xlsx(
    "./data/input/NK_cell_killing_experiment/Kura-NK-20220417.xlsx", sheet = 2,
  range = "B3:Z9")
table[1, ] <- t(rep(c("0", "0.5", "1", "2", "3"), each = 5))
table <- t(table)
colnames(table) <- c("rate", "sample", "rep1", "rep2", "rep3", "rep4")
table <- table %>% 
  as.data.frame() %>% 
  filter(sample %in% c("Kura-GFP-H", "Kura-B2-GFP-H")) %>% 
  pivot_longer(cols = starts_with("rep"), values_to = "read", names_to = "replicates") %>% 
  na.omit()
table$read <- as.numeric(table$read)
NC_Kura_GFP_M <- mean(table$read[table$sample == "Kura-GFP-H" & table$rate == "0"])
NC_Kura_B1_GFP_M <- mean(table$read[table$sample == "Kura-B2-GFP-H" & table$rate == "0"])
table <- table %>% 
  mutate(
    cell_viability = ifelse(sample == "Kura-GFP-H", read/NC_Kura_GFP_M, read/NC_Kura_B1_GFP_M) * 100
  )
table$sample[table$sample == "Kura-GFP-H"] <- "Kuramochi + ctrl vector"
table$sample[table$sample == "Kura-B2-GFP-H"] <- "Kuramochi + BRCA2"
table$sample <- factor(table$sample, levels = c("Kuramochi + ctrl vector", "Kuramochi + BRCA2"))
write.csv(table, paste0(sub_output_dir, "Figure3E_cell_viablility_Kura.csv"))
```

```{r}
col = c(hrd_vector_color, hrp_color)

stat.test <- table %>%
  group_by(rate) %>%
  t_test(cell_viability ~ sample) %>%
  add_xy_position(
    fun = "mean_sd", x = "rate", dodge = 0.8,
    step.increase = 0.05) %>%
  add_significance("p")

p <- ggbarplot(
  table,
  x = "rate",
  y = "cell_viability",
  add = "mean_sd",
  fill = "sample",
  palette = col,
  position = position_dodge(0.8)) +
  stat_pvalue_manual(
    stat.test,  label = "p.signif", tip.length = 0.01,
    bracket.nudge.y = - 2, hide.ns = TRUE,
    label.size = 12, bracket.size = 0.8
  ) +
  labs(
    x = "NK : target cell",
    y = "Cell viability (%)",
    fill = ""
  ) +
  scale_y_continuous(limits = c(0, 110), breaks=seq(0, 100, 25), expand = c(0, 0)) +
  theme_classic() +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 15),
    text = element_text(size = 22.5, color = "black"),
    axis.text = element_text(size = 21, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "Figure3E_KURAMOCHI_NK_killing.pdf"),
  width = 4.5, height = 4
)

# line plot
table$rate <- as.numeric(table$rate)
plot_table_line <- data_summary(table, varname="cell_viability", 
                    groupnames=c("sample", "rate"))
sig_label <- stat.test %>% 
  select(rate, sample = group1, p.signif)
sig_label$rate <- as.numeric(sig_label$rate)
sig_label <- sig_label %>% 
  left_join(
    plot_table_line %>% select(sample, cell_viability, rate),
    by = c("rate", "sample")
  )
sig_label$p.signif[sig_label$p.signif == "ns"] <- NA
p <- ggplot(
  plot_table_line,
  aes(
    x = rate,
    y = cell_viability,
    color = sample,
    group = sample
  )
) +
  geom_errorbar(
    aes(ymin = cell_viability - sd, ymax = cell_viability + sd),
    width = 0.2,
    size = 1) +
  geom_line(size = 1) +
  geom_point() +
  scale_color_manual(values = col) +
  scale_y_continuous(limits = c(0, 110), breaks=seq(0, 100, 25), expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0)) +
  labs(
    x = "NK : target cell",
    y = "Cell viability (%)"
  ) +
  annotate(
    "text",
    x = sig_label$rate,
    y = sig_label$cell_viability + 15,
    label = sig_label$p.signif,
    color = "black",
    size = 10
  ) +
  theme_classic() +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 21),
    legend.title = element_blank(),
    text = element_text(size = 21, color = "black"),
    axis.text = element_text(size = 21, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "Figure3E_KURAMOCHI_NK_line_plot.pdf"),
  width = 8, height = 3.5
)
ggsave(
  paste0(sub_output_dir, "Figure3E_KURAMOCHI_NK_line_plot.png"),
  width = 7.1, height = 3.5
)
```

## F: Hazard ratios of overall survival for the interaction between NK transcriptional signature, and patients’ HRD/HRP status

This figure is generated by Matias.

## G: STING activator and inhibitor with NK killing

The % inhibition of drug combination of NK + STING activators/inhibitors are from Jun and Jing in Anna's group. They are saved in the file "./data/input/ReSisTrace_drug_combination_validatiion/Drug_combination_inhibition_ReSisTrace_20221025.xlsx". It is already in the synergyfinder input table format.

```{r}
combo_sub_dir <- paste0(sub_output_dir, "C_drug_combination/")
if(!dir.exists(combo_sub_dir)){
  dir.create(combo_sub_dir)
}
```

```{r eval=FALSE}
table <- read_xlsx("./data/input/ReSisTrace_drug_combination_validatiion/Drug_combination_inhibition_ReSisTrace_20221025.xlsx")

table$conc_unit1 <- table$conc_unit
table$conc_unit2 <- table$conc_unit
table <- table[, colnames(table) != "conc_unit"]
table$conc_unit1[table$drug1 == "NK-FM2931"] <- "ratio"
table$drug1[table$drug1 == "NK-FM2931"] <- "NK : target cell"
table <- ReshapeData(table, data_type = "inhibition")
table <- CalculateSynergy(table)
table <- CalculateSensitivity(table)

saveRDS(table, file = paste0(combo_sub_dir, "synergyfinder_object_with_score.RDS"))
```

### Visualization

```{r}
table <- readRDS(paste0(combo_sub_dir, "synergyfinder_object_with_score.RDS"))

synergys <- c("HSA", "ZIP", "Loewe", "Bliss")

for (b in unique(table$drug_pairs$block_id)){
  low_color <- min(table$synergy_scores[, paste0(s,"_synergy")]) - 5
  high_color <- max(table$synergy_scores[, paste0(s,"_synergy")]) + 5
  for (s in synergys){
    low_z <- min(table$synergy_scores[table$synergy_scores$block_id == b, paste0(s,"_synergy")]) - 5
    high_z <- max(table$synergy_scores[table$synergy_scores$block_id == b, paste0(s,"_synergy")]) + 5
    pdf(
      paste0(
        combo_sub_dir,
        gsub(" ", "_", gsub(" : ", "-", table$drug_pairs$drug1[table$drug_pairs$block_id == b])),
        "_",
        table$drug_pairs$drug2[table$drug_pairs$block_id == b],
        "_", s, ".pdf"
      ),
      width = 3,
      height = 3
    )
    Plot2DrugSurface(
      table,
      plot_block = b,
      plot_value = paste0(s, "_synergy"),
      summary_statistic = "mean",
      high_value_color = red,
      low_value_color = blue,
      z_range = c(low_z, high_z),
      color_range = c(low_color, high_color),
      axis_line = FALSE
    )
    dev.off()
    }
}



```

# Figure S4

## A: Heatmap showing the subclonal CNV profiles of the carboplatin samples inferred from their scRNA-seq data.

This figure is generated by Matias.

## B: Sankey plot for CNV subgroup transition

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS4/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
data <- data@meta.data
```

```{r}
# Merge replicates
samples <- c("Carboplatin", "Control", "NK", "Olaparib")
for (s in samples) {
  cat(s, "\n")
  data_tmp <- data %>% 
    filter(treatment == s)

  # Sankey diagram
  CNV_subclone <- data_tmp %>% 
    select(
      drugSens, sisters, prer_r_group, CNV_subclone, time_point, treatment) %>% 
    filter(!is.na(prer_r_group))
  CNV_subclone$CNV_subclone <- as.character(CNV_subclone$CNV_subclone)
  
  prer_r_group <- unique(CNV_subclone$prer_r_group)
  
  ## sankeyNetwork ggplot
  
  links <- data.frame(
    prer_r_group = prer_r_group,
    source = NA,
    target = NA
  )
  for (i in prer_r_group){
    # subclone in BT
    subclone_bt <- CNV_subclone %>% 
      filter(time_point == "BeforeTreatment" & prer_r_group == i) %>% 
      select(CNV_subclone) %>% 
      unique() %>% 
      unlist()
    if (length(subclone_bt) == 0 ){
      links$source[links$prer_r_group == i] <- NA
    } else if (length(subclone_bt) > 1) {
      links$source[links$prer_r_group == i] <- paste0(sort(subclone_bt), collapse = "")
    } else {
      links$source[links$prer_r_group == i] <- paste0(rep(subclone_bt, sum(links$prer_r_group == i)))
    }
    # subclone in AT
    subclone_at <- CNV_subclone %>% 
      filter(time_point == "AfterTreatment" & prer_r_group == i) %>% 
      select(CNV_subclone) %>% 
      unique() %>% 
      unlist()
    if (length(subclone_at) == 0 ){
      links$target[links$prer_r_group == i] <- NA
    } else if (length(subclone_at) > 1) {
      links$target[links$prer_r_group == i] <- paste0(sort(subclone_at), collapse = "")
    } else {
      links$target[links$prer_r_group == i] <- paste0(rep(subclone_at, sum(links$prer_r_group == i)))
    }
  }
  cols <- c(
    "A" = "#DC0000FF", "B" = "#B09C85FF", "C" = "#00A087FF", "BC" = "#B09C85FF",
    "AB" = "#F39B7FFF", "D" = "#91D1C2FF",  "E" = "#3C5488FF", "AD" = "#4DBBD5FF",
    "AC" = "#8491B4FF", "ACD" = "#E64B35FF")
  # remove NA subclone
  links <- na.omit(links)
  df <- links %>%
    select(`pre-treatment` = "source", `post-treatment` = "target") %>% 
    make_long(`pre-treatment`, `post-treatment`)
  ggplot(
    df,
    aes(x = x, 
        next_x = next_x, 
        node = node, 
        next_node = next_node,
        fill = factor(node),
        label = node)) +
  geom_sankey(flow.alpha = 0.5) +
  scale_fill_manual(
    values = cols) +
  geom_sankey_text(
    mapping = aes(x = ifelse(as.numeric(x) == 1, 0.9, 2.25)),
    size = 5,
    color = 1,
    show.legend = FALSE,
    width = 0.01,
    hjust = 1) +
  theme_sankey(base_size = 20) +
  ggtitle(s) +
  theme(
    legend.position = "none",
    plot.title = element_text(hjust = 0.5, size = 20, face = "bold"),
    text = element_text(size = 20),
    axis.text = element_text(size = 20, color = "black"),
    axis.title.x = element_blank()
  )
  ggsave(paste0(sub_output_dir, "FigureS4B_CNV_subclone_transfer_", s, ".pdf"), width = 6, height = 6)
  print(nrow(links))
}
```

# Figure S5 cell-cell distances between drug-sensitive groups within CNV subclones

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS5/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
data2 = data@assays$RNA@scale.data  # scale data is for dimension reduction or clustering
dim(data2) # 2000 77724
treatments <- c("Olaparib", "Carboplatin", "NK", "Control")
subclone <- LETTERS[1:5]
```

Calculate cell-cell distance

It costs a lot of time to calculate the cell-cell distances. Here I comment out the codes for calculation and directly read the previously calculated matrix. 

```{r eval = FALSE}
# data2.dist = distances(t(data2), normalize = "mahalanobize")
# saveRDS(data2.dist, file = paste0(sub_output_dir, "all_cell_distance.RDS"))
data2.dist <- readRDS(paste0(sub_output_dir, "all_cell_distance.RDS"))
```

```{r eval = FALSE}
res <- NULL
res_large <- NULL
for (t in treatments){
  tmp_res <- vector(mode = "list", length = 5)
  names(tmp_res) <- subclone
  tmp_large <- vector(mode = "list", length = 5)
  names(tmp_large) <- subclone
  for(i in subclone){
    # cat("\r",i)
    preS = which(
      data$drugSens == "pre-sensitive" &
      data$CNV_subclone == i &
      data$treatment == t)
    preR = which(
      data$drugSens == "pre-resistant" &
      data$CNV_subclone == i &
      data$treatment == t
    )
    posT = which(
      data$drugSens == "resistant" &
      data$CNV_subclone == i &
      data$treatment == t
    )
    tmp = data2.dist[preS, preS]
    distS = tmp[upper.tri(tmp, diag = F)]
    tmp = data2.dist[preR, preR]
    distR = tmp[upper.tri(tmp, diag = F)]
    tmp = data2.dist[posT, posT]
    distT = tmp[upper.tri(tmp, diag = F)]
    tmp = cbind(dist = c(distS, distR, distT),
    drugSens = c(rep(1, length(distS)), rep(2, length(distR)), rep(3, length(distT))),
    subclone = i)
    tmp = data.frame(tmp)
    tmp$drugSens[which(tmp$drugSens == 1)] = "pre-sensitive"
    tmp$drugSens[which(tmp$drugSens == 2)] = "pre-resistant"
    tmp$drugSens[which(tmp$drugSens == 3)] = "post-treatment"
    tmp_large[[i]] = tmp
    if (nrow(tmp) > 100000){
      set.seed(123)
      tmp_small = tmp[sample(1:nrow(tmp), 100000),]
    } else {
      tmp_small <- tmp
    }
    # tmp_small = data.frame(tmp_small)
    # tmp_small$drugSens[which(tmp_small$drugSens == 1)] = "pre-sensitive"
    # tmp_small$drugSens[which(tmp_small$drugSens == 2)] = "pre-resistant"
    # tmp_small$drugSens[which(tmp_small$drugSens == 3)] = "post-treatment"
    tmp_res[[i]] = tmp_small
  }
  tmp_res = do.call("rbind", tmp_res)
  tmp_res$drugSens <- factor(tmp_res$drugSens, levels = c("pre-sensitive", "pre-resistant", "post-treatment"))
  tmp_res$subclone <- factor(tmp_res$subclone, levels = c("A","B","C","D","E"))
  tmp_res$dist <- as.numeric(tmp_res$dist)
  tmp_res$treatment <- t
  tmp_large = do.call("rbind", tmp_large)
  tmp_large$drugSens <- factor(tmp_large$drugSens, levels = c("pre-sensitive", "pre-resistant", "post-treatment"))
  tmp_large$subclone <- factor(tmp_large$subclone, levels = c("A","B","C","D","E"))
  tmp_large$dist <- as.numeric(tmp_large$dist)
  tmp_large$treatment <- t
  if (is.null(res)){
    res <- tmp_res
    res_large <- tmp_large
  } else {
    res <- rbind.data.frame(res, tmp_res)
    res_large <- rbind.data.frame(res_large, tmp_large)
  }
}
View(res)

write.csv(res, paste0(sub_output_dir, "cell_distance_table_by_treatment_for_plot.csv"))
write.csv(res_large, paste0(sub_output_dir, "large_cell_distance_table_by_treatment_for_plot.csv"))
```

```{r}
res <- read.csv(paste0(sub_output_dir, "cell_distance_table_by_treatment_for_plot.csv"))
color = c("pre-sensitive" = "snow2", "pre-resistant" = "snow4", "post-treatment" = "slategray4")
res$drugSens <- factor(res$drugSens, levels = c("pre-sensitive", "pre-resistant", "post-treatment"))
res$subclone <- factor(res$subclone, levels = c("A","B","C","D","E"))
res <- res %>% 
  filter(drugSens != "post-treatment")
for (t in treatments) {
  tmp_plot <- res[res$treatment == t,]
  test <- rep(NA, 5)
  names(test) <- subclone
  for (i in subclone){
    tmp_test <- t.test(
        tmp_plot$dist[tmp_plot$drugSens == "pre-resistant" & tmp_plot$subclone == i],
        tmp_plot$dist[tmp_plot$drugSens == "pre-sensitive" & tmp_plot$subclone == i]
      )
    test[i] <- tmp_test$p.value
  }
  test <- data.frame(
    subclone = names(test),
    p_value = test,
    sig = case_when(
    test < .1 & test >= .05 ~ "*", 
    test < .05 & test >= .01 ~ "**", 
    test < .01 ~ "***", 
    TRUE ~ ""
    ),
    x = 1:5 - 0.1,
    y = rep(90, 5)
  )
  fig <- ggplot(
    data = tmp_plot,
    aes(y = dist, x = subclone, group = interaction(subclone, drugSens))
    ) +
    geom_violin(bw = 2, width = 0.8, aes(fill = drugSens)) +
    geom_boxplot(
      position = position_dodge(width = 0.8),
      fill = "white", width = 0.1,
      outlier.shape = NA,# alpha = 1
    ) +
    scale_y_continuous(name = "Normalized distance", limits = c(30, 90)) +  # There are 4.71% cell pairs has the distance larger than 100
    scale_x_discrete(name = "Subclone") +
    scale_fill_manual(values = color) + 
    # annotate("text", x = test$x, y = test$y, label = test$sig) +
    ggtitle(t) +
    theme_bw() +
    theme(
      text = element_text(size = 20, colour = "black"),
      plot.title = element_text(hjust = 0.5, size = 20),
      axis.text = element_text(size = 18, color = "black"),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      legend.position = "none"
    )
  ggsave(
    paste0(sub_output_dir, t, "_dist_cnv.pdf"),
    fig, width = 9, height = 4
  )
  ggsave(
    paste0(sub_output_dir, t, "_dist_cnv.svg"),
    fig, width = 9, height = 4
  )
}
```

two-way ANOVA

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS5/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

```{r}
res_large <- read.csv(paste0(sub_output_dir, "large_cell_distance_table_by_treatment_for_plot.csv"))
res_large <- filter(res_large, drugSens != "post-treatment")
test <- aov(dist ~ drugSens + subclone + drugSens:subclone, data = res_large[res_large$treatment== "Carboplatin", ])
summary(test)
```

```{r}
test <- aov(dist ~ drugSens + subclone + drugSens:subclone, data = res_large[res_large$treatment== "NK", ])
summary(test)
```

```{r}
test <- aov(dist ~ drugSens + subclone + drugSens:subclone, data = res_large[res_large$treatment== "Control", ])
summary(test)
```

```{r}
test <- aov(dist ~ drugSens + subclone + drugSens:subclone, data = res_large[res_large$treatment== "Olaparib", ])
summary(test)
```

# Figure 4 Drug combination discovery

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "Figure4/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
treatments <- c("Carbo", "NK", "Olaparib")
```

```{r}
data <- read.csv("./data/input/consensus_signature_drugs.csv", row.names = 1) %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("symbol")

drug_anno <- read.csv("./data/input/drug_repurposing_hub_alldrugsanno.csv")

length(setdiff(colnames(data)[-1], c(drug_anno$deprecated_broad_id, drug_anno$broad_id)))

## Match drug by "broad_id"

drug_anno_match <- drug_anno %>% 
  filter(broad_id %in% colnames(data)) %>% 
  mutate(match_id = broad_id)
drug_anno_match
drug_anno_match_deprecated_id <- drug_anno %>% 
  filter(deprecated_broad_id %in% colnames(data),
         !broad_id %in% drug_anno_match$broad_id) %>% 
  mutate(match_id = deprecated_broad_id)
drug_anno_match <- rbind.data.frame(drug_anno_match, drug_anno_match_deprecated_id)
drug_anno_match$deprecated_broad_id[drug_anno_match$deprecated_broad_id == "NA-NA"] <- NA
drug_anno_match <- drug_anno_match[!duplicated(drug_anno_match$broad_id), ]
```

```{r}
drug_without_anno <- setdiff(colnames(data)[-1], c(drug_anno_match$broad_id, drug_anno_match_deprecated_id$deprecated_broad_id))
write.csv(
  data.frame(drug = drug_without_anno),
  paste0(sub_output_dir, "drug_without_anno.csv"),
  row.names = FALSE
)
write.csv(
  drug_anno_match,
  paste0(sub_output_dir, "drug_matched_anno.csv"),
  row.names = FALSE
)
```

Table for log2FCs

```{r}
# log2FC for preR-preS-adjusted
log2FC_table <- read.csv("./data/output/Figure3/Figure3D_logFC_table_for_plots.csv")

# Bootstrap and permutation P values
# p <- read.csv("./data/input/DEG_filter_by_bootstrap_permutation_in_singleton/prer_pres_Olaparib_control_bootstrap_permutation_summary.csv") %>% 
#   select(ensembl, symbol, boot_Olaparib_ctrl_p_value, perm_Olaparib_ctrl_p_value) %>% 
#   left_join(
#     read.csv("./data/input/DEG_filter_by_bootstrap_permutation_in_singleton/deg_prer_pres_Carbo_control_bootstrap_permutation.csv") %>% 
#       select(ensembl, symbol, boot_Carbo_ctrl_p_value, perm_Carbo_ctrl_p_value),
#     by = c("ensembl", "symbol")
#   ) %>% 
#   left_join(
#     read.csv("./data/input/DEG_filter_by_bootstrap_permutation_in_singleton/deg_prer_pres_NK_control_bootstrap_permutation.csv") %>% 
#       select(ensembl, symbol, boot_NK_ctrl_p_value, perm_NK_ctrl_p_value),
#     by = c("ensembl", "symbol")
#   )
# log2FC_table <- log2FC_table %>% 
#   left_join(p, by = c("ensembl", "symbol"))

# posT-preT signature

posT <- read.csv("./data/input/DEG/deg_at_bt_Carbo_control_wilcox.csv") %>% 
  mutate(
    Carbo_log2FC_posT_preT = (Carbo_1_avg_log2FC + Carbo_2_avg_log2FC) / 2,
    Carbo_log2FC_posT_preT_adj = (Carbo_1_avg_log2FC + Carbo_2_avg_log2FC) / 2 - mean_log2FC_ctrl,
    Carbo_posT_preT_p = (Carbo_1_p_val + Carbo_2_p_val) / 2,
    Carbo_posT_preT_p_adj = (Carbo_1_p_val_adj + Carbo_2_p_val_adj) / 2
  ) %>% 
  select(
    symbol, ensembl, Carbo_log2FC_posT_preT, Carbo_log2FC_posT_preT_adj,
    Carbo_posT_preT_p, Carbo_posT_preT_p_adj
  ) %>% 
  left_join(
    read.csv("./data/input/DEG/deg_at_bt_Olaparib_control_wilcox.csv") %>% 
    mutate(
      Olaparib_log2FC_posT_preT = (Olaparib_1_avg_log2FC + Olaparib_2_avg_log2FC) / 2,
      Olaparib_log2FC_posT_preT_adj = (Olaparib_1_avg_log2FC + Olaparib_2_avg_log2FC) / 2 - mean_log2FC_ctrl,
      Olaparib_posT_preT_p = (Olaparib_1_p_val + Olaparib_2_p_val) / 2,
      Olaparib_posT_preT_p_adj = (Olaparib_1_p_val_adj + Olaparib_2_p_val_adj) / 2
    ) %>% 
    select(
      symbol, ensembl, Olaparib_log2FC_posT_preT, Olaparib_log2FC_posT_preT_adj,
      Olaparib_posT_preT_p, Olaparib_posT_preT_p_adj
    ),
    by = c("symbol", "ensembl")
  ) %>% 
  left_join(
    read.csv("./data/input/DEG/deg_at_bt_NK_control_wilcox.csv") %>% 
    mutate(
      NK_log2FC_posT_preT = (NK_1_avg_log2FC + NK_2_avg_log2FC) / 2,
      NK_log2FC_posT_preT_adj = (NK_1_avg_log2FC + NK_2_avg_log2FC) / 2 - mean_log2FC_ctrl,
      NK_posT_preT_p = (NK_1_p_val + NK_2_p_val) / 2,
      NK_posT_preT_p_adj = (NK_1_p_val_adj + NK_2_p_val_adj) / 2
    ) %>% 
    select(
      symbol, ensembl, NK_log2FC_posT_preT, NK_log2FC_posT_preT_adj,
      NK_posT_preT_p, NK_posT_preT_p_adj
    ),
    by = c("symbol", "ensembl")
  )

log2FC_table <- log2FC_table %>% 
  left_join(posT, by = c("symbol", "ensembl"))

# preR-preS signature without adjusting growth control

preR <- read.csv("./data/input/DEG_filter_by_bootstrap_permutation_in_singleton/deg_prer_pres_Carbo_control_bootstrap_permutation.csv") %>% 
  mutate(
    Carbo_log2FC_preR_preS = (Carbo_1_avg_log2FC + Carbo_2_avg_log2FC) / 2,
    Carbo_preR_preS_p = (Carbo_1_p_val + Carbo_2_p_val) / 2,
    Carbo_preR_preS_p_adj = (Carbo_1_p_val_adj + Carbo_2_p_val_adj) / 2
  ) %>% 
  select(
    symbol, ensembl, Carbo_log2FC_preR_preS, Carbo_preR_preS_p,
    Carbo_preR_preS_p_adj
  ) %>% 
  left_join(
    read.csv("./data/input/DEG_filter_by_bootstrap_permutation_in_singleton/deg_prer_pres_Olaparib_control_bootstrap_permutation.csv") %>% 
    mutate(
      Olaparib_log2FC_preR_preS = (Olaparib_1_avg_log2FC + Olaparib_2_avg_log2FC) / 2,
      Olaparib_preR_preS_p = (Olaparib_1_p_val + Olaparib_2_p_val) / 2,
      Olaparib_preR_preS_p_adj = (Olaparib_1_p_val_adj + Olaparib_2_p_val_adj) / 2
    ) %>% 
    select(
      symbol, ensembl, Olaparib_log2FC_preR_preS, Olaparib_preR_preS_p,
      Olaparib_preR_preS_p_adj
    ),
    by = c("symbol", "ensembl")
  ) %>% 
  left_join(
    read.csv("./data/input/DEG_filter_by_bootstrap_permutation_in_singleton/deg_prer_pres_NK_control_bootstrap_permutation.csv") %>% 
    mutate(
      NK_log2FC_preR_preS = (NK_1_avg_log2FC + NK_2_avg_log2FC) / 2,
      NK_preR_preS_p = (NK_1_p_val + NK_2_p_val) / 2,
      NK_preR_preS_p_adj = (NK_1_p_val_adj + NK_2_p_val_adj) / 2
    ) %>% 
    select(
      symbol, ensembl, NK_log2FC_preR_preS, NK_preR_preS_p, NK_preR_preS_p_adj
    ),
    by = c("symbol", "ensembl")
  )

log2FC_table <- log2FC_table %>% 
  left_join(preR, by = c("symbol", "ensembl"))

write.csv(
  log2FC_table,
  paste0(sub_output_dir, "Figure4B_all_signatures_logFC_p.csv"),
  row.names = FALSE)
```

```{r}
length(intersect(data$symbol, log2FC_table$symbol))
```

## B Connectivity score (GSEA) on sister conserved genes

```{r include=FALSE}
treatments <- c("Carbo", "NK", "Olaparib")

data_merge <- data %>% 
  left_join(
    log2FC_table,
    by = "symbol"
  ) %>% 
  filter(conserve_gene == TRUE)
drugs <- colnames(data)[-1]

for (t in treatments) {
  df <- NULL
  for (i in drugs){
    
    drug_signature <- data_merge[[i]]
    names(drug_signature) <- data_merge$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    preR_signature <- data_merge[[paste0(t, "_log2FC_preR_preS_adj")]]
    names(preR_signature) <- data_merge$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    tmp <- data.frame(
      drug = i,
      connect_score_log2FC_preR_preS_adj = connect[1],
      connect_p_log2FC_preR_preS_adj = connect[2],
      pearson_n_log2FC_preR_preS_adj = length(preR_signature)
    )
    
    # preR-preS
    
    preR_signature <- data_merge[[paste0(t, "_log2FC_preR_preS")]]
    names(preR_signature) <- data_merge$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
   
    tmp$connectivity_score_log2FC_preR_preS = connect[1]
    tmp$connectivity_p_log2FC_preR_preS = connect[2]
    tmp$connectivity_n_log2FC_preR_preS = length(preR_signature)
    
    # filter by bootstrap p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_boot_p")]] < 0.05, ]
    
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]],
      method = "pearson",
      use = "complete"
    )
    tmp$connectivity_score_log2FC_preR_preS_adj_boot_p = connect[1]
    tmp$connectivity_p_log2FC_preR_preS_adj_boot_p = connect[2]
    tmp$connectivity_n_log2FC_preR_preS_adj_boot_p = length(preR_signature)
    
    # filter by permutation p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_perm_p")]] < 0.05, ]
    
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    tmp$connectivity_score_log2FC_preR_preS_adj_perm_p = connect[1]
    tmp$connectivity_p_log2FC_preR_preS_adj_perm_p = connect[2]
    tmp$connectivity_n_log2FC_preR_preS_adj_perm_p = length(preR_signature)

    # filter by permutation p < 0.05 and bootstrap p < 0.05
    data_tmp <- data_merge[
      data_merge[[paste0(t, "_perm_p")]] < 0.05 &
        data_merge[[paste0(t, "_boot_p")]] < 0.05,
      ]
    
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    tmp$connectivity_score_log2FC_preR_preS_adj_perm_boot_p = connect[1]
    tmp$connectivity_p_log2FC_preR_preS_adj_perm_boot_p = connect[2]
    tmp$connectivity_n_log2FC_preR_preS_adj_perm_boot_p = length(preR_signature)
    
    # filter by preR-preS adjusted p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_preR_preS_p_adj")]] < 0.05, ]
    
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_preR_preS")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    tmp$connectivity_score_log2FC_preR_preS_p_adj = connect[1]
    tmp$connectivity_p_log2FC_preR_preS_p_adj = connect[2]
    tmp$connectivity_n_log2FC_preR_preS_p_adj = length(preR_signature)
    
    # filter by preR-preS p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_preR_preS_p")]] < 0.05, ]
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_preR_preS")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    tmp$connectivity_score_log2FC_preR_preS_p = connect[1]
    tmp$connectivity_p_log2FC_preR_preS_p = connect[2]
    tmp$connectivity_n_log2FC_preR_preS_p = length(preR_signature)
    
    # posT-preT
    drug_signature <- data_merge[[i]]
    names(drug_signature) <- data_merge$symbol
    drug_signature <- na.omit(drug_signature)
    
    preR_signature <- data_merge[[paste0(t, "_log2FC_posT_preT")]]
    names(preR_signature) <- data_merge$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )

    tmp$connectivity_score_log2FC_posT_preT = connect[1]
    tmp$connectivity_p_log2FC_posT_preT = connect[2]
    tmp$connectivity_n_log2FC_posT_preT = length(preR_signature)
    
    # posT-preT adjusted growth control
    drug_signature <- data_merge[[i]]
    names(drug_signature) <- data_merge$symbol
    drug_signature <- na.omit(drug_signature)
    
    preR_signature <- data_merge[[paste0(t, "_log2FC_posT_preT")]]
    names(preR_signature) <- data_merge$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )

    tmp$connectivity_score_log2FC_posT_preT_adj = connect[1]
    tmp$connectivity_p_log2FC_posT_preT_adj = connect[2]
    tmp$connectivity_n_log2FC_posT_preT_adj = length(preR_signature)
    
    # filter by posT-preT adjusted p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_posT_preT_p_adj")]] < 0.05, ]
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_posT_preT")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    tmp$connectivity_score_log2FC_posT_preT_p_adj = connect[1]
    tmp$connectivity_p_log2FC_posT_preT_p_adj = connect[2]
    tmp$connectivity_n_log2FC_posT_preT_p_adj = length(preR_signature)
    
    # filter by posT-preT p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_posT_preT_p")]] < 0.05, ]
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_posT_preT")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    tmp$connectivity_score_log2FC_posT_preT_p = connect[1]
    tmp$connectivity_p_log2FC_posT_preT_p = connect[2]
    tmp$connectivity_n_log2FC_posT_preT_p = length(preR_signature)
    
    df <- rbind.data.frame(df, tmp)
  }
  df <- df %>%
    left_join(
      drug_anno_match,
      by = c("drug" = "match_id")
    )
  write.csv(
    df,
    paste0(
      sub_output_dir,
      t,
      "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      ),
    row.names = FALSE)
}
```

### Visualization - Volcano plot

```{r}
volcano_sub_dir <- paste0(sub_output_dir, "volcano_plot/")
if(!dir.exists(volcano_sub_dir)){
  dir.create(volcano_sub_dir)
}

drugs_NK <- c("nocodazole","clofarabine","pevonedistat","GW-843682X")
drugs_Olaparib <- c(
  "pyrrolidine-dithiocarbamate", "bortezomib", "carfilzomib", "pevonedistat",
  "ixazomib-citrate", "sitagliptin", "delanzomib"
)
drugs_Cabo <- "pevonedistat"
low_color <- blue2
high_color <- red2

for (t in treatments){
  highlight_drug <- read_xlsx(
    paste0(
      sub_output_dir,
      t,
      "_connectivity_score_preR_vs_drug_signature_sister_conserved.xlsx"
    ),
    sheet = 2
  )
  
  if (t == "Carbo"){
    highlight_drug_black <- filter(highlight_drug, pert_iname %in% drugs_Cabo)
    highlight_drug_grey <- filter(highlight_drug, !pert_iname %in% drugs_Cabo)
  } else if (t == "NK"){
    highlight_drug_black <- filter(highlight_drug, pert_iname %in% drugs_NK)
    highlight_drug_grey <- filter(highlight_drug, !pert_iname %in% drugs_NK)
  } else if (t == "Olaparib"){
    highlight_drug_black <- filter(highlight_drug, pert_iname %in% drugs_Olaparib)
    highlight_drug_grey <- filter(highlight_drug, !pert_iname %in% drugs_Olaparib)
  }
  
  highlight_drug_black <- highlight_drug_black %>% 
  select(drug) %>% 
  unlist()
  highlight_drug_grey <- highlight_drug_grey %>% 
  select(drug) %>% 
  unlist()
  
  # preR_preS_adj
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS_adj`,
        p = `connectivity_p_log2FC_preR_preS_adj`,
        n = `connectivity_n_log2FC_preR_preS_adj`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]
  
  highlight_black <- filter(plot_table, drug %in% highlight_drug_black)
  highlight_black$color <- "black"
  highlight_grey <- filter(plot_table, drug %in% highlight_drug_grey)
  highlight_grey$color <- "grey40"
  highlight <- rbind.data.frame(highlight_black, highlight_grey)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(), 
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          #title="PreR-preS log2FC adjusted control \n No filter",
          # subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color), 
      ) +
      geom_point(
        aes(x = x, y = y),
        shape = 1,
        size = 5,
        data = highlight_grey,
        color = "grey40",
        show.legend = FALSE) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_sister_conserved_volcano.pdf"),
    p,
    width = 4.5,
    height = 4.5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_sister_conserved_volcano.png"),
    p,
    width = 4.5,
    height = 4.5
  )
  
  # preR_preS
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS`,
        p = `connectivity_p_log2FC_preR_preS`,
        n = `connectivity_n_log2FC_preR_preS`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]

  highlight_black <- filter(plot_table, drug %in% highlight_drug_black)
  highlight_black$color <- "black"
  highlight_grey <- filter(plot_table, drug %in% highlight_drug_grey)
  highlight_grey$color <- "grey40"
  highlight <- rbind.data.frame(highlight_black, highlight_grey)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          #title="PreR-preS log2FC \n No filter",
          #subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_sister_conserved_volcano.pdf"),
    p,
    width = 4.5,
    height = 4.5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_sister_conserved_volcano.png"),
    p,
    width = 4.5,
    height = 4.5
  )
  
  # preR_preS_adj_boot_p
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS_adj_boot_p`,
        p = `connectivity_p_log2FC_preR_preS_adj_boot_p`,
        n = `connectivity_n_log2FC_preR_preS_adj_boot_p`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  # plot_table$Term <- paste0("bold('", plot_table$Term, "')")
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]

  highlight_black <- filter(plot_table, drug %in% highlight_drug_black)
  highlight_black$color <- "black"
  highlight_grey <- filter(plot_table, drug %in% highlight_drug_grey)
  highlight_grey$color <- "grey40"
  highlight <- rbind.data.frame(highlight_black, highlight_grey)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          # title="PreR-preS log2FC adjusted control \n Bootstrap P < 0.05",
          # subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_boot_p_sister_conserved_volcano.pdf"),
    p,
    width = 4.5,
    height = 4.5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_boot_p_sister_conserved_volcano.png"),
    p,
    width = 4.5,
    height = 4.5
  )
  
    # preR_preS_adj_perm_p
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS_adj_perm_p`,
        p = `connectivity_p_log2FC_preR_preS_adj_perm_p`,
        n = `connectivity_n_log2FC_preR_preS_adj_perm_p`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  # plot_table$Term <- paste0("bold('", plot_table$Term, "')")
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]
  
  highlight_black <- filter(plot_table, drug %in% highlight_drug_black)
  highlight_black$color <- "black"
  highlight_grey <- filter(plot_table, drug %in% highlight_drug_grey)
  highlight_grey$color <- "grey40"
  highlight <- rbind.data.frame(highlight_black, highlight_grey)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          # title="PreR-preS log2FC adjusted control \n Permutation P < 0.05",
          # subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_perm_p_sister_conserved_volcano.pdf"),
    p,
    width = 4.5,
    height = 4.5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_perm_p_sister_conserved_volcano.png"),
    p,
    width = 4.5,
    height = 4.5
  )
  
    # preR_preS_adj_perm_boot_p
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS_adj_perm_boot_p`,
        p = `connectivity_p_log2FC_preR_preS_adj_perm_boot_p`,
        n = `connectivity_n_log2FC_preR_preS_adj_perm_boot_p`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  # plot_table$Term <- paste0("bold('", plot_table$Term, "')")
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]
  
  highlight_black <- filter(plot_table, drug %in% highlight_drug_black)
  highlight_black$color <- "black"
  highlight_grey <- filter(plot_table, drug %in% highlight_drug_grey)
  highlight_grey$color <- "grey40"
  highlight <- rbind.data.frame(highlight_black, highlight_grey)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          # title="PreR-preS log2FC adjust control \n Bootstrap, permutation P < 0.05",
          # subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_perm_boot_p_sister_conserved_volcano.pdf"),
    p,
    width = 4.5,
    height = 4.5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_perm_boot_p_sister_conserved_volcano.png"),
    p,
    width = 4.5,
    height = 4.5
  )
  
    # preR_preS_p_adj
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS_p_adj`,
        p = `connectivity_p_log2FC_preR_preS_p_adj`,
        n = `connectivity_n_log2FC_preR_preS_p_adj`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  # plot_table$Term <- paste0("bold('", plot_table$Term, "')")
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]
  highlight <- read_xlsx(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.xlsx"
      ),
      sheet = 2
    ) %>% 
    select(drug) %>% 
    unlist()
  highlight <- filter(plot_table, drug %in% highlight)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          # title="PreR-preS log2FC \n Adjust P < 0.05",
          # subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_p_adj_sister_conserved_volcano.pdf"),
    p,
    width = 4.5,
    height = 4.5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_p_adj_sister_conserved_volcano.png"),
    p,
    width = 4.5,
    height = 4.5
  )
  
    # preR_preS_p
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS_p`,
        p = `connectivity_p_log2FC_preR_preS_p`,
        n = `connectivity_n_log2FC_preR_preS_p`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]

  highlight_black <- filter(plot_table, drug %in% highlight_drug_black)
  highlight_black$color <- "black"
  highlight_grey <- filter(plot_table, drug %in% highlight_drug_grey)
  highlight_grey$color <- "grey40"
  highlight <- rbind.data.frame(highlight_black, highlight_grey)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          # title="PreR-preS log2FC \n P < 0.05",
          # subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_p_sister_conserved_volcano.pdf"),
    p,
    width = 4.5,
    height = 4.5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_p_sister_conserved_volcano.png"),
    p,
    width = 4.5,
    height = 4.5
  )
}
```

Validate with drug combination screening data from DrugComb

```{r}
DrugComb_drug <- read.csv("./data/input/DrugComb_all_drug.csv")
olaparib_combo <- read.csv("./data/input/DrugComb_summary_Olaparib.csv") %>% 
  filter(drug_col != "NULL")
olaparib_combo <- olaparib_combo %>% 
  mutate(drug = ifelse(drug_row == "Olaparib", drug_col, drug_row))
drugs_combo <- DrugComb_drug %>% 
  filter(dname %in% unique(olaparib_combo$drug)) %>% 
  filter(inchikey %in% drug_anno_match$InChIKey)
olaparib_combo <- olaparib_combo %>% 
  select(block_id, drug, cell_line_name, css, synergy_zip, synergy_bliss, synergy_loewe, synergy_hsa, S_sum, quality, S_mea, S_max) %>% 
  filter(drug %in% drugs_combo$dname) %>% 
  left_join(select(drugs_combo, dname, inchikey), by = c("drug" = "dname"))

olaparib_correlation <- read.csv("./data/output/test/Olaparib_correlation_preR_vs_drug_signature_sister_conserved.csv") %>% 
  filter(InChIKey %in% olaparib_combo$inchikey) %>% 
  select(InChIKey, starts_with("pearson"))

olaparib_combo <- olaparib_combo %>% 
  left_join(olaparib_correlation, by = c("inchikey" = "InChIKey"))

plot_table <- olaparib_combo %>% 
  select(
    css, synergy_zip, synergy_bliss, synergy_loewe, synergy_hsa, S_sum,
    S_mea, S_max, starts_with("pearson_r"))
pdf(file = paste0(sub_output_dir, "Olaparib_DrugComb_pearson_correlation_sister_conserved_heatmap.pdf"))
heatmap(
  as.matrix(plot_table),
  labRow = paste0(olaparib_combo$drug, "_", olaparib_combo$cell_line_name),
  labCol = gsub("pearson_r_log2FC_", "", colnames(plot_table)),
  cexCol = 0.8,
  margins = c(10, 5)
)
dev.off()
```

```{r}
write.csv(
  cor(plot_table),
  paste0(sub_output_dir, "Olaparib_DrugComb_pearson_correlation_sister_conserved.csv")
)
```

```{r}
# summarize across cell lines
plot_table <- olaparib_combo %>% 
  select(
    drug, css, synergy_zip, synergy_bliss, synergy_loewe, synergy_hsa, S_sum,
    S_mea, S_max, starts_with("pearson_r")) %>% 
  group_by(drug) %>% 
  summarise_if(is.numeric, mean, na.rm = TRUE)
pdf(file = paste0(sub_output_dir, t, "_DrugComb_pearson_correlation_summarize_response_across_cells_sister_conserved_heatmap.pdf"))
heatmap(
  as.matrix(plot_table[, -1]),
  labRow = plot_table$drug,
  labCol = gsub("pearson_r_log2FC_", "", colnames(plot_table)[-1]),
  cexCol = 0.8,
  margins = c(10, 5)
  )
dev.off()
```

```{r}
DrugComb_drug <- read.csv("./data/input/DrugComb_all_drug.csv")
olaparib_combo <- read.csv("./data/input/DrugComb_summary_Olaparib.csv") %>% 
  filter(drug_col != "NULL")
olaparib_combo <- olaparib_combo %>% 
  mutate(drug = ifelse(drug_row == "Olaparib", drug_col, drug_row))
drugs_combo <- DrugComb_drug %>% 
  filter(dname %in% unique(olaparib_combo$drug)) %>% 
  filter(inchikey %in% drug_anno_match$InChIKey)
olaparib_combo <- olaparib_combo %>% 
  select(block_id, drug, cell_line_name, css, synergy_zip, synergy_bliss, synergy_loewe, synergy_hsa, S_sum, quality, S_mea, S_max) %>% 
  filter(drug %in% drugs_combo$dname) %>% 
  left_join(select(drugs_combo, dname, inchikey), by = c("drug" = "dname"))

olaparib_correlation <- read.csv("./data/output/test/Olaparib_correlation_preR_vs_drug_signature.csv") %>% 
  filter(InChIKey %in% olaparib_combo$inchikey) %>% 
  select(InChIKey, starts_with("pearson"))

olaparib_combo <- olaparib_combo %>% 
  left_join(olaparib_correlation, by = c("inchikey" = "InChIKey"))

plot_table <- olaparib_combo %>% 
  select(
    css, synergy_zip, synergy_bliss, synergy_loewe, synergy_hsa, S_sum,
    S_mea, S_max, starts_with("pearson_r"))
pdf(file = paste0(sub_output_dir, "Olaparib_DrugComb_pearson_correlation_heatmap.pdf"))
heatmap(
  as.matrix(plot_table),
  labRow = paste0(olaparib_combo$drug, "_", olaparib_combo$cell_line_name),
  labCol = gsub("pearson_r_log2FC_", "", colnames(plot_table)),
  cexCol = 0.8,
  margins = c(10, 5)
)
dev.off()
```

```{r}
write.csv(
  cor(plot_table),
  paste0(sub_output_dir, "Olaparib_DrugComb_pearson_correlation.csv")
)
```

```{r}
# summarize across cell lines
plot_table <- olaparib_combo %>% 
  select(
    drug, css, synergy_zip, synergy_bliss, synergy_loewe, synergy_hsa, S_sum,
    S_mea, S_max, starts_with("pearson_r")) %>% 
  group_by(drug) %>% 
  summarise_if(is.numeric, mean, na.rm = TRUE)
pdf(file = paste0(sub_output_dir, "Olaparib_DrugComb_pearson_correlation_summarize_response_across_cells_heatmap.pdf"))
heatmap(
  as.matrix(plot_table[, -1]),
  labRow = plot_table$drug,
  labCol = gsub("pearson_r_log2FC_", "", colnames(plot_table)[-1]),
  cexCol = 0.8,
  margins = c(10, 5)
  )
dev.off()
```

```{r}
write.csv(
  cor(plot_table[-1]),
  paste0(sub_output_dir, "Olaparib_DrugComb_pearson_correlation_summarize_response_across_cells.csv")
)
```

## C Drug combination validation

```{r}
combo_sub_dir <- paste0(sub_output_dir, "C_drug_combination/")
if(!dir.exists(combo_sub_dir)){
  dir.create(combo_sub_dir)
}
```

### First batch

The % inhibition of drug combination validation experiments are from Jun Dai. They are saved in the file "./data/input/ReSisTrace_drug_combination_validatiion/Drug_combination_inhibition_ReSisTrace_20221025.xlsx". It is already in the synergyfinder input table format.

```{r eval=FALSE, include=FALSE}
table <- read_xlsx("./data/input/ReSisTrace_drug_combination_validatiion/Drug_combination_inhibition_ReSisTrace_20221025.xlsx")

table$conc_unit1 <- table$conc_unit
table$conc_unit2 <- table$conc_unit
table <- table[, colnames(table) != "conc_unit"]
table$conc_unit1[table$drug1 == "NK-FM2931"] <- "ratio"
table$drug1[table$drug1 == "NK-FM2931"] <- "NK : target cell"
table <- ReshapeData(table, data_type = "inhibition")
table <- CalculateSynergy(table)
table <- CalculateSensitivity(table)

saveRDS(table, file = paste0(combo_sub_dir, "synergyfinder_object_with_score_first_batch.RDS"))
```

### Second batch

The % inhibition of drug combination validation experiments are from Jing Jiang. They are saved in the file "./data/input/drug_combination_raw_experiment_data/KURU-29-olaparibb.xlsx". It is already in the synergyfinder input table format.

```{r eval=FALSE, include=FALSE}
table <- read.csv("./data/input/drug_combination_raw_experiment_data/merged_template_2022_11_9.csv")

table <- ReshapeData(table, data_type = "inhibition")
table <- CalculateSynergy(table)
table <- CalculateSensitivity(table)

saveRDS(table, file = paste0(combo_sub_dir, "synergyfinder_object_with_score_second_batch.RDS"))
```

### Third batch

The % inhibition of drug combination validation experiments are from Jun. They are saved in the file "./data/input/ReSisTrace_drug_combination_validatiion/Drug_combination_inhibition_ReSisTrace_20221025.xlsx". It is already in the synergyfinder input table format.

```{r eval=FALSE, include=FALSE}
table <- read_xlsx("./data/input/drug_combination_raw_experiment_data/KURU-29-olaparibb.xlsx")

table <- ReshapeData(table, data_type = "inhibition")
table <- CalculateSynergy(table)
table <- CalculateSensitivity(table)

saveRDS(table, file = paste0(combo_sub_dir, "synergyfinder_object_with_score_third_batch.RDS"))
```

### Merged template

```{r}
selected_blocks <- list(
  "batch1" = 1:8,
  "batch2" = c(5, 6, 7, 18, 19),
  "batch3" = 1)
template <- NULL
tmp <- read_xlsx("./data/input/ReSisTrace_drug_combination_validatiion/Drug_combination_inhibition_ReSisTrace_20221025.xlsx") %>% 
  filter(block_id %in% selected_blocks[[1]]) %>% 
  mutate(block_id = paste0("1-", block_id)) %>% 
  mutate(conc_unit1 = conc_unit, conc_unit2 = conc_unit) %>% 
  select(-conc_unit)
template <- rbind.data.frame(template, tmp)
tmp <- read.csv("./data/input/drug_combination_raw_experiment_data/merged_template_2022_11_9.csv") %>% 
  filter(block_id %in% selected_blocks[[2]]) %>% 
  mutate(block_id = paste0("2-", block_id))
template <- rbind.data.frame(template, tmp)

tmp <- read_xlsx("./data/input/drug_combination_raw_experiment_data/KURU-29-olaparibb.xlsx") %>% 
  filter(block_id %in% selected_blocks[[3]]) %>% 
  mutate(block_id = paste0("3-", block_id)) %>% 
  mutate(conc_unit1 = conc_unit, conc_unit2 = conc_unit) %>% 
  select(-conc_unit)
template <- rbind.data.frame(template, tmp)

write.csv(template, paste0(combo_sub_dir, "all_selected_combos.csv"))
```

### Visualization

```{r}
# First batch
table <- readRDS(paste0(combo_sub_dir, "synergyfinder_object_with_score_first_batch.RDS"))
# Second batch
table2 <- readRDS(paste0(combo_sub_dir, "synergyfinder_object_with_score_second_batch.RDS"))
# Third batch
table3 <- readRDS(paste0(combo_sub_dir, "synergyfinder_object_with_score_third_batch.RDS"))
synergys <- c("HSA", "ZIP", "Loewe", "Bliss")

# dose-response range
min_score_response <- min(c(table$response$response, table2$response$response, table3$response$response))
max_score_response <- max(c(table$response$response, table2$response$response, table3$response$response))
low_color_response <- min_score_response - 5
high_color_response <- max_score_response + 5

# First batch
for (b in unique(table$drug_pairs$block_id)){
  # dose-response map
  Plot2DrugHeatmap(
    table,
    plot_block = b,
    plot_value = "response",
    summary_statistic = "mean",
    high_value_color = red,
    low_value_color = blue,
    text_label_size_scale = 1.4,
    statistic = "sd",
    color_range = c(low_color_response, high_color_response)
  )
  ggsave(
     paste0(
      combo_sub_dir,
      gsub(" ", "_", gsub(" : ", "-", table$drug_pairs$drug1[table$drug_pairs$block_id == b])),
      "_",
      table$drug_pairs$drug2[table$drug_pairs$block_id == b],
      "_dose_response.pdf"
    ),
    # width = 3, # Science
    # height = 3 # Science
    width = 4,
    height = 2.8
  )
  
  # synergy map
  for (s in synergys){
  min_score <- min(c(table$synergy_scores[[paste0(s,"_synergy")]], table2$synergy_scores[[paste0(s,"_synergy")]], table3$synergy_scores[[paste0(s,"_synergy")]]))
  max_score <- max(c(table$synergy_scores[[paste0(s,"_synergy")]], table2$synergy_scores[[paste0(s,"_synergy")]], table3$synergy_scores[[paste0(s,"_synergy")]]))
  low_color <- min_score - 5
  high_color <- max_score + 5
  low_z <- min(table$synergy_scores[table$synergy_scores$block_id == b, paste0(s,"_synergy")]) - 5
  high_z <- max(table$synergy_scores[table$synergy_scores$block_id == b, paste0(s,"_synergy")]) + 5
    pdf(
      paste0(
        combo_sub_dir,
        gsub(" ", "_", gsub(" : ", "-", table$drug_pairs$drug1[table$drug_pairs$block_id == b])),
        "_",
        table$drug_pairs$drug2[table$drug_pairs$block_id == b],
        "_", s, ".pdf"
      ),
      # width = 3, # Science
      # height = 3 # Science
      width = 5,
      height = 5
    )
    Plot2DrugSurface(
      table,
      plot_block = b,
      plot_value = paste0(s, "_synergy"),
      summary_statistic = "mean",
      high_value_color = red,
      low_value_color = blue,
      z_range = c(low_z, high_z),
      color_range = c(low_color, high_color),
      axis_line = FALSE
    )
    dev.off()
    }
}

# Second batch

for (b in c(5, 6, 7, 18, 19)){
  # dose-response map
  Plot2DrugHeatmap(
    table2,
    plot_block = b,
    plot_value = "response",
    summary_statistic = "mean",
    high_value_color = red,
    low_value_color = blue,
    text_label_size_scale = 1.4,
    statistic = "sd",
    color_range = c(low_color_response, high_color_response)
  )
  ggsave(
    paste0(
      combo_sub_dir,
      gsub(" ", "_", gsub(" : ", "-", table2$drug_pairs$drug1[table2$drug_pairs$block_id == b])),
      "_",
      table2$drug_pairs$drug2[table2$drug_pairs$block_id == b],
      "_dose_response.pdf"
    ),
    # width = 3, # Science
    # height = 3 # Science
    width = 4,
    height = 2.8
  )
  
  # synergy map
  for (s in synergys){
    min_score <- min(c(table$synergy_scores[[paste0(s,"_synergy")]], table2$synergy_scores[[paste0(s,"_synergy")]], table3$synergy_scores[[paste0(s,"_synergy")]]))
    max_score <- max(c(table$synergy_scores[[paste0(s,"_synergy")]], table2$synergy_scores[[paste0(s,"_synergy")]], table3$synergy_scores[[paste0(s,"_synergy")]]))
    low_color <- min_score - 5
  high_color <- max_score + 5
    low_z <- min(table2$synergy_scores[table2$synergy_scores$block_id == b, paste0(s,"_synergy")]) - 5
    high_z <- max(table2$synergy_scores[table2$synergy_scores$block_id == b, paste0(s,"_synergy")]) + 5
    pdf(
      paste0(
        combo_sub_dir,
        gsub(" ", "_", gsub(" : ", "-", table2$drug_pairs$drug1[table2$drug_pairs$block_id == b])),
        "_",
        table2$drug_pairs$drug2[table2$drug_pairs$block_id == b],
        "_", s, ".pdf"
      ),
      # width = 3, # Science
      # height = 3 # Science
      width = 5,
      height = 5
    )
    Plot2DrugSurface(
      table2,
      plot_block = b,
      plot_value = paste0(s, "_synergy"),
      summary_statistic = "mean",
      high_value_color = red,
      low_value_color = blue,
      z_range = c(low_z, high_z),
      color_range = c(low_color, high_color),
      axis_line = FALSE
    )
    dev.off()
    }
}

# Third batch
for (b in 1){
  # dose-response map
  Plot2DrugHeatmap(
    table3,
    plot_block = b,
    plot_value = "response",
    summary_statistic = "mean",
    high_value_color = red,
    low_value_color = blue,
    text_label_size_scale = 1.4,
    statistic = "sd",
    color_range = c(low_color_response, high_color_response)
  )
  ggsave(
    paste0(
      combo_sub_dir,
      gsub(" ", "_", gsub(" : ", "-", table3$drug_pairs$drug1[table3$drug_pairs$block_id == b])),
      "_",
      table3$drug_pairs$drug2[table3$drug_pairs$block_id == b],
      "_dose_response.pdf"
    ),
    # width = 3, # Science
    # height = 3 # Science
    width = 4,
    height = 2.8
  )
  
  for (s in synergys){
    min_score <- min(c(table$synergy_scores[[paste0(s,"_synergy")]], table2$synergy_scores[[paste0(s,"_synergy")]], table3$synergy_scores[[paste0(s,"_synergy")]]))
    max_score <- max(c(table$synergy_scores[[paste0(s,"_synergy")]], table2$synergy_scores[[paste0(s,"_synergy")]], table3$synergy_scores[[paste0(s,"_synergy")]]))
    low_color <- min_score - 5
  high_color <- max_score + 5
    low_z <- min(table3$synergy_scores[table3$synergy_scores$block_id == b, paste0(s,"_synergy")]) - 5
    high_z <- max(table3$synergy_scores[table3$synergy_scores$block_id == b, paste0(s,"_synergy")]) + 5
    pdf(
      paste0(
        combo_sub_dir,
        "third_batch_",
        gsub(" ", "_", gsub(" : ", "-", table3$drug_pairs$drug1[table3$drug_pairs$block_id == b])),
        "_",
        table3$drug_pairs$drug2[table3$drug_pairs$block_id == b],
        "_", s, ".pdf"
      ),
      # width = 3, # Science
      # height = 3 # Science
      width = 5,
      height = 5
    )
    Plot2DrugSurface(
      table3,
      plot_block = b,
      plot_value = paste0(s, "_synergy"),
      summary_statistic = "mean",
      high_value_color = red,
      low_value_color = blue,
      z_range = c(low_z, high_z),
      color_range = c(low_color, high_color),
      axis_line = FALSE
    )
    dev.off()
    }
}
```

### Heat map

```{r}
min_HSA_score <- min(c(table$synergy_scores[["HSA_synergy"]], table2$synergy_scores[["HSA_synergy"]]))
max_HSA_score <- max(c(table$synergy_scores[["HSA_synergy"]], table2$synergy_scores[["HSA_synergy"]]))
```

```{r}
# First batch
max_synergy_scores <- table$synergy_scores %>% 
  group_by(block_id) %>% 
  summarise(
    HSA_synergy_max = max(HSA_synergy), 
    ZIP_synergy_max = max(ZIP_synergy), 
    Bliss_synergy_max = max(Bliss_synergy), 
    Loewe_synergy_max = max(Loewe_synergy)
    )

synergy_table <- table$drug_pairs %>%
  filter(drug2 != "Cilengitide") %>% 
  select(
    block_id, drug1, drug2, HSA = HSA_synergy, HSA_P = HSA_synergy_p_value,
    ZIP = ZIP_synergy, ZIP_P = ZIP_synergy_p_value,
    Loewe = Loewe_synergy, Loewe_P = Loewe_synergy_p_value,
    Bliss = Bliss_synergy, Bliss_P = Bliss_synergy_p_value
  ) %>% 
  left_join(max_synergy_scores, by = "block_id")

# Second batch
max_synergy_scores2 <- table2$synergy_scores %>% 
  filter(block_id %in% c(5, 6, 7, 18, 19)) %>% 
  group_by(block_id) %>% 
  summarise(
    HSA_synergy_max = max(HSA_synergy), 
    ZIP_synergy_max = max(ZIP_synergy), 
    Bliss_synergy_max = max(Bliss_synergy), 
    Loewe_synergy_max = max(Loewe_synergy)
  )

synergy_table2 <- table2$drug_pairs %>% 
  filter(block_id %in% c(5, 6, 18, 19)) %>%
  select(
    block_id, drug1, drug2, HSA = HSA_synergy, HSA_P = HSA_synergy_p_value,
    ZIP = ZIP_synergy, ZIP_P = ZIP_synergy_p_value,
    Loewe = Loewe_synergy, Loewe_P = Loewe_synergy_p_value,
    Bliss = Bliss_synergy, Bliss_P = Bliss_synergy_p_value
  ) %>% 
  left_join(max_synergy_scores2, by = "block_id")

# Third batch
max_synergy_scores3 <- table3$synergy_scores %>% 
  group_by(block_id) %>% 
  summarise(
    HSA_synergy_max = max(HSA_synergy), 
    ZIP_synergy_max = max(ZIP_synergy), 
    Bliss_synergy_max = max(Bliss_synergy), 
    Loewe_synergy_max = max(Loewe_synergy)
    )

synergy_table3 <- table3$drug_pairs %>%
  select(
    block_id, drug1, drug2, HSA = HSA_synergy, HSA_P = HSA_synergy_p_value,
    ZIP = ZIP_synergy, ZIP_P = ZIP_synergy_p_value,
    Loewe = Loewe_synergy, Loewe_P = Loewe_synergy_p_value,
    Bliss = Bliss_synergy, Bliss_P = Bliss_synergy_p_value
  ) %>% 
  left_join(max_synergy_scores, by = "block_id")

synergy_table <- synergy_table %>% 
  rbind.data.frame(synergy_table2) %>% 
  rbind.data.frame(synergy_table3)

synergy_table$block_id <- seq(1, nrow(synergy_table))
synergy_table$drug1 <- gsub("NK:target", "NK", synergy_table$drug1)
synergy_table$drug1 <- gsub("NK : target cell", "NK", synergy_table$drug1)
write.csv(
  synergy_table,
  paste0(combo_sub_dir, "drug_combination_table.csv"),
  row.names = FALSE
)
```


```{r}
table <- read.csv(paste0(combo_sub_dir, "drug_combination_table.csv"))

table <- table %>% 
  mutate(combo = paste0(drug1, " - ", drug2))
score <- table %>% 
  select(combo, HSA, ZIP, Loewe, Bliss) %>% 
  pivot_longer(!combo, names_to = "method", values_to = "synergy_score")

table[table == "< 2e-324"] <- 0
table[, 4:nrow(table)] <- sapply(table[, 4:nrow(table)], as.numeric)
table$sort_value <- table$HSA # rowMeans(table[, c("HSA", "ZIP", "Loewe", "Bliss")])
p_value <- table %>% 
  select(combo, HSA_P, ZIP_P, Loewe_P, Bliss_P) %>% 
  pivot_longer(!combo, names_to = "method", values_to = "p_value")
p_value <- p_value %>% 
  mutate(
    signif = case_when(
      p_value < .05 & p_value >= .01 ~ "*", 
      p_value < .01 & p_value >= .001 ~ "**", 
      p_value < .001 & p_value >= .0001 ~ "***", 
      p_value < .0001 ~ "****",
      TRUE ~ ""
    )
  )
p_value$method <- gsub("_P", "", p_value$method)
plot_table <- score %>% left_join(p_value, by = c("combo", "method"))
plot_table$drug2 <- gsub(".* \\- ", "", plot_table$combo)
plot_table$drug1 <- gsub(" \\- .*", "", plot_table$combo)
plot_table <- plot_table %>% 
  left_join(select(table, combo, sort_value), by = "combo")
plot_table$drug1 <- factor(plot_table$drug1, levels = c("Carboplatin", "Olaparib", "NK"))
plot_table$combo <- factor(plot_table$combo, levels = unique(plot_table$combo[order(plot_table$sort_value + (as.numeric(plot_table$drug1) * 100))]))
    
start_point <- min_HSA_score - 5
end_point <- max_HSA_score + 5
colour_breaks <- c(start_point, 0, end_point)
colours <- c("#3C5488FF", "#FFFFFF", "#DC0000FF")
  
p <- ggplot2::ggplot(
      data = plot_table,
      aes(x = method, y = combo, fill = synergy_score)#, group = drug1)
    ) +
      ggplot2::geom_tile() +
      ggplot2::geom_text(
        ggplot2::aes(label = signif),
        size = 18/.pt,
        nudge_y = -0.2,
        fontface = "plain"
      ) +
  labs(fill = "Synergy score") +
  scale_y_discrete(
    breaks=levels(plot_table$combo),
    labels=gsub(".* \\- ", "", levels(plot_table$combo))) +
  # scale_fill_gradient2(low = "#3C5488FF", high = "#DC0000FF") +
  scale_fill_gradientn(
    limits  = c(start_point, end_point),#range(plot_table$value),
    colours = colours[c(1, seq_along(colours), length(colours))],
    values  = c(
      0,
      scales::rescale(colour_breaks, from = c(start_point, end_point)), #range(plot_table$value)),
      1
    ),
    oob = scales::oob_squish_any
  ) +
  ggplot2::guides(
    fill = ggplot2::guide_colorbar(
      barheight = 1.5,
      barwidth = 10,
      ticks = FALSE
    )
  ) +
  facet_grid(
    rows = vars(drug1), scales = "free", space="free_y"
  ) +
  # Add the title for heatmap
  ggplot2::theme(
    panel.background = ggplot2::element_blank(),
    # Set label's style of heatmap
    axis.text.x = ggplot2::element_text(
      size = 21,
      face = "plain",
      color = "black"
    ),
    axis.text.y= ggplot2::element_text(
      size = 21,
      face = "plain",
      color = "black"
    ),
    axis.ticks = element_blank(),
    axis.title =  element_blank(),
    legend.title = ggplot2::element_text(
      size = 21,
      color = "black"
    ),
    legend.text = ggplot2::element_text(
      size = 21,
      color = "black"
    ),
    legend.position = "top",
    legend.background = element_rect(color = NA),
    strip.text.y.right = element_text(angle = 0, color = "white", size = 21),
    strip.background = element_rect(fill = "#B3B3B3")
  )
# ggsave(paste0(combo_sub_dir, "heatmap_for_all_drugs.pdf"), height = 7, width = 14) # Science format
ggsave(paste0(combo_sub_dir, "heatmap_for_all_drugs.pdf"), height = 6, width = 15)
```

# Figure S6 scRNA sequence

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS6/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

```{r}
scRNA <- readRDS("./data/input/scRNA_seq_for_drug_signature/drug_induced_sample_all_merged.RDS")
sig <- read.csv("./data/input/scRNA_seq_for_drug_signature/preR_preS_drug_signature_all.csv")
con_score <- list(
  Carbo = read.csv("./data/input/scRNA_seq_for_drug_signature/Carbo_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"),
  Olaparib = read.csv("./data/input/scRNA_seq_for_drug_signature/Olaparib_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"),
  NK = read.csv("./data/input/scRNA_seq_for_drug_signature/NK_connectivity_score_preR_vs_drug_signature_sister_conserved.csv")
  )
```

### UMAP

```{r}
scRNA$treatment <- gsub("_", " ", scRNA$treatment)

Idents(scRNA) <- "treatment"
colors <- c(
  "DMSO" = "grey",
  "Pevonedistat hydrochloride H" = "#00A087B2", 
  "Pevonedistat hydrochloride L" = "#91D1C2B2",
  
  "Ixazomib citrate" = "#DC0000B2", # proteasome inhibitor
  "Carfilzomib" = "#E64B35B2", # proteasome inhibitor
  "Delanzomib" = "#F39B7FB2", # proteasome inhibitor
  "Bortezomib" = "#EE6677B2", # proteasome inhibitor
  "Pyrrolidinedithiocarbamate ammonium" = "#8491B4B2", 
  "Sitagliptin" = "#AA4499B2",
  
  "Nocodazole" = "#7E6148B2",
  "Clofarabine" = "#4DBBD5B2",
  "GW843682X" = "#DDAA33B2"
)
levels(scRNA) <- c(
  "DMSO", "Pevonedistat hydrochloride H", "Pevonedistat hydrochloride L",
  "Ixazomib citrate", "Carfilzomib", "Delanzomib", "Bortezomib",
  "Pyrrolidinedithiocarbamate ammonium", "Sitagliptin", "Nocodazole",
  "Clofarabine", "GW843682X"
)
DimPlot(scRNA, reduction = "umap") +
  scale_color_manual(values = colors)

ggsave(paste0(sub_output_dir, "umap_all_drugs.pdf"), width = 6.5, height = 3)
```
```{r}
treatments <- names(colors)[-1]
for (t in treatments){
  DimPlot(subset(scRNA, cells = colnames(scRNA)[scRNA$treatment %in% c("DMSO", t)]), reduction = "umap") +
  scale_color_manual(values = colors) +
    ggtitle(t) +
    theme(
      legend.position="none",
      axis.ticks.x = element_blank(),
      axis.text = element_text(size = 21, colour = "black"),
      axis.title = element_text(size = 21, colour = "black"),
      plot.title = element_text(size = 21, hjust = 0.5, colour = "black", face = "plain")
    )
   if (t == "Pevonedistat hydrochloride L"){
    ggsave(paste0(sub_output_dir, "umap_", t, "_DMSO.pdf"), width = 4.5, height = 4.5)
    ggsave(paste0(sub_output_dir, "umap_", t, "_DMSO.png"), width = 4.5, height = 4.5)
    ggsave(paste0(sub_output_dir, "umap_", t, "_DMSO_supplement.pdf"), width = 5, height = 5)
    ggsave(paste0(sub_output_dir, "umap_", t, "_DMSO_supplement.png"), width = 5, height = 5)
   } else if (t == "Ixazomib citrate"| t == "GW843682X") {
    ggsave(paste0(sub_output_dir, "umap_", t, "_DMSO.pdf"), width = 4.5, height = 4.5)
    ggsave(paste0(sub_output_dir, "umap_", t, "_DMSO.png"), width = 4.5, height = 4.5)
   } else {
    ggsave(paste0(sub_output_dir, "umap_", t, "_DMSO.pdf"), width = 5, height = 5)
    ggsave(paste0(sub_output_dir, "umap_", t, "_DMSO.png"), width = 5, height = 5)
   }
}
```

### Scatter plot

```{r}
combo <- data.frame(
  drug1 = c(rep("Carbo", 2), rep("Olaparib", 8), rep("NK", 5)),
  drug2 = c(
    "Pevonedistat_hydrochloride_H",
    "Pevonedistat_hydrochloride_L",
    
    "Pevonedistat_hydrochloride_H",
    "Pevonedistat_hydrochloride_L",
  "Ixazomib_citrate", "Carfilzomib", "Delanzomib", "Bortezomib",
  "Pyrrolidinedithiocarbamate_ammonium", "Sitagliptin",
  
    "Pevonedistat_hydrochloride_H",
    "Pevonedistat_hydrochloride_L",
    "Nocodazole", "Clofarabine",
    "GW843682X")
)
for (i in 1:nrow(combo)){
  data_tmp <- sig[sig[[paste0(combo$drug1[i], "_boot_p")]] < 0.05, ]
  plot_table <- data_tmp %>% 
    dplyr::select(
      preR = paste0(combo$drug1[i], "_log2FC_preR_preS_adj"),
      drug = paste0(combo$drug2[i], "_DMSO_log2FC")
    ) %>% 
    na.omit()
  
  connect_score <- con_score[[combo$drug1[i]]] %>% 
    filter(drug == combo$drug2[i]) %>% 
    select(connectivity_score_log2FC_preR_preS_adj_boot_p) %>% 
    unlist()
  connect_p <- con_score[[combo$drug1[i]]] %>% 
    filter(drug == combo$drug2[i]) %>% 
    select(connectivity_p_log2FC_preR_preS_adj_boot_p) %>% 
    unlist()
  
  range <- range(
      plot_table
  )
  range_x <- range(plot_table$preR)
  range_y <- range(plot_table$drug)
  p <- ggplot(
    data = plot_table,
    aes(
      x = preR,
      y = drug
    )
    ) +
    geom_point(
      color = "#3C548855",
      size = 5,
      show.legend = FALSE
    ) +
    # ylim(range) +
    # xlim(range) +
    # geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
    geom_vline(xintercept = 0, linetype = "dashed") + 
    geom_hline(yintercept = 0, linetype = "dashed") +
    annotate(
      geom = "text",
      x = range_x[2] - (range_x[2] - range_x[1]) * 0.4,
      y = range_y[2] - (range_y[2] - range_y[1]) * 0.1,
      label = paste0(
        "Connectivity score = ", signif(connect_score, 2), "\n",
        ifelse(
          connect_p == 0, "P < 2.2E-16",
          paste0("P =", format(connect_p, scientific = T))
        )
      ),
      size = 5.5
    ) + 
    labs(
      x = paste0(gsub("Carbo", "Carboplatin", combo$drug1[i]), " preR vs preS"),
      y = paste0(gsub("_", " ", combo$drug2[i]), " vs DMSO")
    ) +
    theme_classic() +
    theme(
      legend.position="none",
      axis.ticks.x = element_blank(),
      axis.text = element_text(size = 21, colour = "black"),
      axis.title = element_text(size = 21, colour = "black"),
      plot.title = element_text(size = 21, hjust = 0.5, colour = "black", face = "bold")
    )
  p
  if (
    (combo$drug1[i] == "Carbo" & combo$drug2[i] == "Pevonedistat_hydrochloride_L")|
    (combo$drug1[i] == "Olaparib" & combo$drug2[i] == "Ixazomib_citrate")|
    (combo$drug1[i] == "NK" & combo$drug2[i] == "GW843682X")) {
    ggsave(paste0(sub_output_dir, "scatter_plot_", combo$drug1[i],"_", combo$drug2[i], ".png"), p,
         width = 4.5, height = 4.5)
  ggsave(paste0(sub_output_dir, "scatter_plot_", combo$drug1[i],"_", combo$drug2[i], ".pdf"), p,
         width = 4.5, height = 4.5)
  } else {
        ggsave(paste0(sub_output_dir, "scatter_plot_", combo$drug1[i],"_", combo$drug2[i], ".png"), p,
         width = 5, height = 5)
  ggsave(paste0(sub_output_dir, "scatter_plot_", combo$drug1[i],"_", combo$drug2[i], ".pdf"), p,
         width = 5, height = 5)
  }
  
}
```

# Figure S7

## A: COV362 survival plot

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS7/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

```{r}
data_clinical_patient = read.delim("./data/input/BRCAness/data_clinical_patient.txt", skip = 4)
data_clinical_sample = read.delim("./data/input/BRCAness/data_clinical_sample.txt", skip = 4)
data_mrna = readRDS("./data/input/BRCAness/EOC_deconv_TCGA_expr_mat.RDS")
colnames(data_mrna) = gsub("\\.","\\-", colnames(data_mrna))
colnames(data_mrna) = substr(colnames(data_mrna), 1, 15)
data_mrna <- data_mrna[apply(data_mrna, 1, function(x){
  sum(is.na(x))
}) < 300, ]

identical(data_clinical_patient$PATIENT_ID, data_clinical_sample$PATIENT_ID) # data aligned already
table(data_clinical_sample$GRADE)
high_grade = grepl("G2|G3|G4", data_clinical_sample$GRADE)

data_clinical_sample = data_clinical_sample[high_grade, ]
data_clinical_patient = data_clinical_patient[high_grade, ]

patients = names(data_mrna) # 307 patients
time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]
table(status, exclude = NULL)
status[which(status == "")] = NA

length(which(is.na(time)==F)) # 270 has survival
length(which(is.na(status)==F)) # 271 has status
cov362 <- read.csv("./data/input/BRCAness/deg_COV362.csv")
```

#### Test for good p value

```{r}
p_threshold = c(1*10^-(2:14), 5*10^-(2:14))
fc_threshold = 0
conserve_gene_only <- FALSE

res <- mat.or.vec(length(p_threshold), 3)
colnames(res) <- c("p_threshold", "n_gene", "survival_p")

for (i in 1:length(p_threshold)){
  data_mrna2 <- data_mrna
  if (conserve_gene_only){
    cov362_test <- filter(cov362, ensembl %in% sister_conserve_gene)
  } else {
    cov362_test <- cov362
  }
  res[i, 1] <- p_threshold[i]
  
  index1 = which(cov362_test$avg_log2FC <= -fc_threshold & cov362_test$p_val_adj <= p_threshold[i]) # downregulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(cov362_test$avg_log2FC > fc_threshold & cov362_test$p_val_adj <= p_threshold[i]) # upregulated genes when BRCA1 is knocked out
  index2_symbol = cov362_test$symbol[index2]
  
  index = c(index1, index2)
  length(index)
  cov362_test = cov362_test[index, ]
  # cov362_test$symbol
  
  cov362_test = cov362_test[match(rownames(data_mrna2), cov362_test$symbol), ]
  
  res[i, 2] <- nrow(na.omit(cov362_test))
  
  # reverse index2 gene expressions
  data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ] = -data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ]
  
  # gene selection
  index = which(is.na(cov362_test$symbol) == F)
  
  res1 = c()
  for (j in 1:length(patients)){
    cat('\r', j)
    res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
    res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])) # rank of the BRCAness genes
    # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])*(-log10(cov362_test$p_val_adj[index]))) # rank of the BRCAness genes
    
  }
  time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
  status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]
  
  group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
  data_plot = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
  set.seed(123)
  fit <- survfit(Surv(time, status) ~ group, data = data_plot)
  fit.coxph = coxph(Surv(time, status) ~ group, data = data_plot)
  s <- summary(fit.coxph)
  res[i, 3] <- s$logtest[3]
}

write.csv(
  res,
  paste0(sub_output_dir, "FigureS7A_BRCA_survival_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

#### Survival plot with selected parameters

```{r}
conserve_gene_only <- FALSE

if (conserve_gene_only) {
  p_threshold = 1*10^(-8)
  fc_threshold = 0
} else {
  p_threshold = 1*10^(-8)
  fc_threshold = 0
}
data_mrna2 <- data_mrna
  
if (conserve_gene_only){
  cov362_test <- filter(cov362, ensembl %in% sister_conserve_gene)
} else {
  cov362_test <- cov362
}

index1 = which(cov362_test$avg_log2FC <= -fc_threshold & cov362_test$p_val_adj <= p_threshold) # downregulated genes when BRCA1 is knocked out (BRCA-)
index2 = which(cov362_test$avg_log2FC > fc_threshold & cov362_test$p_val_adj <= p_threshold) # upregulated genes when BRCA1 is knocked out
index2_symbol = cov362_test$symbol[index2]

index = c(index1, index2)
length(index)
cov362_test = cov362_test[index, ]
# cov362_test$symbol

cov362_test = cov362_test[match(rownames(data_mrna2), cov362_test$symbol), ]

# reverse index2 gene expressions
data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ] = -data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ]

# gene selection
index = which(is.na(cov362_test$symbol) == F)

res1 = c()
for (j in 1:length(patients)){
  cat('\r', j)
  res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
  res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])) # rank of the BRCAness genes
  # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])*(-log10(cov362_test$p_val_adj[index]))) # rank of the BRCAness genes
  
}
time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]

group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
data_plot = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
set.seed(123)
fit <- survfit(Surv(time, status) ~ group, data = data_plot)
fit.coxph = coxph(Surv(time, status) ~ group, data = data_plot)
s <- summary(fit.coxph)
# res[i, 3] <- s$logtest[3]
  
pdf(paste0(sub_output_dir,"FigureS7A_threshold_", p_threshold, ifelse(conserve_gene_only, "BRCA_COV_OS_conserve_gene.pdf", "BRCA_COV_OS.pdf")), width = 4, height = 5.5, pointsize = 20)
fig <- ggsurvplot(
  fit,
  data = data_plot,
  palette = c(red, blue),
  risk.table = T,
  legend.title = "HRD (COV362)", # - undeconvoluted",
  legend.labs = c("high", "low"),
  xlab = "Time (months)", ylab = "Overall survival",
  ggtheme = theme_survminer(
    base_size = 14,
    font.main = c(14, "plain", "black"),
    font.x = c(14, "plain", "black"),
    font.y = c(14, "plain", "black"),
    font.caption = c(14, "plain", "black"),
    font.legend = c(14, "plain", "black"),
    font.tickslab = c(14, "plain", "black"),
    font.submain = c(14, "plain", "black")
  ),
  pval = signif(s$logtest[3], 2),
  pval.coord = c(100, 0.75)
  ) +
  guides(colour = guide_legend(nrow = 2))
fig
dev.off()
```

## B: Bar plot for TCGA BRCAness signatute's log2FC

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "FigureS7/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}

lseq <- function(from=1, to=100000, length.out=6) {
  # logarithmic spaced sequence
  # blatantly stolen from library("emdbook"), because need only this
  exp(seq(log(from), log(to), length.out = length.out))
}

conserve_gene_only <- TRUE
```

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
log2FC_table <- read.csv("./data/output/Figure3/Figure3D_logFC_table_for_plots.csv")
```

```{r}
conserve_gene_only <- TRUE

if (conserve_gene_only){
  plot_table <- filter(log2FC_table, conserve_gene)
} else {
  plot_table <- log2FC_table
}
```

```{r}
selected_p_threshold <- 0.005

table_tests <- plot_table %>% 
  filter(BRCAness_tcga_p_adj <= selected_p_threshold) %>% 
  select(ensembl, starts_with("BRCAness_tcga"), ends_with("log2FC_preR_preS_adj")) %>% 
  na.omit() %>% 
  arrange(BRCAness_tcga_logFC)

# BRCAness
pdf(
  paste0(
    sub_output_dir, "FigureS7B_BRCA_signature_tcga_p_", signif(selected_p_threshold, 2),
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(table_tests$BRCAness_tcga_logFC > 0, red, blue)
fig = barplot(
  table_tests$BRCAness_tcga_logFC,
  names.arg = table_tests$ensembl,
  las = 2,
  main = "BRCAness signature (TCGA)",
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# Olaparib
r <- cor(
  table_tests$BRCAness_tcga_logFC,
  -table_tests$Olaparib_log2FC_preR_preS_adj,
  method = "spearman", use = "complete"
  )
pdf(paste0(sub_output_dir,"FigureS7B_Olaparib_signature_tcga_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$Olaparib_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$Olaparib_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("Olaparib to TCGA, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# Carboplatin
r <- cor(
  table_tests$BRCAness_tcga_logFC,
  -table_tests$Carbo_log2FC_preR_preS_adj,
  method = "spearman", use = "complete"
)
pdf(paste0(sub_output_dir,"FigureS7B_Carboplatin_signature_tcga_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$Carbo_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$Carbo_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("Carboplatin to TCGA, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# NK
r <- cor(
  table_tests$BRCAness_tcga_logFC,
  -table_tests$NK_log2FC_preR_preS_adj,
  method = "spearman", use = "complete"
  )
pdf(paste0(sub_output_dir,"FigureS7B_NK_signature_tcga_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$NK_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$NK_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("NK to TCGA, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()
```

## C: Bar plot for COV362 BRCAness signatute's log2FC

```{r}
selected_p_threshold <- 1 * 10^-8

table_tests <- plot_table %>% 
  filter(BRCAness_cov_p_adj <= selected_p_threshold) %>% 
  select(ensembl, starts_with("BRCAness_cov"), ends_with("log2FC_preR_preS_adj")) %>% 
  na.omit() %>% 
  arrange(BRCAness_cov_logFC)

# BRCAness
pdf(
  paste0(
    sub_output_dir, "FigureS7C_BRCA_signature_cov_p_", signif(selected_p_threshold, 2),
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(table_tests$BRCAness_cov_logFC > 0, red, blue)
fig = barplot(
  table_tests$BRCAness_cov_logFC,
  names.arg = table_tests$ensembl,
  las = 2,
  main = "BRCAness signature (COV362)",
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# Olaparib
r <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Olaparib_log2FC_preR_preS_adj)
pdf(paste0(sub_output_dir,"FigureS7C_Olaparib_signature_cov_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$Olaparib_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$Olaparib_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("Olaparib to COV362, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# Carboplatin
r <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Carbo_log2FC_preR_preS_adj)
pdf(paste0(sub_output_dir,"FigureS7C_Carboplatin_signature_cov_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$Carbo_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$Carbo_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("Carboplatin to COV362, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# NK
r <- cor(table_tests$BRCAness_cov_logFC, -table_tests$NK_log2FC_preR_preS_adj)
pdf(paste0(sub_output_dir,"FigureS7C_NK_signature_cov_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$NK_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$NK_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("NK to COV362, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()
```

# test

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "test/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
treatments <- c("Carbo", "NK", "Olaparib")
```

```{r}
data <- read.csv("./data/input/consensus_signature_drugs.csv", row.names = 1) %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("symbol")

drug_anno <- read.csv("./data/input/drug_repurposing_hub_alldrugsanno.csv")

length(setdiff(colnames(data)[-1], c(drug_anno$deprecated_broad_id, drug_anno$broad_id)))

## Match drug by "broad_id"

drug_anno_match <- drug_anno %>% 
  filter(broad_id %in% colnames(data)) %>% 
  mutate(match_id = broad_id)
drug_anno_match
drug_anno_match_deprecated_id <- drug_anno %>% 
  filter(deprecated_broad_id %in% colnames(data),
         !broad_id %in% drug_anno_match$broad_id) %>% 
  mutate(match_id = deprecated_broad_id)
drug_anno_match <- rbind.data.frame(drug_anno_match, drug_anno_match_deprecated_id)
drug_anno_match$deprecated_broad_id[drug_anno_match$deprecated_broad_id == "NA-NA"] <- NA
drug_anno_match <- drug_anno_match[!duplicated(drug_anno_match$broad_id), ]
```


```{r}
drug_without_anno <- setdiff(colnames(data)[-1], c(drug_anno_match$broad_id, drug_anno_match_deprecated_id$deprecated_broad_id))
write.csv(
  data.frame(drug = drug_without_anno),
  paste0(sub_output_dir, "drug_without_anno.csv"),
  row.names = FALSE
)
```

Table for log2FCs

```{r}
# log2FC for preR-preS-adjusted
log2FC_table <- read.csv("./data/output/Figure4/Figure4B_all_signatures_logFC_p.csv")

length(intersect(data$symbol, log2FC_table$symbol))
```

## Pearson correlation coefficient

```{r}
data_merge <- data %>% 
  left_join(
    log2FC_table,
    by = "symbol"
  )
drugs <- colnames(data)[-1]
for (t in treatments) {
  df <- NULL
  for (i in drugs){
    # preR-preS adjusted growth control
    cor <- cor.test(
      data_merge[[i]], data_merge[[paste0(t, "_log2FC_preR_preS_adj")]],
      method = "pearson",
      use = "complete"
    )
    tmp <- data.frame(
      drug = i,
      pearson_r_log2FC_preR_preS_adj = cor$estimate,
      pearson_p_log2FC_preR_preS_adj = cor$p.value,
      pearson_n_log2FC_preR_preS_adj = cor$parameter + 2
    )
    # preR-preS
    cor <- cor.test(
      data_merge[[i]], data_merge[[paste0(t, "_log2FC_preR_preS")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS = cor$estimate
    tmp$pearson_p_log2FC_preR_preS = cor$p.value
    tmp$pearson_n_log2FC_preR_preS = cor$parameter + 2
    
    # filter by bootstrap p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_boot_p")]] < 0.05, ]
    # preR-preS adjusted growth control
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS_adj_boot_p = cor$estimate
    tmp$pearson_p_log2FC_preR_preS_adj_boot_p = cor$p.value
    tmp$pearson_n_log2FC_preR_preS_adj_boot_p = cor$parameter + 2
    # preR-preS
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS_boot_p = cor$estimate
    tmp$pearson_p_log2FC_preR_preS_boot_p = cor$p.value
    tmp$pearson_n_log2FC_preR_preS_boot_p = cor$parameter + 2
    
    # filter by permutation p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_perm_p")]] < 0.05, ]
    # preR-preS adjusted growth control
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS_adj_perm_p = cor$estimate
    tmp$pearson_p_log2FC_preR_preS_adj_perm_p = cor$p.value
    tmp$pearson_n_log2FC_preR_preS_adj_perm_p = cor$parameter + 2
    # preR-preS
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS_perm_p = cor$estimate
    tmp$pearson_p_log2FC_preR_preS_perm_p = cor$p.value
    tmp$pearson_n_log2FC_preR_preS_perm_p = cor$parameter + 2
    
    # filter by preR-preS adjusted p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_preR_preS_p_adj")]] < 0.05, ]
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS_p_adj = cor$estimate
    tmp$pearson_p_log2FC_preR_preS_p_adj = cor$p.value
    tmp$pearson_n_log2FC_preR_preS_p_adj = cor$parameter + 2
    
    # filter by preR-preS p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_preR_preS_p")]] < 0.05, ]
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS_p = cor$estimate
    tmp$pearson_p_log2FC_preR_preS_p = cor$p.value
    tmp$pearson_n_log2FC_preR_preS_p = cor$parameter + 2
    
    # posT-preT
    cor <- cor.test(
      data_merge[[i]], data_merge[[paste0(t, "_log2FC_posT_preT")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_posT_preT = cor$estimate
    tmp$pearson_p_log2FC_posT_preT = cor$p.value
    tmp$pearson_n_log2FC_posT_preT = cor$parameter + 2
    
    # posT-preT adjusted growth control
    cor <- cor.test(
      data_merge[[i]], data_merge[[paste0(t, "_log2FC_posT_preT_adj")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_posT_preT_adj = cor$estimate
    tmp$pearson_p_log2FC_posT_preT_adj = cor$p.value
    tmp$pearson_n_log2FC_posT_preT_adj = cor$parameter + 2
    
    # filter by posT-preT adjusted p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_posT_preT_p_adj")]] < 0.05, ]
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_posT_preT")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_posT_preT_p_adj = cor$estimate
    tmp$pearson_p_log2FC_posT_preT_p_adj = cor$p.value
    tmp$pearson_n_log2FC_posT_preT_p_adj = cor$parameter + 2
    
    # filter by posT-preT p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_posT_preT_p")]] < 0.05, ]
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_posT_preT")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_posT_preT_p = cor$estimate
    tmp$pearson_p_log2FC_posT_preT_p = cor$p.value
    tmp$pearson_n_log2FC_posT_preT_p = cor$parameter + 2
    
    df <- rbind.data.frame(df, tmp)
  }
  df <- df %>%
    left_join(
      drug_anno_match,
      by = c("drug" = "match_id")
    )
  write.csv(
    df,
    paste0(
      sub_output_dir,
      t,
      "_correlation_preR_vs_drug_signature.csv"
      ),
    row.names = FALSE)
}
```

## Pearson correlation coefficient on sister conserved genes

```{r}
treatments <- c("Carbo", "NK", "Olaparib")
data_merge <- data %>% 
  left_join(
    log2FC_table,
    by = "symbol"
  ) %>% 
  filter(conserve_gene == TRUE)
drugs <- colnames(data)[-1]
for (t in treatments) {
  df <- NULL
  for (i in drugs){
    # preR-preS adjusted growth control
    cor <- cor.test(
      data_merge[[i]], data_merge[[paste0(t, "_log2FC_preR_preS_adj")]],
      method = "pearson",
      use = "complete"
    )
    tmp <- data.frame(
      drug = i,
      pearson_r_log2FC_preR_preS_adj = cor$estimate,
      pearson_p_log2FC_preR_preS_adj = cor$p.value,
      pearson_n_log2FC_preR_preS_adj = cor$parameter + 2
    )
    # preR-preS
    cor <- cor.test(
      data_merge[[i]], data_merge[[paste0(t, "_log2FC_preR_preS")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS = cor$estimate
    tmp$pearson_p_log2FC_preR_preS = cor$p.value
    tmp$pearson_n_log2FC_preR_preS = cor$parameter + 2
    
    # filter by bootstrap p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_boot_p")]] < 0.05, ]
    # preR-preS adjusted growth control
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS_adj_boot_p = cor$estimate
    tmp$pearson_p_log2FC_preR_preS_adj_boot_p = cor$p.value
    tmp$pearson_n_log2FC_preR_preS_adj_boot_p = cor$parameter + 2
    
    # filter by permutation p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_perm_p")]] < 0.05, ]
    # preR-preS adjusted growth control
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS_adj_perm_p = cor$estimate
    tmp$pearson_p_log2FC_preR_preS_adj_perm_p = cor$p.value
    tmp$pearson_n_log2FC_preR_preS_adj_perm_p = cor$parameter + 2

    # filter by permutation p < 0.05 and bootstrap p < 0.05
    data_tmp <- data_merge[
      data_merge[[paste0(t, "_perm_p")]] < 0.05 &
        data_merge[[paste0(t, "_boot_p")]] < 0.05,
      ]
    # preR-preS adjusted growth control
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS_adj_perm_boot_p = cor$estimate
    tmp$pearson_p_log2FC_preR_preS_adj_perm_boot_p = cor$p.value
    tmp$pearson_n_log2FC_preR_preS_adj_perm_boot_p = cor$parameter + 2
    
    # filter by preR-preS adjusted p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_preR_preS_p_adj")]] < 0.05, ]
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS_p_adj = cor$estimate
    tmp$pearson_p_log2FC_preR_preS_p_adj = cor$p.value
    tmp$pearson_n_log2FC_preR_preS_p_adj = cor$parameter + 2
    
    # filter by preR-preS p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_preR_preS_p")]] < 0.05, ]
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_preR_preS_p = cor$estimate
    tmp$pearson_p_log2FC_preR_preS_p = cor$p.value
    tmp$pearson_n_log2FC_preR_preS_p = cor$parameter + 2
    
    # posT-preT
    cor <- cor.test(
      data_merge[[i]], data_merge[[paste0(t, "_log2FC_posT_preT")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_posT_preT = cor$estimate
    tmp$pearson_p_log2FC_posT_preT = cor$p.value
    tmp$pearson_n_log2FC_posT_preT = cor$parameter + 2
    
    # posT-preT adjusted growth control
    cor <- cor.test(
      data_merge[[i]], data_merge[[paste0(t, "_log2FC_posT_preT_adj")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_posT_preT_adj = cor$estimate
    tmp$pearson_p_log2FC_posT_preT_adj = cor$p.value
    tmp$pearson_n_log2FC_posT_preT_adj = cor$parameter + 2
    
    # filter by posT-preT adjusted p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_posT_preT_p_adj")]] < 0.05, ]
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_posT_preT")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_posT_preT_p_adj = cor$estimate
    tmp$pearson_p_log2FC_posT_preT_p_adj = cor$p.value
    tmp$pearson_n_log2FC_posT_preT_p_adj = cor$parameter + 2
    
    # filter by posT-preT p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_posT_preT_p")]] < 0.05, ]
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_posT_preT")]],
      method = "pearson",
      use = "complete"
    )
    tmp$pearson_r_log2FC_posT_preT_p = cor$estimate
    tmp$pearson_p_log2FC_posT_preT_p = cor$p.value
    tmp$pearson_n_log2FC_posT_preT_p = cor$parameter + 2
    
    df <- rbind.data.frame(df, tmp)
  }
  df <- df %>%
    left_join(
      drug_anno_match,
      by = c("drug" = "match_id")
    )
  write.csv(
    df,
    paste0(
      sub_output_dir,
      t,
      "_correlation_preR_vs_drug_signature_sister_conserved.csv"
      ),
    row.names = FALSE)
}
```

## Connectivity score (GSEA) on sister conserved genes

```{r include=FALSE}
treatments <- c("Carbo", "NK", "Olaparib")

data_merge <- data %>% 
  left_join(
    log2FC_table,
    by = "symbol"
  ) %>% 
  filter(conserve_gene == TRUE)
drugs <- colnames(data)[-1]

for (t in treatments) {
  df <- NULL
  for (i in drugs){
    
    drug_signature <- data_merge[[i]]
    names(drug_signature) <- data_merge$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    preR_signature <- data_merge[[paste0(t, "_log2FC_preR_preS_adj")]]
    names(preR_signature) <- data_merge$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    tmp <- data.frame(
      drug = i,
      connect_score_log2FC_preR_preS_adj = connect[1],
      connect_p_log2FC_preR_preS_adj = connect[2],
      pearson_n_log2FC_preR_preS_adj = length(preR_signature)
    )
    
    # preR-preS
    
    preR_signature <- data_merge[[paste0(t, "_log2FC_preR_preS")]]
    names(preR_signature) <- data_merge$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
   
    tmp$connectivity_score_log2FC_preR_preS = connect[1]
    tmp$connectivity_p_log2FC_preR_preS = connect[2]
    tmp$connectivity_n_log2FC_preR_preS = length(preR_signature)
    
    # filter by bootstrap p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_boot_p")]] < 0.05, ]
    
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    cor <- cor.test(
      data_tmp[[i]], data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]],
      method = "pearson",
      use = "complete"
    )
    tmp$connectivity_score_log2FC_preR_preS_adj_boot_p = connect[1]
    tmp$connectivity_p_log2FC_preR_preS_adj_boot_p = connect[2]
    tmp$connectivity_n_log2FC_preR_preS_adj_boot_p = length(preR_signature)
    
    # filter by permutation p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_perm_p")]] < 0.05, ]
    
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    tmp$connectivity_score_log2FC_preR_preS_adj_perm_p = connect[1]
    tmp$connectivity_p_log2FC_preR_preS_adj_perm_p = connect[2]
    tmp$connectivity_n_log2FC_preR_preS_adj_perm_p = length(preR_signature)

    # filter by permutation p < 0.05 and bootstrap p < 0.05
    data_tmp <- data_merge[
      data_merge[[paste0(t, "_perm_p")]] < 0.05 &
        data_merge[[paste0(t, "_boot_p", t, "_ctrl_p_value")]] < 0.05,
      ]
    
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_preR_preS_adj")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    tmp$connectivity_score_log2FC_preR_preS_adj_perm_boot_p = connect[1]
    tmp$connectivity_p_log2FC_preR_preS_adj_perm_boot_p = connect[2]
    tmp$connectivity_n_log2FC_preR_preS_adj_perm_boot_p = length(preR_signature)
    
    # filter by preR-preS adjusted p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_preR_preS_p_adj")]] < 0.05, ]
    
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_preR_preS")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    tmp$connectivity_score_log2FC_preR_preS_p_adj = connect[1]
    tmp$connectivity_p_log2FC_preR_preS_p_adj = connect[2]
    tmp$connectivity_n_log2FC_preR_preS_p_adj = length(preR_signature)
    
    # filter by preR-preS p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_preR_preS_p")]] < 0.05, ]
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_preR_preS")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    tmp$connectivity_score_log2FC_preR_preS_p = connect[1]
    tmp$connectivity_p_log2FC_preR_preS_p = connect[2]
    tmp$connectivity_n_log2FC_preR_preS_p = length(preR_signature)
    
    # posT-preT
    drug_signature <- data_merge[[i]]
    names(drug_signature) <- data_merge$symbol
    drug_signature <- na.omit(drug_signature)
    
    preR_signature <- data_merge[[paste0(t, "_log2FC_posT_preT")]]
    names(preR_signature) <- data_merge$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )

    tmp$connectivity_score_log2FC_posT_preT = connect[1]
    tmp$connectivity_p_log2FC_posT_preT = connect[2]
    tmp$connectivity_n_log2FC_posT_preT = length(preR_signature)
    
    # posT-preT adjusted growth control
    drug_signature <- data_merge[[i]]
    names(drug_signature) <- data_merge$symbol
    drug_signature <- na.omit(drug_signature)
    
    preR_signature <- data_merge[[paste0(t, "_log2FC_posT_preT")]]
    names(preR_signature) <- data_merge$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )

    tmp$connectivity_score_log2FC_posT_preT_adj = connect[1]
    tmp$connectivity_p_log2FC_posT_preT_adj = connect[2]
    tmp$connectivity_n_log2FC_posT_preT_adj = length(preR_signature)
    
    # filter by posT-preT adjusted p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_posT_preT_p_adj")]] < 0.05, ]
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_posT_preT")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    
    tmp$connectivity_score_log2FC_posT_preT_p_adj = connect[1]
    tmp$connectivity_p_log2FC_posT_preT_p_adj = connect[2]
    tmp$connectivity_n_log2FC_posT_preT_p_adj = length(preR_signature)
    
    # filter by posT-preT p < 0.05
    data_tmp <- data_merge[data_merge[[paste0(t, "_posT_preT_p")]] < 0.05, ]
    drug_signature <- data_tmp[[i]]
    names(drug_signature) <- data_tmp$symbol
    drug_signature <- na.omit(drug_signature)
    
    # preR-preS adjusted growth control
    
    preR_signature <- data_tmp[[paste0(t, "_log2FC_posT_preT")]]
    names(preR_signature) <- data_tmp$symbol
    preR_signature <- na.omit(preR_signature)
    
    connect <- connectivityScore(
      x = drug_signature,
      y = preR_signature,
      method = "fgsea",
      nperm = 1000
    )
    tmp$connectivity_score_log2FC_posT_preT_p = connect[1]
    tmp$connectivity_p_log2FC_posT_preT_p = connect[2]
    tmp$connectivity_n_log2FC_posT_preT_p = length(preR_signature)
    
    df <- rbind.data.frame(df, tmp)
  }
  df <- df %>%
    left_join(
      drug_anno_match,
      by = c("drug" = "match_id")
    )
  write.csv(
    df,
    paste0(
      sub_output_dir,
      t,
      "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      ),
    row.names = FALSE)
}
```

## Connectivity score (GSEA) for drug-induced signature and BRCAness signature

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "test/drug_BRCAness/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}

lseq <- function(from=1, to=100000, length.out=6) {
  # logarithmic spaced sequence
  # blatantly stolen from library("emdbook"), because need only this
  exp(seq(log(from), log(to), length.out = length.out))
}

```

```{r}
log2FC_table <- read.csv("./data/input/consensus_signature_drugs.csv", row.names = 1) %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("symbol")
```

TCGA BRCAness signature

```{r}
tcga <- read.csv("./data/input/BRCAness/TCGA_deconvoluted_HRD_DEGs.tsv", sep = "\t") %>% 
  rownames_to_column("symbol") %>% 
    select(symbol, BRCAness_tcga_logFC = logFC, BRCAness_tcga_p_adj = adj.P.Val)
```

```{r}
log2FC_table <- left_join(
  log2FC_table,
  tcga,
  by = c("symbol")
)
```

COV362 BRCAness signature

```{r}
cov362 <- read.csv("./data/input/BRCAness/deg_COV362.csv") %>% 
  group_by(symbol) %>% 
  summarise(
    BRCAness_cov_logFC = mean(avg_log2FC),
    BRCAness_cov_p_adj = mean(p_val_adj)
  )
```

```{r}
log2FC_table <- left_join(
  log2FC_table,
  cov362 ,
  by = c("symbol")
)
```

Sister concordant genes

```{r}
conserve <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  select(symbol, conserve_gene) %>% 
  unique()
log2FC_table <- left_join(
  log2FC_table,
  conserve,
  by = c("symbol")
)
```

```{r}
write.csv(log2FC_table, paste0(sub_output_dir, "logFC_table_for_drug_induced_BRCAness.csv"), row.names = FALSE)
```

```{r include=FALSE}
data_merge <- log2FC_table %>% 
  filter(conserve_gene == TRUE)
# drugs <- colnames(data_merge)[startsWith(colnames(data_merge), "BRD-")]
drugs <- read_xlsx("./data/output/test/drug_target_expression.xlsx") %>% 
  select(drug) %>% 
  unlist() %>% 
  unique()
df <- NULL
for (i in drugs){
  drug_signature <- data_merge[[i]]
  names(drug_signature) <- data_merge$symbol
  drug_signature <- na.omit(drug_signature)
  
  # BRCAness COV362
  BRCA_cov_signature <- data_merge[["BRCAness_cov_logFC"]]
  names(BRCA_cov_signature) <- data_merge$symbol
  BRCA_cov_signature <- na.omit(BRCA_cov_signature)
  
  connect <- connectivityScore(
    x = drug_signature,
    y = BRCA_cov_signature,
    method = "fgsea",
    nperm = 1000
  )
  
  tmp <- data.frame(
    drug = i,
    connectivity_score_BRCA_cov = connect[1],
    connectivity_p_BRCA_cov = connect[2],
    connectivity_n_BRCA_cov = length(BRCA_cov_signature)
  )
  
  # BRCAness TCGA
  BRCA_tcga_signature <- data_merge[["BRCAness_tcga_logFC"]]
  names(BRCA_tcga_signature) <- data_merge$symbol
  BRCA_tcga_signature <- na.omit(BRCA_tcga_signature)
  
  connect <- connectivityScore(
    x = drug_signature,
    y = BRCA_tcga_signature,
    method = "fgsea",
    nperm = 1000
  )
  
  tmp$connectivity_score_BRCA_tcga <- connect[1]
  tmp$connectivity_p_BRCA_tcga <- connect[2]
  tmp$connectivity_n_BRCA_tcga <- length(BRCA_tcga_signature)
  
  df <- rbind.data.frame(df, tmp)
}
```

Annotate drug

```{r include=FALSE}
drug_anno <- read.csv("./data/input/drug_repurposing_hub_alldrugsanno.csv")
drug_anno$pert_iname[is.na(drug_anno$pert_iname)] <- drug_anno$broad_id[is.na(drug_anno$pert_iname)] 
df <- df %>%
  left_join(
    drug_anno,
    by = c("drug" = "broad_id")
  ) %>% 
  unique()
write.csv(
  df,
  paste0(
    sub_output_dir,
    "connectivity_score_BRCAness_vs_drug_signature_sister_conserved.csv"
    ),
  row.names = FALSE)
```
## Visualization

### Correlation coefficient

```{r}
for (t in treatments){
  # preR_preS
  plot_table <- read.csv(
    paste0(
      sub_output_dir,
      t,
      "_correlation_preR_vs_drug_signature.csv"
    )
  ) %>% 
    filter_at(vars(starts_with("pearson_p")), any_vars(. < 0.05)) %>% 
    select(drug, starts_with("pearson_r")) %>% 
    select(drug, contains("preR_preS"))
  colnames(plot_table) <- gsub("pearson_r_log2FC_", "", colnames(plot_table))
  pdf(file = paste0(sub_output_dir, t, "_preR_preS_heatmap.pdf"))
  heatmap(
    as.matrix(plot_table[, -1]), 
    labRow = plot_table$drug,
    cexCol = 0.8,
    margins = c(10, 5)
  )
  dev.off()
  
  # preR_preS sister concordant gene
  plot_table <- read.csv(
    paste0(
      sub_output_dir,
      t,
      "_correlation_preR_vs_drug_signature_sister_conserved.csv"
    )
  ) %>% 
    filter_at(vars(starts_with("pearson_p")), any_vars(. < 0.05)) %>% 
    select(drug, starts_with("pearson_r")) %>% 
    select(drug, contains("preR_preS"))
  colnames(plot_table) <- gsub("pearson_r_log2FC_", "", colnames(plot_table))
  pdf(file = paste0(sub_output_dir, t, "_preR_preS_sister_conserved_heatmap.pdf"))
  heatmap(
    as.matrix(plot_table[, -1]),
    labRow = plot_table$drug,
    labCol = gsub("pearson_r_log2FC_", "", colnames(plot_table)[-1]),
    cexCol = 0.8,
    margins = c(10, 5)
  )
  dev.off()
  
  # posT_preT
  plot_table <- read.csv(
    paste0(
      sub_output_dir,
      t,
      "_correlation_preR_vs_drug_signature.csv"
    )
  ) %>% 
    filter_at(vars(starts_with("pearson_p")), any_vars(. < 0.05)) %>% 
    select(drug, starts_with("pearson_r")) %>% 
    select(drug, contains("posT_preT"))
  colnames(plot_table) <- gsub("pearson_r_log2FC_", "", colnames(plot_table))
  pdf(file = paste0(sub_output_dir, t, "_posT_preT_connectivity_p_heatmap.pdf"))
  heatmap(
    as.matrix(plot_table[, -1]),
    labRow = plot_table$drug,
    labCol = gsub("pearson_r_log2FC_", "", colnames(plot_table)[-1]),
    cexCol = 0.8,
    margins = c(10, 5)
  )
  dev.off()
  
  # posT_preT sister concordant gene
  plot_table <- read.csv(
    paste0(
      sub_output_dir,
      t,
      "_correlation_preR_vs_drug_signature_sister_conserved.csv"
    )
  ) %>% 
    filter_at(vars(starts_with("pearson_p")), any_vars(. < 0.05)) %>% 
    select(drug, starts_with("pearson_r")) %>% 
    select(drug, contains("posT_preT"))
  colnames(plot_table) <- gsub("pearson_r_log2FC_", "", colnames(plot_table))
  pdf(file = paste0(sub_output_dir, t, "_posT_preT_sister_conserved_heatmap.pdf"))
  heatmap(
    as.matrix(plot_table[, -1]),
    labRow = plot_table$drug,
    labCol = gsub("pearson_r_log2FC_", "", colnames(plot_table)[-1]),
    cexCol = 0.8,
    margins = c(10, 5)
  )
  dev.off()
  
  # Venn
  ## preR_preS
  plot_table <- read.csv(
    paste0(
      sub_output_dir,
      t,
      "_correlation_preR_vs_drug_signature.csv"
    )
  ) %>% 
    select(drug, contains("preR_preS")) %>% 
    filter_at(vars(starts_with("pearson_p")), all_vars(. < 0.05)) %>% 
    filter_at(vars(starts_with("pearson_r")), all_vars(. < 0))
}
```

### Connectivity score

#### heatmap

```{r}
for (t in treatments){
  plot_table <- read.csv(
    paste0(
      sub_output_dir,
      t,
      "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
    )
  ) %>% 
    filter_at(vars(starts_with("connectivity_p")), any_vars(. < 0.05))
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]
  plot_table <- plot_table %>% 
    select(drug = pert_iname, starts_with("connectivity_score")) %>% 
    select(drug, contains("preR_preS"))
  colnames(plot_table) <- gsub("connectivity_score_log2FC_", "", colnames(plot_table))
  pdf(file = paste0(sub_output_dir, t, "_preR_preS_connectivity_score_heatmap.pdf"))
  heatmap(
    as.matrix(plot_table[, -1]), 
    labRow = plot_table$drug,
    cexCol = 0.8,
    margins = c(10, 5)
  )
  dev.off()
}
```

#### Volcano plot

```{r}
volcano_sub_dir <- paste0(sub_output_dir, "volcano_plot/")
if(!dir.exists(volcano_sub_dir)){
  dir.create(volcano_sub_dir)
}

for (t in treatments){
  highlight_drug <- read_xlsx(
    paste0(
      sub_output_dir,
      t,
      "_connectivity_score_preR_vs_drug_signature_sister_conserved.xlsx"
    ),
    sheet = 2
  ) %>% 
  select(drug) %>% 
  unlist()
  # preR_preS_adj
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS_adj`,
        p = `connectivity_p_log2FC_preR_preS_adj`,
        n = `connectivity_n_log2FC_preR_preS_adj`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]
  
  highlight_black <- filter(plot_table, drug %in% highlight_drug_black)
  highlight_black$color <- "black"
  highlight_grey <- filter(plot_table, drug %in% highlight_drug_grey)
  highlight_grey$color <- "grey40"
  highlight <- rbind.data.frame(highlight_black, highlight_grey)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          #title="PreR-preS log2FC adjusted control \n No filter",
          # subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_sister_conserved_volcano.pdf"),
    p,
    width = 5,
    height = 5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_sister_conserved_volcano.png"),
    p,
    width = 5,
    height = 5
  )
  
  # preR_preS
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS`,
        p = `connectivity_p_log2FC_preR_preS`,
        n = `connectivity_n_log2FC_preR_preS`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]

  highlight_black <- filter(plot_table, drug %in% highlight_drug_black)
  highlight_black$color <- "black"
  highlight_grey <- filter(plot_table, drug %in% highlight_drug_grey)
  highlight_grey$color <- "grey40"
  highlight <- rbind.data.frame(highlight_black, highlight_grey)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          #title="PreR-preS log2FC \n No filter",
          #subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_sister_conserved_volcano.pdf"),
    p,
    width = 5,
    height = 5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_sister_conserved_volcano.png"),
    p,
    width = 5,
    height = 5
  )
  
  # preR_preS_adj_boot_p
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS_adj_boot_p`,
        p = `connectivity_p_log2FC_preR_preS_adj_boot_p`,
        n = `connectivity_n_log2FC_preR_preS_adj_boot_p`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  # plot_table$Term <- paste0("bold('", plot_table$Term, "')")
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]

  highlight_black <- filter(plot_table, drug %in% highlight_drug_black)
  highlight_black$color <- "black"
  highlight_grey <- filter(plot_table, drug %in% highlight_drug_grey)
  highlight_grey$color <- "grey40"
  highlight <- rbind.data.frame(highlight_black, highlight_grey)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          # title="PreR-preS log2FC adjusted control \n Bootstrap P < 0.05",
          # subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_boot_p_sister_conserved_volcano.pdf"),
    p,
    width = 5,
    height = 5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_boot_p_sister_conserved_volcano.png"),
    p,
    width = 5,
    height = 5
  )
  
    # preR_preS_adj_perm_p
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS_adj_perm_p`,
        p = `connectivity_p_log2FC_preR_preS_adj_perm_p`,
        n = `connectivity_n_log2FC_preR_preS_adj_perm_p`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  # plot_table$Term <- paste0("bold('", plot_table$Term, "')")
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]
  
  highlight_black <- filter(plot_table, drug %in% highlight_drug_black)
  highlight_black$color <- "black"
  highlight_grey <- filter(plot_table, drug %in% highlight_drug_grey)
  highlight_grey$color <- "grey40"
  highlight <- rbind.data.frame(highlight_black, highlight_grey)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          # title="PreR-preS log2FC adjusted control \n Permutation P < 0.05",
          # subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_perm_p_sister_conserved_volcano.pdf"),
    p,
    width = 5,
    height = 5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_perm_p_sister_conserved_volcano.png"),
    p,
    width = 5,
    height = 5
  )
  
    # preR_preS_adj_perm_boot_p
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS_adj_perm_boot_p`,
        p = `connectivity_p_log2FC_preR_preS_adj_perm_boot_p`,
        n = `connectivity_n_log2FC_preR_preS_adj_perm_boot_p`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  # plot_table$Term <- paste0("bold('", plot_table$Term, "')")
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]
  
  highlight_black <- filter(plot_table, drug %in% highlight_drug_black)
  highlight_black$color <- "black"
  highlight_grey <- filter(plot_table, drug %in% highlight_drug_grey)
  highlight_grey$color <- "grey40"
  highlight <- rbind.data.frame(highlight_black, highlight_grey)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          # title="PreR-preS log2FC adjust control \n Bootstrap, permutation P < 0.05",
          # subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_perm_boot_p_sister_conserved_volcano.pdf"),
    p,
    width = 5,
    height = 5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_adj_perm_boot_p_sister_conserved_volcano.png"),
    p,
    width = 5,
    height = 5
  )
  
    # preR_preS_p_adj
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS_p_adj`,
        p = `connectivity_p_log2FC_preR_preS_p_adj`,
        n = `connectivity_n_log2FC_preR_preS_p_adj`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  # plot_table$Term <- paste0("bold('", plot_table$Term, "')")
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]
  highlight <- read_xlsx(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.xlsx"
      ),
      sheet = 2
    ) %>% 
    select(drug) %>% 
    unlist()
  highlight <- filter(plot_table, drug %in% highlight)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          # title="PreR-preS log2FC \n Adjust P < 0.05",
          # subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_p_adj_sister_conserved_volcano.pdf"),
    p,
    width = 5,
    height = 5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_p_adj_sister_conserved_volcano.png"),
    p,
    width = 5,
    height = 5
  )
  
    # preR_preS_p
  plot_table <- read.csv(
      paste0(
        sub_output_dir,
        t,
        "_connectivity_score_preR_vs_drug_signature_sister_conserved.csv"
      )
    ) %>% 
      select(
        pert_iname,
        drug,
        x = `connectivity_score_log2FC_preR_preS_p`,
        p = `connectivity_p_log2FC_preR_preS_p`,
        n = `connectivity_n_log2FC_preR_preS_p`) %>% 
      mutate(
        y = -log10(p),
        sig = p < 0.05
      )
  plot_table$color <- "grey"
  plot_table$color[plot_table$x < 0 & plot_table$p < 0.05] <- "blue"
  plot_table$color[plot_table$x > 0 & plot_table$p < 0.05] <- "red"
  plot_table$pert_iname[is.na(plot_table$pert_iname)] <- plot_table$drug[is.na(plot_table$pert_iname)]

  highlight_black <- filter(plot_table, drug %in% highlight_drug_black)
  highlight_black$color <- "black"
  highlight_grey <- filter(plot_table, drug %in% highlight_drug_grey)
  highlight_grey$color <- "grey40"
  highlight <- rbind.data.frame(highlight_black, highlight_grey)
  
  p <- ggplot() +
     theme_classic() +
     theme(
         legend.key = element_blank(),
         legend.position = "none",
         axis.text = element_text(size = 21, colour = "black"),
         axis.title = element_text(size = 21),
         plot.title = element_text(size = 21, hjust = 0.5),
         plot.subtitle = element_text(size = 15, hjust = 0.5)
         ) +
     labs(x = "Connectivity Score",
          y = expression("-log"[10]~"(P value)"),
          # title="PreR-preS log2FC \n P < 0.05",
          # subtitle = paste0(unique(plot_table$n), " genes in PreR signature")
          title = ifelse(t == "Carbo", "Carboplatin", t)) +
     geom_point(
         aes(x = x, y = y, colour=color),
         alpha = 0.5,
         size = 5,
         data = plot_table,
         show.legend = FALSE) +
      scale_color_manual(
        values = c("black", low_color,"grey70", "grey40", high_color),
      ) +
      geom_point(
        aes(x = x, y = y, color = color),
        shape = 1,
        size = 5,
        data = highlight,
        show.legend = FALSE) +
      geom_text_repel(
        aes(x = x, y = y, label = pert_iname, color = color),
        size = 18/.pt,
        data = highlight,
        # parse = TRUE,
        box.padding = 0.5
      )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_p_sister_conserved_volcano.pdf"),
    p,
    width = 5,
    height = 5
  )
  ggsave(
    paste0(volcano_sub_dir, t, "_connectivity_preR_preS_p_sister_conserved_volcano.png"),
    p,
    width = 5,
    height = 5
  )
}
```

Validate with drug combination screening data from DrugComb

```{r}
DrugComb_drug <- read.csv("./data/input/DrugComb_all_drug.csv")
olaparib_combo <- read.csv("./data/input/DrugComb_summary_Olaparib.csv") %>% 
  filter(drug_col != "NULL")
olaparib_combo <- olaparib_combo %>% 
  mutate(drug = ifelse(drug_row == "Olaparib", drug_col, drug_row))
drugs_combo <- DrugComb_drug %>% 
  filter(dname %in% unique(olaparib_combo$drug)) %>% 
  filter(inchikey %in% drug_anno_match$InChIKey)
olaparib_combo <- olaparib_combo %>% 
  select(block_id, drug, cell_line_name, css, synergy_zip, synergy_bliss, synergy_loewe, synergy_hsa, S_sum, quality, S_mea, S_max) %>% 
  filter(drug %in% drugs_combo$dname) %>% 
  left_join(select(drugs_combo, dname, inchikey), by = c("drug" = "dname"))

olaparib_correlation <- read.csv("./data/output/test/Olaparib_correlation_preR_vs_drug_signature_sister_conserved.csv") %>% 
  filter(InChIKey %in% olaparib_combo$inchikey) %>% 
  select(InChIKey, starts_with("pearson"))

olaparib_combo <- olaparib_combo %>% 
  left_join(olaparib_correlation, by = c("inchikey" = "InChIKey"))

plot_table <- olaparib_combo %>% 
  select(
    css, synergy_zip, synergy_bliss, synergy_loewe, synergy_hsa, S_sum,
    S_mea, S_max, starts_with("pearson_r"))
pdf(file = paste0(sub_output_dir, "Olaparib_DrugComb_pearson_correlation_sister_conserved_heatmap.pdf"))
heatmap(
  as.matrix(plot_table),
  labRow = paste0(olaparib_combo$drug, "_", olaparib_combo$cell_line_name),
  labCol = gsub("pearson_r_log2FC_", "", colnames(plot_table)),
  cexCol = 0.8,
  margins = c(10, 5)
)
dev.off()
```

```{r}
write.csv(
  cor(plot_table),
  paste0(sub_output_dir, "Olaparib_DrugComb_pearson_correlation_sister_conserved.csv")
)
```

```{r}
# summarize across cell lines
plot_table <- olaparib_combo %>% 
  select(
    drug, css, synergy_zip, synergy_bliss, synergy_loewe, synergy_hsa, S_sum,
    S_mea, S_max, starts_with("pearson_r")) %>% 
  group_by(drug) %>% 
  summarise_if(is.numeric, mean, na.rm = TRUE)
pdf(file = paste0(sub_output_dir, t, "_DrugComb_pearson_correlation_summarize_response_across_cells_sister_conserved_heatmap.pdf"))
heatmap(
  as.matrix(plot_table[, -1]),
  labRow = plot_table$drug,
  labCol = gsub("pearson_r_log2FC_", "", colnames(plot_table)[-1]),
  cexCol = 0.8,
  margins = c(10, 5)
  )
dev.off()
```

```{r}
DrugComb_drug <- read.csv("./data/input/DrugComb_all_drug.csv")
olaparib_combo <- read.csv("./data/input/DrugComb_summary_Olaparib.csv") %>% 
  filter(drug_col != "NULL")
olaparib_combo <- olaparib_combo %>% 
  mutate(drug = ifelse(drug_row == "Olaparib", drug_col, drug_row))
drugs_combo <- DrugComb_drug %>% 
  filter(dname %in% unique(olaparib_combo$drug)) %>% 
  filter(inchikey %in% drug_anno_match$InChIKey)
olaparib_combo <- olaparib_combo %>% 
  select(block_id, drug, cell_line_name, css, synergy_zip, synergy_bliss, synergy_loewe, synergy_hsa, S_sum, quality, S_mea, S_max) %>% 
  filter(drug %in% drugs_combo$dname) %>% 
  left_join(select(drugs_combo, dname, inchikey), by = c("drug" = "dname"))

olaparib_correlation <- read.csv("./data/output/test/Olaparib_correlation_preR_vs_drug_signature.csv") %>% 
  filter(InChIKey %in% olaparib_combo$inchikey) %>% 
  select(InChIKey, starts_with("pearson"))

olaparib_combo <- olaparib_combo %>% 
  left_join(olaparib_correlation, by = c("inchikey" = "InChIKey"))

plot_table <- olaparib_combo %>% 
  select(
    css, synergy_zip, synergy_bliss, synergy_loewe, synergy_hsa, S_sum,
    S_mea, S_max, starts_with("pearson_r"))
pdf(file = paste0(sub_output_dir, "Olaparib_DrugComb_pearson_correlation_heatmap.pdf"))
heatmap(
  as.matrix(plot_table),
  labRow = paste0(olaparib_combo$drug, "_", olaparib_combo$cell_line_name),
  labCol = gsub("pearson_r_log2FC_", "", colnames(plot_table)),
  cexCol = 0.8,
  margins = c(10, 5)
)
dev.off()
```

```{r}
write.csv(
  cor(plot_table),
  paste0(sub_output_dir, "Olaparib_DrugComb_pearson_correlation.csv")
)
```

```{r}
# summarize across cell lines
plot_table <- olaparib_combo %>% 
  select(
    drug, css, synergy_zip, synergy_bliss, synergy_loewe, synergy_hsa, S_sum,
    S_mea, S_max, starts_with("pearson_r")) %>% 
  group_by(drug) %>% 
  summarise_if(is.numeric, mean, na.rm = TRUE)
pdf(file = paste0(sub_output_dir, "Olaparib_DrugComb_pearson_correlation_summarize_response_across_cells_heatmap.pdf"))
heatmap(
  as.matrix(plot_table[, -1]),
  labRow = plot_table$drug,
  labCol = gsub("pearson_r_log2FC_", "", colnames(plot_table)[-1]),
  cexCol = 0.8,
  margins = c(10, 5)
  )
dev.off()
```

```{r}
write.csv(
  cor(plot_table[-1]),
  paste0(sub_output_dir, "Olaparib_DrugComb_pearson_correlation_summarize_response_across_cells.csv")
)
```

## Drug target expression in BT cells

```{r}
# scRNA expression
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
```

```{r}
# Drug target
treatments <- c("Carbo", "NK", "Olaparib")
drugs <- NULL
for (t in treatments){
  tmp <- read_xlsx(
    paste0(
      "./data/output/test/", t,
      "_connectivity_score_preR_vs_drug_signature_sister_conserved.xlsx"),
    sheet = 2) %>% 
    mutate(treatment = t) %>% 
    select(
      treatment, drug, pert_iname, clinical_phase, moa, target, broad_id, InChIKey,
      pubchem_cid, deprecated_broad_id
    )
  drugs <- rbind.data.frame(drugs, tmp)
  }
```

```{r}
target_expression <- NULL
for (i in 1:nrow(drugs)){
  treatment <- drugs$treatment[i]
  if (treatment == "Carbo"){
    treatment <- "Carboplatin"
  }
  drug_targets <- drugs$target[i]
  drug_targets <- unlist(strsplit(drug_targets, "\\|"))
  drug_targets <- drug_targets[drug_targets != "NA"]
  if (length(drug_targets) == 0){
    df <- data.frame(
      treatment = treatment,
      drug = drugs$drug[i],
      symbol = NA,
      ensembl = NA,
      mean_exp = NA,
      non_zero = NA,
      mean_exp_preR = NA,
      non_zero_preR  = NA,
      mean_exp_preS = NA,
      non_zero_preS = NA
    )
  } else {
    ensembls <- data@misc$gene_meta$ensembl[data@misc$gene_meta$symbol %in% drug_targets]
    ensembls <- intersect(ensembls, rownames(data))
    exp <- data@assays$RNA@data[ 
      rownames(data) %in% ensembls,
      which(data$time_point == "BeforeTreatment" & data$treatment == treatment)
    ]
    exp_preR <- data@assays$RNA@data[
      rownames(data) %in% ensembls,
      which(data$drugSens == "pre-resistant" & data$treatment == treatment)
    ]
    exp_preS <- data@assays$RNA@data[
      rownames(data) %in% ensembls,
      which(data$drugSens == "pre-sensitive" & data$treatment == treatment)
    ]
    if (length(ensembls) > 1){
      mean_exp <- rowMeans(exp)
      non_zero <- rowSums(exp > 0)/ncol(exp)
      mean_exp_preR <- rowMeans(exp_preR)
      non_zero_preR <- rowSums(exp_preR > 0)/ncol(exp_preR)
      mean_exp_preS <- rowMeans(exp_preS)
      non_zero_preS <- rowSums(exp_preS > 0)/ncol(exp_preS)
      df <- data.frame(
        ensembl = names(mean_exp),
        mean_exp = mean_exp,
        non_zero = non_zero[names(mean_exp)],
        mean_exp_preR = mean_exp_preR[names(mean_exp)],
        non_zero_preR  = non_zero_preR [names(mean_exp)],
        mean_exp_preS = mean_exp_preS[names(mean_exp)],
        non_zero_preS = non_zero_preS[names(mean_exp)]
      )
      df <- df %>% 
        left_join(data@misc$gene_meta, by = "ensembl") %>% 
        mutate(drug = drugs$drug[i], treatment = treatment)
    } else if (length(ensembls) == 1) {
      mean_exp <- mean(exp)
      non_zero <- sum(exp > 0)/length(exp)
      mean_exp_preR <- mean(exp_preR)
      non_zero_preR <- sum(exp_preR > 0)/length(exp_preR)
      mean_exp_preS <- mean(exp_preS)
      non_zero_preS <- sum(exp_preS > 0)/length(exp_preS)
      df <- data.frame(
        ensembl = ensembls,
        mean_exp = mean_exp,
        non_zero = non_zero,
        mean_exp_preR = mean_exp_preR,
        non_zero_preR  = non_zero_preR,
        mean_exp_preS = mean_exp_preS,
        non_zero_preS = non_zero_preS,
        symbol = data@misc$gene_meta$symbol[data@misc$gene_meta$ensembl == ensembls],
        drug = drugs$drug[i],
        treatment = treatment
      )
    } else {
      df <- data.frame(
        ensembl = NA,
        mean_exp = NA,
        non_zero = NA,
        mean_exp_preR = NA,
        non_zero_preR  = NA,
        mean_exp_preS = NA,
        non_zero_preS = NA,
        symbol = drug_targets,
        drug = drugs$drug[i],
        treatment = treatment
      )
    }
    left_target <- setdiff(drug_targets, df$symbol)
    if (length(left_target) > 0){
      tmp <- data.frame(
        ensembl = NA,
        mean_exp = NA,
        non_zero = NA,
        mean_exp_preR = NA,
        non_zero_preR  = NA,
        mean_exp_preS = NA,
        non_zero_preS = NA,
        symbol = setdiff(drug_targets, df$symbol),
        drug = drugs$drug[i],
        treatment = treatment
      )
      df <- rbind.data.frame(df, tmp)
    }
  }
  target_expression <- rbind.data.frame(target_expression, df)
}

drug_target_expression <- drugs %>% 
  select(drug, pert_iname) %>% 
  full_join(target_expression,by = c("drug")) %>% 
  select(
    treatment, drug, pert_iname, target_symbol = symbol, target_ensembl = ensembl,
    everything()) %>% 
  arrange(treatment, drug) %>% 
  unique()
openxlsx::write.xlsx(drug_target_expression, file = paste0(sub_output_dir, "drug_target_expression.xlsx"))
```

## L100 experiment condition for selected drugs

```{r}
candidates <- readxl::read_xlsx("./data/output/test/NK_correlation_preR_vs_drug_signature_sister_conserved.xlsx", sheet = 2) %>% 
  rbind.data.frame(readxl::read_xlsx("./data/output/test/Olaparib_correlation_preR_vs_drug_signature_sister_conserved.xlsx", sheet = 2))

L1000_cell <- data.table::fread("./data/input/L1000_treatment_information/phase_one/GSE92742_Broad_LINCS_cell_info.txt", header = TRUE) %>% 
  rbind.data.frame(
    data.table::fread("./data/input/L1000_treatment_information/phase_two/GSE92742_Broad_LINCS_cell_info.txt", header = TRUE))
    
L1000 <- data.table::fread("./data/input/L1000_treatment_information/phase_one/GSE92742_Broad_LINCS_inst_info.txt", header = TRUE) %>% 
  select(pert_id, pert_iname, pert_dose, pert_dose_unit, pert_time, pert_time_unit, cell_id) %>% 
  rbind.data.frame(
    data.table::fread("./data/input/L1000_treatment_information/phase_two/GSE70138_Broad_LINCS_inst_info.txt", header = TRUE) %>% 
      select(pert_id, pert_iname, pert_dose, pert_dose_unit, pert_time, pert_time_unit, cell_id)) %>%
  filter(
    pert_id %in% candidates$drug
  ) %>% left_join(
    L1000_cell %>% 
      select(cell_id, primary_site, modification, sample_type, primary_site, original_growth_pattern),
    by = "cell_id") %>% 
  unique()

L1000[L1000 == -666] <- NA
writexl::write_xlsx(L1000, paste0(sub_output_dir, "L1000_treatment_condition.xlsx"))
```

## Survival plot for PreR-preS signature

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "test/survival_plot_on_preR_preS/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}

deg <- read.csv("./data/output/Figure4/Figure4B_all_signatures_logFC_p.csv")
data_clinical_patient = read.delim("./data/input/BRCAness/data_clinical_patient.txt", skip = 4)
data_clinical_sample = read.delim("./data/input/BRCAness/data_clinical_sample.txt", skip = 4)
sister_conserve_gene <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  filter(conserve_gene == TRUE) %>% 
  select(ensembl)
sister_conserve_gene <- sister_conserve_gene$ensembl
```


```{r}
deconvolute <- TRUE

if (deconvolute){
  # deconvoluted data
  data_mrna = readRDS("./data/input/BRCAness/EOC_deconv_TCGA_expr_mat.rds")
  colnames(data_mrna) = gsub("\\.","\\-", colnames(data_mrna))
  colnames(data_mrna) = substr(colnames(data_mrna), 1, 15)
} else {
  # Non-deconvoluted data
  data_mrna <- read.delim("./data/input/BRCAness/data_mrna_seq_v2_rsem.txt", check.names = F) %>% 
    select(-Entrez_Gene_Id) %>% 
    filter(Hugo_Symbol != "") %>% 
    dplyr::rename(symbol = Hugo_Symbol) %>% 
    group_by(symbol) %>% 
    summarise_all("mean") %>% 
    column_to_rownames("symbol")
}

data_mrna <- data_mrna[apply(data_mrna, 1, function(x){
  sum(is.na(x))
}) < 300, ]

identical(data_clinical_patient$PATIENT_ID, data_clinical_sample$PATIENT_ID) # data aligned already
table(data_clinical_sample$GRADE)
high_grade = grepl("G2|G3|G4", data_clinical_sample$GRADE)

data_clinical_sample = data_clinical_sample[high_grade, ]
data_clinical_patient = data_clinical_patient[high_grade, ]

patients = names(data_mrna)
time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]
table(status, exclude = NULL)
status[which(status == "")] = NA

length(which(is.na(time)==F))
length(which(is.na(status)==F))
```

#### Test for good p value

```{r}
p_threshold = c(1*10^-(0:10), 5*10^-(1:10))
fc_threshold = 0
treatments <- c("Carbo", "Olaparib", "NK")
conserve_gene_only <- TRUE

signature_selection <- data.frame(
  name = c(
    "log2FC_preR_preS_adj_boot_p", 
    "log2FC_preR_preS_adj_perm_p",
    "log2FC_preR_preS_p_adj",
    "log2FC_preR_preS_p"
  ),
  logFC = c(
    "_log2FC_preR_preS_adj",
    "_log2FC_preR_preS_adj",
    "_log2FC_preR_preS",
    "_log2FC_preR_preS"
  ),
  p = c(
    "_boot_p",
    "_perm_p",
    "_preR_preS_p_adj",
    "_preR_preS_p"
  )
)

res <- NULL
for (t in treatments){
  for (s in 1:nrow(signature_selection)){
    for (i in 1:length(p_threshold)){
      tmp <- NULL
      data_mrna2 <- data_mrna
      if (conserve_gene_only){
        deg_test <- filter(deg, ensembl %in% sister_conserve_gene)
      } else {
        deg_test <- deg
      }
      index1 = which(
        deg_test[[paste0(t, signature_selection$logFC[s])]] < -fc_threshold & 
          deg_test[[paste0(t, signature_selection$p[s])]] <= p_threshold[i]
        ) # downregulated genes in pre-resistant cell
      index2 = which(
        deg_test[[paste0(t, signature_selection$logFC[s])]] > fc_threshold &
          deg_test[[paste0(t, signature_selection$p[s])]] <= p_threshold[i]
        ) # upregulated genes in pre-resistant cell
      index1_symbol = deg_test$symbol[index1]
      
      index = c(index1, index2)
      length(index)
      deg_test = deg_test[index, ]
      # deg_test$symbol
      
      deg_test = deg_test[match(rownames(data_mrna2), deg_test$symbol), ]
      n_gene <- nrow(na.omit(deg_test))
      
      if (n_gene < 50) {
        break
      }
      
      # reverse index1 gene expressions
      data_mrna2[which(rownames(data_mrna2) %in% index1_symbol), ] = -data_mrna2[which(rownames(data_mrna2) %in% index1_symbol), ]
      
      # gene selection
      index = which(is.na(deg_test$symbol) == F)
      
      res1 = c()
      for (j in 1:length(patients)){
        cat('\r', j)
        res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
        res1[j] = weighted.mean(res.tmp[index], w = abs(deg_test[[paste0(t, signature_selection$logFC[s])]][index]), na.rm = TRUE) # rank of the preR signature genes
        
      }
      time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
      status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]
      
      group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
      data_plot = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
      set.seed(123)
      fit <- survfit(Surv(time, status) ~ group, data = data_plot)
      fit.coxph = coxph(Surv(time, status) ~ group, data = data_plot)
      survival <- summary(fit.coxph)
      
      tmp <- data.frame(
        p_threshold = p_threshold[i],
        n_gene = n_gene,
        survival_p = survival$logtest[3],
        signature_type = signature_selection$name[s],
        treatment = t
      )

      res <- rbind.data.frame(res, tmp)
    }
  }
}


write.csv(
  res,
  paste0(
    sub_output_dir,
    "preR_preS_survival",
    ifelse(conserve_gene_only, "_conserve_gene", ""),
    ifelse(deconvolute, "_deconvoluted.csv", "_non_deconvoluted.csv")),
  row.names = FALSE
)
```

#### Survival plot with selected parameters

```{r}
conserve_gene_only <- FALSE

if (conserve_gene_only) {
  p_threshold = 1*10^(-8)
  fc_threshold = 0
} else {
  p_threshold = 1*10^(-8)
  fc_threshold = 0
}
data_mrna2 <- data_mrna
  
if (conserve_gene_only){
  deg_test <- filter(deg, ensembl %in% sister_conserve_gene)
} else {
  deg_test <- deg
}

index1 = which(deg_test$avg_log2FC <= -fc_threshold & deg_test$p_val_adj <= p_threshold) # downregulated genes when BRCA1 is knocked out (BRCA-)
index2 = which(deg_test$avg_log2FC > fc_threshold & deg_test$p_val_adj <= p_threshold) # upregulated genes when BRCA1 is knocked out
index2_symbol = deg_test$symbol[index2]

index = c(index1, index2)
length(index)
deg_test = deg_test[index, ]
# deg_test$symbol

deg_test = deg_test[match(rownames(data_mrna2), deg_test$symbol), ]

# reverse index2 gene expressions
data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ] = -data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ]

# gene selection
index = which(is.na(deg_test$symbol) == F)

res1 = c()
for (j in 1:length(patients)){
  cat('\r', j)
  res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
  res1[j] = weighted.mean(res.tmp[index], w = abs(deg_test$avg_log2FC[index])) # rank of the BRCAness genes
  # res1[j] = weighted.mean(res.tmp[index], w = abs(deg_test$avg_log2FC[index])*(-log10(deg_test$p_val_adj[index]))) # rank of the BRCAness genes
  
}
time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]

group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
data_plot = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
set.seed(123)
fit <- survfit(Surv(time, status) ~ group, data = data_plot)
fit.coxph = coxph(Surv(time, status) ~ group, data = data_plot)
s <- summary(fit.coxph)
# res[i, 3] <- s$logtest[3]
  
pdf(paste0(sub_output_dir,"FigureS7A_threshold_", p_threshold, ifelse(conserve_gene_only, "BRCA_COV_OS_conserve_gene.pdf", "BRCA_COV_OS.pdf")), width = 4, height = 5.5, pointsize = 20)
fig <- ggsurvplot(
  fit,
  data = data_plot,
  palette = c(red, blue),
  risk.table = T,
  legend.title = "HRD (deg)", # - undeconvoluted",
  legend.labs = c("high", "low"),
  xlab = "Time (months)", ylab = "Overall survival",
  ggtheme = theme_survminer(
    base_size = 14,
    font.main = c(14, "plain", "black"),
    font.x = c(14, "plain", "black"),
    font.y = c(14, "plain", "black"),
    font.caption = c(14, "plain", "black"),
    font.legend = c(14, "plain", "black"),
    font.tickslab = c(14, "plain", "black"),
    font.submain = c(14, "plain", "black")
  ),
  pval = signif(s$logtest[3], 2),
  pval.coord = c(100, 0.75)
  ) +
  guides(colour = guide_legend(nrow = 2))
fig
dev.off()
```

# preR-preS signature in subclone A

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "purple", "red2", "blue2", "brown2", "green2","red2", "blue3", "hrd_color", "hrp_color", "hrd_vector_color")))
sub_output_dir <- paste0(output_dir, "test/DEG_within_subclone_A")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
treatments <- c("Carbo", "NK", "Olaparib")
```

```{r}
data <- readRDS("./data/input/merged_KURAMOCHI_samples1_scaled_new.RDS")
cell_anno <- data@meta.data
write.csv(cell_anno, paste0(sub_output_dir, "cell_annotation_all.csv"))
rm(data)
```

```{r}
data <- readRDS("./data/input/all_bt_sample_merge_filtered_gene_sisters.RDS")
length(setdiff(colnames(data), rownames(cell_anno)))
length(setdiff(rownames(cell_anno), colnames(data)))
length(intersect(rownames(cell_anno), colnames(data)))
meta_data <- cell_anno$CNV_subclone
names(meta_data) <- rownames(cell_anno)
data <- AddMetaData(data, meta_data, "CNV_subclone")
data <- subset(data, cells = colnames(data)[data$CNV_subclone == "A"])
```

# Tests

```{r}
# preR-preS in singleton
samples <- unique(data$sample)
Idents(data) <- "drugSens"
for (s in samples) {
  deg_prer_pres <- FindMarkers(
      subset(data, cells = colnames(data)[is.na(data$sisters) & data$sample == s]),
      ident.1 = "pre-resistant",
      ident.2 = "pre-sensitive",
      test.use = "wilcox", # old version is "wilcox"
      logfc.threshold = 0,
      min.pct = 0,
      verbose = TRUE
    )
  deg_prer_pres <- deg_prer_pres %>%
    rownames_to_column(var = "ensembl") %>%
    left_join(as.data.frame(data@misc$gene_meta), by = "ensembl")
  
  write.csv(deg_prer_pres, paste0(sub_output_dir, "/deg_prer_pres_in_singleton_wilcox_", s, ".csv"), row.names = FALSE)
}

```

## Integrate DEGs

```{r}
treatments <- c("NK", "Carbo", "Olaparib", "ctrl")
Idents(data) <- "drugSens"
conserved_gene <- read.csv("./data/input/sister_concordant_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  select(ensembl, symbol, conserve_gene)
for (t in treatments) {
  deg1 <- read.csv(paste0("./data/output/test/DEG_within_subclone_A/deg_prer_pres_in_singleton_wilcox_", t, "1.csv"),
                   stringsAsFactors = FALSE)
  deg2 <- read.csv(paste0("./data/output/test/DEG_within_subclone_A/deg_prer_pres_in_singleton_wilcox_", t, "2.csv"),
                   stringsAsFactors = FALSE)
  colnames(deg1)[!colnames(deg1) %in% c("ensembl", "symbol")] <- paste0(t,"_1_", colnames(deg1)[!colnames(deg1) %in% c("ensembl", "symbol")])
  colnames(deg2)[!colnames(deg2) %in% c("ensembl", "symbol")] <- paste0(t, "_2_", colnames(deg2)[!colnames(deg2) %in% c("ensembl", "symbol")])
  deg_merge <- deg1 %>% 
    full_join(deg2, by = c("ensembl", "symbol")) %>% 
    full_join(conserved_gene, by = c("ensembl", "symbol"))
  write.csv(deg_merge, paste0(sub_output_dir, "/deg_prer_pres_in_singleton_wilcox_overlap_", t, ".csv"), row.names = FALSE)
}
```

## Adjust growth control

```{r}
treatments <- c("NK", "Carbo", "Olaparib")
control <- read.csv(paste0(sub_output_dir, "/deg_prer_pres_in_singleton_wilcox_overlap_ctrl.csv")) %>% 
  mutate(mean_avg_log2FC_ctrl = (ctrl_1_avg_log2FC + ctrl_2_avg_log2FC) / 2) %>% 
  mutate(mean_p_val_adj_ctrl = (ctrl_1_p_val_adj + ctrl_2_p_val_adj) / 2) %>% 
  mutate(mean_p_val_ctrl = (ctrl_1_p_val + ctrl_2_p_val) / 2) %>% 
  select(ensembl, symbol, mean_avg_log2FC_ctrl, mean_p_val_adj_ctrl, mean_p_val_ctrl)

deg_adjust_merge <- NULL

for (t in treatments) {
  deg <-  read.csv(paste0(sub_output_dir, "/deg_prer_pres_in_singleton_wilcox_overlap_", t, ".csv"))
  deg[[paste0("mean_avg_log2FC_", t)]] <- (deg[[paste0(t, "_1_avg_log2FC")]] + deg[[paste0(t, "_2_avg_log2FC")]]) / 2
  deg[[paste0("mean_p_val_adj_", t)]] <- (deg[[paste0(t, "_1_p_val_adj")]] + deg[[paste0(t, "_2_p_val_adj")]]) / 2
  deg[[paste0("mean_p_val_", t)]] <- (deg[[paste0(t, "_1_p_val")]] + deg[[paste0(t, "_2_p_val")]]) / 2
  
  deg <- deg[, c(
      "ensembl", "symbol", paste0("mean_avg_log2FC_", t),
      paste0("mean_p_val_adj_", t), paste0("mean_p_val_", t))
    ]
  
  deg_adjust <- deg %>% 
    full_join(control, by = c("ensembl", "symbol"))
  
  deg_adjust[[paste0("avg_log2FC_", t, "_ctrl_adj")]] <- deg_adjust[[paste0("mean_avg_log2FC_", t)]] - deg_adjust[["mean_avg_log2FC_ctrl"]]
  
  if (is.null(deg_adjust_merge)){
    deg_adjust_merge <- deg_adjust
  } else {
    deg_adjust_merge <- full_join(
      deg_adjust_merge,
      select(deg_adjust, -mean_avg_log2FC_ctrl, -mean_p_val_adj_ctrl, -mean_p_val_ctrl),
      by = c("ensembl", "symbol"))
  }
}

deg_adjust_merge <- deg_adjust_merge %>% 
  full_join(conserved_gene, by = c("ensembl", "symbol"))
write.csv(deg_adjust_merge, paste0(sub_output_dir, "/deg_prer_pres_in_singleton_wilcox_adjusted_by_control.csv"), row.names = FALSE)

deg_adjust_log2FC <- deg_adjust_merge %>%
  dplyr::select(ensembl, symbol, conserve_gene, ends_with("_ctrl_adj"))
write.csv(deg_adjust_log2FC, paste0(sub_output_dir, "/deg_prer_pres_in_singleton_wilcox_adjusted_by_control_log2FC_only.csv"), row.names = FALSE)
```


```{r}

```

# Session information

```{r}
sessionInfo()
```


