---
title: "Figures in KURAMOCHI manuscript"
author: "Shuyu Zheng and Jing Tang"
data: "`r Sys.Date()`"
output: html_notebook
---

# Load pakcages

```{r}
library(Seurat)
library(plyr)
library(tidyverse)
library(readxl)
library(ggpubr)
library(ggalt)
library(ggrepel)
library(ggforce)
library(wordspace)
require(scales)
library(viridis)
library(ggsci)
library(synergyfinder)
library(ggsankey)
library(survminer)
library(survival)
library(scales)
library(sparseMatrixStats) # sparse matrix calculation
library(distances)
library(vegan) # distance function
library(irlba) # for fast pca
library(gridExtra) # Merge multiple ggplots with 'grid.arrange' function
library(reshape2)
library(car) # scatterplot combined with barplot
output_dir <- "../data/output/Figures_for_manuscript/"
if (!dir.exists(output_dir)){
  dir.create(output_dir, recursive = TRUE)
}
```

# Color code

```{r}
blue = "#3C5488FF"
red =  "#DC0000FF"
green = "#00A087FF"
brown = "#B09C85FF"

blue2 <- "#4DBBD5FF"
red2 <- "#E64B35FF"
green2 <- "#91D1C2FF"
brown2 <- "#7E6148FF"

blue3 <- "#8491B4FF"
red3 <- "#F39B7FFF"
```

# Figure 1: whether linear tracing is working as expected?

## B2: Sister distance in BT sample (violin plot)

Load auxiliary codes

```{r}
source("./auxiliary_codes.R")
sub_output_dir <- paste0(output_dir, "Figure1B/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir, recursive = TRUE)
}
```

Load data

```{r}
merge <- readRDS("../data/output_from_puhti/KURAMOCHI_before_treatment/all_bt_sample_merge_filtered_gene_sisters.RDS")

# Relabel sisters with the info of sample
all_sister_pairs <- merge$sisters_sample[!endsWith(merge$sisters_sample, "NA")]
# Percentage of sister cells' progenitor

t <- table(table(all_sister_pairs))
t
```

```{r}

2424/(sum(!is.na(merge$drugSens)) - sum(!is.na(merge$sisters)) + sum(t))
2424 * 2/sum(!is.na(merge$drugSens))

sum(!is.na(merge$sisters))/sum(!is.na(merge$drugSens))
sum(t)/(sum(!is.na(merge$drugSens)) - sum(!is.na(merge$sisters)) + sum(t))
```

% preR/(preR + preS)

```{r}
# Cell count
sum(merge$drugSens == "pre-resistant", na.rm = TRUE) / sum(merge$drugSens %in% c("pre-resistant", "pre-sensitive"), na.rm = TRUE) * 100
```

```{r}
# Lineages all
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens) %>% 
  filter(!is.na(drugSens) & drugSens == "pre-resistant")
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens) %>% 
  filter(!is.na(drugSens) & drugSens %in% c("pre-resistant", "pre-sensitive"))
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

```{r}
# Lineages Carbo
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens == "pre-resistant",
    sample %in% c("Carbo1", "Carbo2")
  )
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens %in% c("pre-resistant", "pre-sensitive"),
    sample %in% c("Carbo1", "Carbo2")
    )
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

```{r}
# Lineages Olaparib
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens == "pre-resistant",
    sample %in% c("Olaparib1", "Olaparib2")
  )
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens %in% c("pre-resistant", "pre-sensitive"),
    sample %in% c("Olaparib1", "Olaparib2")
    )
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

```{r}
# Lineages NK
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens == "pre-resistant",
    sample %in% c("NK1", "NK2")
  )
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens %in% c("pre-resistant", "pre-sensitive"),
    sample %in% c("NK1", "NK2")
    )
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

```{r}
# Lineages control
preR <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens == "pre-resistant",
    sample %in% c("ctrl1", "ctrl2")
  )
t <- table(table(preR$sisters_sample))
preR_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preRS <- merge@meta.data %>% 
  select(sisters_sample, drugSens, sample) %>% 
  filter(
    !is.na(drugSens),
    drugSens %in% c("pre-resistant", "pre-sensitive"),
    sample %in% c("ctrl1", "ctrl2")
    )
t <- table(table(preRS$sisters_sample))
preRS_l <- sum(t[t != 1]) + sum(as.numeric(names(t)[t == 1]))

preR_l/preRS_l * 100
```

1. Prepare expression matrix

```{r}
expression_mat <- as.matrix(GetAssayData(merge, slot = "data"))
```

2. Prepare sister table (including all sisters)

```{r}
sister_table <- data.frame(cell_barcode = colnames(merge), 
                     sisters = merge$sisters_sample,
                     stringsAsFactors = FALSE) %>% 
  dplyr::filter(sisters %in% all_sister_pairs) # remove singletons

sisters <- NULL
for (i in unique(sister_table$sisters)) {
  tmp_set <- sister_table$cell_barcode[which(sister_table$sisters  == i)]
  tmp_set <- combn(tmp_set, 2)
  tmp_df <- data.frame(group = rep(i, ncol(tmp_set)),
                       sister1 = tmp_set[1, ],
                       sister2 = tmp_set[2, ],
                       stringsAsFactors = FALSE)
  sisters <- rbind.data.frame(sisters, tmp_df)
  tmp_df <- NULL
  tmp_set <- NULL
}
```

```{r}
rm(merge) # remove large seurat object to release memory
rm(pca, tsne, umap)
```

3. Statistic analysis 

We can use function `CalculateSisterSimilarity` (defined in 'script/Auxiliary_codes.R" file) to generate analysis sister similarities by setting the parameters:

* **expression_mat**: the gene expression matrix in log2(normalized_count + 1) scale containing gene names as rows and cell barcodes as columns.
* **n_gene**: number of top variable genes used to calculate the similarity.
* **sister_table**: a data frame containing the group number of twins and corresponding sisters' cell barcodes.

Following is an example using Euclidean distance and top 1000 variable genes.

1. Calculate difference matrix

**Note** This function will take a long time to calculate the difference matrix. Some times it takes hours when there are a lot genes and cells.

```{r}
# time_stamp <- Sys.time()
plot_data <- CalculateSisterSimilarity(
  expression_mat = expression_mat,
  n_gene = 1000,
  sister_table = sisters,
  similarity = "euclidean"
  )
# Sys.time() - time_stamp
```

```{r}
sister_d <- data.frame(
    sister_id = names(plot_data$sister_distance),
    distance = plot_data$sister_distance,
    stringsAsFactors = FALSE
    )
```

```{r}
rm(expression_mat)
```

4. (Batch job) Generate the violin plot

This part is done with batch job. The codes are in: "./plot_sister_similarity.R" and "./plot_sister_similarity.sh"

```{r eval=FALSE}
sister_d <- readRDS("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_distance.RDS")
# p1 <- PlotSisterSimilarity(distance = sister_d$distance,
#                           similarity = sister_d$similarity,
#                                  n_gene = sister_d$n_gene,
#                                  cutoff = NULL)
m <- switch (sister_d$similarity,
             "euclidean" = "Euclidean Distance",
             "pearson" = "Pearson Coefficient",
             "spearman" = "Spearman Coefficient"
)

cutoff <- 14.94
test <- t.test(sister_d$distance$similarity[sister_d$distance$group == "sisters"], sister_d$distance$similarity[sister_d$distance$group == "non-sisters"])
print(test)

p1 <- ggplot(data = sister_d$distance, aes(x = group, y = similarity, fill = group)) +
  geom_violin(trim = FALSE) + 
  labs(title=paste0(m, " (genes: ", sister_d$n_gene, ")"),x="", y = m) +
  geom_boxplot(width=0.1, fill="white", outlier.shape = NA) + 
  # geom_segment(aes(y = cutoff, x = 1.75, xend = 2.25, yend = 14.93)) + 
  scale_fill_manual(values=c("#3C5488FF", "#DC0000FF")) +
  theme_classic() + 
  theme(
    legend.position = "none",
    plot.title = element_text(hjust = 0.5),
    axis.title.y = element_text(size = 30),
    axis.text.x = element_text(size = 30)
  ) +
  annotate(
    geom = "text",
    x = 1.5,
    y = 40,
    label = ifelse(
      test$p.value == 0,
      "P < 2.2E-16",
      paste0("P = ", test$p.value)
    ))
  # ggpubr::stat_compare_means(
  #   method = "t.test",
  #   vjust = 2,
  #   label.x.npc = "center"
  # ) +
  # ggpubr::stat_compare_means(
  #   method = "wilcox.test",
  #   vjust = 4,
  #   label.x.npc = "center"
  # ) +
  # geom_text(
  #   data = means,
  #   aes(
  #     label = signif(similarity, 4),
  #     y = similarity
  #   ), 
  #   nudge_x = 0.15
  # )
ggsave(
  paste0(
    sub_output_dir, "Figure1B2_",
    sister_d$sister_d, "_", sister_d$n_gene, "_genes.png"
  ), 
  plot = p1, device = "png", width = 6, height = 6, dpi = 300
)
ggsave(
  paste0(
    sub_output_dir, "Figure1B2_",
    sister_d$sister_d, "_", sister_d$n_gene, "_genes.pdf"
  ),
  plot = p1, device = "pdf", width = 6, height = 6
)
```

## B3: UMAP for sister pairs

```{r}
merge <- readRDS("../data/output_from_puhti/KURAMOCHI_before_treatment/all_bt_sample_merge_filtered_gene_sisters.RDS")
```


```{r}
all_sister_pairs <- merge$sisters_sample[!endsWith(merge$sisters_sample, "NA")]
all_sister_pairs <- unique(all_sister_pairs)
set.seed(123)
n <- 10
groups <- sample(all_sister_pairs, n)
highlight_group <- vector("list", length(groups))
names(highlight_group) <- groups
for (i in groups){
  highlight_group[[i]] <- colnames(merge)[which(merge$sisters_sample == i)]
}

myCol <- ggsci::pal_npg("nrc")(n)
# myCol <- c('#88CCEE', '#44AA99', '#117733', '#332288', '#DDCC77',
#            '#999933', '#CC6677', '#882255', '#AA4499', "#000000")
names(myCol) <- groups
# umap <- UMAPPlot(merge, cells.highlight = highlight_group,
#                cols.highlight = myCol[groups], sizes.highlight = 4,
#                cols = "grey70", na.value = "grey70") +
#   scale_color_manual(values = myCol, labels = seq(1, length(myCol)), na.value = "grey70") +
#   labs(
#     x = "UMAP 1",
#     y = "UMAP 2",
#     color = "Lineage label"
#   )
plot_table <- as.data.frame(merge@reductions$umap@cell.embeddings)
plot_table$sisters <- merge$sisters_sample
plot_table$sisters[!plot_table$sisters %in% groups] <- NA
umap <- ggplot(data = plot_table, mapping = aes(x = UMAP_1, y = UMAP_2)) +
  geom_point(
    data = plot_table,
    mapping = aes(x = UMAP_1, y = UMAP_2),
    color = "grey90") +
  geom_point(
    data = plot_table %>% filter(!is.na(sisters)),
    size = 6,
    mapping = aes(x = UMAP_1, y = UMAP_2, color=sisters)
  ) +
  scale_color_manual(values = myCol, labels = seq(1, length(myCol))) + 
  geom_line(
    data = plot_table %>% filter(!is.na(sisters)),
    mapping = aes(color = sisters),
    size = 2
  ) +
  labs(
    x = "UMAP 1",
    y = "UMAP 2",
    color = "Lineage label"
  ) +
  theme_classic() +
  theme(
    axis.text = element_text(size = 20, colour = "black"),
    axis.title = element_text(size = 20),
    legend.text = element_text(size = 20),
    legend.title = element_text(size = 25)
  )
umap
ggsave(paste0(sub_output_dir, "/Figure1B3_highlight_sister_group_",paste0(groups, collapse = "."), "_filtered_sisters_umap.png"), plot = umap, device = "png", width = 8.5, height = 6, dpi = 300)
ggsave(paste0(sub_output_dir, "/Figure1B3_highlight_sister_group_",paste0(groups, collapse = "."), "_filtered_sisters_umap.pdf"), plot = umap, device = "pdf", width = 7.5, height = 4.5)
```

## C: (Batch job) Sister conserve genes

This part is calculated with batch job. The codes are in: "./calculate_conserve_gene.R" and "./calculate_conserve_gene.sh"

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure1C/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

```{r}
conserve_gene <- read.csv(
  "../data/output_from_puhti/KURAMOCHI_before_treatment/sister_conserve_gene/sister_conserve_gene_all_bt_sample_twin.csv",
  stringsAsFactors = FALSE)
```

```{r}
highlight_gene_symbols <- conserve_gene %>% 
  arrange(desc(log2FC_dif)) %>% 
  head(10) %>% 
  select(symbol) %>% 
  unlist()
data_plot <- conserve_gene %>% 
  mutate(log10_p_value = -log10(adj_p_value)) %>% 
  mutate(
    highlight = ifelse(
      log10_p_value > -log10(0.05),
      "sister-concordant gene",
      "sister-disconcordant gene"
    ),
    label = (
      symbol %in% highlight_gene_symbols
    )
  )
data_plot$highlight <- factor(data_plot$highlight, levels = c("sister-concordant gene", "sister-disconcordant gene"))
cor <- cor.test(data_plot$mean_log2FC_sister, data_plot$mean_log2FC_non_sister)
p <- 
  # ggscatter(
  # data_plot, x = "mean_log2FC_sister", y = "mean_log2FC_non_sister",
  #             #size = "log10_p_value", 
  #             color = "highlight", # shape = "sister_conserve_gene"), # Customize reg. line
  # order=as.numeric(data_plot$highlight)
  #  ) +
  ggplot(
    data_plot,
    aes(
      x = "mean_log2FC_sister",
      y = "mean_log2FC_non_sister",
      color = "highlight"
    )
  ) +
  geom_point() +
  lims(x = c(0, 1.5), y = c(0, 1.5)) +
  scale_color_manual(values = c("#E64B3555", "#3C5488")) + #"#55555555", "#3C548855")) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
  annotate(
    geom = "text",
    x = 1.0,
    y = 0.2,
    label = paste0("R = ", signif(cor$estimate, 2), "\n", ifelse(cor$p.value == 0, "P < 2.2E-16", paste0("P =", format(cor$p.value, scientific = T)))),
    size = 6.5
  ) + 
  # stat_cor(
  #   method = "pearson",
  #   label.y.npc = "bottom",
  #   label.x = 0.4,
  #   size = 8) +
  geom_point(
    data=dplyr::filter(data_plot, label==TRUE),
    size = 6,
    color = "#DC0000") +
  geom_point(
    data=dplyr::filter(data_plot, label==TRUE),
    shape = 1,
    size = 6,
    show.legend = FALSE
  ) +
  ggrepel::geom_text_repel(
    data=dplyr::filter(
      data_plot, label==TRUE),
    aes(label = symbol),
    box.padding = 0.5,
    colour="black", point.padding = 0.25, max.overlaps = 30,
    size = 6.5,
    fontface="bold",
    show.legend=FALSE) +
  labs(
       x = expression("Sister cell pair log"[2]~"(FC)"),
       y = expression("Random cell pair log"[2]~"(FC)")#,
       # shape = "Sister Conserve Gene",
       # size = expression("-log"[10]~"(Average P value)")
       ) +
  theme(legend.position="right",
        # plot.title = element_text(hjust = 0.5, size = 20),
        axis.ticks.x = element_blank(),
        axis.text = element_text(size = 20),
        axis.title = element_text(size = 20)
      ) +
  guides(colour = "none")
p
ggsave(paste0(sub_output_dir, "conserve_gene_adjust_p.png"), p, device = "png",
       dpi = 300, width = 6, height = 6)
ggsave(paste0(sub_output_dir, "conserve_gene_adjust_p.svg"), p,
       width = 6, height = 6)
ggsave(paste0(sub_output_dir, "conserve_gene_adjust_p.pdf"), p,
       width = 6, height = 6)
```

## D: RNA dynamics from RNA velocity analysis

The alignment is done with velocyto. The codes are in: "./runVelo.sh"
The scVelo is used to do the RNA velocity analysis. The codes are in "./runscvelo.py" and "./runscvelo.sh"

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure1D/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

load data

```{r}
dynamic <- read.csv("../data/output_from_puhti/scvelo/RNA_velocity_var_all_BT.csv")
conserve_gene <- read.csv("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_conserve_gene/sister_conserve_gene_all_bt_sample_twin.csv")
selected_genes <- read.csv("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_conserve_gene/sum_exp_sister_conserve_gene_second_bin.csv")
plot_table <- selected_genes %>%
  mutate(conserve_gene = ifelse(Sister.conserved.gene, "Sister-concordant", "Sister-discordant")) %>% 
  right_join(dynamic, by = c("ensembl" = "X")) %>% 
  select(
    ensembl, 
    transcription = fit_alpha, 
    splicing = fit_beta, 
    degradation = fit_gamma, 
    conserve_gene) %>% 
  na.omit()
plot_table$conserve_gene <- factor(plot_table$conserve_gene, levels = c("Sister-concordant", "Sister-discordant"))
```

Visualization

```{r}
t <- plot_table %>% 
  select(ensembl, transcription, value = splicing, conserve_gene) %>%
  mutate(dynamic = "splicing") %>% 
  rbind.data.frame(
    plot_table %>% 
      select(ensembl, transcription, value = degradation, conserve_gene) %>% 
      mutate(dynamic = "degradation"))
t$dynamic <- str_to_title(t$dynamic)
t$dynamic <- factor(t$dynamic, levels = c("Splicing", "Degradation"))
min_value <- min(c(t$transcription, t$value))
max_value <- max(c(t$transcription, t$value))
p <- ggplot(t, aes(x = transcription, y = value)) +
  stat_density_2d(aes(fill = after_stat(ndensity)), geom = "raster", contour = FALSE,
                  ) +
  scale_fill_viridis() +
  labs(x = "Transcription") +
  geom_abline(
    intercept = 0,
    slope = 1,
    color = "#FFFFFFFF",
    linetype = "dashed",
    size = 1) + 
  scale_y_continuous(
    expand = c(0, 0),
    limits = c(min_value, max_value),
    trans = log2_trans(),
    breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x))) +
  scale_x_continuous(
    expand = c(0, 0),
    limits = c(min_value, max_value),
    trans = log2_trans(),
    breaks = trans_breaks("log2", function(x) 2^x),
    labels = trans_format("log2", math_format(2^.x))) +
  coord_equal(xlim=c(min_value, max_value),ylim=c(min_value, max_value)) +
  facet_grid(vars(dynamic), vars(conserve_gene)) +
  theme(
    legend.position = "none",
    strip.text.x = element_text(
      size = 20
    ),
    strip.text.y = element_text(
      size = 20
    ),
    axis.text = element_text(size = 20, colour = "black"),
    axis.title.y = element_blank(),
    axis.title.x = element_text(
      size = 20
    )
  )
p
ggsave(
  paste0(sub_output_dir, "sister_conserve_gene_dynamic_log_scale_second_bin.pdf"),
  p, width = 6, height = 6)
ggsave(
  paste0(sub_output_dir, "sister_conserve_gene_dynamic_log_scale_second_bin.svg"),
  p, width = 6, height = 6)
ggsave(
  paste0(sub_output_dir, "sister_conserve_gene_dynamic_log_scale_second_bin.png"),
  dpi = 300,
  p, width = 6, height = 6)
```

# Figure S1A number divisions after dubouling

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

The data table comes from Jun. He manually counted the number of cells which divided once, twice or more times after "1 x doubling time". Totally 6 views are counted.

```{r}
data <- read_xlsx("../data/raw/ReSisTrace_figS1 - Copy.xlsx", range = "B1:G4")
data <- t(data)
colnames(data) <- c("0", "1", "> 1")
data <- data * 100
data <- data %>% 
  as.data.frame() %>% 
  mutate(view = 1:6) %>% 
  pivot_longer(!view, names_to = "Division", values_to = "Proportion")
data$Division <- factor(data$Division, levels = c("0", "1", "> 1"))
```

```{r}
plot_table <- data %>%
  dplyr::group_by(Division) %>% 
  dplyr::summarise(
    mean = mean(Proportion),
    sd = sd(Proportion)
  )
write.csv(plot_table, paste0(sub_output_dir, "A_cell_division_times.csv"), row.names = FALSE)
col = c("#d1d7e3", "#8f9dbb", "#3c5488")
p <- ggplot(plot_table, aes(x = Division, y = mean, fill = Division)) +
  geom_bar(
    stat="identity", color="black",
    position=position_dodge()
  ) +
  geom_errorbar(
    aes(ymin = mean - sd, ymax = mean + sd, width = 0.2), color="black"
  ) +
  scale_fill_manual(values = col) +
  labs(
    x = "Division(s)",
    y = "Proportion of clones (%)"
  ) +
  theme_classic() +
  theme(
    legend.position = "None",
    text = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 20, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "A_cell_division_times.pdf"),
  width = 5, height = 5
)
```

# Figure S1B Counts of lineages in the BT sample

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

```{r}
plot_table <- NULL

# Control 1
load("../data/output_from_puhti/scLT_ctrl_1/preprocess/lineage_label_tables.RData")
tmp <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table() %>% 
  prop.table()
tmp2 <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table()
cat("Control 1\n", tmp2, "\n")
tmp <- data.frame(
  Size_of_lineage = tmp,
  Sample = "Control_1",
  label_at = length(unique(l_a_fre_merge_filtered$lineage_id)),
  label_bt = length(unique(l_b_fre_merge_filtered$lineage_id)),
  label_btat = length(unique(intersect(l_b_fre_merge_filtered$lineage_id, l_a_fre_merge_filtered$lineage_id)))
)
plot_table <- tmp

# Control 2
load("../data/output_from_puhti/scLT_ctrl_2/preprocess/lineage_label_tables.RData")
tmp <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table() %>% 
  prop.table()
tmp2 <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table()
cat("Control 2\n", tmp2, "\n")
tmp <- data.frame(
  Size_of_lineage = tmp,
  Sample = "Control_2",
  label_at = length(unique(l_a_fre_merge_filtered$lineage_id)),
  label_bt = length(unique(l_b_fre_merge_filtered$lineage_id)),
  label_btat = length(unique(intersect(l_b_fre_merge_filtered$lineage_id, l_a_fre_merge_filtered$lineage_id)))
)
plot_table <- rbind.data.frame(plot_table, tmp)

# Carboplatin 1
load("../../2021.1.3_KURAMOCHI_NK_Ola_Carbo/data/output/Carbo_1/preprocess/lineage_label_tables.RData")
tmp <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table() %>% 
  prop.table()
tmp2 <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table()
cat("Carboplatin 1\n", tmp2, "\n")
tmp <- data.frame(
  Size_of_lineage = tmp,
  Sample = "Carboplatin_1",
  label_at = length(unique(l_a_fre_merge_filtered$lineage_id)),
  label_bt = length(unique(l_b_fre_merge_filtered$lineage_id)),
  label_btat = length(unique(intersect(l_b_fre_merge_filtered$lineage_id, l_a_fre_merge_filtered$lineage_id)))
)
plot_table <- rbind.data.frame(plot_table, tmp)

# Carboplatin 2
load("../../2021.1.3_KURAMOCHI_NK_Ola_Carbo/data/output/Carbo_2/preprocess/lineage_label_tables.RData")
tmp <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table() %>% 
  prop.table()
tmp2 <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table()
cat("Carboplatin 2\n", tmp2, "\n")
tmp <- data.frame(
  Size_of_lineage = tmp,
  Sample = "Carboplatin_2",
  label_at = length(unique(l_a_fre_merge_filtered$lineage_id)),
  label_bt = length(unique(l_b_fre_merge_filtered$lineage_id)),
  label_btat = length(unique(intersect(l_b_fre_merge_filtered$lineage_id, l_a_fre_merge_filtered$lineage_id)))
)
plot_table <- rbind.data.frame(plot_table, tmp)

# Olaparib 1
load("../../2021.1.3_KURAMOCHI_NK_Ola_Carbo/data/output/Olaparib_1/preprocess/lineage_label_tables.RData")
tmp <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table() %>% 
  prop.table()
tmp2 <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table()
cat("Olaparib 1\n", tmp2, "\n")
tmp <- data.frame(
  Size_of_lineage = tmp,
  Sample = "Olaparib_1",
  label_at = length(unique(l_a_fre_merge_filtered$lineage_id)),
  label_bt = length(unique(l_b_fre_merge_filtered$lineage_id)),
  label_btat = length(unique(intersect(l_b_fre_merge_filtered$lineage_id, l_a_fre_merge_filtered$lineage_id)))
)
plot_table <- rbind.data.frame(plot_table, tmp)

# Olaparib 2
load("../../2021.1.3_KURAMOCHI_NK_Ola_Carbo/data/output/Olaparib_2/preprocess/lineage_label_tables.RData")
tmp <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table() %>% 
  prop.table()
tmp2 <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table()
cat("Olaparib 2\n", tmp2, "\n")
tmp <- data.frame(
  Size_of_lineage = tmp,
  Sample = "Olaparib_2",
  label_at = length(unique(l_a_fre_merge_filtered$lineage_id)),
  label_bt = length(unique(l_b_fre_merge_filtered$lineage_id)),
  label_btat = length(unique(intersect(l_b_fre_merge_filtered$lineage_id, l_a_fre_merge_filtered$lineage_id)))
)
plot_table <- rbind.data.frame(plot_table, tmp)

# NK 1
load("../../2021.1.3_KURAMOCHI_NK_Ola_Carbo/data/output/NK_1/preprocess/lineage_label_tables.RData")
tmp <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table() %>% 
  prop.table()
tmp2 <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table()
cat("NK 1\n", tmp2, "\n")
tmp <- data.frame(
  Size_of_lineage = tmp,
  Sample = "NK_1",
  label_at = length(unique(l_a_fre_merge_filtered$lineage_id)),
  label_bt = length(unique(l_b_fre_merge_filtered$lineage_id)),
  label_btat = length(unique(intersect(l_b_fre_merge_filtered$lineage_id, l_a_fre_merge_filtered$lineage_id)))
)
plot_table <- rbind.data.frame(plot_table, tmp)

# NK 2
load("../../2021.1.3_KURAMOCHI_NK_Ola_Carbo/data/output/NK_2/preprocess/lineage_label_tables.RData")
tmp <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table() %>% 
  prop.table()
tmp2 <- l_b_fre_merge_filtered %>% 
  select(lineage_id) %>% 
  table() %>% 
  table()
cat("NK 2\n", tmp2, "\n")
tmp <- data.frame(
  Size_of_lineage = tmp,
  Sample = "NK_2",
  label_at = length(unique(l_a_fre_merge_filtered$lineage_id)),
  label_bt = length(unique(l_b_fre_merge_filtered$lineage_id)),
  label_btat = length(unique(intersect(l_b_fre_merge_filtered$lineage_id, l_a_fre_merge_filtered$lineage_id)))
)
plot_table <- rbind.data.frame(plot_table, tmp)

colnames(plot_table) <- c("Size_of_lineage", "Frequency", "Sample", "labelAT", "labelBT", "labelBTAT")
write.csv(
  plot_table,
  paste0(sub_output_dir, "B_frequence_of_lineage_lable.csv"),
  row.names = FALSE
)
```

```{r}
plot_table$Frequency <- plot_table$Frequency * 100
plot_table <- plot_table %>%
  dplyr::group_by(Size_of_lineage) %>% 
  dplyr::summarise(
    mean = mean(Frequency),
    sd = sd(Frequency)
  )
col = c("#d1d7e3", "#8897b7", "#5e729d", "#3c5488")
p <- ggplot(plot_table, aes(x = Size_of_lineage, y = mean, fill = Size_of_lineage)) +
  geom_bar(
    stat="identity", color="black",
    position=position_dodge()
  ) +
  geom_errorbar(
    aes(ymin = mean - sd, ymax = mean + sd, width = 0.2), color="black"
  ) +
  scale_fill_manual(values = col) +
  labs(
    x = "Clone size",
    y = "Proportion of lineages (%)"
  ) +
  theme_classic() +
  theme(
    legend.position = "None",
    text = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 20, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "B_frequence_of_lineage_lable.pdf"),
  width = 5, height = 5
)
```

# Figure S1C DEGs correlation between replicates

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir, recursive = TRUE)
}
```

## Carboplatin

```{r}
deg <- read.csv("../../2021.04.08_KURAMOCHI_Olaparib_NK_Carbo/data/output/Before_treatment/preS_preR_in_singleton/Carbo_DEG/deg_prer_pres_in_singleton_Carbo_intersection.csv")
cat("Number of genes ", nrow(deg))
cor <- cor.test(deg$Carbo_1_avg_log2FC, deg$Carbo_2_avg_log2FC)
range <- range(c(deg$Carbo_1_avg_log2FC, deg$Carbo_2_avg_log2FC))
p <- ggplot(
  data = deg,
  aes(
    x = Carbo_1_avg_log2FC,
    y = Carbo_2_avg_log2FC)
  ) +
  geom_point(
    color = "#3C548855",
    size = 6,
    show.legend = FALSE
  ) +
  ylim(range) +
  xlim(range) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
  annotate(
    geom = "text",
    x = 0.15,
    y = - 0.3,
    label = paste0("R = ", signif(cor$estimate, 2), "\n", ifelse(cor$p.value == 0, "P < 2.2E-16", paste0("P =", format(cor$p.value, scientific = T)))),
    size = 6.5
  ) + 
  labs(
    x = expression("Replicate 1 log"[2]~"(FC)"),
    y = expression("Replicate 2 log"[2]~"(FC)"),
    title = "Carboplatin"
  ) +
  theme_classic() +
  theme(
    legend.position="none",
    axis.ticks.x = element_blank(),
    axis.text = element_text(size = 20, colour = "black"),
    axis.title = element_text(size = 20, colour = "black"),
    plot.title = element_text(size = 20, hjust = 0.5, colour = "black", face = "bold")
  )
p
ggsave(paste0(sub_output_dir, "correlation_between_replicates_carboplatin.png"), p, device = "png",
       dpi = 300, width = 4.5, height = 4.5)
ggsave(paste0(sub_output_dir, "correlation_between_replicates_carboplatin.svg"), p,
       width = 4.5, height = 4.5)
ggsave(paste0(sub_output_dir, "correlation_between_replicates_carboplatin.pdf"), p,
       width = 4.5, height = 4.5)
```

## Olaparib

```{r}
deg <- read.csv("../../2021.04.08_KURAMOCHI_Olaparib_NK_Carbo/data/output/Before_treatment/preS_preR_in_singleton/Olaparib_DEG/deg_prer_pres_in_singleton_Olaparib_intersection.csv")
cat("Number of genes ", nrow(deg))
cor <- cor.test(deg$Olaparib_1_avg_log2FC, deg$Olaparib_2_avg_log2FC)
range <- range(c(deg$Olaparib_1_avg_log2FC, deg$Olaparib_2_avg_log2FC))
p <- ggplot(
  data = deg,
  aes(
    x = Olaparib_1_avg_log2FC,
    y = Olaparib_2_avg_log2FC)
  ) +
  geom_point(
    color = "#3C548855",
    size = 6,
    show.legend = FALSE
  ) +
  ylim(range) +
  xlim(range) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
  annotate(
    geom = "text",
    x = 0.1,
    y = - 0.3,
    label = paste0("R = ", signif(cor$estimate, 2), "\n", ifelse(cor$p.value == 0, "P < 2.2E-16", paste0("P =", format(cor$p.value, scientific = T)))),
    size = 6.5
  ) + 
  labs(
    x = expression("Replicate 1 log"[2]~"(FC)"),
    y = expression("Replicate 2 log"[2]~"(FC)"),
    title = "Olaparib"
  ) +
  theme_classic() +
  theme(
    legend.position="none",
    axis.ticks.x = element_blank(),
    axis.text = element_text(size = 20, colour = "black"),
    axis.title = element_text(size = 20, colour = "black"),
    plot.title = element_text(size = 20, colour = "black", hjust = 0.5, face = "bold")
  )
p
ggsave(paste0(sub_output_dir, "correlation_between_replicates_Olaparib.png"), p, device = "png",
       dpi = 300, width = 4.5, height = 4.5)
ggsave(paste0(sub_output_dir, "correlation_between_replicates_Olaparib.svg"), p,
       width = 4.5, height = 4.5)
ggsave(paste0(sub_output_dir, "correlation_between_replicates_Olaparib.pdf"), p,
       width = 4.5, height = 4.5)
```

## NK

```{r}
deg <- read.csv("../../2021.04.08_KURAMOCHI_Olaparib_NK_Carbo/data/output/Before_treatment/preS_preR_in_singleton/NK_DEG/deg_prer_pres_in_singleton_NK_intersection.csv")
cat("Number of genes ", nrow(deg))
cor <- cor.test(deg$NK_1_avg_log2FC, deg$NK_2_avg_log2FC)
range <- range(c(deg$NK_1_avg_log2FC, deg$NK_2_avg_log2FC))
p <- ggplot(
  data = deg,
  aes(
    x = NK_1_avg_log2FC,
    y = NK_2_avg_log2FC)
  ) +
  geom_point(
    color = "#3C548855",
    size = 6,
    show.legend = FALSE
  ) +
  ylim(range) +
  xlim(range) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
  annotate(
    geom = "text",
    x = 0.2,
    y = - 0.3,
    label = paste0("R = ", signif(cor$estimate, 2), "\n", ifelse(cor$p.value == 0, "P < 2.2E-16", paste0("P =", format(cor$p.value, scientific = T)))),
    size = 6.5
  ) + 
  labs(
    x = expression("Replicate 1 log"[2]~"(FC)"),
    y = expression("Replicate 2 log"[2]~"(FC)"),
    title = "NK"
  ) +
  theme_classic() +
  theme(
    legend.position="none",
    axis.ticks.x = element_blank(),
    axis.text = element_text(size = 20, colour = "black"),
    axis.title = element_text(size = 20, colour = "black"),
    plot.title = element_text(size = 20, colour = "black", hjust = 0.5, face = "bold")
  )
p
ggsave(paste0(sub_output_dir, "correlation_between_replicates_NK.png"), p, device = "png",
       dpi = 300, width = 4.5, height = 4.5)
ggsave(paste0(sub_output_dir, "correlation_between_replicates_NK.svg"), p,
       width = 4.5, height = 4.5)
ggsave(paste0(sub_output_dir, "correlation_between_replicates_NK.pdf"), p,
       width = 4.5, height = 4.5)
```

## Control

```{r}
deg <- read.csv("../data/output/DEG/deg_pres_prer_scLT_ctrl_intersection_in_singleton_wilcox.csv")
cat("Number of genes ", nrow(deg))
cor <- cor.test(deg$scLT_ctrl_1_avg_log2FC, deg$scLT_ctrl_2_avg_log2FC)
range <- range(c(deg$scLT_ctrl_1_avg_log2FC, deg$scLT_ctrl_2_avg_log2FC))
p <- ggplot(
  data = deg,
  aes(
    x = scLT_ctrl_1_avg_log2FC,
    y = scLT_ctrl_2_avg_log2FC)
  ) +
  geom_point(
    color = "#3C548855",
    size = 6,
    show.legend = FALSE
  ) +
  ylim(range) +
  xlim(range) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
  annotate(
    geom = "text",
    x = 0.2,
    y = - 0.3,
    label = paste0("R = ", signif(cor$estimate, 2), "\n", ifelse(cor$p.value == 0, "P < 2.2E-16", paste0("P =", format(cor$p.value, scientific = T)))),
    size = 6.5
  ) + 
  labs(
    x = expression("Replicate 1 log"[2]~"(FC)"),
    y = expression("Replicate 2 log"[2]~"(FC)"),
    title = "Control"
  ) +
  theme_classic() +
  theme(
    legend.position="none",
    axis.ticks.x = element_blank(),
    axis.text = element_text(size = 20, colour = "black"),
    axis.title = element_text(size = 20, colour = "black"),
    plot.title = element_text(size = 20, colour = "black", hjust = 0.5, face = "bold")
  )
p
ggsave(paste0(sub_output_dir, "correlation_between_replicates_Control.png"), p, device = "png",
       dpi = 300, width = 4.5, height = 4.5)
ggsave(paste0(sub_output_dir, "correlation_between_replicates_Control.svg"), p,
       width = 4.5, height = 4.5)
ggsave(paste0(sub_output_dir, "correlation_between_replicates_Control.pdf"), p,
       width = 4.5, height = 4.5)
```

# Figure S1C Histogram for gene expressions of sister-conserved/non-conserved genes

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir, recursive = TRUE)
}
```

```{r}
plot_table <- read.csv("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_conserve_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv")

t_test <- t.test(plot_table$mean_expression[plot_table$conserve_gene == TRUE], plot_table$mean_expression[plot_table$conserve_gene == FALSE])
print(t_test)

plot_table <-  plot_table %>%  
  mutate(`Sister-conserved gene` = ifelse(conserve_gene, "Sister-concordant gene", "Sister-discordant gene")) %>% 
  select(ensembl, `Sister-conserved gene`, sum_expression, mean_expression)

mylog10_trans <- function (base = 10)
{
  trans <- function(x) log(x + 1, base)
  inv <- function(x) base^x
  trans_new(paste0("log-", format(base)), trans, inv, log_breaks(base = base),
            domain = c(1e-100, Inf))
}

p <- ggplot(
  plot_table,
  aes(
    x = sum_expression,
    fill = `Sister-conserved gene`,
    color = `Sister-conserved gene`
    )
  ) +
  geom_histogram(
    alpha=0.2,
    bins = 50,
    position="identity"
  ) +
  scale_fill_manual(values = c("#DC000055", "#3C548855")) +
  scale_color_manual(values = c("#DC0000", "#3C5488")) +
  # ?(
  #   breaks = trans_breaks("log10", function(x) 10^x),
  #   labels = trans_format("log10", math_format(10^.x))
  # ) +
  geom_segment(
    data = NULL, 
    aes(x = 10000, y = 5 * 10^3, xend = 5000, yend = 2.5* 10^3),
    arrow = arrow(length = unit(0.25, "cm"), type = "closed"),
    show.legend = FALSE) +
  scale_y_continuous(
    trans = "mylog10",
    breaks = trans_breaks("log10", function(x) 10^(x - 1)),
    labels = trans_format("log10", function(x) math_format()(round(x)))
    ) +
  labs(
    x = "Expression in pre-treatment sample",
    y = "Number of genes",
    fill = "",
    color = "") +
  theme_classic() +
  theme(
    legend.position = "top",
    axis.text = element_text(size = 20, colour = "black"),
    axis.title = element_text(size = 20, colour = "black"),
    legend.title = element_text(size = 20, colour = "black"),
    legend.text = element_text(size = 20, colour = "black"),
    axis.line = element_line(colour = "black"),
    axis.ticks = element_line(colour = "black")
  )
p
ggsave(
  paste0(sub_output_dir, "C_sister_conserve_gene_exp_distribution.pdf"), p,
  width = 9, height = 5
)
ggsave(
  paste0(sub_output_dir, "C_sister_conserve_gene_exp_distribution.png"), p,
  width = 9, height = 5, dpi = 300
)
ggsave(paste0(sub_output_dir, "C_sister_conserve_gene_exp_distribution.svg"), p,
  width = 9, height = 5
)
```

# Figure S1D Unspliced RNA propotion for sister-conserved/non-conserved genes

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
selected_genes <- read.csv("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_conserve_gene/sum_exp_sister_conserve_gene_second_bin.csv")
```

```{r eval = FALSE}
unsplice <- read.csv("../data/output_from_puhti/scvelo/RNA_velocity_unspliced_all_BT.csv", row.names = 1)
unsplice <- as.matrix(unsplice)
unsplice  <- unsplice[, colnames(unsplice) %in% selected_genes$ensembl]
splice <- read.csv("../data/output_from_puhti/scvelo/RNA_velocity_spliced_all_BT.csv", row.names = 1)
splice <- as.matrix(splice)
splice <- splice[, colnames(splice) %in% selected_genes$ensembl]
u_s_propotion <- unsplice / (unsplice + splice)
write.csv(u_s_propotion, paste0(sub_output_dir, "unspliced_RNA_propotion_second_bin.csv"))
```

```{r}
u_s_propotion <- read.csv(
  paste0(sub_output_dir, "unspliced_RNA_propotion_second_bin.csv"),
  row.names = 1
)
```

```{r}
u_s_propotion <- as.matrix(u_s_propotion)
selected_genes$conserve_gene <- ifelse(selected_genes$Sister.conserved.gene, "Sister-concordant\ngene", "Sister-discordant\ngene")
selected_genes$conserve_gene <- factor(selected_genes$conserve_gene, levels = c("Sister-concordant\ngene", "Sister-discordant\ngene"))
u_s_propotion_mean <- colMeans(u_s_propotion, na.rm = T)
u_s_propotion_mean <- data.frame(
  gene = names(u_s_propotion_mean), 
  unsplice_propotion = u_s_propotion_mean
) %>% 
  filter(gene %in% selected_genes$ensembl) %>% 
  left_join(select(selected_genes, gene = ensembl, conserve_gene), by = "gene") %>% 
  filter(!is.na(conserve_gene))
  
means <- aggregate(unsplice_propotion ~ conserve_gene, u_s_propotion_mean, mean)
p <- ggplot(
  data = u_s_propotion_mean,
  aes(x = conserve_gene, y = unsplice_propotion, fill = conserve_gene)
) + 
  geom_violin(trim = FALSE) + 
  labs(x="", y = "Unspliced RNA propotion") +
  geom_boxplot(width=0.1, fill="white", outlier.shape = NA) + 
  scale_fill_manual(values=c(red, blue)) +
  theme_classic() + 
  theme(
    legend.position = "none",
    plot.title = element_text(hjust = 0.5),
    axis.title.y = element_text(size = 20, color = "black"),
    axis.text.x = element_text(size = 20, color = "black"),
    axis.text.y = element_text(size = 20, color = "black")
  ) + 
  ggpubr::stat_compare_means(
    method = "t.test",
    label.y = 1.3,
    label.x.npc = "center",
    size = 6.5,
    hjust = 0.5) +
  geom_text(
    data = means,
    aes(label = signif(unsplice_propotion, 3),y = unsplice_propotion), 
    nudge_x = 0.18, color = "white", size = 6.5, fontface = "bold")

ggsave(
  paste0(sub_output_dir, "D_unspliced_RNA_propotion_second_bin.png"), 
  plot = p, width = 8, height = 5, dpi = 300
)
ggsave(
  paste0(sub_output_dir, "D_unspliced_RNA_propotion_second_bin.pdf"), 
  plot = p, width = 8, height = 5
)
ggsave(
  paste0(sub_output_dir, "D_unspliced_RNA_propotion_second_bin.svg"), 
  plot = p, width = 8, height = 5
)
```

# Figure S1E Degradation/Transcription violin plot for sister-conserved/non-conserved genes

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
selected_genes <- read.csv("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_conserve_gene/sum_exp_sister_conserve_gene_second_bin.csv")
```

```{r}
dynamic <- read.csv("../data/output_from_puhti/scvelo/RNA_velocity_var_all_BT.csv")
selected_genes$conserve_gene <- ifelse(selected_genes$Sister.conserved.gene, "Sister-concordant\ngene", "Sister-discordant\ngene")
selected_genes$conserve_gene <- factor(selected_genes$conserve_gene, levels = c("Sister-concordant\ngene", "Sister-discordant\ngene"))
plot_table <- selected_genes %>%
  select(ensembl, conserve_gene) %>% 
  right_join(dynamic, by = c("ensembl" = "X")) %>% 
  mutate(
    Degradation_Transcription = fit_gamma/fit_alpha
  ) %>% 
  select(
    ensembl, 
    Degradation_Transcription, 
    conserve_gene) %>% 
  na.omit()
means <- aggregate(Degradation_Transcription ~ conserve_gene, plot_table, mean)
```

```{r}
p <- ggplot(
  data = plot_table,
  aes(x = conserve_gene, y = Degradation_Transcription, fill = conserve_gene)
) + 
  geom_violin(trim = FALSE) + 
  labs(x="", y = "Degradation rate / Transcription rate") +
  geom_boxplot(width=0.1, fill="white", outlier.shape = NA) + 
  scale_fill_manual(values=c(red, blue)) +
  theme_classic() + 
  theme(
    legend.position = "none",
    axis.title.y = element_text(size = 20, color = "black"),
    axis.text.x = element_text(size = 20, color = "black"),
    axis.text.y = element_text(size = 20, color = "black")
  ) + 
  ggpubr::stat_compare_means(
    method = "t.test",
    label.x.npc = "center",
    size = 6.5,
    hjust = 0.5
  ) +
  geom_text(
    data = means,
    aes(label = signif(Degradation_Transcription, 3), y = Degradation_Transcription), 
    nudge_x = 0.15, color = "white", size = 6.5, fontface = "bold")

ggsave(
  paste0(sub_output_dir, "E_Degradation_Transcription_second_bin.png"), 
  plot = p, width = 8, height = 5.5, dpi = 300
)
ggsave(
  paste0(sub_output_dir, "E_Degradation_Transcription_second_bin.pdf"), 
  plot = p, width = 8, height = 5.5
)
ggsave(
  paste0(sub_output_dir, "E_Degradation_Transcription_second_bin.svg"), 
  plot = p, width = 8, height = 5.5
)
```

## E: gene set analysis for sister-conserved genes

```{r}
sub_output_dir <- paste0(output_dir, "Figure1E/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir, recursive = TRUE)
}
```

```{r}
cons <- data.table::fread("../data/output/enrichr/KEGG_2021_Human_table_non_conserved_genes.txt")
plot_table <- cons %>% 
    select(Term, x = `Odds Ratio`, p = `Adjusted P-value`) %>% 
    mutate(y = -log10(p),
           sig = p < 0.05)
# plot_table$Term <- paste0("bold('", plot_table$Term, "')")
highlight <- plot_table %>% 
    # filter(Term %in% c("Inositol phosphate metabolism", "Phosphatidylinositol signaling system"))
    filter(sig)

p <- ggplot() +
   theme_classic() +
   theme(
       legend.key = element_blank(), 
       axis.text = element_text(size = 20, colour = "black"),
       axis.title = element_text(size = 20),
       plot.title = element_text(size = 20, hjust = 0.5)
       ) +
   labs(x = expression("Odds Ratio"),
        y = expression("-log"[10]~"(P value)"),
        title="") +
   # geom_hline(yintercept=-log10(0.05), linetype="dashed") +
   # geom_vline(xintercept = 0, linetype="dashed") +
 geom_point(
     aes(x = x, y = y, colour=factor(sig)),
     colour = "grey70",
     size = 6,
     data = plot_table,
     show.legend = FALSE) +
  geom_point(
    aes(x = x, y = y),
    size = 8,
    data = highlight,
    color = "#3C5488FF",
    show.legend = FALSE) +
  geom_point(
    aes(x = x, y = y),
    shape = 1,
    size = 8,
    data = highlight,
    show.legend = FALSE) +
  geom_text_repel(
    aes(x = x, y = y, label = Term),
    size = 6.5,
    data = highlight,
    # parse = TRUE,
    box.padding = 1
  )
p
ggsave(paste0(sub_output_dir, "non_conserved_KEGG_volcano.pdf"), p,
       width = 6, height = 6)
ggsave(paste0(sub_output_dir, "non_conserved_KEGG_volcano.svg"), p,
       width = 6, height = 6)
ggsave(paste0(sub_output_dir, "non_conserved_KEGG_volcano.png"), p,
       width = 6, height = 6, dpi = 300)
```

# Figure S1F Non-conserve gene

```{r}
sub_output_dir <- paste0(output_dir, "FigureS1/")
if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```
sister-conserve gene

```{r}
cons <- data.table::fread("../data/output/enrichr/KEGG_2021_Human_table_sister_conserved_gene.txt")
plot_table <- cons %>% 
    select(Term, x = `Odds Ratio`, p = `Adjusted P-value`) %>% 
    mutate(y = -log10(p),
           sig = p < 0.05)
# plot_table$Term <- paste0("bold('", plot_table$Term, "')")
highlight <- plot_table %>% 
    # filter(Term %in% c("TGF-beta signaling pathway", "Fanconi anemia pathway",
    #            "TNF signaling pathway"))
    filter(sig)

p <- ggplot() +
   theme_classic() +
   theme(
       legend.key = element_blank(), 
       axis.text = element_text(size = 20, colour = "black"),
       axis.title = element_text(size = 20),
       plot.title = element_text(size = 20, hjust = 0.5)
       ) +
   labs(x = expression("Odds Ratio"),
        y = expression("-log"[10]~"(P value)"),
        title="") +
   # geom_hline(yintercept=-log10(0.05), linetype="dashed") +
   # geom_vline(xintercept = 0, linetype="dashed") +
 geom_point(
     aes(x = x, y = y, colour=factor(sig)),
     colour = "grey70",
     size = 6,
     data = plot_table,
     show.legend = FALSE) +
  geom_point(
    aes(x = x, y = y),
    size = 8,
    data = highlight,
    color = "#3C5488FF",
    show.legend = FALSE) +
  geom_point(
    aes(x = x, y = y),
    shape = 1,
    size = 8,
    data = highlight,
    show.legend = FALSE) +
  geom_text_repel(
    aes(x = x, y = y, label = Term),
    size = 6.5,
    data = highlight,
    # parse = TRUE,
    box.padding = 1
  )
p
ggsave(paste0(sub_output_dir, "D_sister_conserved_KEGG_volcano.pdf"), p,
       width = 8, height = 5.5)
ggsave(paste0(sub_output_dir, "D_sister_conserved_KEGG_volcano.svg"), p,
       width = 8, height = 5.5)
ggsave(paste0(sub_output_dir, "D_sister_conserved_KEGG_volcano.png"), p,
       width = 8, height = 5.5, dpi = 300)
```



# Figure 2: how cells manage to survive during treatment?

## B1_old: DEG preR-pres filter growth control drug targets

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure2B/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
highlight_gene_symbols <- c(
  "HMGB1", "PRKDC", "MDK", "FIS1"
)
conserve_gene <- read.csv("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_conserve_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  select(conserve_gene, ensembl)
```

### Carbo

```{r}
data <- read.csv("../data/output/DEG_filter_by_bootstrap_permutation_in_singleton/deg_prer_pres_Carbo_control_bootstrap_permutation.csv") %>% 
  filter(
    Carbo_1_pct.1 > 0.1 &
    Carbo_1_pct.2 > 0.1 &
    Carbo_2_pct.1 > 0.1 &
    Carbo_2_pct.2 > 0.1 &
    Carbo_1_p_val_adj < 0.05 &
    Carbo_2_p_val_adj < 0.05 &
    !conflict
  ) %>% 
  select(-sister_conserve_gene) %>% 
  left_join(conserve_gene, by = "ensembl") %>% 
  filter(
    conserve_gene
  )

data$boot_Carbo_ctrl_adj_p_value <- p.adjust(data$boot_Carbo_ctrl_p_value, method = "fdr")
data$perm_Carbo_ctrl_adj_p_value <- p.adjust(data$perm_Carbo_ctrl_p_value, method = "fdr")
```

```{r}
data %>% 
  filter(
    boot_Carbo_ctrl_adj_p_value < 0.05 &
    perm_Carbo_ctrl_adj_p_value < 0.05
  )
```

```{r}
df <- data
df$boot_Carbo_ctrl_adj_p_value[df$boot_Carbo_ctrl_adj_p_value == 0] <- min(df$boot_Carbo_ctrl_adj_p_value[df$boot_Carbo_ctrl_adj_p_value > 0])
df$perm_Carbo_ctrl_adj_p_value[df$perm_Carbo_ctrl_adj_p_value == 0] <- min(df$perm_Carbo_ctrl_adj_p_value[df$perm_Carbo_ctrl_adj_p_value > 0])

df <- df %>% 
  mutate(
    y_boot = -log10(boot_Carbo_ctrl_adj_p_value),
    y_perm = -log10(perm_Carbo_ctrl_adj_p_value),
    avg_logFC = (Carbo_1_avg_log2FC + Carbo_2_avg_log2FC) / 2) %>% 
  dplyr::select(symbol, x = "avg_logFC", y_boot, y_perm)
color <- NULL
for (i in 1:nrow(df)){
  if (df$y_boot[i] < -log10(0.05)){
    tmp <- 0
  } else if (df$x[i] > 0) {
    tmp <- 1
  } else {
    tmp <- 2
  }
  color <- c(color, tmp)
}
df$color <- color
color2 <- NULL
for (i in 1:nrow(df)){
  if (df$y_perm[i] < -log10(0.05)){
    tmp <- 0
  } else if (df$x[i] > 0) {
    tmp <- 1
  } else {
    tmp <- 2
  }
  color2 <- c(color2, tmp)
}
df$color2 <- color2
highlight_genes <- df %>% 
  filter(symbol %in% highlight_gene_symbols)
p <- ggplot() +
  theme_classic() +
  theme(
    legend.key=element_blank(),
    axis.text = element_text(face = "plain", size = 25),
    axis.title = element_text(size = 30),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 30),
    plot.margin = margin(10, 30, 10, 10)) +
  scale_colour_manual(
    values=c(
      "grey90",
      # "#E64B35FF", "#4DBBD5FF",
      "#DC0000FF", "#3C5488FF")
  ) +
  labs(
    x=expression("log(FC) PreR vs PreS cells"),
    y=expression("-log"[10]~"(P value) Bootstrap"),
    title="Carboplatin") +
  geom_hline(yintercept=-log10(0.05), linetype="dashed") +
  geom_vline(xintercept=0, linetype="dashed") +
  geom_point(
    aes(x=x, y=y_boot, colour=factor(color)),
    data=df,
    size=3,
    show.legend = FALSE,
    alpha = 0.5
  ) +
  geom_text_repel(
    data = highlight_genes,
    aes(x=x, y=y_boot, label = symbol),
    size = 8,
    fontface = "bold",
    box.padding = 1.5) +
  geom_point(
    aes(x=x, y=y_boot, colour=factor(color)),
    data=highlight_genes,
    size = 6,
    show.legend = FALSE
  ) +
  geom_point(
    aes(x=x, y=y_boot),
    data=highlight_genes,
    color = "black",
    shape = 1,
    size=6,
    show.legend = FALSE
  )
p
ggsave(
  paste0(sub_output_dir, "Carbo_preR_preS_bootstrap.pdf"),
  p,
  width = 6,
  height = 6)
ggsave(
  paste0(sub_output_dir, "Carbo_preR_preS_bootstrap.png"),
  p,
  width = 6,
  height = 6,
  dpi = 300)
```

### Olaparib

```{r}
data <- read.csv("../data/output/DEG_filter_by_bootstrap_permutation_in_singleton/deg_prer_pres_Olaparib_control_bootstrap_permutation.csv") %>% 
  filter(
    Olaparib_1_pct.1 > 0.1 &
    Olaparib_1_pct.2 > 0.1 &
    Olaparib_2_pct.1 > 0.1 &
    Olaparib_2_pct.2 > 0.1 &
    Olaparib_1_p_val_adj < 0.05 &
    Olaparib_2_p_val_adj < 0.05 &
    ! conflict
  ) %>% 
  select(-sister_conserve_gene) %>% 
  left_join(conserve_gene, by = "ensembl") %>% 
  filter(
    conserve_gene
  )
data$boot_Olaparib_ctrl_adj_p_value <- p.adjust(data$boot_Olaparib_ctrl_p_value, method = "fdr")
data$perm_Olaparib_ctrl_adj_p_value <- p.adjust(data$perm_Olaparib_ctrl_p_value, method = "fdr")
```

```{r}
data %>% 
  filter(
    boot_Olaparib_ctrl_adj_p_value < 0.05 &
    perm_Olaparib_ctrl_adj_p_value < 0.05
  )
```
```{r}
df <- data
df$boot_Olaparib_ctrl_adj_p_value[df$boot_Olaparib_ctrl_adj_p_value == 0] <- min(df$boot_Olaparib_ctrl_adj_p_value[df$boot_Olaparib_ctrl_adj_p_value > 0])
df$perm_Olaparib_ctrl_adj_p_value[df$perm_Olaparib_ctrl_adj_p_value == 0] <- min(df$perm_Olaparib_ctrl_adj_p_value[df$perm_Olaparib_ctrl_adj_p_value > 0])

df <- df %>% 
  mutate(
    y_boot = -log10(boot_Olaparib_ctrl_adj_p_value),
    y_perm = -log10(perm_Olaparib_ctrl_adj_p_value),
    avg_logFC = (Olaparib_1_avg_log2FC + Olaparib_2_avg_log2FC) / 2) %>% 
  dplyr::select(symbol, x = "avg_logFC", y_boot, y_perm)
color <- NULL
for (i in 1:nrow(df)){
  if (df$y_boot[i] < -log10(0.05)){
    tmp <- 0
  } else if (df$x[i] > 0) {
    tmp <- 1
  } else {
    tmp <- 2
  }
  color <- c(color, tmp)
}
df$color <- color
color2 <- NULL
for (i in 1:nrow(df)){
  if (df$y_perm[i] < -log10(0.05)){
    tmp <- 0
  } else if (df$x[i] > 0) {
    tmp <- 1
  } else {
    tmp <- 2
  }
  color2 <- c(color2, tmp)
}
df$color2 <- color2
highlight_genes <- df %>% 
  filter(symbol %in% highlight_gene_symbols)
p <- ggplot() +
  theme_classic() +
  theme(
    legend.key=element_blank(),
    axis.text = element_text(face = "plain", size = 25),
    axis.title = element_text(size = 30),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 30),
    plot.margin = margin(10, 30, 10, 10)) +
  scale_colour_manual(
    values=c(
      "grey90",
      # "#E64B35FF", "#4DBBD5FF",
      "#DC0000FF", "#3C5488FF")
  ) +
  labs(
    x=expression("log(FC) PreR vs PreS cells"),
    y=expression("-log"[10]~"(P value) Bootstrap"),
    title="Olaparib") +
  geom_hline(yintercept=-log10(0.05), linetype="dashed") +
  geom_vline(xintercept=0, linetype="dashed") +
  geom_point(
    aes(x=x, y=y_boot, colour=factor(color)),
    data=df,
    size=3,
    show.legend = FALSE,
    alpha = 0.5
  ) +
  geom_text_repel(
    data = highlight_genes,
    aes(x=x, y=y_boot, label = symbol),
    size = 8,
    fontface = "bold",
    box.padding = 1.5) +
  geom_point(
    aes(x=x, y=y_boot, colour=factor(color)),
    data=highlight_genes,
    size = 6,
    show.legend = FALSE
  ) +
  geom_point(
    aes(x=x, y=y_boot),
    data=highlight_genes,
    color = "black",
    shape = 1,
    size=6,
    show.legend = FALSE
  )
p
ggsave(
  paste0(sub_output_dir, "Olaparib_preR_preS_bootstrap.pdf"),
  p,
  width = 6,
  height = 6)
ggsave(
  paste0(sub_output_dir, "Olaparib_preR_preS_bootstrap.png"),
  p,
  width = 6,
  height = 6,
  dpi = 300)
```

### NK

```{r}
data <- read.csv("../data/output/DEG_filter_by_bootstrap_permutation_in_singleton/deg_prer_pres_NK_control_bootstrap_permutation.csv") %>% 
  filter(
    NK_1_pct.1 > 0.1 &
    NK_1_pct.2 > 0.1 &
    NK_2_pct.1 > 0.1 &
    NK_2_pct.2 > 0.1 &
    NK_1_p_val_adj < 0.05 &
    NK_2_p_val_adj < 0.05 &
    ! conflict
  ) %>% 
  select(-sister_conserve_gene) %>% 
  left_join(conserve_gene, by = "ensembl") %>% 
  filter(
    conserve_gene
  )
data$boot_NK_ctrl_adj_p_value <- p.adjust(data$boot_NK_ctrl_p_value, method = "fdr")
data$perm_NK_ctrl_adj_p_value <- p.adjust(data$perm_NK_ctrl_p_value, method = "fdr")
```

```{r}
data %>% 
  filter(
    boot_NK_ctrl_adj_p_value < 0.05 &
    perm_NK_ctrl_adj_p_value < 0.05
  )
```
```{r}
df <- data
df$boot_NK_ctrl_adj_p_value[df$boot_NK_ctrl_adj_p_value == 0] <- min(df$boot_NK_ctrl_adj_p_value[df$boot_NK_ctrl_adj_p_value > 0])
df$perm_NK_ctrl_adj_p_value[df$perm_NK_ctrl_adj_p_value == 0] <- min(df$perm_NK_ctrl_adj_p_value[df$perm_NK_ctrl_adj_p_value > 0])

df <- df %>% 
  mutate(
    y_boot = -log10(boot_NK_ctrl_adj_p_value),
    y_perm = -log10(perm_NK_ctrl_adj_p_value),
    avg_logFC = (NK_1_avg_log2FC + NK_2_avg_log2FC) / 2) %>% 
  dplyr::select(symbol, x = "avg_logFC", y_boot, y_perm)
color <- NULL
for (i in 1:nrow(df)){
  if (df$y_boot[i] < -log10(0.05)){
    tmp <- 0
  } else if (df$x[i] > 0) {
    tmp <- 1
  } else {
    tmp <- 2
  }
  color <- c(color, tmp)
}
df$color <- color
color2 <- NULL
for (i in 1:nrow(df)){
  if (df$y_perm[i] < -log10(0.05)){
    tmp <- 0
  } else if (df$x[i] > 0) {
    tmp <- 1
  } else {
    tmp <- 2
  }
  color2 <- c(color2, tmp)
}
df$color2 <- color2
highlight_genes <- df %>% 
  filter(symbol %in% highlight_gene_symbols)
p <- ggplot() +
  theme_classic() +
  theme(
    legend.key=element_blank(),
    axis.text = element_text(face = "plain", size = 25),
    axis.title = element_text(size = 30),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 30),
    plot.margin = margin(10, 30, 10, 10)) +
  scale_colour_manual(
    values=c(
      "grey90",
      # "#E64B35FF", "#4DBBD5FF",
      "#DC0000FF", "#3C5488FF")
  ) +
  labs(
    x=expression("log(FC) PreR vs PreS cells"),
    y=expression("-log"[10]~"(P value) Bootstrap"),
    title="NK cell") +
  geom_hline(yintercept=-log10(0.05), linetype="dashed") +
  geom_vline(xintercept=0, linetype="dashed") +
  geom_point(
    aes(x=x, y=y_boot, colour=factor(color)),
    data=df,
    size=3,
    show.legend = FALSE,
    alpha = 0.5
  ) +
  geom_text_repel(
    data = highlight_genes,
    aes(x=x, y=y_boot, label = symbol),
    size = 8,
    fontface = "bold",
    box.padding = 1.5) +
  geom_point(
    aes(x=x, y=y_boot, colour=factor(color)),
    data=highlight_genes,
    size = 6,
    show.legend = FALSE
  ) +
  geom_point(
    aes(x=x, y=y_boot),
    data=highlight_genes,
    color = "black",
    shape = 1,
    size=6,
    show.legend = FALSE
  )
p
ggsave(
  paste0(sub_output_dir, "NK_preR_preS_bootstrap.pdf"),
  p,
  width = 6,
  height = 6)
ggsave(
  paste0(sub_output_dir, "NK_preR_preS_bootstrap.png"),
  p,
  width = 6,
  height = 6,
  dpi = 300)
```

## B2_old: HMGB1/PRKDC inhibitors in combination with olaparib/Carboplatin (synergy box plots)

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure2B/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

```{r}
load("../../2022.2.9_drug_combination_screening/data/output/analysis/SynergyFinder_object.RData")
second <- data$drug_pairs %>% 
  filter(
    drug1 %in% c("Olaparib", "Carboplatin"),
    (grepl("HMGB1", drug2) | grepl("PRKDC", drug2)),
    !drug2 %in% c("Glycyrrhizic.acid-HMGB1", "RAGE.antagonist-HMGB1")
    ) %>% 
  select(drug1, drug2, HSA_synergy)
second$target <- sapply(second$drug2, function(x) unlist(strsplit(x, "-"))[2])
second$group <- paste(second$drug1, second$target, sep = " + ")
```

```{r}
ggplot(data = second, aes(x = drug1, y = HSA_synergy, fill = target)) + 
  scale_fill_manual(values=c("#3C5488FF", "#DC0000FF")) + 
  geom_boxplot() +
  labs(
    x = "",
    y = "HSA synergy score",
    fill = "Drug target") +
  theme_classic() +
  theme(
    axis.text = element_text(size = 35),
    axis.title = element_text(size = 35),
    legend.title = element_text(size = 30),
    legend.text = element_text(size = 25),
    legend.key.size = unit(1, "cm")
  )
ggsave(
  paste0(sub_output_dir, "figure2B_box_plot_for_synergy_score.pdf"),
  width = 9, height = 6)
ggsave(
  paste0(sub_output_dir, "figure2B_box_plot_for_synergy_score.png"),
  width = 9, height = 6, dpi = 300)
```

## D_old: HMGB1 inhibitors in combination with Olaparib

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure2D/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

```{r include=FALSE}
synergyfinder_obj <- readxl::read_xlsx("../data/output/drug_combination_screening/drug_targetting_HMGB1_PRKDC.xlsx") %>% 
  filter(
    drug_row == "Olaparib",
    drug_col != "Glycyrrhizic.acid-HMGB1")
synergyfinder_obj$drug_col <- gsub("-.*", "", synergyfinder_obj$drug_col)
synergyfinder_obj <- ReshapeData(synergyfinder_obj, data_type = "inhibition")
blocks <- synergyfinder_obj$drug_pairs$block_id
```

```{r}
for (b in blocks){
  pdf(paste0(sub_output_dir, synergyfinder_obj$drug_pairs$drug2[synergyfinder_obj$drug_pairs$block_id == b], ".pdf"), width = 4, height = 3)
  p <- Plot2DrugSurface(
    synergyfinder_obj,
    plot_block = b,
    plot_title = "",
    summary_statistic = "mean",
    plot_value = "response",
    high_value_color = "#DC0000FF",
    low_value_color = "white",
    color_range = c(0, 100),
    z_range = range(synergyfinder_obj$response$response[synergyfinder_obj$response$block_id == b]),
    dynamic = F
  )
  dev.off()
}
```

## S4 cell-cell distances between drug-sensitive groups within CNV subclones

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure S4/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

```{r}
# library(Seurat)
# library(irlba)
# library(umap)
# library(plyr)
# library(vegan)
# library(reshape2)

data <- readRDS("../data/Jing_code_data/merged_KURAMOCHI_samples1_scaled_new.RDS") # Shuyu update on 1.4.2022 after FindVariableFeatures and Scale, check test.R in the server
data2 = data@assays$RNA@scale.data  # scale data is for dimension reduction or clustering
dim(data2) # 2000 77724
treatments <- c("Olaparib", "Carboplatin", "NK", "Control")
subclone <- LETTERS[1:5]
```

Calculate cell-cell distance

```{r eval = FALSE}
# data2.dist = distances(t(data2), normalize = "mahalanobize")
# saveRDS(data2.dist, file = paste0(sub_output_dir, "all_cell_distance.RDS"))
data2.dist <- readRDS(paste0(sub_output_dir, "all_cell_distance.RDS"))
```

```{r eval = FALSE}
res <- NULL
for (t in treatments){
  tmp_res <- vector(mode = "list", length = 5)
  names(tmp_res) <- subclone
  for(i in subclone){
    # cat("\r",i)
    preS = which(
      data$drugSens == "pre-sensitive" &
      data$CNV_subclone == i &
      data$treatment == t)
    preR = which(
      data$drugSens == "pre-resistant" &
      data$CNV_subclone == i &
      data$treatment == t
    )
    posT = which(
      data$drugSens == "resistant" &
      data$CNV_subclone == i &
      data$treatment == t
    )
    tmp = data2.dist[preS, preS]
    distS = tmp[upper.tri(tmp, diag = F)]
    tmp = data2.dist[preR, preR]
    distR = tmp[upper.tri(tmp, diag = F)]
    tmp = data2.dist[posT, posT]
    distT = tmp[upper.tri(tmp, diag = F)]
    tmp = cbind(dist = c(distS, distR, distT),
    drugSens = c(rep(1, length(distS)), rep(2, length(distR)), rep(3, length(distT))),
    subclone = i)
    if (nrow(tmp) > 100000){
      tmp_small = tmp[sample(1:nrow(tmp), 100000),]
    } else {
      tmp_small <- tmp
    }
    tmp_small = data.frame(tmp_small)
    tmp_small$drugSens[which(tmp_small$drugSens == 1)] = "pre-sensitive"
    tmp_small$drugSens[which(tmp_small$drugSens == 2)] = "pre-resistant"
    tmp_small$drugSens[which(tmp_small$drugSens == 3)] = "post-treatment"
    tmp_res[[i]] = tmp_small
  }
  tmp_res = do.call("rbind", tmp_res)
  tmp_res$drugSens <- factor(tmp_res$drugSens, levels = c("pre-sensitive", "pre-resistant", "post-treatment"))
  tmp_res$subclone <- factor(tmp_res$subclone, levels = c("A","B","C","D","E"))
  tmp_res$dist <- as.numeric(tmp_res$dist)
  tmp_res$treatment <- t
  if (is.null(res)){
    res <- tmp_res
  } else {
    res <- rbind.data.frame(res, tmp_res)
  }
}
View(res)

write.csv(res, paste0(sub_output_dir, "cell_distance_table_by_treatment_for_plot.csv"))
```

```{r}
res <- read.csv(paste0(sub_output_dir, "cell_distance_table_by_treatment_for_plot.csv"))
color = c("pre-sensitive" = "snow2", "pre-resistant" = "snow4", "post-treatment" = "slategray4")
res$drugSens <- factor(res$drugSens, levels = c("pre-sensitive", "pre-resistant", "post-treatment"))
res$subclone <- factor(res$subclone, levels = c("A","B","C","D","E"))
res <- res %>% 
  filter(drugSens != "post-treatment")
for (t in treatments) {
  tmp_plot <- res[res$treatment == t,]
  test <- rep(NA, 5)
  names(test) <- subclone
  for (i in subclone){
    tmp_test <- t.test(
        tmp_plot$dist[tmp_plot$drugSens == "pre-resistant" & tmp_plot$subclone == i],
        tmp_plot$dist[tmp_plot$drugSens == "pre-sensitive" & tmp_plot$subclone == i]
      )
    test[i] <- tmp_test$p.value
  }
  test <- data.frame(
    subclone = names(test),
    p_value = test,
    sig = case_when(
    test < .1 & test >= .05 ~ "*", 
    test < .05 & test >= .01 ~ "**", 
    test < .01 ~ "***", 
    TRUE ~ ""
    ),
    x = 1:5 - 0.1,
    y = rep(90, 5)
  )
  fig <- ggplot(
    data = tmp_plot,
    aes(y = dist, x = subclone, group = interaction(subclone, drugSens))
    ) +
    geom_violin(bw = 2, width = 0.8, aes(fill = drugSens)) +
    geom_boxplot(
      position = position_dodge(width = 0.8),
      fill = "white", width = 0.1,
      outlier.shape = NA,# alpha = 1
    ) +
    scale_y_continuous(name = "Normalized distance", limits = c(30, 90)) +  # There are 4.71% cell pairs has the distance larger than 100
    scale_x_discrete(name = "Subclone") +
    scale_fill_manual(values = color) + 
    # annotate("text", x = test$x, y = test$y, label = test$sig) +
    ggtitle(t) +
    theme_bw() +
    theme(
      text = element_text(size = 20, colour = "black"),
      plot.title = element_text(hjust = 0.5, size = 20),
      axis.text = element_text(size = 18, color = "black"),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      legend.position = "none"
    )
  ggsave(
    paste0(sub_output_dir, t, "_dist_cnv.pdf"),
    fig, width = 9, height = 4
  )
  ggsave(
    paste0(sub_output_dir, t, "_dist_cnv.svg"),
    fig, width = 9, height = 4
  )
}
```

## A: UMAP drug sensitivity groups

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure 2/")

if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

```{r}
data <- readRDS("../data/Jing_code_data/merged_KURAMOCHI_samples1_scaled_new.RDS") # Shuyu update on 1.4.2022 after FindVariableFeatures and Scale, check test.R in the server
data2 = data@assays$RNA@scale.data  # scale data is for dimension reduction or clustering
dim(data2) # 2000 77724

### UMAP visualization ----------------------------------------------------

data$drugSens[which(data$drugSens == "resistant")] = "post-treatment"
data2.umap = readRDS("../data/Jing_code_data/merged_KURAMOCHI_samples1_SCT_umap_2000_n50_scaleT.RDS") # run prcom_irlba with n = 50 ans scale = T, check test.R in the server

index = which(data$drugSens %in% c("pre-sensitive", "pre-resistant","post-treatment"))
# index = which(data$drugSens %in% c("pre-resistant"))
pdf(paste0(sub_output_dir, "A_umap_all.pdf"), width = 5, height = 5, pointsize = 12)
col = c("snow2", "snow4", "slategray4")
size = c(1,3,3)
tmp = factor(data$drugSens[index], levels = c("pre-sensitive", "pre-resistant", "post-treatment"))
plot(data2.umap$layout[index,], col = col[tmp], pch = '.', cex = size[tmp], xlab = "UMAP 1", ylab = "UMAP 2")
legend("bottomright", legend = levels(tmp), pch = 16, col = col[1:length(tmp)], cex = 1, pt.cex = 1)
dev.off()
table(data$drugSens[index])
```

## B: UMAP CNV subclone

```{r}
index = which(data$drugSens %in% c("pre-sensitive", "pre-resistant","post-treatment"))

pdf(paste0(sub_output_dir, "B_umap_all_cnv.pdf"), width = 5, height = 5, pointsize = 12)
col <- c(
  "A" = "#DC0000FF", "B" = "#B09C85FF", "C" = "#00A087FF",
  "D" = "#91D1C2FF",  "E" = "#3C5488FF"
  )
tmp = factor(data$CNV_subclone[index], levels = c("A", "B", "C", "D", "E"))
plot(data2.umap$layout[index,], col = col[tmp], pch = '.', xlab = "UMAP 1", ylab = "UMAP 2")
legend("topright", legend = levels(tmp), pch = 16, col = col[1:length(tmp)], cex = 1, pt.cex = 1)
dev.off()
table(data$CNV_subclone[index])
```

## Figure S2A

```{r}
# individual treatment conditions
for(i in unique(data$treatment)){
  for(j in c("pre-sensitive", "pre-resistant", "post-treatment")){
    pdf(
      paste(sub_output_dir, paste(i, j, sep="_"),".pdf",sep=""),
      width = 3.7, height = 3.5, pointsize = 10)
    
    index1 = which(data$drugSens == j & data$treatment == i)
    # if(j == "pre-resistant") color = red
    # if(j == "pre-sensitive") color = blue
    # if(j == "post-treatment") color = brown
    if(i == "Olaparib") color = red
    if(i == "Carboplatin") color = blue
    if(i == "NK") color = brown
    if(i == "Control") color = green
    scatterplot(
      data2.umap$layout[index1, 1],
      data2.umap$layout[index1, 2], cex = 0.5, cex.main = 1.4, cex.axis = 1.2,
      col = color, main = paste(i,j),
      boxplots = FALSE, las = 1, smooth = FALSE, grid = TRUE,
      regLine = FALSE, ellipse = TRUE, pch = 20, 
      xlim = c(-10,10), ylim = c(-10,10),
      legend = list(title = "treatment", coords = "topleft"),
      # xlab = "UMAP 1", ylab = "UMAP 2"
      xlab = "", ylab = ""
      )
    dev.off()
    cat(i, ", ", j, ", ", sum(index1), "\n")
  }
  print(i)
  print(t.test(data2.umap$layout[which(data$drugSens == "pre-resistant" & data$treatment == i), 1], data2.umap$layout[which(data$drugSens == "pre-sensitive" & data$treatment == i),1]))
  print(t.test(data2.umap$layout[which(data$drugSens == "pre-resistant" & data$treatment == i), 2], data2.umap$layout[which(data$drugSens == "pre-sensitive" & data$treatment == i),2]))
  
}
```

## C: Proportion of CNV subclones

```{r}
# proportion of sub-clones in pre-S
subclone = sort(unique(data$CNV_subclone))
# treatment = unique(data$treatment)
treatment = c("Carboplatin", "Olaparib", "NK", "Control") 
res2 = data.frame(mat.or.vec(length(treatment), length(subclone)))
count2 = data.frame(mat.or.vec(length(treatment), length(subclone)))
for(i in 1:length(treatment)){
  tmp = data$CNV_subclone[which(data$drugSens == "pre-sensitive" & data$treatment == treatment[i])]
  count2[i,] = table(tmp)
  x = table(tmp)/length(tmp)
  cat(treatment[i], "\n", table(tmp), "\n")
  res2[i,] = x
  names(res2) = names(x)
  names(count2) <- names(x)
}

rownames(res2) = treatment
res2 = melt(as.matrix(res2))
colnames(res2) <- c("treatment", "subclone", "proportion")
res2$drugSens <- "pre-sensitive"

rownames(count2) = treatment
count2 = melt(as.matrix(count2))
colnames(count2) <- c("treatment", "subclone", "count")
count2$drugSens <- "pre-sensitive"

values1 = c(red, brown, green, green2, blue)

p <- ggplot(res2, aes(x = treatment, y = proportion, fill = subclone)) +
  geom_col(position = "fill") +
  labs(
    y = "Proportion",
    fill = ""
  ) + 
  ggtitle("pre-sensitive") +
  scale_fill_manual(values = values1) +
  theme_bw() +
  theme(
    text = element_text(size = 24),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 21,
      color = "black"
    ),
    axis.text.y = element_text(
      size = 21,
      color = "black"
    ),
    axis.title.x = element_blank(),
    legend.text = element_text(size = 21),
    plot.title = element_text(hjust = 0.5, size = 28),
    panel.grid = element_blank()
  )
ggsave(paste0(sub_output_dir, "C_BRCA_CNV_preS.pdf"), width = 5, height = 5)

# pre-R
res3 = data.frame(mat.or.vec(length(treatment), length(subclone)))
count3 = data.frame(mat.or.vec(length(treatment), length(subclone)))
for(i in 1:length(treatment)){
  tmp = data$CNV_subclone[which(data$drugSens == "pre-resistant" & data$treatment == treatment[i])]
  count3[i, ] <- table(tmp)
  x = table(tmp)/length(tmp)
  res3[i,] = x
  names(res3) = names(x)
  names(count3) <- names(x)
}
rownames(res3) = treatment
res3
res3 = melt(as.matrix(res3))
colnames(res3) <- c("treatment", "subclone", "proportion")
res3$drugSens <- "pre-resistant"

rownames(count3) = treatment
count3 = melt(as.matrix(count3))
colnames(count3) <- c("treatment", "subclone", "count")
count3$drugSens <- "pre-resistant"

p <- ggplot(res3, aes(x = treatment, y = proportion, fill = subclone)) +
  geom_col(position = "fill") +
  labs(
    y = "Proportion",
    fill = "") + 
  ggtitle("pre-resistant") +
  scale_fill_manual(values = values1) +
  theme_bw() +
  theme(
    text = element_text(size = 24),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 21,
      color = "black"
    ),
    axis.text.y = element_text(
      size = 21,
      color = "black"
    ),
    axis.title.x = element_blank(),
    plot.title = element_text(hjust = 0.5, size = 28),
    panel.grid = element_blank()
  )
ggsave(paste0(sub_output_dir, "C_BRCA_CNV_preR.pdf"), width = 5, height = 5)

# Post-treatment
res4 = data.frame(mat.or.vec(length(treatment), length(subclone)))
count4 = data.frame(mat.or.vec(length(treatment), length(subclone)))
for(i in 1:length(treatment)){
  tmp = data$CNV_subclone[which(data$drugSens == "post-treatment" & data$treatment == treatment[i])]
  count4[i, ] <- table(tmp)
  x = table(tmp)/length(tmp)
  res4[i,] = x
  names(res4) = names(x)
  names(count4) <- names(x)
}
rownames(res4) = treatment
res4
res4 = melt(as.matrix(res4))
colnames(res4) <- c("treatment", "subclone", "proportion")
res4$drugSens <- "post-treatment"

rownames(count4) = treatment
count4 = melt(as.matrix(count4))
colnames(count4) <- c("treatment", "subclone", "count")
count4$drugSens <- "post-treatment"

p <- ggplot(res4, aes(x = treatment, y = proportion, fill = subclone)) +
  geom_col(position = "fill") +
  labs(
    y = "Proportion",
    fill = "") + 
  ggtitle("post-treatment") +
  scale_fill_manual(values = values1) +
  theme_bw() +
  theme(
    text = element_text(size = 24),
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 21,
      color = "black"
    ),
    axis.text.y = element_text(
      size = 21,
      color = "black"
    ),
    axis.title.x = element_blank(),
    plot.title = element_text(hjust = 0.5, size = 28),
    panel.grid = element_blank()
  )
ggsave(paste0(sub_output_dir, "C_BRCA_CNV_postT.pdf"), width = 5, height = 5)
```

```{r}
# paired t tes
res <- res2 %>% 
  rbind.data.frame(res3) %>% 
  rbind.data.frame(res4)
write.csv(res, paste0(sub_output_dir, "C_cell_proportion_table.csv"), row.names = FALSE)

preR_A <- res$proportion[res$subclone == "A" & res$drugSens == "pre-resistant"]
preS_A <- res$proportion[res$subclone == "A" & res$drugSens == "pre-sensitive"]
t_test <- t.test(preR_A, preS_A, paired = TRUE)
t_test
```

```{r}
# proportion test
count <- count2 %>% 
  rbind.data.frame(count3) %>% 
  rbind.data.frame(count4)
write.csv(count, paste0(sub_output_dir, "C_cell_count_table.csv"), row.names = FALSE)

preR_A_control <- count$count[count$subclone == "A" & count$drugSens == "pre-resistant" & treatment == "Control"]
preR_control <- sum(count$count[count$drugSens == "pre-resistant" & treatment == "Control"])
preS_A_control <- count$count[count$subclone == "A" & count$drugSens == "pre-sensitive" & treatment == "Control"]
preS_control <- sum(count$count[count$drugSens == "pre-sensitive" & treatment == "Control"])
cat("preR vs preS control", "\n")

for (t in c("Carboplatin", "Olaparib", "NK")){
  cat(t, "\n")
  cat("pre-resistant", "\n")
  preR_A <- count$count[count$subclone == "A" & count$drugSens == "pre-resistant" & treatment == t]
  preR_treatment <- sum(count$count[count$drugSens == "pre-resistant" & treatment == t])
  p <- prop.test(x = c(preR_A, preR_A_control), n = c(preR_treatment, preR_control))
  print(p)
  # cat("pre-sensitive", "\n")
  # preS_A <- count$count[count$subclone == "A" & count$drugSens == "pre-sensitive" & treatment == t]
  # preS_treatment <- sum(count$count[count$drugSens == "pre-sensitive" & treatment == t])
  # prop.test(x = c(preS_A, preS_A_control), n = c(preS_treatment, preS_control))
  # cat("preR vs preS", "\n")
  # prop.test(x = c(preR_A, preS_A), n = c(preS_treatment, preR_treatment))
}

```

## D correlation between CNV subclone DEG and preR vs preS DEG

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure 2/")

if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
conserve <- read.csv("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_conserve_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  select(ensembl, conserve_gene)
```

```{r}
treatment <- c("Carbo", "Olaparib", "NK")
deg <- NULL
for (t in treatment){
  tmp <- read.csv(
  paste0(
    "../data/output/DEG_filter_by_bootstrap_permutation_in_singleton/pres_prer_",
    t,
    "_control_bootstrap_permutation_summary.csv")
  )
  tmp[[paste0(t, "_log2FC_preR_preS_adj")]] <- tmp$mean_log2FC_treat - tmp$mean_log2FC_ctrl
  tmp <- tmp[, c("ensembl", "symbol", paste0(t, "_log2FC_preR_preS_adj"))]
  if (is.null(deg)){
    deg <- tmp
  } else {
    deg <- left_join(deg, tmp, by = c("ensembl", "symbol"))
  }
}

# DEG for CNV
for (s in LETTERS[1:5]){
  CNV_deg <- read.csv(
    paste0(
      "../data/output_from_puhti/CNV_DEG/",
      s,
      "_vs_others_DEG.csv")
    ) %>% 
    select(X, avg_log2FC)
  colnames(CNV_deg) <- c("ensembl", paste0("log2FC_", s))
  deg <- deg %>% 
    left_join(CNV_deg, by = "ensembl")
}

deg <- deg %>% 
  left_join(conserve, by = "ensembl")
write.csv(deg, paste0(sub_output_dir, "D_log2FC_preR_preS_CNV_subclone.csv"), row.names = FALSE)

deg <- deg %>% 
  filter(conserve_gene)
```

```{r}
treatments <- c("NK", "Carbo", "Olaparib")
subclones <- c("A", "B", "C", "D", "E")
plot_table <- NULL
for (t in treatments){
  for (s in subclones){
    tmp <- cor.test(
      deg[[paste0(t, "_log2FC_preR_preS_adj")]],
      deg[[paste0("log2FC_", s)]]
    )
    df <- data.frame(
      treatment = case_when(
        t == "Carbo" ~ "Carboplatin", 
        TRUE ~ t
      ),
      subclone = s,
      correlation = tmp$estimate,
      CI_lower = tmp$conf.int[1],
      CI_upper = tmp$conf.int[2],
      p_value = tmp$p.value,
      sig = case_when(
        tmp$p.value < .05 & tmp$p.value >= .01 ~ "*", 
        tmp$p.value < .01 & tmp$p.value >= .001 ~ "**", 
        tmp$p.value < .001 ~ "***", 
        TRUE ~ ""
      ),
      n = sum(
        !is.na(deg[[paste0("log2FC_", s)]]) & 
        !is.na(deg[[paste0(t, "_log2FC_preR_preS_adj")]])
      )
    )
    if (is.null(plot_table)){
      plot_table <- df
    } else {
      plot_table <- rbind.data.frame(plot_table, df)
    }
  }
}
plot_table$sig_y <- plot_table$CI_upper + 0.05
plot_table$sig_y[plot_table$correlation < 0] <- plot_table$CI_lower[plot_table$correlation < 0] - 0.05
write.csv(plot_table, paste0(sub_output_dir, "D_Pearson_corelation_coefficient.csv"),
          row.names = FALSE)
plot_table$treatment <- factor(plot_table$treatment, levels = c("Carboplatin", "Olaparib", "NK"))
```

```{r}
color <- c("A" = "#DC0000FF", "B" = "#B09C85FF", "C" = "#00A087FF", "D" = "#91D1C2FF", "E" = "#3C5488FF")

ggplot(
  plot_table,
  aes(y = correlation, x = treatment, fill = subclone)
  ) +
  scale_fill_manual(
    values = color
  ) + 
  geom_bar(stat="identity", position="dodge") + 
  geom_errorbar(
    aes(ymin=CI_lower, ymax=CI_upper),
    position=position_dodge(width=.9),
    width=.15
  ) + 
  geom_text(
    aes(y=sig_y, label=sig),
    position=position_dodge(width=.9),
    size = 6
  ) + 
  labs(x="", y="Correlation coefficient", fill = "") +
  theme_classic() +
  theme(
     legend.key = element_blank(), 
     axis.text = element_text(size = 21, colour = "black"),
     axis.title = element_text(size = 21),
     legend.text = element_text(size = 21)
     )
ggsave(
  paste0(sub_output_dir, "D_Pearson_correlation_coefficients_color_by_subclones.pdf"),
  width = 10, height = 5)
```

### Spearman

```{r}
treatments <- c("NK", "Carbo", "Olaparib")
subclones <- c("A", "B", "C", "D", "E")
plot_table <- NULL
for (t in treatments){
  for (s in subclones){
    tmp <- cor.test(
      deg[[paste0(t, "_log2FC_preR_preS_adj")]],
      deg[[paste0("log2FC_", s)]],
      method = "spearman"
    )
    df <- data.frame(
      treatment = case_when(
        t == "Carbo" ~ "Carboplatin", 
        TRUE ~ t
      ),
      subclone = s,
      correlation = tmp$estimate,
      p_value = tmp$p.value,
      sig = case_when(
        tmp$p.value < .1 & tmp$p.value >= .05 ~ "*", 
        tmp$p.value < .05 & tmp$p.value >= .01 ~ "**", 
        tmp$p.value < .01 ~ "***", 
        TRUE ~ ""
      )
    )
    if (is.null(plot_table)){
      plot_table <- df
    } else {
      plot_table <- rbind.data.frame(plot_table, df)
    }
  }
}
plot_table$sig_y <- plot_table$correlation + 0.05
plot_table$sig_y[plot_table$correlation < 0] <- plot_table$correlation[plot_table$correlation < 0] - 0.05
write.csv(plot_table, paste0(sub_output_dir, "Spearman_corelation_coefficient.csv"),
          row.names = FALSE)
```

```{r}
color <- c("A" = "#DC0000FF", "B" = "#B09C85FF", "C" = "#00A087FF", "D" = "#91D1C2FF", "E" = "#3C5488FF")
plot_table$treatment <- factor(plot_table$treatment, levels = c("Carboplatin", "Olaparib", "NK"))
ggplot(
  plot_table,
  aes(y = correlation, x = treatment, fill = subclone)
  ) +
  scale_fill_manual(
    values = color) + 
  geom_bar(stat="identity", position="dodge") + 
  # geom_errorbar(aes(ymin=CI_lower, ymax=CI_upper), position=position_dodge(width=.9), width=.15) + 
  geom_text(aes(y=sig_y, label=sig), position=position_dodge(width=.9)) + 
  labs(x="", y="Correlation coefficient", fill = "CNV subclone") +
  theme_classic() +
  theme(
     legend.key = element_blank(), 
     axis.text = element_text(size = 21, colour = "black"),
     axis.title = element_text(size = 21),
     legend.text = element_text(size = 21)
     )
ggsave(
  paste0(sub_output_dir, "Spearman_correlation_coefficients_color_by_subclones.pdf"),
  width = 10, height = 5)
```

# Figure S2

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "FigureS2/")

if (!dir.exists(sub_output_dir)) {
  dir.create(sub_output_dir)
}
```

## C-deleted: PCA 

```{r}
data <- readRDS("../data/Jing_code_data/merged_KURAMOCHI_samples1_scaled_new.RDS") # Shuyu update on 1.4.2022 after FindVariableFeatures and Scale, check test.R in the server
```

```{r}
## averaging cells for each group and make a PCA plot ----------------------
data$drugSens[which(data$drugSens == "resistant")] = "post-treatment"

data2 = data@assays$RNA@scale.data
data3 = data.frame(t(data2))
data3$sample = data$sample
data3$timepoint = data$time_point
data3$treatment = data$treatment
data3 = data3[which(data$drugSens %in% c("pre-sensitive", "pre-resistant","post-treatment")), ]

df = ddply(data3, c("sample","timepoint"), numcolwise(median))
tmp = ddply(data3, c("sample","timepoint"), function(x) treatment = unique(x$treatment))
df = cbind(tmp, df)

P = prcomp_irlba(t(df[,-c(1:5)]), n = 3, scale. = T)
tmp = data.frame(PC1 = P$rotation[,1], PC2 = P$rotation[,2], Treatment = df$V1, Timepoint = df$timepoint)

tmp$Treatment[tmp$Treatment == "Carbo"] <- "Carboplatin"
tmp$Treatment[tmp$Treatment == "ctrl"] <- "Control"
tmp$Treatment = factor(tmp$Treatment, levels = c("Carboplatin", "Olaparib", "NK", "Control"))
tmp$Timepoint[tmp$Timepoint == "BeforeTreatment"] <- "pre-treatment"
tmp$Timepoint[tmp$Timepoint == "AfterTreatment"] <- "post-treatment"
tmp$Timepoint = factor(tmp$Timepoint, levels = c("pre-treatment", "post-treatment"))

color = c("pre-treatment" = blue, "post-treatment" = red)
p <- ggplot(tmp, aes(x = PC1, y = PC2, color = Timepoint, shape = Treatment)) +
  geom_point(size = 4) +
  scale_color_manual(values = color) +
  theme_classic() +
  theme(
    text = element_text(size = 20),
    axis.title = element_text(size = 25),
    axis.text = element_text(color = "black", size = 20),
    legend.background = element_blank())
ggsave(paste0(sub_output_dir, "C_all_pca.pdf"), p,  width = 6.5, height = 4)
```

## B: Shannon diversity

```{r}
# library(reshape2)
# library(ggplot2)
# library(drc)
# # install.packages("devtools")
# # devtools::install_github("analyxcompany/ForceAtlas2")
# library(Seurat)
# # library(ForceAtlas2)
# # library(tidyverse)
# # library(destiny)
# # library(cccd)
# library(stringr)
# library(umap)
# library(vegan) # distance function
# library(preprocessCore) # for quantile normalization
# library(plyr)
# library(car) # scatterplot combined with barplot
# # if (!require(devtools)) install.packages("devtools")
# # devtools::install_github("yanlinlin82/ggvenn")
# library(ggvenn) # venn diagram
# library(sparseMatrixStats) # sparse matrix calculation
# library(openxlsx)
data <- readRDS("../data/Jing_code_data/merged_KURAMOCHI_samples1_scaled_new.RDS") 

# options(scipen = 999)
# memory.limit(size = 54000)
data2 = data@assays$RNA@data

sink(paste0(sub_output_dir, "shannon.txt"))
res = list()
k = 1
for (i in unique(data$sample)){
  for (j in unique(data$time_point)){
    index = intersect(grep(i, data@meta.data$prer_r_group), which(data$time_point == j))
    tmp = data.frame(table(data@meta.data$prer_r_group[index]))
    shannon = diversity(tmp$Freq)
    print(c(gsub('[0-9]+', '', i), i,j,shannon))
    
    # png(paste(paste(i, j, sep="_"),".png",sep=""), width = 3, height = 3, units = "in", res = 300, pointsize = 10)
    # barplot(height = table(factor(tmp$Freq, levels=min(tmp$Freq):max(tmp$Freq)))/length(tmp$Freq),
    #         ylab = "proportion",
    #         xlab = "values"
    #        )
    # dev.off()
    tmp$sample = i
    tmp$time_point = j
    tmp$treatment = gsub('[0-9]+', '', i)
    res[[k]] = tmp
    k = k + 1
  }
}
sink()
```

### Shannon index

```{r}
shannon = read.table(paste0(sub_output_dir, "shannon.txt"))
names(shannon) = c("V1", "Treatment","sample","time_point","value")
shannon$Treatment[shannon$Treatment == "Carbo"] <- "Carboplatin"
shannon$Treatment[shannon$Treatment == "ctrl"] <- "Control"
shannon$Treatment = factor(shannon$Treatment, levels = c("Carboplatin", "Olaparib", "NK", "Control"))
shannon$time_point[shannon$time_point == "BeforeTreatment"] <- "pre-treatment"
shannon$time_point[shannon$time_point == "AfterTreatment"] <- "post-treatment"
shannon$time_point = factor(shannon$time_point, levels = c("pre-treatment", "post-treatment"))
values1 = c(blue, red, brown, green)

p <- ggplot(shannon, aes(x = time_point, y = value)) +
  geom_boxplot() + 
  geom_point(aes(x = time_point, color = Treatment), size = 4) +
  geom_line(aes(group = sample, color = Treatment)) +
  xlab("") + 
  ylab("Shannon diversity index") +
  theme_classic() +
  theme(
    text = element_text(size = 20),
    axis.title = element_text(size = 20),
    axis.text = element_text(color = "black", size = 20),
    legend.text = element_text(size = 20),
    legend.title = element_blank(),
    # legend.position = c(0.9, 0.9),
    legend.background = element_blank()
  ) +
  scale_color_manual(values = values1)
ggsave(
  paste0(sub_output_dir, "B_shannon.pdf"),
  p,
  width = 9, height = 5
  )

# Paired t test

pre <- shannon$value[shannon$time_point == "pre-treatment"]
post <- shannon$value[shannon$time_point == "post-treatment"]
t_test <- t.test(pre, post, paired = TRUE)
t_test
```

## C: Cumulative distribution
https://rgraphs.com/how-to-create-a-cumulative-frequency-graph-in-r/

```{r}
tmp = do.call(rbind, res)
tmp$treatment[tmp$treatment == "Carbo"] <- "Carboplatin"
tmp$treatment[tmp$treatment == "ctrl"] <- "Control"
tmp$treatment = factor(tmp$treatment, levels = c("Carboplatin", "Olaparib", "NK", "Control"))
tmp$time_point[tmp$time_point == "BeforeTreatment"] <- "pre-treatment"
tmp$time_point[tmp$time_point == "AfterTreatment"] <- "post-treatment"
tmp$time_point = factor(tmp$time_point, levels = c("pre-treatment", "post-treatment"))
values1 = c(blue, red, brown, green)

fig1 = ggplot(tmp, aes(Freq, y = 1 - ..y.., color = time_point)) +
  stat_ecdf(geom = "step", size = 1) + 
  labs(
    x = "Clone size",
    y = "Cumulative proportion",
    color = "") +
  scale_y_continuous(trans = 'log10') + scale_x_log10() +
  theme_classic() +
  theme(
    text = element_text(size = 20),
    axis.title = element_text(size = 20),
    axis.text = element_text(color = "black", size = 20),
    legend.text = element_text(size = 20),
    legend.title = element_blank(),
    # legend.position=c(.7,.9)
  ) +
  scale_color_manual(values = values1)
ggsave(
  paste0(sub_output_dir, "B_cumulate_all.pdf"),
  fig1, width = 9, height = 5
)

fig2 = ggplot(
  tmp[which(tmp$time_point == "pre-treatment"), ],
  aes(Freq, y = 1 - ..y.., color = treatment)
  ) +
  stat_ecdf(geom = "step", size = 1) + 
  labs(
    x = "Clone size",
    y = "Cumulative proportion",
    color = "") +
  scale_y_continuous(trans = 'log10') +
  scale_x_log10() +
  theme_classic() +
  theme(
    text = element_text(size = 20),
    axis.title = element_text(size = 20),
    axis.text = element_text(color = "black", size = 20),
    legend.text = element_text(size = 20),
    legend.title = element_blank(),
    legend.position=c(.75,.85)
  ) +
  scale_color_manual(values = values1)
ggsave(
  paste0(sub_output_dir, "cumulate_pre_treatment.pdf"),
  fig2, width = 6, height = 5
)

fig3 = ggplot(tmp[which(tmp$time_point == "post-treatment"), ], aes(Freq, y = 1 - ..y.., color = treatment)) +
  stat_ecdf(geom = "step", size = 1) + 
  labs(
    x = "Clone size",
    y = "Cumulative proportion",
    color = "") +
  scale_y_continuous(trans = 'log10') +
  scale_x_log10() +
  theme_classic() +
  theme(
    text = element_text(size = 20),
    axis.title = element_text(size = 20),
    axis.text = element_text(color = "black", size = 20),
    legend.text = element_text(size = 20),
    legend.title = element_blank(),
    legend.position=c(.75,.85)
  ) +
  scale_color_manual(values = values1)
ggsave(
  paste0(sub_output_dir, "cumulate_post_treatment.pdf"),
  fig3, width = 6, height = 5
)
```

# Figure 3: Functional validations/interpretations

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure3B/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

## B-deleted: Box plot: gene expression for targets of tested drugs in preR-preS-R (t test)

```{r}
data <- readRDS("../data/output/KURAMOCHI_force_directed_layout/merged_KURAMOCHI_samples.RDS")
```

```{r}
drug_targets <- data@misc$gene_meta %>% 
  filter(symbol %in% c("HMGB1", "PRKDC", "MDK", "FIS1"))
```

```{r}
for (i in 1:nrow(drug_targets)) {
  plot_table <- data.frame(
  expression = data@assays$RNA@data[drug_targets$ensembl[i], ],
  drugSens = data$drugSens,
  treatment = data$treatment) %>% 
  filter(drugSens %in% c("pre-sensitive", "pre-resistant", "resistant"))
  plot_table$drugSens <- factor(plot_table$drugSens, levels = c("pre-sensitive", "pre-resistant", "resistant"))
  t <- "Olaparib"
  for (t in unique(plot_table$treatment)) {
    tmp <- plot_table[plot_table$treatment == t, ]
    means <- aggregate(expression ~ drugSens, tmp, mean)
    my_comparisons <- list(
      c("pre-sensitive", "pre-resistant"),
      c("pre-resistant", "resistant"),
      c("pre-sensitive", "resistant"))
    p <- ggplot(
      data = tmp,
      aes(x = drugSens, y = expression, fill = drugSens)
      ) +
      geom_violin(trim = FALSE) + 
      labs(
        title = paste0(t, " : ", drug_targets$symbol[i]),
        x = "",
        y = "Expression Level"
      ) +
      geom_boxplot(width = 0.1, fill = "white", outlier.shape = NA) + 
      # scale_fill_manual(values=c("#317DF7", "#CC3311")) + 
      ggpubr::stat_compare_means(
        comparisons = my_comparisons,
        method = "wilcox",
        size = 5
      ) +
      ggpubr::stat_compare_means(
        label.y = max(tmp$expression) * 1.5,
        label.y.npc = "top",
        label.x = 0.8,
        size = 5
      ) +
      geom_text(
        data = means, size = 5,
        aes(label = signif(expression, 4), y = expression),
        nudge_x = 0.25) +
      theme_classic() + 
      theme(legend.position = "none",
            plot.title = element_text(hjust = 0.5),
            axis.text.x = element_text(size = 15, face = "bold"),
            text = element_text(size = 20))
    p
    ggsave(
      paste0(
        output_dir, "Figure3B/",
        t ,"_", drug_targets$symbol[i],
        "_expression_in_preR_preS_R.png"),
      p,
      width = 6, height = 7, dpi = 300
    )
  }
}

```

## B: BRCAness/HRD survival plot

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure3B/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

Load lineage tracing data

```{r}
data <- readRDS("../data/Jing_code_data/merged_KURAMOCHI_samples1_scaled_new.RDS") # Shuyu update on 1.4.2022 after FindVariableFeatures and Scale, check test.R in the server
sister_conserve_gene <- read.csv("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_conserve_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  filter(conserve_gene) %>% 
  select(ensembl) %>% 
  unlist()
data2 = data@assays$RNA@data[rownames(data) %in% sister_conserve_gene, ]
dim(data2)
unique(data$drugSens)
unique(data$sample)
unique(data$treatment)
unique(data$time_point)
```

load TCGA data

```{r}
data_clinical_patient = read.delim("../data/Jing_code_data/data_clinical_patient.txt", skip = 4)
data_clinical_sample = read.delim("../data/Jing_code_data/data_clinical_sample.txt", skip = 4)
data_mrna = read.delim("../data/Jing_code_data/data_mrna_seq_v2_rsem.txt", check.names = F)
data_mrna <- data_mrna[apply(data_mrna, 1, function(x){
  sum(is.na(x))
}) < 300, ]

identical(data_clinical_patient$PATIENT_ID, data_clinical_sample$PATIENT_ID) # data aligned already
table(data_clinical_sample$GRADE)
high_grade = grepl("G2|G3|G4", data_clinical_sample$GRADE)

data_clinical_sample = data_clinical_sample[high_grade, ]
data_clinical_patient = data_clinical_patient[high_grade, ]

patients = names(data_mrna)[-c(1:2)] # 300 patients
time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]
table(status, exclude = NULL)
status[which(status == "")] = NA

length(which(is.na(time)==F)) # 271 has survival
length(which(is.na(status)==F)) # 272 has status
```

### TCGA BRCAness survival plot 

```{r}
tcga = read.csv("../data/Jing_code_data/TCGA_deconvoluted_HRD_DEGs.tsv", sep = "\t")
tcga[which(rownames(tcga) == "BRCA1"), ] # should be negative logFC
tcga$ensembl = data@misc$gene_meta$ensembl[match(rownames(tcga), data@misc$gene_meta$symbol)]
tcga$symbol = rownames(tcga)
length(which(is.na(tcga$ensembl)==T)) # 383 genes have no emsembl
```

#### Test for good p value

```{r}
p_threshold = c(1*10^-(2:4), 5*10^-(2:4))
fc_threshold = 0
conserve_gene_only <- FALSE

res <- mat.or.vec(length(p_threshold), 3)
colnames(res) <- c("p_threshold", "n_gene", "survival_p")

for (i in 1:length(p_threshold)){
  if (conserve_gene_only){
    tcga_test <- filter(tcga, ensembl %in% sister_conserve_gene)
  } else {
    tcga_test <- tcga
  }
  
  res[i, 1] <- p_threshold[i]
  index1 = which(tcga_test$logFC <= -fc_threshold & tcga_test$adj.P.Val <= p_threshold[i]) # down-regulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(tcga_test$logFC > fc_threshold & tcga_test$adj.P.Val <= p_threshold[i]) # up-regulated genes when BRCA1 is knocked out
  index2_symbol = tcga_test$symbol[index2]
  
  index = c(index1, index2)
  length(index)
  tcga_test = tcga_test[index, ]
  # tcga_test$symbol
  
  tcga_test = tcga_test[match(data_mrna$Hugo_Symbol, tcga_test$symbol), ]
  data_mrna2 = data_mrna[, -c(1:2)]
  
  res[i, 2] <- nrow(na.omit(tcga_test))
  
  # reverse index2 gene expressions
  data_mrna2[which(data_mrna$Hugo_Symbol %in% index2_symbol), ] = -data_mrna2[which(data_mrna$Hugo_Symbol %in% index2_symbol), ]
  
  # gene selection
  index = which(is.na(tcga_test$symbol) == F)
  res1 = c()
  for (j in 1:length(patients)){
    cat('\r', j)
    res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
    res1[j] = weighted.mean(res.tmp[index], w = abs(tcga_test$logFC[index]), na.rm = T) # rank of the BRCAness genes
    # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362$avg_log2FC[index])*(-log10(cov362$p_val_adj[index]))) # rank of the BRCAness genes
    
  }
  
  group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
  # fiveyear = which(time < 60)
  plot_data = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
  fit <- survfit(Surv(time, status) ~ group, data = plot_data)
  fit.coxph = coxph(Surv(time, status) ~ group, data = plot_data)
  s <- summary(fit.coxph)
  res[i, 3] <- s$logtest[3]
}

write.csv(
  res,
  paste0(sub_output_dir, "BRCA_survival_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

#### Survival plot with selected parameters

```{r}
conserve_gene_only <- FALSE

if (conserve_gene_only) {
  p_threshold = 0.01
  fc_threshold = 0
} else {
  p_threshold = 0.01
  fc_threshold = 0
}

if (conserve_gene_only){
  tcga_test <- filter(tcga, ensembl %in% sister_conserve_gene)
} else {
  tcga_test <- tcga
}

index1 = which(tcga_test$logFC <= -fc_threshold & tcga_test$adj.P.Val <= p_threshold) # down-regulated genes when BRCA1 is knocked out (BRCA-)
index2 = which(tcga_test$logFC > fc_threshold & tcga_test$adj.P.Val <= p_threshold) # up-regulated genes when BRCA1 is knocked out
index2_symbol = tcga_test$symbol[index2]

index = c(index1, index2)
length(index)
tcga_test = tcga_test[index, ]
# tcga_test$symbol

tcga_test = tcga_test[match(data_mrna$Hugo_Symbol, tcga_test$symbol), ]
data_mrna2 = data_mrna[, -c(1:2)]

# reverse index2 gene expressions
data_mrna2[which(data_mrna$Hugo_Symbol %in% index2_symbol), ] = -data_mrna2[which(data_mrna$Hugo_Symbol %in% index2_symbol), ]

# gene selection
index = which(is.na(tcga_test$symbol) == F)
res1 = c()
for (j in 1:length(patients)){
  cat('\r', j)
  res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
  res1[j] = weighted.mean(res.tmp[index], w = abs(tcga_test$logFC[index]), na.rm = T) # rank of the BRCAness genes
  # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362$avg_log2FC[index])*(-log10(cov362$p_val_adj[index]))) # rank of the BRCAness genes
  
}

group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
# fiveyear = which(time < 60)
plot_data = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
fit <- survfit(Surv(time, status) ~ group, data = plot_data)
fit.coxph = coxph(Surv(time, status) ~ group, data = plot_data)

summary(fit.coxph)
  
dim(tcga_test)
concordance(fit.coxph)$concordance

pdf(paste0(sub_output_dir, ifelse(conserve_gene_only, "BRCA_TCGA_OS_conserve_gene.pdf", "BRCA_TCGA_OS.pdf")), width = 4, height = 5, pointsize = 20)
ggsurvplot(
  fit,
  data = plot_data,
  palette = c(red, blue),
  risk.table = T,
  legend.title = "BRCAness (TCGA)",# - undeconvoluted",
  legend.labs = c("high","low"),
  xlab = "Time (months)", ylab = "Overall survival",
  ggtheme = theme_survminer(
    base_size = 12, font.main = 12, font.x = 12, font.y = 12,
    font.caption = 12, font.legend = 12, font.tickslab = 12, font.submain = 12
  ),
  pval = TRUE,
  pval.coord = c(100, 0.75)
)+
  # annotate("text", x = 150, y =0.75, label = paste0("P = ", round(s$logtest[3], 2))) +
  guides(colour = guide_legend(nrow = 2))
dev.off()
```

### COV362 survival plot

#### Test for good p value

```{r}
cov362 = read.csv("../data/Jing_code_data/deg_COV362.csv") # JUN data

# min_p = 0
# max_p = 1 * 10 ^ (- 12) # 0.000000000001
# p_threshold = lseq(min_p, max_p, 20)
# p_threshold = p_threshold[6] # the max correlation
# p_threshold = 1 * 10 ^ (- 12) # 0.000000000001

p_threshold = c(1*10^-(2:14), 5*10^-(2:14))
fc_threshold = 0
conserve_gene_only <- TRUE

res <- mat.or.vec(length(p_threshold), 3)
colnames(res) <- c("p_threshold", "n_gene", "survival_p")

for (i in 1:length(p_threshold)){
  if (conserve_gene_only){
    cov362_test <- filter(cov362, ensembl %in% sister_conserve_gene)
  } else {
    cov362_test <- cov362
  }
  res[i, 1] <- p_threshold[i]
  
  index1 = which(cov362_test$avg_log2FC <= -fc_threshold & cov362_test$p_val_adj <= p_threshold[i]) # downregulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(cov362_test$avg_log2FC > fc_threshold & cov362_test$p_val_adj <= p_threshold[i]) # upregulated genes when BRCA1 is knocked out
  index2_symbol = cov362_test$symbol[index2]
  
  index = c(index1, index2)
  length(index)
  cov362_test = cov362_test[index, ]
  # cov362_test$symbol
  
  cov362_test = cov362_test[match(data_mrna$Hugo_Symbol, cov362_test$symbol), ]
  data_mrna2 = data_mrna[, -c(1:2)]
  
  res[i, 2] <- nrow(na.omit(cov362_test))
  
  # reverse index2 gene expressions
  data_mrna2[which(data_mrna$Hugo_Symbol %in% index2_symbol), ] = -data_mrna2[which(data_mrna$Hugo_Symbol %in% index2_symbol), ]
  
  # gene selection
  index = which(is.na(cov362_test$symbol) == F)
  
  res1 = c()
  for (j in 1:length(patients)){
    cat('\r', j)
    res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
    res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])) # rank of the BRCAness genes
    # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])*(-log10(cov362_test$p_val_adj[index]))) # rank of the BRCAness genes
    
  }
  time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
  status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]
  
  group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
  data_plot = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
  fit <- survfit(Surv(time, status) ~ group, data = data_plot)
  fit.coxph = coxph(Surv(time, status) ~ group, data = data_plot)
  s <- summary(fit.coxph)
  res[i, 3] <- s$logtest[3]
}

write.csv(
  res,
  paste0(sub_output_dir, "BRCA_survival_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

#### Survival plot with selected parameters

```{r}
conserve_gene_only <- TRUE

if (conserve_gene_only) {
  p_threshold = 1*10^(-11)
  fc_threshold = 0
} else {
  p_threshold = 1*10^(-13)
  fc_threshold = 0
}

if (conserve_gene_only){
  cov362_test <- filter(cov362, ensembl %in% sister_conserve_gene)
} else {
  cov362_test <- cov362
}

index1 = which(cov362_test$avg_log2FC <= -fc_threshold & cov362_test$p_val_adj <= p_threshold) # downregulated genes when BRCA1 is knocked out (BRCA-)
index2 = which(cov362_test$avg_log2FC > fc_threshold & cov362_test$p_val_adj <= p_threshold) # upregulated genes when BRCA1 is knocked out
index2_symbol = cov362_test$symbol[index2]

index = c(index1, index2)
length(index)
cov362_test = cov362_test[index, ]

# cov362_test$symbol

cov362_test = cov362_test[match(data_mrna$Hugo_Symbol, cov362_test$symbol), ]
data_mrna2 = data_mrna[, -c(1:2)]

# reverse index2 gene expressions
data_mrna2[which(data_mrna$Hugo_Symbol %in% index2_symbol), ] = -data_mrna2[which(data_mrna$Hugo_Symbol %in% index2_symbol), ]

# gene selection
index = which(is.na(cov362_test$symbol) == F)

res1 = c()
for (j in 1:length(patients)){
  cat('\r', j)
  res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
  res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])) # rank of the BRCAness genes
  # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])*(-log10(cov362_test$p_val_adj[index]))) # rank of the BRCAness genes
  
}
time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]

group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
data_plot = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
fit <- survfit(Surv(time, status) ~ group, data = data_plot)
fit.coxph = coxph(Surv(time, status) ~ group, data = data_plot)
summary(fit.coxph)
  
# names(test) <- 5*10^(-seq(2, 12))
concordance(fit.coxph)$concordance

pdf(paste0(sub_output_dir, ifelse(conserve_gene_only, "BRCA_COV_OS_conserve_gene.pdf", "BRCA_COV_OS.pdf")), width = 4, height = 5, pointsize = 20)
fig <- ggsurvplot(
  fit,
  data = data_plot,
  palette = c(red, blue),
  risk.table = T,
  legend.title = "BRCAness (COV362)", # - undeconvoluted",
  legend.labs = c("high", "low"),
  xlab = "Time (months)", ylab = "Overall survival",
  ggtheme = theme_survminer(
    base_size = 12, font.main = 12, font.x = 12, font.y = 12, font.caption = 12,
    font.legend = 12, font.tickslab = 12, font.submain = 12),
  pval = TRUE,
  pval.coord = c(100, 0.75)
  ) +
  guides(colour = guide_legend(nrow = 2))
fig
dev.off()
```

## B_deconvoluted: BRCAness/HRD survival plot

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure3B_deconvolutated/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

Load lineage tracing data

```{r}
data <- readRDS("../data/Jing_code_data/merged_KURAMOCHI_samples1_scaled_new.RDS") # Shuyu update on 1.4.2022 after FindVariableFeatures and Scale, check test.R in the server
sister_conserve_gene <- read.csv("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_conserve_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  filter(conserve_gene) %>% 
  select(ensembl) %>% 
  unlist()
data2 = data@assays$RNA@data[rownames(data) %in% sister_conserve_gene, ]
dim(data2)
unique(data$drugSens)
unique(data$sample)
unique(data$treatment)
unique(data$time_point)
```

load TCGA data

```{r}
data_clinical_patient = read.delim("../data/Jing_code_data/data_clinical_patient.txt", skip = 4)
data_clinical_sample = read.delim("../data/Jing_code_data/data_clinical_sample.txt", skip = 4)
data_mrna = readRDS("../data/Jing_code_data/EOC_deconv_TCGA_expr_mat.RDS")
colnames(data_mrna) = gsub("\\.","\\-", colnames(data_mrna))
colnames(data_mrna) = substr(colnames(data_mrna), 1, 15)
data_mrna <- data_mrna[apply(data_mrna, 1, function(x){
  sum(is.na(x))
}) < 300, ]

identical(data_clinical_patient$PATIENT_ID, data_clinical_sample$PATIENT_ID) # data aligned already
table(data_clinical_sample$GRADE)
high_grade = grepl("G2|G3|G4", data_clinical_sample$GRADE)

data_clinical_sample = data_clinical_sample[high_grade, ]
data_clinical_patient = data_clinical_patient[high_grade, ]

patients = names(data_mrna) # 307 patients
time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]
table(status, exclude = NULL)
status[which(status == "")] = NA

length(which(is.na(time)==F)) # 270 has survival
length(which(is.na(status)==F)) # 271 has status
```

### TCGA BRCAness survival plot 

```{r}
tcga = read.csv("../data/Jing_code_data/TCGA_deconvoluted_HRD_DEGs.tsv", sep = "\t")
tcga[which(rownames(tcga) == "BRCA1"), ] # should be negative logFC
tcga$ensembl = data@misc$gene_meta$ensembl[match(rownames(tcga), data@misc$gene_meta$symbol)]
tcga$symbol = rownames(tcga)
length(which(is.na(tcga$ensembl)==T)) # 383 genes have no emsembl
```

#### Test for good p value

```{r}
p_threshold = c(1*10^-(2:4), 5*10^-(2:4))
fc_threshold = 0
conserve_gene_only <- TRUE

res <- mat.or.vec(length(p_threshold), 3)
colnames(res) <- c("p_threshold", "n_gene", "survival_p")

for (i in 1:length(p_threshold)){
  data_mrna2 <- data_mrna
  if (conserve_gene_only){
    tcga_test <- filter(tcga, ensembl %in% sister_conserve_gene)
  } else {
    tcga_test <- tcga
  }
  
  res[i, 1] <- p_threshold[i]
  index1 = which(tcga_test$logFC <= -fc_threshold & tcga_test$adj.P.Val <= p_threshold[i]) # down-regulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(tcga_test$logFC > fc_threshold & tcga_test$adj.P.Val <= p_threshold[i]) # up-regulated genes when BRCA1 is knocked out
  index2_symbol = tcga_test$symbol[index2]
  
  index = c(index1, index2)
  length(index)
  tcga_test = tcga_test[index, ]
  # tcga_test$symbol
  
  tcga_test = tcga_test[match(rownames(data_mrna2), tcga_test$symbol), ]
  
  res[i, 2] <- nrow(na.omit(tcga_test))
  
  # reverse index2 gene expressions
  data_mrna2[which(rownames(rownames(data_mrna2)) %in% index2_symbol), ] = -data_mrna2[which(rownames(rownames(data_mrna2)) %in% index2_symbol), ]
  
  # gene selection
  index = which(is.na(tcga_test$symbol) == F)
  res1 = c()
  for (j in 1:length(patients)){
    cat('\r', j)
    res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
    res1[j] = weighted.mean(res.tmp[index], w = abs(tcga_test$logFC[index]), na.rm = T) # rank of the BRCAness genes
    # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362$avg_log2FC[index])*(-log10(cov362$p_val_adj[index]))) # rank of the BRCAness genes
    
  }
  
  group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
  # fiveyear = which(time < 60)
  plot_data = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
  fit <- survfit(Surv(time, status) ~ group, data = plot_data)
  fit.coxph = coxph(Surv(time, status) ~ group, data = plot_data)
  s <- summary(fit.coxph)
  res[i, 3] <- s$logtest[3]
}

write.csv(
  res,
  paste0(sub_output_dir, "BRCA_survival_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

#### Survival plot with selected parameters

```{r}
conserve_gene_only <- FALSE

if (conserve_gene_only) {
  selected_p_threshold = 0.005
  selected_fc_threshold = 0
} else {
  selected_p_threshold = 0.005
  selected_fc_threshold = 0
}

if (conserve_gene_only){
  tcga_test <- filter(tcga, ensembl %in% sister_conserve_gene)
} else {
  tcga_test <- tcga
}

data_mrna2 <- data_mrna
index1 = which(tcga_test$logFC <= -selected_fc_threshold & tcga_test$adj.P.Val <= selected_p_threshold) # down-regulated genes when BRCA1 is knocked out (BRCA-)
index2 = which(tcga_test$logFC > selected_fc_threshold & tcga_test$adj.P.Val <= selected_p_threshold) # up-regulated genes when BRCA1 is knocked out
index2_symbol = tcga_test$symbol[index2]

index = c(index1, index2)
length(index)
tcga_test = tcga_test[index, ]
# tcga_test$symbol

tcga_test = tcga_test[match(rownames(data_mrna2), tcga_test$symbol), ]

# reverse index2 gene expressions
data_mrna2[which(rownames(rownames(data_mrna2)) %in% index2_symbol), ] = -data_mrna2[which(rownames(rownames(data_mrna2)) %in% index2_symbol), ]

# gene selection
index = which(is.na(tcga_test$symbol) == F)
res1 = c()
for (j in 1:length(patients)){
  cat('\r', j)
  res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
  res1[j] = weighted.mean(res.tmp[index], w = abs(tcga_test$logFC[index]), na.rm = T) # rank of the BRCAness genes
  # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362$avg_log2FC[index])*(-log10(cov362$p_val_adj[index]))) # rank of the BRCAness genes
  
}

group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
# fiveyear = which(time < 60)
plot_data = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
fit <- survfit(Surv(time, status) ~ group, data = plot_data)
fit.coxph = coxph(Surv(time, status) ~ group, data = plot_data)

s <- summary(fit.coxph)
  
dim(tcga_test)
concordance(fit.coxph)$concordance

pdf(paste0(sub_output_dir, "threshold_", selected_p_threshold, "_", ifelse(conserve_gene_only, "BRCA_TCGA_OS_conserve_gene.pdf", "BRCA_TCGA_OS.pdf")), width = 4, height = 5, pointsize = 20)
ggsurvplot(
  fit,
  data = plot_data,
  palette = c(red, blue),
  risk.table = T,
  legend.title = "BRCAness (TCGA)",# - undeconvoluted",
  legend.labs = c("high","low"),
  xlab = "Time (months)", ylab = "Overall survival",
  ggtheme = theme_survminer(
    base_size = 12,
    font.main = c(12, "plain", "black"),
    font.x = c(12, "plain", "black"),
    font.y = c(12, "plain", "black"),
    font.caption = c(12, "plain", "black"),
    font.legend = c(12, "plain", "black"),
    font.tickslab = c(12, "plain", "black"),
    font.submain = c(12, "plain", "black")
  ),
  pval = signif(s$logtest[3], 2),
  pval.coord = c(100, 0.75)
)+
  # annotate("text", x = 150, y =0.75, label = paste0("P = ", round(s$logtest[3], 2))) +
  guides(colour = guide_legend(nrow = 2))
dev.off()
```

### COV362 survival plot

#### Test for good p value

```{r}
cov362 = read.csv("../data/Jing_code_data/deg_COV362.csv") # JUN data

# min_p = 0
# max_p = 1 * 10 ^ (- 12) # 0.000000000001
# p_threshold = lseq(min_p, max_p, 20)
# p_threshold = p_threshold[6] # the max correlation
# p_threshold = 1 * 10 ^ (- 12) # 0.000000000001

p_threshold = c(1*10^-(2:14), 5*10^-(2:14))
fc_threshold = 0
conserve_gene_only <- FALSE

res <- mat.or.vec(length(p_threshold), 3)
colnames(res) <- c("p_threshold", "n_gene", "survival_p")

for (i in 1:length(p_threshold)){
  data_mrna2 <- data_mrna
  if (conserve_gene_only){
    cov362_test <- filter(cov362, ensembl %in% sister_conserve_gene)
  } else {
    cov362_test <- cov362
  }
  res[i, 1] <- p_threshold[i]
  
  index1 = which(cov362_test$avg_log2FC <= -fc_threshold & cov362_test$p_val_adj <= p_threshold[i]) # downregulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(cov362_test$avg_log2FC > fc_threshold & cov362_test$p_val_adj <= p_threshold[i]) # upregulated genes when BRCA1 is knocked out
  index2_symbol = cov362_test$symbol[index2]
  
  index = c(index1, index2)
  length(index)
  cov362_test = cov362_test[index, ]
  # cov362_test$symbol
  
  cov362_test = cov362_test[match(rownames(data_mrna2), cov362_test$symbol), ]
  
  res[i, 2] <- nrow(na.omit(cov362_test))
  
  # reverse index2 gene expressions
  data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ] = -data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ]
  
  # gene selection
  index = which(is.na(cov362_test$symbol) == F)
  
  res1 = c()
  for (j in 1:length(patients)){
    cat('\r', j)
    res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
    res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])) # rank of the BRCAness genes
    # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])*(-log10(cov362_test$p_val_adj[index]))) # rank of the BRCAness genes
    
  }
  time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
  status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]
  
  group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
  data_plot = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
  set.seed(123)
  fit <- survfit(Surv(time, status) ~ group, data = data_plot)
  fit.coxph = coxph(Surv(time, status) ~ group, data = data_plot)
  s <- summary(fit.coxph)
  res[i, 3] <- s$logtest[3]
}

write.csv(
  res,
  paste0(sub_output_dir, "BRCA_survival_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

#### Survival plot with selected parameters

```{r}
conserve_gene_only <- FALSE

if (conserve_gene_only) {
  p_threshold = 1*10^(-9)
  fc_threshold = 0
} else {
  p_threshold = 1*10^(-9)
  fc_threshold = 0
}
data_mrna2 <- data_mrna
  
if (conserve_gene_only){
  cov362_test <- filter(cov362, ensembl %in% sister_conserve_gene)
} else {
  cov362_test <- cov362
}

index1 = which(cov362_test$avg_log2FC <= -fc_threshold & cov362_test$p_val_adj <= p_threshold) # downregulated genes when BRCA1 is knocked out (BRCA-)
index2 = which(cov362_test$avg_log2FC > fc_threshold & cov362_test$p_val_adj <= p_threshold) # upregulated genes when BRCA1 is knocked out
index2_symbol = cov362_test$symbol[index2]

index = c(index1, index2)
length(index)
cov362_test = cov362_test[index, ]
# cov362_test$symbol

cov362_test = cov362_test[match(rownames(data_mrna2), cov362_test$symbol), ]

# reverse index2 gene expressions
data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ] = -data_mrna2[which(rownames(data_mrna2) %in% index2_symbol), ]

# gene selection
index = which(is.na(cov362_test$symbol) == F)

res1 = c()
for (j in 1:length(patients)){
  cat('\r', j)
  res.tmp = rank(data_mrna2[, j], na.last = "keep") # min = r1, max = rmax
  res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])) # rank of the BRCAness genes
  # res1[j] = weighted.mean(res.tmp[index], w = abs(cov362_test$avg_log2FC[index])*(-log10(cov362_test$p_val_adj[index]))) # rank of the BRCAness genes
  
}
time = data_clinical_patient$OS_MONTHS[match(patients, data_clinical_sample$SAMPLE_ID)]
status = data_clinical_patient$OS_STATUS[match(patients, data_clinical_sample$SAMPLE_ID)]

group = ifelse(res1 > quantile(res1, 0.5), 1, 0) # lower rank survives better
data_plot = data.frame(time = time, status = as.numeric(as.factor(status)), group = group)
set.seed(123)
fit <- survfit(Surv(time, status) ~ group, data = data_plot)
fit.coxph = coxph(Surv(time, status) ~ group, data = data_plot)
s <- summary(fit.coxph)
# res[i, 3] <- s$logtest[3]
  
pdf(paste0(sub_output_dir,"threshold_", p_threshold, ifelse(conserve_gene_only, "BRCA_COV_OS_conserve_gene.pdf", "BRCA_COV_OS.pdf")), width = 4, height = 5, pointsize = 20)
fig <- ggsurvplot(
  fit,
  data = data_plot,
  palette = c(red, blue),
  risk.table = T,
  legend.title = "BRCAness (COV362)", # - undeconvoluted",
  legend.labs = c("high", "low"),
  xlab = "Time (months)", ylab = "Overall survival",
  ggtheme = theme_survminer(
    base_size = 12,
    font.main = c(12, "plain", "black"),
    font.x = c(12, "plain", "black"),
    font.y = c(12, "plain", "black"),
    font.caption = c(12, "plain", "black"),
    font.legend = c(12, "plain", "black"),
    font.tickslab = c(12, "plain", "black"),
    font.submain = c(12, "plain", "black")
  ),
  pval = signif(s$logtest[3], 2),
  pval.coord = c(100, 0.75)
  ) +
  guides(colour = guide_legend(nrow = 2))
fig
dev.off()
```

## C: BRCAness/HRD box plot

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure3C/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}

lseq <- function(from=1, to=100000, length.out=6) {
  # logarithmic spaced sequence
  # blatantly stolen from library("emdbook"), because need only this
  exp(seq(log(from), log(to), length.out = length.out))
}

conserve_gene_only <- TRUE
```

### load data

```{r}
data <- readRDS("../data/Jing_code_data/merged_KURAMOCHI_samples1_scaled_new.RDS") # Shuyu update on 1.4.2022 after FindVariableFeatures and Scale, check test.R in the server
sister_conserve_gene <- read.csv("../data/output_from_puhti/KURAMOCHI_before_treatment/sister_conserve_gene/sum_exp_sister_conserve_gene_all_bt_sample_twin.csv") %>% 
  filter(conserve_gene) %>% 
  select(ensembl) %>% 
  unlist()
```

### TCGA signature

```{r}
tcga = read.csv("../data/Jing_code_data/TCGA_deconvoluted_HRD_DEGs.tsv", sep = "\t")
tcga[which(rownames(tcga) == "BRCA1"), ] # should be negative logFC
tcga$ensembl = data@misc$gene_meta$ensembl[match(rownames(tcga), data@misc$gene_meta$symbol)]
length(which(is.na(tcga$ensembl)==T)) # 383 genes have no emsembl

summary(tcga$adj.P.Val)

treatments <- c("Olaparib", "Carboplatin", "NK")
```

#### Test for good P threshold

##### preS vs preR

```{r}
group1 = "pre-resistant"
group2 = "pre-sensitive"
# min_p = 0.0002
min_p = 0.005 # around 200 sister conserved BRCAness genes 
# max_p = 0.1
max_p = 0.05

p_threshold = c(lseq(min_p, max_p, 20), 1*10^-(2:4), 5*10^-(2:4))
# p_threshold = p_threshold[4] #

res = mat.or.vec(length(p_threshold), 6)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK", "Control")

if (conserve_gene_only){
  data2 = data@assays$RNA@data[rownames(data) %in% sister_conserve_gene, ]
} else {
  data2 = data@assays$RNA@data
}

for(i in 1:length(p_threshold)){
  if (conserve_gene_only){
    tcga_test <- filter(tcga, ensembl %in% sister_conserve_gene)
  } else {
    tcga_test <- tcga
  }
  
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  index1 = which(tcga_test$logFC < 0 & tcga_test$adj.P.Val <= p_threshold[i]) # downregulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(tcga_test$logFC > 0 & tcga_test$adj.P.Val <= p_threshold[i]) # upregulated genes when BRCA1 is knocked out
  
  # whether these genes show the similar trend in the treatment?
  match1 = which(tcga_test$ensembl[index1] %in% rownames(data2))
  match2 = which(tcga_test$ensembl[index2] %in% rownames(data2))
  match = c(index1[match1], index2[match2])
  
  signature = data.frame(FC = -tcga_test$logFC[match], name = rownames(tcga_test)[match])
  anchor = order(signature$FC)
  signature = signature[anchor,]
  
  signature_all = signature
  
  # # adjust on control
  # exp_ctrl_group1 <- data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == "Control")]
  # mean_ctrl_group1 <- log2(x = rowMeans(x = expm1(exp_ctrl_group1)) + 1)
  # exp_ctrl_group2 <- data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == "Control")]
  # mean_ctrl_group2 <- log2(x = rowMeans(x = expm1(exp_ctrl_group2)) + 1)
  # # tmp1 = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == "Control")]) - 
  # #   rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == "Control")])
  # tmp1 <- mean_ctrl_group1 - mean_ctrl_group2
  # 
  # res[i, 6] = cor(tmp1, tcga_test$logFC[match], method = "spearman") # should be positive
  # 
  # for (treatment in treatments){
  #   exp_group1 <- data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]
  #   mean_group1 <- log2(x = rowMeans(x = expm1(exp_group1)) + 1)
  #   exp_group2 <- data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)]
  #   mean_group2 <- log2(x = rowMeans(x = expm1(exp_group2)) + 1)
  #   # tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
  #   #   rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
  #   tmp <- mean_group1 - mean_group2
  #   tmp <- tmp - tmp1
  #   res[i, treatment] = cor(tmp, -tcga_test$logFC[match], method = "spearman") # should be positive
  #   
  #   signature = data.frame(FC = tmp, name = rownames(tcga_test)[match])
  #   signature = signature[anchor,]
  #   
  #   signature_all = cbind(signature_all, signature)
  # }

  # mean log2FC from control for adjustment
  tmp1 = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == "Control")]) -
    rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == "Control")])

  # Olaparib
  treatment = "Olaparib"
  tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) -
    rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
  tmp = tmp - tmp1
  res[i, 3] = cor(tmp, -tcga_test$logFC[match], method = "spearman") # should be positive

  signature = data.frame(FC = tmp, name = rownames(tcga_test)[match])
  signature = signature[anchor,]

  signature_all = cbind(signature_all, signature)

  treatment = "Carboplatin"
  tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) -
    rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
  tmp = tmp - tmp1
  res[i, 4] = cor(tmp, -tcga_test$logFC[match], method = "spearman") # should be positive

  signature = data.frame(FC = tmp, name = rownames(tcga_test)[match])
  signature = signature[anchor,]

  signature_all = cbind(signature_all, signature)

  treatment = "NK"
  tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) -
    rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
  tmp = tmp - tmp1
  res[i, 5] = cor(tmp, -tcga_test$logFC[match], method = "spearman") # should be positive

  signature = data.frame(FC = tmp, name = rownames(tcga_test)[match])
  signature = signature[anchor,]

  signature_all = cbind(signature_all, signature)

  treatment = "Control"
  tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) -
    rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
  res[i, 6] = cor(tmp, -tcga_test$logFC[match], method = "spearman") # should be positive

  signature = data.frame(FC = tmp, name = rownames(tcga_test)[match])
  signature = signature[anchor,]

  signature_all = cbind(signature_all, signature)
  res[i, 2] <- nrow(signature_all)
  print(res[i,])
}
write.csv(
  res,
  paste0(
    sub_output_dir,
    "BRCA_cor_tcga",
    ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")
  ),
  row.names = FALSE
)
```

##### BT vs AT

```{r}
# min_p = 0.0002
min_p = 0.005 # around 200 sister conserved BRCAness genes 
# max_p = 0.1
max_p = 0.05 # around 200 sister conserved BRCAness genes 

p_threshold = c(lseq(min_p, max_p, 20), 1*10^-(2:4), 5*10^-(2:4))
# p_threshold = p_threshold[4] #

res = mat.or.vec(length(p_threshold), 6)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK", "Control")

group1 = "BeforeTreatment"
group2 = "AfterTreatment"

if (conserve_gene_only){
  data2 = data@assays$RNA@data[rownames(data) %in% sister_conserve_gene, ]
} else {
  data2 = data@assays$RNA@data
}

for(i in 1:length(p_threshold)){
  if (conserve_gene_only){
    tcga_test <- filter(tcga, ensembl %in% sister_conserve_gene)
  } else {
    tcga_test <- tcga
  }
  
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  
  index1 = which(tcga_test$logFC < 0 & tcga_test$adj.P.Val <= p_threshold[i]) # downregulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(tcga_test$logFC > 0 & tcga_test$adj.P.Val <= p_threshold[i]) # upregulated genes when BRCA1 is knocked out
  
  # whether these genes show the similar trend in the treatment?
  match1 = which(tcga_test$ensembl[index1] %in% rownames(data2))
  match2 = which(tcga_test$ensembl[index2] %in% rownames(data2))
  match = c(index1[match1], index2[match2])
  res[i, 2] <- length(match)
  # adjust on control
  # exp_ctrl_group1 <- data2[tcga_test$ensembl[match], which(data$time_point == group1 & data$treatment == "Control")]
  # mean_ctrl_group1 <- log2(x = rowMeans(x = expm1(exp_ctrl_group1)) + 1)
  # exp_ctrl_group2 <- data2[tcga_test$ensembl[match], which(data$time_point == group2 & data$treatment == "Control")]
  # mean_ctrl_group2 <- log2(x = rowMeans(x = expm1(exp_ctrl_group2)) + 1)
    # tmp1 <- mean_ctrl_group1 - mean_ctrl_group2
  tmp1 = rowMeans2(data2[tcga_test$ensembl[match], which(data$time_point == group1 & data$treatment == "Control")]) -
    rowMeans2(data2[tcga_test$ensembl[match], which(data$time_point == group2 & data$treatment == "Control")])
  
  res[i, 6] = cor(tmp1, tcga_test$logFC[match], method = "spearman") # should be positive
  
  for (treatment in treatments){
    # exp_group1 <- data2[tcga_test$ensembl[match], which(data$time_point == group1 & data$treatment == treatment)]
    # mean_group1 <- log2(x = rowMeans(x = expm1(exp_group1)) + 1)
    # exp_group2 <- data2[tcga_test$ensembl[match], which(data$time_point == group2 & data$treatment == treatment)]
    # mean_group2 <- log2(x = rowMeans(x = expm1(exp_group2)) + 1)
    # tmp <- mean_group1 - mean_group2
    tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$time_point == group1 & data$treatment == treatment)]) -
      rowMeans2(data2[tcga_test$ensembl[match], which(data$time_point == group2 & data$treatment == treatment)])
    tmp <- tmp - tmp1
    res[i, treatment] = cor(tmp, tcga_test$logFC[match], method = "spearman") # should be positive
  }
  
  print(res[i,])
}

write.csv(
  res,
  paste0(
    sub_output_dir,
    "BRCA_cor_tcga",
    ifelse(conserve_gene_only, "_conserve_gene_atbt.csv", "_atbt.csv")
  ),
  row.names = FALSE
)
```

##### (PreS + PreR) vs PosT

```{r}
# min_p = 0.0002
min_p = 0.005 # around 200 sister conserved BRCAness genes 
# max_p = 0.1
max_p = 0.05 # around 200 sister conserved BRCAness genes 

p_threshold = c(lseq(min_p, max_p, 20), 1*10^-(2:4), 5*10^-(2:4))
# p_threshold = p_threshold[4] #

res = mat.or.vec(length(p_threshold), 6)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK", "Control")
group1 = c("pre-sensitive", "pre-resistant")
group2 = "resistant"

if (conserve_gene_only){
  data2 = data@assays$RNA@data[rownames(data) %in% sister_conserve_gene, ]
} else {
  data2 = data@assays$RNA@data
}

for(i in 1:length(p_threshold)){
  if (conserve_gene_only){
    tcga_test <- filter(tcga, ensembl %in% sister_conserve_gene)
  } else {
    tcga_test <- tcga
  }
  
  cat('\r', i)
  
  res[i, 1] <- p_threshold[i]
  
  index1 = which(tcga_test$logFC < 0 & tcga_test$adj.P.Val <= p_threshold[i]) # downregulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(tcga_test$logFC > 0 & tcga_test$adj.P.Val <= p_threshold[i]) # upregulated genes when BRCA1 is knocked out
  
  # whether these genes show the similar trend in the treatment?
  match1 = which(tcga_test$ensembl[index1] %in% rownames(data2))
  match2 = which(tcga_test$ensembl[index2] %in% rownames(data2))
  match = c(index1[match1], index2[match2])
  
  res[i, 2] <- length(match)
  
  # Olaparib
  treatment = "Olaparib"
  tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens %in% group1 & data$treatment == treatment)]) - 
    rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens %in% group2 & data$treatment == treatment)])
  # adjust on control
  tmp1 = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens %in% group1 & data$treatment == "Control")]) - 
    rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens %in% group2 & data$treatment == "Control")])
  tmp = tmp - tmp1
  
  res[i, 3] = cor(tmp, tcga_test$logFC[match], method = "spearman") # should be positive
  
  treatment = "Carboplatin"
  tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens %in% group1 & data$treatment == treatment)]) - 
    rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens %in% group2 & data$treatment == treatment)])
  tmp = tmp - tmp1
  res[i, 4] = cor(tmp, tcga_test$logFC[match], method = "spearman") # should be positive
  
  treatment = "NK"
  tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens %in% group1 & data$treatment == treatment)]) - 
    rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens %in% group2 & data$treatment == treatment)])
  tmp = tmp - tmp1
  res[i, 5] = cor(tmp, tcga_test$logFC[match], method = "spearman") # should be positive
  
  treatment = "Control"
  tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens %in% group1 & data$treatment == treatment)]) - 
    rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens %in% group2 & data$treatment == treatment)])
  res[i, 6] = cor(tmp, tcga_test$logFC[match], method = "spearman") # should be positive
  print(res[i,])
}

write.csv(
  res,
  paste0(
    sub_output_dir,
    "BRCA_cor_tcga",
    ifelse(
      conserve_gene_only,
      "_conserve_gene_preRpreSposT.csv",
      "_preRpreSposT.csv"
    )
  ),
  row.names = FALSE
)
```

#### Bar plot

```{r}
selected_p_threshold <- 0.05

if (conserve_gene_only) {
  data2 = data@assays$RNA@data[rownames(data) %in% sister_conserve_gene, ]
  tcga_test <- filter(tcga, ensembl %in% sister_conserve_gene)
} else {
  data2 = data@assays$RNA@data
  tcga_test <- tcga
}

index1 = which(tcga_test$logFC < 0 & tcga_test$adj.P.Val <= selected_p_threshold) # downregulated genes when BRCA1 is knocked out (BRCA-)
index2 = which(tcga_test$logFC > 0 & tcga_test$adj.P.Val <= selected_p_threshold) # upregulated genes when BRCA1 is knocked out
  
# whether these genes show the similar trend in the treatment?
match1 = which(tcga_test$ensembl[index1] %in% rownames(data2))
match2 = which(tcga_test$ensembl[index2] %in% rownames(data2))
match = c(index1[match1], index2[match2])

signature = data.frame(FC = tcga_test$logFC[match], name = rownames(tcga_test)[match])
anchor = order(signature$FC)
signature = signature[anchor,]

# BRCAness
pdf(paste0(sub_output_dir, "BRCA_signature_tcga_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(signature$FC > 0, red, blue)
fig = barplot(
  signature$FC,
  names.arg = signature$name,
  las = 2,
  main = "BRCAness signature (TCGA)",
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()
group1 <- "pre-sensitive"
group2 <- "pre-resistant"

# mean log2FC from control for adjustment
tmp1 = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == "Control")]) - 
  rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == "Control")])

# Olaparib
treatment = "Olaparib"
tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
  rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
tmp = tmp - tmp1
r = cor(tmp, tcga_test$logFC[match], method = "spearman")
# plot(tmp, -cov362$avg_log2FC[match])
signature = data.frame(FC = tmp, name = rownames(tcga_test)[match])
signature = signature[anchor,]

pdf(paste0(sub_output_dir,"Olaparib_signature_tcga_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(signature$FC > 0, red, blue)
fig = barplot(
  signature$FC,
  names.arg = signature$name,
  las = 2,
  main = paste("Olaparib (TCGA), cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()

treatment = "Carboplatin"
tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
  rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
tmp = tmp - tmp1
# plot(tmp, -cov362$avg_log2FC[match])
r = cor(tmp, tcga_test$logFC[match], method = "spearman") # should be positive

signature = data.frame(FC = tmp, name = rownames(tcga_test)[match])
signature = signature[anchor,]

pdf(paste0(sub_output_dir, "Carboplatin_signature_TCGA_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(signature$FC > 0, red, blue)
fig = barplot(
  signature$FC,
  names.arg = signature$name,
  las = 2,
  main = paste("Carboplatin (TCGA), cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()

treatment = "NK"
tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
  rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
tmp = tmp - tmp1
# plot(tmp, -cov362$avg_log2FC[match])
r = cor(tmp, tcga_test$logFC[match], method = "spearman")
signature = data.frame(FC = tmp, name = rownames(tcga_test)[match])
signature = signature[anchor,]

pdf(paste0(sub_output_dir, "NK_signature_TCGA_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(signature$FC > 0, red, blue)
fig = barplot(
  signature$FC,
  names.arg = signature$name,
  las = 2,
  main = paste("NK (TCGA), cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()

treatment = "Control"
tmp = rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
  rowMeans2(data2[tcga_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
r = cor(tmp, tcga_test$logFC[match], method = "spearman")
signature = data.frame(FC = tmp, name = rownames(tcga_test)[match])
signature = signature[anchor,]

pdf(paste0(sub_output_dir, "ctrl_signature_TCGA_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(signature$FC > 0, red, blue)
fig = barplot(
  signature$FC,
  names.arg = signature$name,
  las = 2,
  main = paste("Control (TCGA), cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()
```

#### Box plot 

##### PreS vs PreR

```{r}
res <- read.csv(
  paste0(sub_output_dir,"BRCA_cor_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "Control", "p_threshold")
res <- res[1:20, ]
# res$p = p_threshold
res.df = reshape2::melt(res, id.vars = "p_threshold", variable.name = "treatment")

values1 = c(blue, red, brown, green)
res.df$treatment = factor(res.df$treatment, levels = c("Olaparib", "Carboplatin", "NK", "Control"))
# ggplot(res.df, aes(p, value)) + geom_line(aes(colour = treatment), size = 2) + xlab("p value threshold") +
#  ylab("Correlation with BRCA signature") + theme(text = element_text(size = 20)) + scale_x_continuous(trans = "log10")
ggplot(res.df[which(res.df$treatment !="Control"),], aes(x = treatment, y = value, fill = treatment)) + geom_boxplot() + xlab("") + theme_bw() + 
  ylab("BRCAness (TCGA)") + theme(text = element_text(size = 20)) + scale_fill_manual(values = values1) 
ggsave(paste0(sub_output_dir, "BRCA_cor_tcga.pdf"), width = 5, height = 3.5)
# res1.df = res.df
```

##### (Pres+PreR)/PosT PreS/preR

```{r}
res1 <- read.csv(
    paste0(
      sub_output_dir,
      "BRCA_cor_tcga",
      ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")
    )
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "Control", "p_threshold")
res1 <- res1[1:20, ]
res1.df = reshape2::melt(res1, id.vars = "p_threshold", variable.name = "treatment")
res1.df$group = "PreS/PreR"

res2 <- read.csv(
    paste0(
      sub_output_dir,
      "BRCA_cor_tcga",
      ifelse(conserve_gene_only, "_conserve_gene_preRpreSposT.csv", "_preRpreSposT.csv")
    )
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "Control", "p_threshold")
res2 <- res2[1:20, ]
res2.df = reshape2::melt(res2, id.vars = "p_threshold", variable.name = "treatment")
res2.df$group = "(Pres+PreR)/PosT"
res.df = rbind.data.frame(res1.df, res2.df)

ggplot(
  res.df[res.df$treatment !="Control",],
  aes(x = group, y = value, fill = treatment)
) +
  geom_boxplot() +
  xlab("") +
  ylab("BRCAness (TCGA)") +
  theme_bw() + 
  theme(text = element_text(size = 20)) +
  scale_fill_manual(values = values1)
ggsave(
  paste0(
    sub_output_dir,
    "BRCA_cor_tcga_merged_preRpreSposT",
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 3.5
)
```

##### BT/AT PreS/preR

```{r}
res1 <- read.csv(
  paste0(sub_output_dir,"BRCA_cor_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "Control", "p_threshold")
res1 <- res1[1:20, ]
res1.df = reshape2::melt(res1, id.vars = "p_threshold", variable.name = "treatment")
res1.df$group = "PreS/PreR"

res2 <- read.csv(
  paste0(sub_output_dir,"BRCA_cor_tcga", ifelse(conserve_gene_only, "_conserve_gene_atbt.csv", "_atbt.csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "Control", "p_threshold")
res2 <- res2[1:20, ]
res2.df = reshape2::melt(res2, id.vars = "p_threshold", variable.name = "treatment")
res2.df$group = "PreT/PosT"

res.df = rbind.data.frame(res1.df, res2.df)
res.df$group <- factor(res.df$group, levels = c("PreT/PosT", "PreS/PreR"))

ggplot(
  res.df[res.df$treatment !="Control",],
  aes(x = group, y = value, fill = treatment)
) +
  geom_boxplot() +
  xlab("") +
  ylab("BRCAness (TCGA)") +
  theme_bw() + 
  theme(text = element_text(size = 20)) +
  scale_fill_manual(values = values1)
ggsave(
  paste0(
    sub_output_dir,
    "BRCA_cor_tcga_merged_atbt",
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 3.5
)
```

### COV362 signature

```{r}
cov362 = read.csv("../data/Jing_code_data/deg_COV362.csv") # JUN data
cov362[which(cov362$symbol == "BRCA1"),] # should be negative logFC
```

#### Test for good P threshold

##### PreS preR

```{r}
# min_p = 0.0000000000000000000001
min_p = 5 * 10^-16 # around 200 sister conserved BRCAness genes
# max_p = 0.000000000001
max_p = 0.05
# p_threshold = p_threshold[6] # the max correlation
p_threshold <- c(lseq(min_p, max_p, 20), 1 * 10 ^ -c(2:14), 5 * 10 ^ -c(2:14))

group1 = "pre-resistant"
group2 = "pre-sensitive"

res = mat.or.vec(length(p_threshold), 6)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK", "Control")

if (conserve_gene_only){
  data2 = data@assays$RNA@data[rownames(data) %in% sister_conserve_gene, ]
} else {
  data2 = data@assays$RNA@data
}

for(i in 1:length(p_threshold)){
  if (conserve_gene_only){
    cov362_test <- filter(cov362, ensembl %in% sister_conserve_gene)
  } else {
    cov362_test <- cov362
  }
  
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  index1 = which(cov362_test$avg_log2FC < 0 & cov362_test$p_val_adj <= p_threshold[i]) # downregulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(cov362_test$avg_log2FC > 0 & cov362_test$p_val_adj <= p_threshold[i]) # upregulated genes when BRCA1 is knocked out
  
  # whether these genes show the similar trend in the treatment?
  match1 = which(cov362_test$ensembl[index1] %in% rownames(data2))
  match2 = which(cov362_test$ensembl[index2] %in% rownames(data2))
  match = c(index1[match1], index2[match2])
  
  signature = data.frame(FC = -cov362_test$avg_log2FC[match], name = cov362_test$symbol[match])
  anchor = order(signature$FC)
  signature = signature[anchor,]

  signature_all = signature
  
  # Olaparib
  treatment = "Olaparib"
  tmp = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
    rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
  # plot(tmp, -cov362_test$avg_log2FC[match])
  tmp1 = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == "Control")]) - 
    rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == "Control")])
  tmp = tmp - tmp1
  res[i, 3] = cor(tmp, -cov362_test$avg_log2FC[match], method = "spearman") # should be positive

  signature = data.frame(FC = tmp, name = cov362_test$symbol[match])
  signature = signature[anchor,]

  signature_all = cbind(signature_all, signature)
  
  treatment = "Carboplatin"
  tmp = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
    rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
  # plot(tmp, -cov362_test$avg_log2FC[match])
  tmp1 = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == "Control")]) - 
    rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == "Control")])
  tmp = tmp - tmp1
  res[i, 4] = cor(tmp, -cov362_test$avg_log2FC[match], method = "spearman") # should be positive
  
  signature = data.frame(FC = tmp, name = cov362_test$symbol[match])
  signature = signature[anchor,]

  signature_all = cbind(signature_all, signature)
  
  treatment = "NK"
  tmp = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
    rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
  # plot(tmp, -cov362_test$avg_log2FC[match])
  tmp1 = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == "Control")]) - 
    rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == "Control")])
  tmp = tmp - tmp1
  res[i, 5] = cor(tmp, -cov362_test$avg_log2FC[match], method = "spearman") # should be positive
  
  signature = data.frame(FC = tmp, name = cov362_test$symbol[match])
  signature = signature[anchor,]

  signature_all = cbind(signature_all, signature)
  
  treatment = "Control"
  tmp = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
    rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
  res[i, 6] = cor(tmp, -cov362_test$avg_log2FC[match], method = "spearman") # should be positive
  
  signature = data.frame(FC = tmp, name = cov362_test$symbol[match])
  signature = signature[anchor,]

  signature_all = cbind(signature_all, signature)
  res[i, 2] = nrow(signature_all)
  # write.xlsx(signature_all, "signature_all.xlsx")
  print(res[i,])
}
write.csv(
  res,
  paste0(sub_output_dir,"BRCA_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE)
```

##### BT vs AT

```{r}
# min_p = 0.0000000000000000000001
min_p = 5 * 10^-16 # around 200 sister conserved BRCAness genes
# max_p = 0.000000000001
max_p = 0.05
# p_threshold = p_threshold[6] # the max correlation
p_threshold <- c(lseq(min_p, max_p, 20), 1 * 10 ^ -c(2:14), 5 * 10 ^ -c(2:14))
res = mat.or.vec(length(p_threshold), 6)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK", "Control")
treatments <- c("Olaparib", "Carboplatin", "NK")
group1 = "BeforeTreatment"
group2 = "AfterTreatment"

res = mat.or.vec(length(p_threshold), 6)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK", "Control")

if (conserve_gene_only){
  data2 = data@assays$RNA@data[rownames(data) %in% sister_conserve_gene, ]
} else {
  data2 = data@assays$RNA@data
}

for(i in 1:length(p_threshold)){
  if (conserve_gene_only){
    cov362_test <- filter(cov362, ensembl %in% sister_conserve_gene)
  } else {
    cov362_test <- cov362
  }
  
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  index1 = which(cov362_test$avg_log2FC < 0 & cov362_test$p_val_adj <= p_threshold[i]) # downregulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(cov362_test$avg_log2FC > 0 & cov362_test$p_val_adj <= p_threshold[i]) # upregulated genes when BRCA1 is knocked out
  
  # whether these genes show the similar trend in the treatment?
  match1 = which(cov362_test$ensembl[index1] %in% rownames(data2))
  match2 = which(cov362_test$ensembl[index2] %in% rownames(data2))
  match = c(index1[match1], index2[match2])
  
  signature_all = data.frame(
    name = cov362_test$symbol[match],
    FC_BRCAness = cov362_test$avg_log2FC[match]
  )

  tmp1 = rowMeans2(data2[cov362_test$ensembl[match], which(data$time_point == group1 & data$treatment == "Control")]) - 
    rowMeans2(data2[cov362_test$ensembl[match], which(data$time_point == group2 & data$treatment == "Control")])
  res[i, "Control"] = cor(tmp1, cov362_test$avg_log2FC[match], method = "spearman") # should be positive

  signature_all[["FC_Control"]] <-tmp1
  
  for (treatment in treatments){
    tmp = rowMeans2(data2[cov362_test$ensembl[match], which(data$time_point == group1 & data$treatment == treatment)]) - 
      rowMeans2(data2[cov362_test$ensembl[match], which(data$time_point == group2 & data$treatment == treatment)])
    # plot(tmp, cov362_test$avg_log2FC[match])
    tmp1 = rowMeans2(data2[cov362_test$ensembl[match], which(data$time_point == group1 & data$treatment == "Control")]) - 
      rowMeans2(data2[cov362_test$ensembl[match], which(data$time_point == group2 & data$treatment == "Control")])
    tmp = tmp - tmp1
    res[i, treatment] = cor(tmp, cov362_test$avg_log2FC[match], method = "spearman") # should be positive
    signature_all[[paste0("FC_", treatment)]] <-tmp
  }
  
  res[i, 2] = nrow(signature_all)
  
  print(res[i,])
}
write.csv(
  res,
  paste0(sub_output_dir,"BRCA_cor_cov", ifelse(conserve_gene_only, "_conserve_gene_atbt.csv", "_atbt.csv")),
  row.names = FALSE
)
```

##### (PreS + PreR) vs PosT

```{r}
# min_p = 0.0000000000000000000001
min_p = 5 * 10^-16 # around 200 sister conserved BRCAness genes
# max_p = 0.000000000001
max_p = 0.05
# p_threshold = p_threshold[6] # the max correlation
p_threshold <- c(lseq(min_p, max_p, 20), 1 * 10 ^ -c(2:14), 5 * 10 ^ -c(2:14))

res = mat.or.vec(length(p_threshold), 6)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK", "Control")

group1 = c("pre-sensitive", "pre-resistant")
group2 = "resistant"

res = mat.or.vec(length(p_threshold), 6)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK", "Control")

if (conserve_gene_only){
  data2 = data@assays$RNA@data[rownames(data) %in% sister_conserve_gene, ]
} else {
  data2 = data@assays$RNA@data
}

for(i in 1:length(p_threshold)){
  if (conserve_gene_only){
    cov362_test <- filter(cov362, ensembl %in% sister_conserve_gene)
  } else {
    cov362_test <- cov362
  }
  
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  index1 = which(cov362_test$avg_log2FC < 0 & cov362_test$p_val_adj <= p_threshold[i]) # downregulated genes when BRCA1 is knocked out (BRCA-)
  index2 = which(cov362_test$avg_log2FC > 0 & cov362_test$p_val_adj <= p_threshold[i]) # upregulated genes when BRCA1 is knocked out
  
  # whether these genes show the similar trend in the treatment?
  match1 = which(cov362_test$ensembl[index1] %in% rownames(data2))
  match2 = which(cov362_test$ensembl[index2] %in% rownames(data2))
  match = c(index1[match1], index2[match2])
  
  signature_all = data.frame(
    name = cov362_test$symbol[match],
    FC_BRCAness = cov362_test$avg_log2FC[match]
  )

  tmp1 = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens %in% group1 & data$treatment == "Control")]) - 
    rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens %in% group2 & data$treatment == "Control")])
  res[i, treatment] = cor(tmp1, cov362_test$avg_log2FC[match], method = "spearman") # should be positive

  signature_all[["FC_Control"]] <-tmp1
  
  for (treatment in treatments){
    tmp = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens %in% group1 & data$treatment == treatment)]) - 
      rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens %in% group2 & data$treatment == treatment)])
    # plot(tmp, cov362_test$avg_log2FC[match])
    tmp1 = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens %in% group1 & data$treatment == "Control")]) - 
      rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens %in% group2 & data$treatment == "Control")])
    tmp = tmp - tmp1
    res[i, treatment] = cor(tmp, cov362_test$avg_log2FC[match], method = "spearman") # should be positive
    signature_all[[paste0("FC_", treatment)]] <-tmp
  }
  
  res[i, 2] = nrow(signature_all)
  
  print(res[i,])
}
write.csv(
  res,
  paste0(sub_output_dir,"BRCA_cor_cov", ifelse(conserve_gene_only, "_conserve_gene_preRpreSposT.csv", "_preRpreSposT.csv")),
  row.names = FALSE
)
```

#### Bar plot

```{r}
selected_p_threshold <- 0.05

if (conserve_gene_only){
  cov362_test <- filter(cov362, ensembl %in% sister_conserve_gene)
  data2 = data@assays$RNA@data[rownames(data) %in% sister_conserve_gene, ]
} else {
  cov362_test <- cov362
   data2 = data@assays$RNA
}

index1 = which(
  cov362_test$avg_log2FC < 0 &
  cov362_test$p_val_adj <= selected_p_threshold
) # downregulated genes when BRCA1 is knocked out (BRCA-)
index2 = which(
  cov362_test$avg_log2FC > 0 &
  cov362_test$p_val_adj <= selected_p_threshold
) # upregulated genes when BRCA1 is knocked out

# whether these genes show the similar trend in the treatment?
match1 = which(cov362_test$ensembl[index1] %in% rownames(data2))
match2 = which(cov362_test$ensembl[index2] %in% rownames(data2))
match = c(index1[match1], index2[match2])

signature = data.frame(
  FC = cov362_test$avg_log2FC[match],
  name = cov362_test$symbol[match]
)
anchor = order(signature$FC)
signature = signature[anchor,]

# BRCAness
pdf(
  paste0(
    sub_output_dir,
    "BRCA_signature_cov_p_",
    selected_p_threshold,
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(signature$FC > 0, red, blue)
fig = barplot(
  signature$FC,
  names.arg = signature$name,
  las = 2,
  main = "BRCAness signature (COV362)",
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()

group1 = "pre-sensitive"
group2 = "pre-resistant"

# Olaparib
treatment = "Olaparib"
tmp = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
  rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
# plot(tmp, cov362_test$avg_log2FC[match])
tmp1 = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == "Control")]) - 
  rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == "Control")])
tmp = tmp - tmp1
r = cor(tmp, cov362_test$avg_log2FC[match], method = "spearman") # should be positive

signature = data.frame(FC = tmp, name = cov362_test$symbol[match])
signature = signature[anchor,]

pdf(
  paste0(
    sub_output_dir,
    "Olaparib_signature_cov_p_",
    selected_p_threshold,
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(signature$FC > 0, red, blue)
fig = barplot(
  signature$FC,
  names.arg = signature$name,
  las = 2,
  main = paste("Olaparib (COV362), cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()

treatment = "Carboplatin"
tmp = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
  rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
# plot(tmp, -cov362_test$avg_log2FC[match])
tmp1 = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == "Control")]) - 
  rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == "Control")])
tmp = tmp - tmp1
r = cor(tmp, cov362_test$avg_log2FC[match], method = "spearman") # should be positive

signature = data.frame(FC = tmp, name = cov362_test$symbol[match])
signature = signature[anchor,]

pdf(
  paste0(
    sub_output_dir,
    "Carboplatin_signature_cov_p_",
    selected_p_threshold,
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(signature$FC > 0, red, blue)
fig = barplot(
  signature$FC,
  names.arg = signature$name,
  las = 2,
  main = paste("Carboplatin (COV362), cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()

treatment = "NK"
tmp = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
  rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
# plot(tmp, -cov362_test$avg_log2FC[match])
tmp1 = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == "Control")]) - 
  rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == "Control")])
tmp = tmp - tmp1
r = cor(tmp, cov362_test$avg_log2FC[match], method = "spearman") # should be positive

signature = data.frame(FC = tmp, name = cov362_test$symbol[match])
signature = signature[anchor,]

pdf(
  paste0(
    sub_output_dir,
    "NK_signature_cov_p_",
    selected_p_threshold,
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(signature$FC > 0, red, blue)
fig = barplot(
  signature$FC,
  names.arg = signature$name,
  las = 2,
  main = paste("NK (COV362), cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()

treatment = "Control"
tmp = rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group1 & data$treatment == treatment)]) - 
  rowMeans2(data2[cov362_test$ensembl[match], which(data$drugSens == group2 & data$treatment == treatment)])
r = cor(tmp, cov362_test$avg_log2FC[match], method = "spearman") # should be positive

signature = data.frame(FC = tmp, name = cov362_test$symbol[match])
signature = signature[anchor,]

pdf(
  paste0(
    sub_output_dir,
    "ctrl_signature_cov_p_",
    selected_p_threshold,
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(signature$FC > 0, red, blue)
fig = barplot(
  signature$FC,
  names.arg = signature$name,
  las = 2,
  main = paste("Control (COV362), cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()
```

#### Box plot

##### PreS vs PreR

```{r}
res <- read.csv(
  paste0(sub_output_dir,"BRCA_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "Control", "p_threshold")
res <- res[1:20, ]
# res$p = p_threshold
res.df = reshape2::melt(res, id.vars = "p_threshold", variable.name = "treatment")
# write.xlsx(res, "BRCA_cor.xlsx")

values1 = c(blue, red, brown, green)
res.df$treatment = factor(res.df$treatment, levels = c("Olaparib", "Carboplatin", "NK", "Control"))
# ggplot(res.df, aes(p, value)) + geom_line(aes(colour = treatment), size = 2) + xlab("p value threshold") +
#  ylab("Correlation with BRCA signature") + theme(text = element_text(size = 20)) + scale_x_continuous(trans = "log10")
ggplot(res.df[which(res.df$treatment !="Control"),], aes(x = treatment, y = value, fill = treatment)) + geom_boxplot() + xlab("") + theme_bw() + 
  ylab("BRCAness (COV362)") + theme(text = element_text(size = 20)) + scale_fill_manual(values = values1) 
ggsave(paste0(sub_output_dir, "BRCA_cor_cov.pdf"), width = 6, height = 3.5)
```

##### (Pres+PreR)/PosT PreS/preR

```{r}
res1 <- read.csv(
  paste0(sub_output_dir,"BRCA_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "Control", "p_threshold")
res1 <- res1[1:20, ]
res1.df = reshape2::melt(res1, id.vars = "p_threshold", variable.name = "treatment")
res1.df$group = "PreS/PreR"

res2 <- read.csv(
  paste0(sub_output_dir,"BRCA_cor_cov", ifelse(conserve_gene_only, "_conserve_gene_preRpreSposT.csv", "_preRpreSposT.csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "Control", "p_threshold")
res2 <- res2[1:20, ]
res2.df = reshape2::melt(res2, id.vars = "p_threshold", variable.name = "treatment")
res2.df$group = "(Pres+PreR)/PosT"
res.df = rbind.data.frame(res1.df, res2.df)

ggplot(
  res.df[res.df$treatment !="Control",],
  aes(x = group, y = value, fill = treatment)
) +
  geom_boxplot() +
  xlab("") +
  ylab("BRCAness (COV362)") +
  theme_bw() + 
  theme(text = element_text(size = 20)) +
  scale_fill_manual(values = values1)
ggsave(paste0(sub_output_dir, "BRCA_cor_cov_merged_preRpreSposT", ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 3.5)
```

##### BT/AT PreS/preR

```{r}
res1 <- read.csv(
  paste0(sub_output_dir,"BRCA_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "Control", "p_threshold")
res1 <- res1[1:20, ]
res1.df = reshape2::melt(res1, id.vars = "p_threshold", variable.name = "treatment")
res1.df$group = "PreS/PreR"

res2 <- read.csv(
  paste0(sub_output_dir,"BRCA_cor_cov", ifelse(conserve_gene_only, "_conserve_gene_atbt.csv", "_atbt.csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "Control", "p_threshold")
res2 <- res2[1:20, ]
res2.df = reshape2::melt(res2, id.vars = "p_threshold", variable.name = "treatment")
res2.df$group = "PreT/PosT"

res.df = rbind.data.frame(res1.df, res2.df)
res.df$group <- factor(res.df$group, levels = c("PreT/PosT", "PreS/PreR"))

ggplot(
  res.df[res.df$treatment !="Control",],
  aes(x = group, y = value, fill = treatment)
) +
  geom_boxplot() +
  xlab("") +
  ylab("BRCAness (COV362)") +
  theme_bw() + 
  theme(text = element_text(size = 20)) +
  scale_fill_manual(values = values1)
ggsave(
  paste0(
    sub_output_dir,
    "BRCA_cor_cov_merged_atbt",
    ifelse(conserve_gene_only,"_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 3.5
)
```

## C: BRCAness/HRD box plot (seurat logFC)

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure3C_Seurat/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}

lseq <- function(from=1, to=100000, length.out=6) {
  # logarithmic spaced sequence
  # blatantly stolen from library("emdbook"), because need only this
  exp(seq(log(from), log(to), length.out = length.out))
}
treatments <- c("Olaparib", "Carboplatin", "NK")
```

### load data

```{r}
data <- readRDS("../data/Jing_code_data/merged_KURAMOCHI_samples1_scaled_new.RDS")
log2FC_table <- read.csv("../data/output/DEG_preR_preS_CNV_subclone_relation/log2FC_preR_preS_CNV_subclone.csv")
```

At vs BT logFC adjust growth control

```{r}
for (t in treatments){
  if (t == "Carboplatin"){
    t <- "Carbo"
  }
  tmp <- read.csv(paste0("../data/output/DEG_filter_by_DEG_list/at_bt/at_bt_", t, "_control_wilcox_summary.csv"))
  tmp[[paste0(t, "_log2FC_at_bt_adj")]] <- tmp$mean_log2FC_treat - tmp$mean_log2FC_ctrl
  tmp <- select(tmp, ensembl, symbol, paste0(t, "_log2FC_at_bt_adj"))
  log2FC_table <- log2FC_table %>% 
    left_join(tmp, by = c("ensembl", "symbol"))
}
```

TCGA BRCAness signature

```{r}
tcga = read.csv("../data/Jing_code_data/TCGA_deconvoluted_HRD_DEGs.tsv", sep = "\t")
tcga[which(rownames(tcga) == "BRCA1"), ] # should be negative logFC
tcga$ensembl = data@misc$gene_meta$ensembl[match(rownames(tcga), data@misc$gene_meta$symbol)]
length(which(is.na(tcga$ensembl)==T)) # 383 genes have no emsembl

summary(tcga$adj.P.Val)
```

```{r}
log2FC_table <- left_join(
  log2FC_table,
  tcga %>% 
    rownames_to_column("symbol") %>% 
    select(ensembl, symbol, BRCAness_tcga_logFC = logFC, BRCAness_tcga_p_adj = adj.P.Val),
  by = c("ensembl", "symbol")
)
```

COV362 BRCAness signature

```{r}
cov362 <- read.csv("../data/Jing_code_data/deg_COV362.csv") # JUN data
cov362[which(cov362$symbol == "BRCA1"),] # should be negative logFC
```

```{r}
log2FC_table <- left_join(
  log2FC_table,
  cov362 %>% 
    select(ensembl, symbol, BRCAness_cov_logFC = avg_log2FC, BRCAness_cov_p_adj = p_val_adj),
  by = c("ensembl", "symbol")
)
```

```{r}
write.csv(log2FC_table, paste0(sub_output_dir, "logFC_table_for_plots.csv"), row.names = FALSE)
```

```{r}
data <- read.csv(paste0(sub_output_dir, "logFC_table_for_plots.csv")) %>%
  filter(BRCAness_tcga_p_adj <= 0.01) %>% 
  select("log2FC_A", "log2FC_B", "log2FC_C", "log2FC_D",
         "log2FC_E", "BRCAness_tcga_logFC", "BRCAness_cov_logFC")

cor(data, method = "spearman", use = "complete")
```

```{r}
data <- read.csv(paste0(sub_output_dir, "logFC_table_for_plots.csv")) %>%
  filter(BRCAness_cov_p_adj <= 1 * 10^-8) %>% 
  select("log2FC_A", "log2FC_B", "log2FC_C", "log2FC_D",
         "log2FC_E", "BRCAness_tcga_logFC", "BRCAness_cov_logFC")

cor(data, method = "spearman", use = "complete")
```

```{r}
conserve_gene_only <- TRUE

if (conserve_gene_only){
  plot_table <- filter(log2FC_table, conserve_gene)
} else {
  plot_table <- table_test
}
```

#### Test for good P threshold

##### PreS preR

TCGA signature

```{r}
min_p = 0.005 # around 200 sister conserved BRCAness genes 
max_p = 0.05

p_threshold = c(lseq(min_p, max_p, 20), 1*10^-(2:4), 5*10^-(2:4))

res = mat.or.vec(length(p_threshold), 5)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK")

for(i in 1:length(p_threshold)){
  table_tests <- plot_table %>% 
    filter(BRCAness_tcga_p_adj <= p_threshold[i]) %>% 
    select(starts_with("BRCAness_tcga"), ends_with("log2FC_preR_preS_adj")) %>% 
    na.omit()
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  res[i, 2] <- nrow(table_tests)
  res[i, 3] <- cor(table_tests$BRCAness_tcga_logFC, -table_tests$Olaparib_log2FC_preR_preS_adj, method = "spearman", use = "complete")
  res[i, 4] <- cor(table_tests$BRCAness_tcga_logFC, -table_tests$Carbo_log2FC_preR_preS_adj, method = "spearman", use = "complete")
  res[i, 5] <- cor(table_tests$BRCAness_tcga_logFC, -table_tests$NK_log2FC_preR_preS_adj, method = "spearman", use = "complete")
}
write.csv(
  res,
  paste0(sub_output_dir,"BRCA_preRpreS_cor_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

COV362 signature

```{r}
min_p = 5 * 10^-16 # around 200 sister conserved BRCAness genes
max_p = 0.05
p_threshold <- c(lseq(min_p, max_p, 20), 1 * 10 ^ -c(2:14), 5 * 10 ^ -c(2:14))

res = mat.or.vec(length(p_threshold), 5)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK")

for(i in 1:length(p_threshold)){
  table_tests <- plot_table %>% 
    filter(BRCAness_cov_p_adj <= p_threshold[i]) %>% 
    select(starts_with("BRCAness_cov"), ends_with("log2FC_preR_preS_adj")) %>% 
    na.omit()
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  res[i, 2] <- nrow(table_tests)
  res[i, 3] <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Olaparib_log2FC_preR_preS_adj, method = "spearman", use = "complete")
  res[i, 4] <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Carbo_log2FC_preR_preS_adj, method = "spearman", use = "complete")
  res[i, 5] <- cor(table_tests$BRCAness_cov_logFC, -table_tests$NK_log2FC_preR_preS_adj, method = "spearman", use = "complete")
}
write.csv(
  res,
  paste0(sub_output_dir,"BRCA_preRpreS_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

##### BT vs AT

TCGA signature

```{r}
min_p = 0.005 # around 200 sister conserved BRCAness genes 
max_p = 0.05

p_threshold = c(lseq(min_p, max_p, 20), 1*10^-(2:4), 5*10^-(2:4))

res = mat.or.vec(length(p_threshold), 5)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK")

for(i in 1:length(p_threshold)){
  table_tests <- plot_table %>% 
    filter(BRCAness_tcga_p_adj <= p_threshold[i]) %>% 
    select(starts_with("BRCAness_tcga"), ends_with("log2FC_at_bt_adj")) %>% 
    na.omit()
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  res[i, 2] <- nrow(table_tests)
  res[i, 3] <- cor(table_tests$BRCAness_tcga_logFC, -table_tests$Olaparib_log2FC_at_bt_adj, method = "spearman", use = "complete")
  res[i, 4] <- cor(table_tests$BRCAness_tcga_logFC, -table_tests$Carbo_log2FC_at_bt_adj, method = "spearman", use = "complete")
  res[i, 5] <- cor(table_tests$BRCAness_tcga_logFC, -table_tests$NK_log2FC_at_bt_adj, method = "spearman", use = "complete")
}
write.csv(
  res,
  paste0(sub_output_dir,"BRCA_atbt_cor_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

COV362 signature

```{r}
min_p = 5 * 10^-16 # around 200 sister conserved BRCAness genes
max_p = 0.05
p_threshold <- c(lseq(min_p, max_p, 20), 1 * 10 ^ -c(2:14), 5 * 10 ^ -c(2:14))

res = mat.or.vec(length(p_threshold), 5)
colnames(res) <- c("p_threshold", "n_gene", "Olaparib", "Carboplatin", "NK")

for(i in 1:length(p_threshold)){
  table_tests <- plot_table %>% 
    filter(BRCAness_cov_p_adj <= p_threshold[i]) %>% 
    select(starts_with("BRCAness_cov"), ends_with("log2FC_at_bt_adj")) %>% 
    na.omit()
  cat('\r', i)
  res[i, 1] <- p_threshold[i]
  res[i, 2] <- nrow(table_tests)
  res[i, 3] <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Olaparib_log2FC_at_bt_adj, method = "spearman", use = "complete")
  res[i, 4] <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Carbo_log2FC_at_bt_adj, method = "spearman", use = "complete")
  res[i, 5] <- cor(table_tests$BRCAness_cov_logFC, -table_tests$NK_log2FC_at_bt_adj, method = "spearman", use = "complete")
}
write.csv(
  res,
  paste0(sub_output_dir,"BRCA_atbt_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv")),
  row.names = FALSE
)
```

#### Figure S5 Bar plot

##### TCGA signature

```{r}
selected_p_threshold <- 0.005

table_tests <- plot_table %>% 
  filter(BRCAness_tcga_p_adj <= selected_p_threshold) %>% 
  select(ensembl, starts_with("BRCAness_tcga"), ends_with("log2FC_preR_preS_adj")) %>% 
  na.omit() %>% 
  arrange(BRCAness_tcga_logFC)

# BRCAness
pdf(
  paste0(
    sub_output_dir, "S5A_BRCA_signature_tcga_p_", signif(selected_p_threshold, 2),
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(table_tests$BRCAness_tcga_logFC > 0, red, blue)
fig = barplot(
  table_tests$BRCAness_tcga_logFC,
  names.arg = table_tests$ensembl,
  las = 2,
  main = "BRCAness signature (TCGA)",
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# Olaparib
r <- cor(
  table_tests$BRCAness_tcga_logFC,
  -table_tests$Olaparib_log2FC_preR_preS_adj,
  method = "spearman", use = "complete"
  )
pdf(paste0(sub_output_dir,"S5A_Olaparib_signature_tcga_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$Olaparib_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$Olaparib_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("Olaparib to TCGA, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# Carboplatin
r <- cor(
  table_tests$BRCAness_tcga_logFC,
  -table_tests$Carbo_log2FC_preR_preS_adj,
  method = "spearman", use = "complete"
)
pdf(paste0(sub_output_dir,"S5A_Carboplatin_signature_tcga_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$Carbo_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$Carbo_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("Carboplatin to TCGA, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# NK
r <- cor(
  table_tests$BRCAness_tcga_logFC,
  -table_tests$NK_log2FC_preR_preS_adj,
  method = "spearman", use = "complete"
  )
pdf(paste0(sub_output_dir,"S5A_NK_signature_tcga_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$NK_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$NK_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("NK to TCGA, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()
```

##### COV362 signature

```{r}
selected_p_threshold <- 1 * 10^-8

table_tests <- plot_table %>% 
  filter(BRCAness_cov_p_adj <= selected_p_threshold) %>% 
  select(ensembl, starts_with("BRCAness_cov"), ends_with("log2FC_preR_preS_adj")) %>% 
  na.omit() %>% 
  arrange(BRCAness_cov_logFC)

# BRCAness
pdf(
  paste0(
    sub_output_dir, "S5B_BRCA_signature_cov_p_", signif(selected_p_threshold, 2),
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(table_tests$BRCAness_cov_logFC > 0, red, blue)
fig = barplot(
  table_tests$BRCAness_cov_logFC,
  names.arg = table_tests$ensembl,
  las = 2,
  main = "BRCAness signature (COV362)",
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# Olaparib
r <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Olaparib_log2FC_preR_preS_adj)
pdf(paste0(sub_output_dir,"S5B_Olaparib_signature_cov_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$Olaparib_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$Olaparib_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("Olaparib to COV362, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# Carboplatin
r <- cor(table_tests$BRCAness_cov_logFC, -table_tests$Carbo_log2FC_preR_preS_adj)
pdf(paste0(sub_output_dir,"S5B_Carboplatin_signature_cov_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$Carbo_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$Carbo_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("Carboplatin to COV362, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()

# NK
r <- cor(table_tests$BRCAness_cov_logFC, -table_tests$NK_log2FC_preR_preS_adj)
pdf(paste0(sub_output_dir,"S5B_NK_signature_cov_p_", signif(selected_p_threshold, 2) , ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")), width = 8, height = 4, pointsize = 20)
cor = ifelse(-table_tests$NK_log2FC_preR_preS_adj > 0, red, blue)
fig = barplot(
  -table_tests$NK_log2FC_preR_preS_adj,
  names.arg = table_tests$ensembl,
  las = 2,
  main = paste("NK to COV362, cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
dev.off()
```

#### Box plot BT/AT PreS/preR

```{r}
values1 = c(blue, red, brown, green)
```

##### TCGA signature

```{r}
res1 <- read.csv(
  paste0(sub_output_dir,"BRCA_preRpreS_cor_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "p_threshold")
res1 <- res1[1:20, ]
res1.df = reshape2::melt(res1, id.vars = "p_threshold", variable.name = "treatment")
res1.df$group = "PreS/PreR"

res2 <- read.csv(
  paste0(sub_output_dir,"BRCA_atbt_cor_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "p_threshold")
res2 <- res2[1:20, ]
res2.df = reshape2::melt(res2, id.vars = "p_threshold", variable.name = "treatment")
res2.df$group = "PreT/PosT"

res.df = rbind.data.frame(res1.df, res2.df)
res.df$group <- factor(res.df$group, levels = c("PreT/PosT", "PreS/PreR"))

ggplot(
  res.df[res.df$treatment !="Control",],
  aes(x = group, y = value, fill = treatment)
) +
  geom_boxplot() +
  xlab("") +
  ylab("BRCAness (TCGA)") +
  theme_bw() + 
  theme(text = element_text(size = 20)) +
  scale_fill_manual(values = values1)
ggsave(
  paste0(
    sub_output_dir,
    "BRCA_atbt_preRpreS_cor_tcga",
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 3.5
)
```

##### COV362 signature

```{r}
res1 <- read.csv(
  paste0(sub_output_dir,"BRCA_preRpreS_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "p_threshold")
res1 <- res1[1:20, ]
res1.df = reshape2::melt(res1, id.vars = "p_threshold", variable.name = "treatment")
res1.df$group = "PreS/PreR"

res2 <- read.csv(
  paste0(sub_output_dir,"BRCA_atbt_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "p_threshold")
res2 <- res2[1:20, ]
res2.df = reshape2::melt(res2, id.vars = "p_threshold", variable.name = "treatment")
res2.df$group = "PreT/PosT"

res.df = rbind.data.frame(res1.df, res2.df)
res.df$group <- factor(res.df$group, levels = c("PreT/PosT", "PreS/PreR"))

ggplot(
  res.df[res.df$treatment !="Control",],
  aes(x = group, y = value, fill = treatment)
) +
  geom_boxplot() +
  xlab("") +
  ylab("BRCAness (COV362)") +
  theme_bw() + 
  theme(text = element_text(size = 20)) +
  scale_fill_manual(values = values1)
ggsave(
  paste0(
    sub_output_dir,
    "BRCA_atbt_preRpreS_cor_cov",
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 8, height = 3.5
)
```

##### preS/preR two signatures

```{r}
res1 <- read.csv(
  paste0(sub_output_dir,"BRCA_preRpreS_cor_tcga", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "p_threshold")
res1 <- res1[1:20, ]
res1.df = reshape2::melt(res1, id.vars = "p_threshold", variable.name = "treatment")
res1.df$group = "TCGA"

res2 <- read.csv(
  paste0(sub_output_dir,"BRCA_preRpreS_cor_cov", ifelse(conserve_gene_only, "_conserve_gene.csv", ".csv"))
  ) %>% 
  select("Olaparib", "Carboplatin", "NK", "p_threshold")
res2 <- res2[1:20, ]
res2.df = reshape2::melt(res2, id.vars = "p_threshold", variable.name = "treatment")
res2.df$group = "COV362"

res.df = rbind.data.frame(res1.df, res2.df)
res.df$group <- factor(res.df$group, levels = c("TCGA", "COV362"))
res.df$treatment <- factor(res.df$treatment, levels = c("Carboplatin", "Olaparib", "NK"))

ggplot(
  res.df[res.df$treatment !="Control",],
  aes(x = group, y = value, fill = treatment)
) +
  geom_boxplot() +
  labs(x = "", y = "Correlation coefficient", fill = "")+
  theme_bw() + 
  theme(
    text = element_text(size = 20),
    axis.title = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 20, color = "black"),
    legend.text = element_text(size = 20, color = "black")
  ) +
  scale_fill_manual(values = values1)
ggsave(
  paste0(
    sub_output_dir,
    "BRCA_preRpreS_cor_cov_tcga",
    ifelse(conserve_gene_only, "_conserve_gene.pdf", ".pdf")
  ),
  width = 9, height = 5
)
```

## E: NK killing rate on COV362 and KURAMOCHI
```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure 3E/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```
Tables come from Jun. The background signal has been removed from the data.

### COV362

```{r}
table <- read_xlsx(
  "../data/raw/COV362-NK-20220417.xlsx", sheet = 2,
  range = "B1:Z7")
table[1, ] <- t(rep(c("0", "0.5", "1", "2", "3"), each = 5))
table <- t(table)
colnames(table) <- c("rate", "sample", "rep1", "rep2", "rep3", "rep4")
table <- table %>% 
  as.data.frame() %>% 
  filter(sample %in% c("COV362-GFP-M", "COV362-B1-GFP-M")) %>% 
  pivot_longer(cols = starts_with("rep"), values_to = "read", names_to = "replicates") %>% 
  na.omit()
table$read <- as.numeric(table$read)
NC_COV362_GFP_M <- mean(table$read[table$sample == "COV362-GFP-M" & table$rate == "0"])
NC_COV362_B1_GFP_M <- mean(table$read[table$sample == "COV362-B1-GFP-M" & table$rate == "0"])
table <- table %>% 
  mutate(
    cell_viability = ifelse(sample == "COV362-GFP-M", read/NC_COV362_GFP_M, read/NC_COV362_B1_GFP_M) * 100
  )
write.csv(table, paste0(sub_output_dir, "cell_viablility_COV362.csv"))
```

```{r}
plot_table <- table %>%
  dplyr::group_by(rate, sample) %>% 
  dplyr::summarise(
    mean = mean(cell_viability),
    sd = sd(cell_viability)
  )
plot_table$sample[plot_table$sample == "COV362-GFP-M"] <- "COV362-control"
plot_table$sample[plot_table$sample == "COV362-B1-GFP-M"] <- "COV362-BRCA1 restored"
plot_table$sample <- factor(plot_table$sample, levels = c("COV362-control", "COV362-BRCA1 restored"))
col = c("#B36981", "#659698")
p <- ggplot(plot_table, aes(x = rate, y = mean, fill = sample)) +
  geom_bar(
    stat="identity",
    color = "white",
    position=position_dodge()
  ) +
  geom_errorbar(
    aes(ymin=mean-sd, ymax=mean+sd),
    width=.2,
    position=position_dodge(.9)
  ) +
  scale_fill_manual(values = col) +
  labs(
    x = "NK : target cell",
    y = "Cell viability (%)"
  ) +
  theme_classic() +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 15),
    text = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 15, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "COV362_NK_killing.pdf"),
  width = 4.5, height = 4
)
```

### KURAMOCHI

```{r}
table <- read_xlsx(
    "../data/raw/Kura-NK-20220417.xlsx", sheet = 2,
  range = "B3:Z9")
table[1, ] <- t(rep(c("0", "0.5", "1", "2", "3"), each = 5))
table <- t(table)
colnames(table) <- c("rate", "sample", "rep1", "rep2", "rep3", "rep4")
table <- table %>% 
  as.data.frame() %>% 
  filter(sample %in% c("Kura-GFP-H", "Kura-B2-GFP-H")) %>% 
  pivot_longer(cols = starts_with("rep"), values_to = "read", names_to = "replicates") %>% 
  na.omit()
table$read <- as.numeric(table$read)
NC_Kura_GFP_M <- mean(table$read[table$sample == "Kura-GFP-H" & table$rate == "0"])
NC_Kura_B1_GFP_M <- mean(table$read[table$sample == "Kura-B2-GFP-H" & table$rate == "0"])
table <- table %>% 
  mutate(
    cell_viability = ifelse(sample == "Kura-GFP-H", read/NC_Kura_GFP_M, read/NC_Kura_B1_GFP_M) * 100
  )
write.csv(table, paste0(sub_output_dir, "cell_viablility_Kura.csv"))
```

```{r}
plot_table <- table %>%
  dplyr::group_by(rate, sample) %>% 
  dplyr::summarise(
    mean = mean(cell_viability),
    sd = sd(cell_viability)
  )
plot_table$sample[plot_table$sample == "Kura-GFP-H"] <- "KURAMOCHI-control"
plot_table$sample[plot_table$sample == "Kura-B2-GFP-H"] <- "KURAMOCHI-BRCA2 restored"
plot_table$sample <- factor(plot_table$sample, levels = c("KURAMOCHI-control", "KURAMOCHI-BRCA2 restored"))
col = c("#B36981", "#659698")
p <- ggplot(plot_table, aes(x = rate, y = mean, fill = sample)) +
  geom_bar(
    stat="identity",
    color = "white",
    position=position_dodge()
  ) +
  geom_errorbar(
    aes(ymin=mean-sd, ymax=mean+sd),
    width=.2,
    position=position_dodge(.9)
  ) +
  scale_fill_manual(values = col) +
  labs(
    x = "NK : target cell",
    y = "Cell viability (%)"
  ) +
  theme_classic() +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 15),
    text = element_text(size = 20, color = "black"),
    axis.text = element_text(size = 15, color = "black")
  )
ggsave(
  paste0(sub_output_dir, "KURAMOCHI_NK_killing.pdf"),
  width = 4.5, height = 4
)
```

## Figure S3B: Sankey plot for CNV subgroup transition

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure S3B/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
```

```{r}
data <- readRDS("../data/output/KURAMOCHI_CNV_subclone_transfer/leiden_resolution_0.3/merge_all_sample_with_CNV_subclone.RDS")

data <- data@meta.data
```

```{r}
# Merge replicates
samples <- c("Carboplatin", "Control", "NK", "Olaparib")
for (s in samples) {
  cat(s, "\n")
  data_tmp <- data %>% 
    filter(treatment == s)

  # Sankey diagram
  CNV_subclone <- data_tmp %>% 
    select(
      drugSens, sisters, prer_r_group, CNV_subclone, time_point, treatment) %>% 
    filter(!is.na(prer_r_group))
  CNV_subclone$CNV_subclone <- as.character(CNV_subclone$CNV_subclone)
  
  prer_r_group <- unique(CNV_subclone$prer_r_group)
  
  ## sankeyNetwork ggplot
  
  links <- data.frame(
    prer_r_group = prer_r_group,
    source = NA,
    target = NA
  )
  for (i in prer_r_group){
    # subclone in BT
    subclone_bt <- CNV_subclone %>% 
      filter(time_point == "BeforeTreatment" & prer_r_group == i) %>% 
      select(CNV_subclone) %>% 
      unique() %>% 
      unlist()
    if (length(subclone_bt) == 0 ){
      links$source[links$prer_r_group == i] <- NA
    } else if (length(subclone_bt) > 1) {
      links$source[links$prer_r_group == i] <- paste0(sort(subclone_bt), collapse = "")
    } else {
      links$source[links$prer_r_group == i] <- paste0(rep(subclone_bt, sum(links$prer_r_group == i)))
    }
    # subclone in AT
    subclone_at <- CNV_subclone %>% 
      filter(time_point == "AfterTreatment" & prer_r_group == i) %>% 
      select(CNV_subclone) %>% 
      unique() %>% 
      unlist()
    if (length(subclone_at) == 0 ){
      links$target[links$prer_r_group == i] <- NA
    } else if (length(subclone_at) > 1) {
      links$target[links$prer_r_group == i] <- paste0(sort(subclone_at), collapse = "")
    } else {
      links$target[links$prer_r_group == i] <- paste0(rep(subclone_at, sum(links$prer_r_group == i)))
    }
  }
  cols <- c(
    "A" = "#DC0000FF", "B" = "#B09C85FF", "C" = "#00A087FF", "BC" = "#B09C85FF",
    "AB" = "#F39B7FFF", "D" = "#91D1C2FF",  "E" = "#3C5488FF", "AD" = "#4DBBD5FF",
    "AC" = "#8491B4FF", "ACD" = "#E64B35FF")
  # remove NA subclone
  links <- na.omit(links)
  df <- links %>%
    select(`pre-treatment` = "source", `post-treatment` = "target") %>% 
    make_long(`pre-treatment`, `post-treatment`)
  ggplot(
    df,
    aes(x = x, 
        next_x = next_x, 
        node = node, 
        next_node = next_node,
        fill = factor(node),
        label = node)) +
  geom_sankey(flow.alpha = 0.5) +
  scale_fill_manual(
    values = cols) +
  geom_sankey_text(
    mapping = aes(x = ifelse(as.numeric(x) == 1, 0.9, 2.25)),
    size = 5,
    color = 1,
    show.legend = FALSE,
    width = 0.01,
    hjust = 1) +
  theme_sankey(base_size = 20) +
  ggtitle(s) +
  theme(
    legend.position = "none",
    plot.title = element_text(hjust = 0.5, size = 20, face = "bold"),
    text = element_text(size = 20),
    axis.text = element_text(size = 20, color = "black"),
    axis.title.x = element_blank()
  )
  ggsave(paste0(sub_output_dir, "CNV_subclone_transfer_", s, ".pdf"), width = 6, height = 6)
  print(nrow(links))
}
```

## I: BRCA pathway

```{r}
rm(list = setdiff(ls(), c("output_dir", "red", "blue", "brown", "green", "red2", "blue2", "brown2", "green2","red2", "blue3")))
sub_output_dir <- paste0(output_dir, "Figure3I/")
if (!dir.exists(sub_output_dir)){
  dir.create(sub_output_dir)
}
gene_set <- c("H", "C2", "C4", "C5", "C6", "C7")
base_breaks_y <- function(x){
  b <- pretty(x)
  d <- data.frame(x=-Inf, xend=-Inf, y=min(b), yend=max(b))
  list(geom_segment(data=d, aes(x=x, y=y, xend=xend, yend=yend), inherit.aes=FALSE),
       scale_y_continuous(breaks=b))
}
```

```{r}
samples <- c(
  # "at_bt_Carbo", "at_bt_Olaparib", "at_bt_NK",
  "prer_pres_Carbo", "prer_pres_Olaparib", "prer_pres_NK","prer_pres_control",
  "BRCA_COV362"
)

# prer_pres_Carbo
prer_pres_Carbo <- NULL
prer_pres_Olaparib <- NULL
prer_pres_NK <- NULL
prer_pres_control <- NULL
BRCA_COV362 <- NULL

for (s in gene_set){
  tmp <- read.csv(paste0("../data/output/GSEA/preR_preS_sister_conserve_gene/Carbo_merge_prer_pres_log2FC_",s,"_dif.csv"))
  prer_pres_Carbo <- rbind.data.frame(prer_pres_Carbo, tmp) %>% 
    na.omit()
  
  tmp <- read.csv(paste0("../data/output/GSEA/preR_preS_sister_conserve_gene/Olaparib_merge_prer_pres_log2FC_",s,"_dif.csv"))
  prer_pres_Olaparib <- rbind.data.frame(prer_pres_Olaparib, tmp) %>% 
    na.omit()
  
  tmp <- read.csv(paste0("../data/output/GSEA/preR_preS_sister_conserve_gene/NK_merge_prer_pres_log2FC_",s,"_dif.csv"))
  prer_pres_NK <- rbind.data.frame(prer_pres_NK, tmp) %>% 
    na.omit()
  
  tmp <- read.csv(paste0("../data/output/GSEA/preR_preS_sister_conserve_gene/ctrl_merge_prer_pres_log2FC_",s,"_dif.csv"))
  prer_pres_control <- rbind.data.frame(prer_pres_control, tmp) %>% 
    na.omit()
  
  tmp <- read.csv(paste0("../data/output/GSEA/BRCA_COV362/BRCA1_neg_pos_log2FC_",s,".csv"))
  BRCA_COV362 <- rbind.data.frame(BRCA_COV362, tmp) %>% 
    na.omit()
}

BRCA_COV362_filter <- BRCA_COV362 %>% 
  filter(padj < 0.05, pathway %in% prer_pres_Carbo$pathway) %>%
  arrange(NES)
overlap_pathway <- Reduce(intersect, list(
  BRCA_COV362_filter$pathway, prer_pres_Carbo$pathway,
  prer_pres_control$pathway, prer_pres_Olaparib$pathway,
  prer_pres_NK$pathway
))

BRCA_COV362_filter <- BRCA_COV362_filter %>% 
  filter(pathway %in% overlap_pathway) %>% 
  unique() %>% 
  mutate(
    rank = 1:n(),
    color = NES < 0
  )

prer_pres_Carbo_filter <- prer_pres_Carbo %>% 
  inner_join(
    BRCA_COV362_filter %>% 
      select(pathway, rank),
    by = "pathway"
  ) %>% 
  arrange(rank)

prer_pres_Olaparib_filter <- prer_pres_Olaparib %>% 
  inner_join(
    BRCA_COV362_filter %>% 
      select(pathway, rank),
    by = "pathway"
  ) %>% 
  arrange(rank)

prer_pres_NK_filter <- prer_pres_NK %>% 
  inner_join(
    BRCA_COV362_filter %>% 
      select(pathway, rank),
    by = "pathway"
  ) %>% 
  arrange(rank)

prer_pres_control_filter <- prer_pres_control %>% 
  inner_join(
    BRCA_COV362_filter %>% 
      select(pathway, rank),
    by = "pathway"
  ) %>% 
  arrange(rank)
```

```{r}
pdf(
  paste0(
    sub_output_dir,
    "BRCA_pathway_bar_plot.pdf"
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(BRCA_COV362_filter$NES > 0, red, blue)
fig = barplot(
  BRCA_COV362_filter$NES,
  names.arg = BRCA_COV362_filter$pathway,
  las = 2,
  main = "BRCAness pathway (COV362)",
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()
```

```{r}
r <- cor(-prer_pres_Olaparib_filter$NES, BRCA_COV362_filter$NES)
pdf(
  paste0(
    sub_output_dir,
    "Olaparib_pathway_bar_plot.pdf"
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(-prer_pres_Olaparib_filter$NES > 0, red, blue)
fig = barplot(
  -prer_pres_Olaparib_filter$NES,
  names.arg = prer_pres_Olaparib_filter$pathway,
  las = 2,
  main = paste("Olaparib (COV362), cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()
```

```{r}
r <- cor(-prer_pres_Carbo_filter$NES, BRCA_COV362_filter$NES)
pdf(
  paste0(
    sub_output_dir,
    "Carboplatin_pathway_bar_plot.pdf"
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(-prer_pres_Carbo_filter$NES > 0, red, blue)
fig = barplot(
  -prer_pres_Carbo_filter$NES,
  names.arg = prer_pres_Carbo_filter$pathway,
  las = 2,
  main = paste("Carboplatin (COV362), cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()
```

```{r}
r <- cor(-prer_pres_NK_filter$NES, BRCA_COV362_filter$NES)
pdf(
  paste0(
    sub_output_dir,
    "NK_pathway_bar_plot.pdf"
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(-prer_pres_NK_filter$NES > 0, red, blue)
fig = barplot(
  -prer_pres_NK_filter$NES,
  names.arg = prer_pres_NK_filter$pathway,
  las = 2,
  main = paste("NK (COV362), cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()
```

```{r}
r <- cor(-prer_pres_control_filter$NES, BRCA_COV362_filter$NES)
pdf(
  paste0(
    sub_output_dir,
    "control_pathway_bar_plot.pdf"
  ),
  width = 8, height = 4, pointsize = 20
)
cor = ifelse(-prer_pres_control_filter$NES > 0, red, blue)
fig = barplot(
  -prer_pres_control_filter$NES,
  names.arg = prer_pres_control_filter$pathway,
  las = 2,
  main = paste("Growth control (COV362), cor =", round(r,2)),
  cex.names = 1,
  col = cor,
  ylab = 'FC',
  xaxt = "n",
  border = NA,
  space = 0
)
# text(fig, par("usr")[3], labels = signature$name, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex = 0.5)
dev.off()
```

```{r}
colnames(BRCA_COV362_filter)[colnames(BRCA_COV362_filter) != "pathway"] <- paste0("COV362_", colnames(BRCA_COV362_filter)[colnames(BRCA_COV362_filter) != "pathway"])

colnames(prer_pres_Carbo_filter)[colnames(prer_pres_Carbo_filter) != "pathway"] <- paste0("Carbo_", colnames(prer_pres_Carbo_filter)[colnames(prer_pres_Carbo_filter) != "pathway"])

colnames(prer_pres_Olaparib_filter)[colnames(prer_pres_Olaparib_filter) != "pathway"] <- paste0("Olaparib_", colnames(prer_pres_Olaparib_filter)[colnames(prer_pres_Olaparib_filter) != "pathway"])

colnames(prer_pres_NK_filter)[colnames(prer_pres_NK_filter) != "pathway"] <- paste0("NK_", colnames(prer_pres_NK_filter)[colnames(prer_pres_NK_filter) != "pathway"])

colnames(prer_pres_control_filter)[colnames(prer_pres_control_filter) != "pathway"] <- paste0("control_", colnames(prer_pres_control_filter)[colnames(prer_pres_control_filter) != "pathway"])

merge_table <- BRCA_COV362_filter %>% 
  left_join(prer_pres_Carbo_filter, by = "pathway")%>% 
  left_join(prer_pres_Olaparib_filter, by = "pathway")%>% 
  left_join(prer_pres_NK_filter, by = "pathway")%>% 
  left_join(prer_pres_control_filter, by = "pathway")

write.csv(
  merge_table,
  paste0(
    sub_output_dir,
    "GSEA_BRCAness_treatment_merged_table.csv"
  ),
  row.names = FALSE)
```

```{r}
sessionInfo()
```


